<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="description" content="Practise scRNA data analysis with Genome Medicine using Seurat package.">
<title>Practise Seurat for single cell data analysis – Guorui Zhong</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting-1b3db88def35042d172274863c1cdcf0.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap-7240976d66d94fd6f6e2300715a3fab7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
</head>
<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html"> 
<span class="menu-text">Wiki</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../linux/index.html"> 
<span class="menu-text">Linux</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../python/index.html"> 
<span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../r/index.html"> 
<span class="menu-text">R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html"> 
<span class="menu-text">Keyboard</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Practise Seurat for single cell data analysis</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
                  <div>
        <div class="description">
          Practise scRNA data analysis with Genome Medicine using Seurat package.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">seurat</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Monday, November 27, 2023</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Wednesday, January 10, 2024</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#packages" id="toc-packages" class="nav-link active" data-scroll-target="#packages">Packages</a></li>
  <li><a href="#retrieve-data" id="toc-retrieve-data" class="nav-link" data-scroll-target="#retrieve-data">Retrieve data</a></li>
  <li><a href="#load-batch-data" id="toc-load-batch-data" class="nav-link" data-scroll-target="#load-batch-data">Load batch data</a></li>
  <li>
<a href="#merage-samples-data" id="toc-merage-samples-data" class="nav-link" data-scroll-target="#merage-samples-data">Merage samples data</a>
  <ul>
<li>
<a href="#directly-merge" id="toc-directly-merge" class="nav-link" data-scroll-target="#directly-merge">Directly merge</a>
  <ul class="collapse">
<li><a href="#filter" id="toc-filter" class="nav-link" data-scroll-target="#filter">Filter</a></li>
  <li><a href="#seurat-workflow" id="toc-seurat-workflow" class="nav-link" data-scroll-target="#seurat-workflow">Seurat workflow</a></li>
  <li><a href="#umap" id="toc-umap" class="nav-link" data-scroll-target="#umap">UMAP</a></li>
  </ul>
</li>
  <li><a href="#harmony-to-remove-batch-effect" id="toc-harmony-to-remove-batch-effect" class="nav-link" data-scroll-target="#harmony-to-remove-batch-effect">Harmony to remove batch effect</a></li>
  </ul>
</li>
  <li><a href="#visualize-marker-genes" id="toc-visualize-marker-genes" class="nav-link" data-scroll-target="#visualize-marker-genes">Visualize marker genes</a></li>
  <li>
<a href="#annotate-clusters" id="toc-annotate-clusters" class="nav-link" data-scroll-target="#annotate-clusters">Annotate clusters</a>
  <ul>
<li><a href="#using-vector" id="toc-using-vector" class="nav-link" data-scroll-target="#using-vector">Using Vector</a></li>
  <li><a href="#directly-assign" id="toc-directly-assign" class="nav-link" data-scroll-target="#directly-assign">Directly assign</a></li>
  <li><a href="#add-annotation-to-meta-data" id="toc-add-annotation-to-meta-data" class="nav-link" data-scroll-target="#add-annotation-to-meta-data">Add annotation to meta data</a></li>
  </ul>
</li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="packages" class="level2"><h2 class="anchored" data-anchor-id="packages">Packages</h2>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">## Loading required package: SeuratObject</span></span>
<span><span class="co">## Loading required package: sp</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SeuratObject'</span></span>
<span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/immunogenomics/harmony">harmony</a></span><span class="op">)</span></span>
<span><span class="co">## Loading required package: Rcpp</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="co">## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──</span></span>
<span><span class="co">## ✔ dplyr     1.1.4     ✔ readr     2.1.4</span></span>
<span><span class="co">## ✔ forcats   1.0.0     ✔ stringr   1.5.1</span></span>
<span><span class="co">## ✔ ggplot2   3.4.4     ✔ tibble    3.2.1</span></span>
<span><span class="co">## ✔ lubridate 1.9.3     ✔ tidyr     1.3.0</span></span>
<span><span class="co">## ✔ purrr     1.0.2</span></span>
<span><span class="co">## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span><span class="co">## ✖ tidyr::expand() masks Matrix::expand()</span></span>
<span><span class="co">## ✖ dplyr::filter() masks stats::filter()</span></span>
<span><span class="co">## ✖ dplyr::lag()    masks stats::lag()</span></span>
<span><span class="co">## ✖ tidyr::pack()   masks Matrix::pack()</span></span>
<span><span class="co">## ✖ tidyr::unpack() masks Matrix::unpack()</span></span>
<span><span class="co">## ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span></span>
<span><span class="co">## here() starts at /Users/zhonggr/Library/CloudStorage/OneDrive-Personal/quarto</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nx10/httpgd">httpgd</a></span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="retrieve-data" class="level2"><h2 class="anchored" data-anchor-id="retrieve-data">Retrieve data</h2>
<ul>
<li>The source data is from <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00776-9">Single-cell transcriptome analysis of tumor and stromal compartments of pancreatic ductal adenocarcinoma primary tumors and metastatic lesions</a>
</li>
<li>Down load data GSE154778_RAW.tar from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154778">here</a> to data and unzip it.</li>
<li>Preprocess sequencing data</li>
</ul>
<p><img src="practise.png" class="img-fluid"></p>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Change working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"./learn/2023_scRNA_Seurat/data/GSE154778_RAW/"</span><span class="op">)</span></span>
<span><span class="co"># getwd()</span></span>
<span><span class="co"># Check files</span></span>
<span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"./"</span>, <span class="st">"^GSM"</span><span class="op">)</span></span>
<span><span class="co"># Get the sample names</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu">str_split</span><span class="op">(</span><span class="va">fs</span>, <span class="st">"_"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="co"># Create folders for each sample, and rename</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">y</span>  <span class="op">&lt;-</span> <span class="va">fs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">fs</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu">str_split</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"_"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">folder</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"barcodes.tsv.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Note the seurat version to check features.tsv.gz or genes.tsv.gz</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"features.tsv.gz"</span><span class="op">)</span><span class="op">)</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="load-batch-data" class="level2"><h2 class="anchored" data-anchor-id="load-batch-data">Load batch data</h2>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Change working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"./learn/2023_scRNA_Seurat/data/GSE154778_RAW/"</span><span class="op">)</span></span>
<span><span class="va">folders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"./"</span><span class="op">)</span></span>
<span><span class="va">folders</span></span>
<span><span class="va">sceList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>  <span class="va">folders</span>, <span class="kw">function</span><span class="op">(</span><span class="va">folder</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">CreateSeuratObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">Read10X</span><span class="op">(</span><span class="va">folder</span><span class="op">)</span>, project <span class="op">=</span> <span class="va">folder</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="merage-samples-data" class="level2"><h2 class="anchored" data-anchor-id="merage-samples-data">Merage samples data</h2>
<section id="directly-merge" class="level3"><h3 class="anchored" data-anchor-id="directly-merge">Directly merge</h3>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Use Seurat merge</span></span>
<span><span class="va">sce.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">sceList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">sceList</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  <span class="va">sceList</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">8</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">11</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  <span class="va">sceList</span><span class="op">[[</span><span class="fl">12</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">13</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">14</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>, <span class="va">sceList</span><span class="op">[[</span><span class="fl">16</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  <span class="co">## Sample names</span></span>
<span>  add.cell.ids <span class="op">=</span> <span class="va">folders</span>,  </span>
<span>  project <span class="op">=</span> <span class="st">"scRNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">sce.all</span>, <span class="fu">here</span><span class="op">(</span><span class="st">"learn"</span>, <span class="st">"2023_scRNA_Seurat"</span>, <span class="st">"sce.all.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="filter" class="level4"><h4 class="anchored" data-anchor-id="filter">Filter</h4>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"learn"</span>, <span class="st">"2023_scRNA_Seurat"</span>, <span class="st">"sce.all.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sce.all</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">##                         orig.ident nCount_RNA nFeature_RNA</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1     K16733       2464          965</span></span>
<span><span class="co">## K16733_AAACCGTGGGTAGG-1     K16733        689          336</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1     K16733       7145         1919</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1     K16733       1655          621</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1     K16733      14272         2771</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1     K16733      13832         2541</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce.all</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## K16733    T10     T2     T3     T4     T5     T6     T8     T9 Y00006 Y00008 </span></span>
<span><span class="co">##    585   1570    837   1026   1826    769   1098   1139    898    786    533 </span></span>
<span><span class="co">## Y00013 Y00014 Y00016 Y00027 </span></span>
<span><span class="co">##    745    526    272   2484</span></span>
<span><span class="co"># Mitochandrial genes</span></span>
<span><span class="va">sce.all</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">sce.all</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="co"># Ribonucleoprotein </span></span>
<span><span class="va">sce.all</span><span class="op">[[</span><span class="st">"percent.rp"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">sce.all</span>, pattern <span class="op">=</span> <span class="st">"^RP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specific gene set</span></span>
<span><span class="va">HB.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HBA1"</span>,<span class="st">"HBA2"</span>,<span class="st">"HBB"</span>,<span class="st">"HBD"</span>,<span class="st">"HBE1"</span>,<span class="st">"HBG1"</span>,<span class="st">"HBG2"</span>,<span class="st">"HBM"</span>,<span class="st">"HBQ1"</span>,<span class="st">"HBZ"</span><span class="op">)</span> </span>
<span><span class="co"># Red blood cells genes</span></span>
<span><span class="va">sce.all</span><span class="op">[[</span><span class="st">"percent.HB"</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span><span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">sce.all</span>, features <span class="op">=</span> <span class="va">HB.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sce.all</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">##                         orig.ident nCount_RNA nFeature_RNA percent.mt</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1     K16733       2464          965 12.6623377</span></span>
<span><span class="co">## K16733_AAACCGTGGGTAGG-1     K16733        689          336  2.1770682</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1     K16733       7145         1919  2.4492652</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1     K16733       1655          621  2.1752266</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1     K16733      14272         2771  1.5414798</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1     K16733      13832         2541  0.3181029</span></span>
<span><span class="co">##                         percent.rp percent.HB</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1   13.35227          0</span></span>
<span><span class="co">## K16733_AAACCGTGGGTAGG-1   29.02758          0</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1   36.72498          0</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1   35.52870          0</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1   38.81026          0</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1   40.73164          0</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span></span>
<span>    <span class="va">sce.all</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span></span>
<span>        <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">5000</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="seurat-workflow" class="level4"><h4 class="anchored" data-anchor-id="seurat-workflow">Seurat workflow</h4>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">## Normalizing layer: counts.1</span></span>
<span><span class="co">## Normalizing layer: counts.2</span></span>
<span><span class="co">## Normalizing layer: counts.3</span></span>
<span><span class="co">## Normalizing layer: counts.4</span></span>
<span><span class="co">## Normalizing layer: counts.5</span></span>
<span><span class="co">## Normalizing layer: counts.6</span></span>
<span><span class="co">## Normalizing layer: counts.7</span></span>
<span><span class="co">## Normalizing layer: counts.8</span></span>
<span><span class="co">## Normalizing layer: counts.9</span></span>
<span><span class="co">## Normalizing layer: counts.10</span></span>
<span><span class="co">## Normalizing layer: counts.11</span></span>
<span><span class="co">## Normalizing layer: counts.12</span></span>
<span><span class="co">## Normalizing layer: counts.13</span></span>
<span><span class="co">## Normalizing layer: counts.14</span></span>
<span><span class="co">## Normalizing layer: counts.15</span></span>
<span><span class="co">## Normalizing layer: counts.16</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">sce</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">## Finding variable features for layer counts.1</span></span>
<span><span class="co">## Finding variable features for layer counts.2</span></span>
<span><span class="co">## Finding variable features for layer counts.3</span></span>
<span><span class="co">## Finding variable features for layer counts.4</span></span>
<span><span class="co">## Finding variable features for layer counts.5</span></span>
<span><span class="co">## Finding variable features for layer counts.6</span></span>
<span><span class="co">## Finding variable features for layer counts.7</span></span>
<span><span class="co">## Finding variable features for layer counts.8</span></span>
<span><span class="co">## Finding variable features for layer counts.9</span></span>
<span><span class="co">## Finding variable features for layer counts.10</span></span>
<span><span class="co">## Finding variable features for layer counts.11</span></span>
<span><span class="co">## Finding variable features for layer counts.12</span></span>
<span><span class="co">## Finding variable features for layer counts.13</span></span>
<span><span class="co">## Finding variable features for layer counts.14</span></span>
<span><span class="co">## Finding variable features for layer counts.15</span></span>
<span><span class="co">## Finding variable features for layer counts.16</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span></span>
<span><span class="co">## Centering and scaling data matrix</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">sce</span>, npcs <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  KRT19, KRT8, KRT18, SMIM22, MAL2, SPINT2, TSPAN8, CLDN4, GPRC5A, PERP </span></span>
<span><span class="co">##     C19orf33, ELF3, TM4SF1, TMC5, LSR, LGALS4, NQO1, TACSTD2, CLDN7, SPINK1 </span></span>
<span><span class="co">##     MUC1, C12orf75, GPX2, TSPAN1, ERBB3, SFTA2, MMP7, CYP3A5, CDH1, TMPRSS4 </span></span>
<span><span class="co">## Negative:  VIM, COL1A2, BGN, COL1A1, SERPINF1, FN1, C1R, MGP, CTHRC1, TAGLN </span></span>
<span><span class="co">##     PMP22, NUPR1, THY1, FBLN1, RARRES2, TIMP3, MXRA8, TCF4, CLEC11A, INHBA </span></span>
<span><span class="co">##     RAB31, CCDC80, ASPN, THBS2, APOD, ISLR, TUBA1A, FSTL1, ANTXR1, MEG3 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  LAPTM5, AIF1, SRGN, LST1, HLA-DPA1, HLA-DRA, HLA-DPB1, MS4A6A, HLA-DQA1, MS4A7 </span></span>
<span><span class="co">##     HLA-DQB1, C1orf162, OLR1, HLA-DRB1, CD53, CD74, CYBB, FCGR2A, CLEC7A, ALOX5AP </span></span>
<span><span class="co">##     CD37, ITGB2, CD14, CD83, MS4A4A, IFI30, RGS1, RNASE6, CD86, HLA-DQA2 </span></span>
<span><span class="co">## Negative:  BGN, C1R, TPM1, COL1A1, RARRES2, COL1A2, NBL1, MXRA8, CTHRC1, FSTL1 </span></span>
<span><span class="co">##     THY1, FBLN1, CCDC80, IGFBP4, TAGLN, NNMT, THBS2, MGP, ASPN, TIMP3 </span></span>
<span><span class="co">##     ISLR, MEG3, APOD, EFEMP2, MFGE8, DKK3, SPON2, FBN1, ANTXR1, EMILIN1 </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  CTSE, VSIG2, AGR3, FOS, MUC5AC, CYSTM1, ATF3, JUN, FOSB, ELF3 </span></span>
<span><span class="co">##     RHOB, LINC01133, IER3, CAPN8, NEAT1, KLF6, TFF3, MUC1, DUSP1, EGR1 </span></span>
<span><span class="co">##     BACE2, PIGR, KLF4, KLF2, CREB3L1, REG4, EDN1, HSPA1B, PLAC8, ZG16B </span></span>
<span><span class="co">## Negative:  TUBA1B, TOP2A, MKI67, PTTG1, UBE2C, H2AFZ, CENPW, TPX2, HMGB2, CDK1 </span></span>
<span><span class="co">##     STMN1, RRM2, RBP1, ASPM, KLK6, PRC1, HMGB1, ATAD2, NUSAP1, GTSE1 </span></span>
<span><span class="co">##     CDKN3, KIF20B, CEP55, HMMR, DTYMK, CDCA3, CLSPN, CENPU, CCNB1, UBE2T </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  MDK, TMEM176B, COL11A1, NBL1, TMEM176A, FAM3C, LYZ, INHBA, THBS2, C1QTNF3 </span></span>
<span><span class="co">##     GCNT3, GPNMB, FBLN1, KLK6, RARRES2, IGFL2, COL8A1, C12orf75, FNDC1, MMP7 </span></span>
<span><span class="co">##     GREM1, PERP, NTM, CLDN3, GJB2, COMP, ISLR, CXCL14, MEG3, RBP1 </span></span>
<span><span class="co">## Negative:  PLVAP, RAMP2, VWF, ECSCR, AQP1, CDH5, CALCRL, BCAM, RAMP3, NOTCH4 </span></span>
<span><span class="co">##     CLDN5, MMRN2, FAM167B, ADAMTS9, EMCN, CD34, CD93, STC1, CYYR1, GPR4 </span></span>
<span><span class="co">##     S1PR1, ANGPT2, PODXL, MYCT1, ARHGAP29, RGS5, CRIP2, ROBO4, GJA4, HIGD1B </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  TOP2A, UBE2C, ATAD2, NUCKS1, PRC1, STMN1, SLPI, GTSE1, CENPW, HMGB1 </span></span>
<span><span class="co">##     TPX2, CTSD, RAD51AP1, PLAT, CDK1, RRM2, DTYMK, CLSPN, ANLN, MYBL2 </span></span>
<span><span class="co">##     FAM83A, CCNB1, NUSAP1, CENPU, ASPM, UBE2T, PRR11, MKI67, CEP55, CRIP2 </span></span>
<span><span class="co">## Negative:  CD3D, CD2, PTPRCAP, CD7, C12orf75, CCL5, CD3G, RBP1, GCNT3, CD27 </span></span>
<span><span class="co">##     KLRB1, LTB, ZFAS1, CA12, GZMM, RHOH, SPINK1, VNN1, ICOS, CLDN3 </span></span>
<span><span class="co">##     CYTIP, TIGIT, PTPN7, GZMB, CDHR2, RUNX3, SLC7A11, CCR7, CD8A, FAM3C</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">## Computing nearest neighbor graph</span></span>
<span><span class="co">## Computing SNN</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span><span class="va">sce</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 11914</span></span>
<span><span class="co">## Number of edges: 436451</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.9419</span></span>
<span><span class="co">## Number of communities: 18</span></span>
<span><span class="co">## Elapsed time: 1 seconds</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">## 15:18:03 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">## 15:18:03 Read 11914 rows and found 30 numeric columns</span></span>
<span><span class="co">## 15:18:03 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">## 15:18:03 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">## 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">## [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">## **************************************************|</span></span>
<span><span class="co">## 15:18:03 Writing NN index file to temp file /var/folders/2c/9q3pg2295195bp3gnrgbzrg40000gn/T//RtmpjzstFL/file2afb9bffbb</span></span>
<span><span class="co">## 15:18:03 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">## 15:18:05 Annoy recall = 100%</span></span>
<span><span class="co">## 15:18:05 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">## 15:18:06 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">## 15:18:06 Commencing optimization for 200 epochs, with 492128 positive edges</span></span>
<span><span class="co">## 15:18:10 Optimization finished</span></span>
<span><span class="co"># sce &lt;- RunTSNE(sce, dims = 1:30)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="umap" class="level4"><h4 class="anchored" data-anchor-id="umap">UMAP</h4>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Rename sample</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"K16733"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P01"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00006"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P02"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P03"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P04"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P05"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T5"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P06"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T6"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P07"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T8"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P08"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T9"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P09"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"T10"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"P10"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00008"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET01"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00013"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET02"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00014"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET03"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00016"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET04"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00019"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET05"</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"Y00027"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"MET06"</span></span>
<span></span>
<span><span class="co"># Add group information</span></span>
<span><span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"MET"</span>,<span class="va">sce</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample</span> <span class="op">)</span> ,<span class="st">"MET"</span> ,<span class="st">"PT"</span> <span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Global</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,repel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Sample</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"sample"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  repel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># group</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"group"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  repel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span><span class="va">p3</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="harmony-to-remove-batch-effect" class="level3"><h3 class="anchored" data-anchor-id="harmony-to-remove-batch-effect">Harmony to remove batch effect</h3>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="va">sce</span>  <span class="op">|&gt;</span>  <span class="fu">RunHarmony</span><span class="op">(</span><span class="st">"sample"</span>, plot_convergence <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## Transposing data matrix</span></span>
<span><span class="co">## Initializing state using k-means centroids initialization</span></span>
<span><span class="co">## Harmony 1/10</span></span>
<span><span class="co">## Harmony 2/10</span></span>
<span><span class="co">## Harmony 3/10</span></span>
<span><span class="co">## Harmony 4/10</span></span>
<span><span class="co">## Harmony 5/10</span></span>
<span><span class="co">## Harmony 6/10</span></span>
<span><span class="co">## Harmony 7/10</span></span>
<span><span class="co">## Harmony 8/10</span></span>
<span><span class="co">## Harmony 9/10</span></span>
<span><span class="co">## Harmony 10/10</span></span>
<span><span class="co">## Harmony converged after 10 iterations</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce2</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 51911 features across 11914 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (51911 features, 2000 variable features)</span></span>
<span><span class="co">##  33 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16, data.1, data.2, data.3, data.4, data.5, data.6, data.7, data.8, data.9, data.10, data.11, data.12, data.13, data.14, data.15, data.16, scale.data</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, umap, harmony</span></span>
<span><span class="co"># Same workflow</span></span>
<span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="va">sce2</span>  <span class="op">|&gt;</span>  </span>
<span>  <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>  <span class="op">|&gt;</span>  </span>
<span>  <span class="fu">FindNeighbors</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>  <span class="op">|&gt;</span>  </span>
<span>  <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>  <span class="op">|&gt;</span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/identity.html">identity</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## 15:19:42 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">## 15:19:42 Read 11914 rows and found 30 numeric columns</span></span>
<span><span class="co">## 15:19:42 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">## 15:19:42 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">## 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">## [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">## **************************************************|</span></span>
<span><span class="co">## 15:19:42 Writing NN index file to temp file /var/folders/2c/9q3pg2295195bp3gnrgbzrg40000gn/T//RtmpjzstFL/file2afb17134535</span></span>
<span><span class="co">## 15:19:42 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">## 15:19:44 Annoy recall = 100%</span></span>
<span><span class="co">## 15:19:45 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">## 15:19:45 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">## 15:19:46 Commencing optimization for 200 epochs, with 516614 positive edges</span></span>
<span><span class="co">## 15:19:50 Optimization finished</span></span>
<span><span class="co">## Computing nearest neighbor graph</span></span>
<span><span class="co">## Computing SNN</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 11914</span></span>
<span><span class="co">## Number of edges: 465972</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.9043</span></span>
<span><span class="co">## Number of communities: 14</span></span>
<span><span class="co">## Elapsed time: 1 seconds</span></span>
<span></span>
<span><span class="va">p11</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p22</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"sample"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p33</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,group.by <span class="op">=</span> <span class="st">"group"</span>, pt.size<span class="op">=</span><span class="fl">0.5</span>, label <span class="op">=</span> <span class="cn">F</span>,repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p11</span> <span class="op">+</span> <span class="va">p22</span> <span class="op">+</span><span class="va">p33</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="visualize-marker-genes" class="level2"><h2 class="anchored" data-anchor-id="visualize-marker-genes">Visualize marker genes</h2>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Marker</span>  <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  Epi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"EPCAM"</span><span class="op">)</span>,</span>
<span>  Endo <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PECAM1"</span>,<span class="st">"PLVAP"</span><span class="op">)</span>,</span>
<span>  Fibroblast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"COL3A1"</span>,<span class="st">"COL1A1"</span>,<span class="st">"COL1A2"</span><span class="op">)</span>,</span>
<span>  IM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PTPRC"</span><span class="op">)</span>,</span>
<span>  B <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD19"</span><span class="op">)</span>,</span>
<span>  T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD8A"</span>,<span class="st">"CD4"</span><span class="op">)</span>,</span>
<span>  Myeloid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"C1QA"</span>,<span class="st">"C1QB"</span>,<span class="st">"CD163"</span>,<span class="st">"CD1C"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">Marker2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"EPCAM"</span>,</span>
<span>    <span class="st">"PECAM1"</span>,<span class="st">"PLVAP"</span>,</span>
<span>    <span class="st">"COL3A1"</span>,<span class="st">"COL1A1"</span>,<span class="st">"COL1A2"</span>,</span>
<span>    <span class="st">'PTPRC'</span>,</span>
<span>    <span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD19"</span>,</span>
<span>    <span class="st">"CD3D"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD8A"</span>,<span class="st">"CD4"</span>,</span>
<span>    <span class="st">"C1QA"</span>,<span class="st">"C1QB"</span>,<span class="st">"CD163"</span>,<span class="st">"CD1C"</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">DotPlot</span><span class="op">(</span><span class="va">sce2</span>, features <span class="op">=</span> <span class="va">Marker2</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.5"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">sce2</span>, features <span class="op">=</span> <span class="va">Marker2</span>, pt.size <span class="op">=</span> <span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>    <span class="va">sce2</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"EPCAM"</span>, <span class="st">"PECAM1"</span>, <span class="st">"COL3A1"</span>, <span class="st">'PTPRC'</span>,</span>
<span>        <span class="st">"CD79A"</span>, <span class="st">"CD79B"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD3E"</span>, <span class="st">"C1QA"</span>, <span class="st">"C1QB"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="annotate-clusters" class="level2"><h2 class="anchored" data-anchor-id="annotate-clusters">Annotate clusters</h2>
<section id="using-vector" class="level3"><h3 class="anchored" data-anchor-id="using-vector">Using Vector</h3>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Myeloid'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'T'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'Endo'</span>,<span class="st">'un'</span>,<span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'un'</span>,<span class="st">'Fibroblast'</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu">RenameIdents</span><span class="op">(</span><span class="va">sce2</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add to metadata, for </span></span>
<span><span class="va">sce2</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu">Idents</span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>      <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="directly-assign" class="level3"><h3 class="anchored" data-anchor-id="directly-assign">Directly assign</h3>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">Idents</span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"seurat_clusters"</span></span>
<span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu">RenameIdents</span><span class="op">(</span><span class="va">sce2</span>,</span>
<span>    <span class="st">"0"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"1"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Myeloid"</span>,</span>
<span>    <span class="st">"3"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>    <span class="st">"4"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"5"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"6"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>    <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"8"</span> <span class="op">=</span> <span class="st">"T"</span>,</span>
<span>    <span class="st">"9"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"10"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>    <span class="st">"11"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"12"</span> <span class="op">=</span> <span class="st">"Endo"</span>,</span>
<span>    <span class="st">"13"</span> <span class="op">=</span> <span class="st">"un"</span>,</span>
<span>    <span class="st">"14"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"15"</span> <span class="op">=</span> <span class="st">"Epi"</span>,</span>
<span>    <span class="st">"16"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span>,</span>
<span>    <span class="st">"17"</span> <span class="op">=</span> <span class="st">"un"</span>,</span>
<span>    <span class="st">"18"</span> <span class="op">=</span> <span class="st">"Fibroblast"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce2</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu">Idents</span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="add-annotation-to-meta-data" class="level3"><h3 class="anchored" data-anchor-id="add-annotation-to-meta-data">Add annotation to meta data</h3>
<div class="cell" data-layout-align="center">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce2</span><span class="op">$</span><span class="va">Anno</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Myeloid'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'T'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'Endo'</span>, <span class="st">'un'</span>, <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'un'</span>, <span class="st">'Fibroblast'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note:cluster start from 0</span></span>
<span><span class="co"># For loop to add</span></span>
<span><span class="va">sub_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">0</span><span class="op">:</span><span class="va">sub_length</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sce2</span><span class="op">$</span><span class="va">Anno</span><span class="op">[</span><span class="va">sce2</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span> <span class="op">=</span> <span class="va">celltype</span><span class="op">[</span><span class="va">i</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span></span>
<span>    <span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">'Anno'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span></span>
<span>  <span class="va">sce2</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>,label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">##                         orig.ident nCount_RNA nFeature_RNA percent.mt</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1     K16733       2464          965 12.6623377</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1     K16733       7145         1919  2.4492652</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1     K16733       1655          621  2.1752266</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1     K16733      14272         2771  1.5414798</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1     K16733      13832         2541  0.3181029</span></span>
<span><span class="co">## K16733_AAATTCGAGACGAG-1     K16733       4207         1183  0.1188495</span></span>
<span><span class="co">##                         percent.rp percent.HB RNA_snn_res.0.5 seurat_clusters</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1   13.35227          0              11              11</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1   36.72498          0               5               5</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1   35.52870          0               2               2</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1   38.81026          0               2               2</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1   40.73164          0               2               2</span></span>
<span><span class="co">## K16733_AAATTCGAGACGAG-1   42.33420          0               6               6</span></span>
<span><span class="co">##                         sample group new.cluster.ids   celltype       Anno</span></span>
<span><span class="co">## K16733_AAACATACTCGTTT-1    P01    PT             Epi        Epi        Epi</span></span>
<span><span class="co">## K16733_AAAGCAGAACGTTG-1    P01    PT             Epi        Epi        Epi</span></span>
<span><span class="co">## K16733_AAAGCAGACTGAGT-1    P01    PT         Myeloid    Myeloid    Myeloid</span></span>
<span><span class="co">## K16733_AAAGGCCTGCTCCT-1    P01    PT         Myeloid    Myeloid    Myeloid</span></span>
<span><span class="co">## K16733_AAATACTGTGGATC-1    P01    PT         Myeloid    Myeloid    Myeloid</span></span>
<span><span class="co">## K16733_AAATTCGAGACGAG-1    P01    PT      Fibroblast Fibroblast Fibroblast</span></span>
<span></span>
<span><span class="co"># save(sce2, here("learn", "2023_scRNA", "sce.anno.RData"))</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>


<!-- -->

</section></section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/grgrzhong\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Practise Seurat for single cell data analysis"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> 2023-11-27</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="an">date-modified:</span><span class="co"> last-modified</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">  - seurat</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="an">image:</span><span class="co"> seurat_banner.jpg</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="an">description:</span><span class="co"> Practise scRNA data analysis with Genome Medicine using Seurat package.</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co">  freeze: true</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">#   # echo: false</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">  warning: false</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co">#   eval: false</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include=FALSE}</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fig.width = 6, </span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fig.height = 3.8,</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">fig.align =</span> <span class="st">"center"</span>, </span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fig.retina = 3,</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">out.width =</span> <span class="st">"100%"</span>, </span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">warning =</span> <span class="cn">FALSE</span>,</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># evaluate = FALSE,</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">collapse =</span> <span class="cn">TRUE</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a><span class="fu">##  Packages</span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(harmony)</span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(httpgd)</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a><span class="fu">## Retrieve data</span></span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The source data is from <span class="co">[</span><span class="ot">Single-cell transcriptome analysis of tumor and stromal compartments of pancreatic ductal adenocarcinoma primary tumors and metastatic lesions</span><span class="co">](https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00776-9)</span></span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Down load data GSE154778_RAW.tar from <span class="co">[</span><span class="ot">here</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154778)</span> to data and unzip it.</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Preprocess sequencing data</span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a><span class="al">![](practise.png)</span></span>
<span id="cb21-51"><a href="#cb21-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-52"><a href="#cb21-52" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-53"><a href="#cb21-53" aria-hidden="true" tabindex="-1"></a><span class="co">#| </span></span>
<span id="cb21-54"><a href="#cb21-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Change working directory</span></span>
<span id="cb21-55"><a href="#cb21-55" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"./projects/2023_scRNA_Seurat/data/GSE154778_RAW/"</span>)</span>
<span id="cb21-56"><a href="#cb21-56" aria-hidden="true" tabindex="-1"></a><span class="co"># getwd()</span></span>
<span id="cb21-57"><a href="#cb21-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Check files</span></span>
<span id="cb21-58"><a href="#cb21-58" aria-hidden="true" tabindex="-1"></a>fs <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"./"</span>, <span class="st">"^GSM"</span>)</span>
<span id="cb21-59"><a href="#cb21-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the sample names</span></span>
<span id="cb21-60"><a href="#cb21-60" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">str_split</span>(fs, <span class="st">"_"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>)[, <span class="dv">2</span>]</span>
<span id="cb21-61"><a href="#cb21-61" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(samples)</span>
<span id="cb21-62"><a href="#cb21-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Create folders for each sample, and rename</span></span>
<span id="cb21-63"><a href="#cb21-63" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(</span>
<span id="cb21-64"><a href="#cb21-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>(samples), <span class="cf">function</span>(x) {</span>
<span id="cb21-65"><a href="#cb21-65" aria-hidden="true" tabindex="-1"></a>    y  <span class="ot">&lt;-</span> fs[<span class="fu">grepl</span>(x, fs)]</span>
<span id="cb21-66"><a href="#cb21-66" aria-hidden="true" tabindex="-1"></a>    folder <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">str_split</span>(y[<span class="dv">1</span>], <span class="st">"_"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>)[, <span class="dv">2</span>], <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb21-67"><a href="#cb21-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-68"><a href="#cb21-68" aria-hidden="true" tabindex="-1"></a>    <span class="fu">file.rename</span>(y[<span class="dv">1</span>], <span class="fu">file.path</span>(folder, <span class="st">"barcodes.tsv.gz"</span>))</span>
<span id="cb21-69"><a href="#cb21-69" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note the seurat version to check features.tsv.gz or genes.tsv.gz</span></span>
<span id="cb21-70"><a href="#cb21-70" aria-hidden="true" tabindex="-1"></a>    <span class="fu">file.rename</span>(y[<span class="dv">2</span>], <span class="fu">file.path</span>(folder, <span class="st">"features.tsv.gz"</span>)) </span>
<span id="cb21-71"><a href="#cb21-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">file.rename</span>(y[<span class="dv">3</span>], <span class="fu">file.path</span>(folder, <span class="st">"matrix.mtx.gz"</span>))</span>
<span id="cb21-72"><a href="#cb21-72" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-73"><a href="#cb21-73" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-74"><a href="#cb21-74" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-75"><a href="#cb21-75" aria-hidden="true" tabindex="-1"></a><span class="fu">## Load batch data</span></span>
<span id="cb21-76"><a href="#cb21-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-79"><a href="#cb21-79" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-80"><a href="#cb21-80" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-81"><a href="#cb21-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Change working directory</span></span>
<span id="cb21-82"><a href="#cb21-82" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"./projects/2023_scRNA_Seurat/data/GSE154778_RAW/"</span>)</span>
<span id="cb21-83"><a href="#cb21-83" aria-hidden="true" tabindex="-1"></a>folders <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"./"</span>)</span>
<span id="cb21-84"><a href="#cb21-84" aria-hidden="true" tabindex="-1"></a>folders</span>
<span id="cb21-85"><a href="#cb21-85" aria-hidden="true" tabindex="-1"></a>sceList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb21-86"><a href="#cb21-86" aria-hidden="true" tabindex="-1"></a>  folders, <span class="cf">function</span>(folder) {</span>
<span id="cb21-87"><a href="#cb21-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">Read10X</span>(folder), <span class="at">project =</span> folder)</span>
<span id="cb21-88"><a href="#cb21-88" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-89"><a href="#cb21-89" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-90"><a href="#cb21-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-91"><a href="#cb21-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## Merage samples data</span></span>
<span id="cb21-92"><a href="#cb21-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-93"><a href="#cb21-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-94"><a href="#cb21-94" aria-hidden="true" tabindex="-1"></a><span class="fu">### Directly merge</span></span>
<span id="cb21-95"><a href="#cb21-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-98"><a href="#cb21-98" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-99"><a href="#cb21-99" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-100"><a href="#cb21-100" aria-hidden="true" tabindex="-1"></a><span class="co"># Use Seurat merge</span></span>
<span id="cb21-101"><a href="#cb21-101" aria-hidden="true" tabindex="-1"></a>sce.all <span class="ot">&lt;-</span> <span class="fu">merge</span>(</span>
<span id="cb21-102"><a href="#cb21-102" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> sceList[[<span class="dv">1</span>]],</span>
<span id="cb21-103"><a href="#cb21-103" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(sceList[[<span class="dv">2</span>]], sceList[[<span class="dv">3</span>]], sceList[[<span class="dv">4</span>]], sceList[[<span class="dv">5</span>]], sceList[[<span class="dv">6</span>]], </span>
<span id="cb21-104"><a href="#cb21-104" aria-hidden="true" tabindex="-1"></a>  sceList[[<span class="dv">7</span>]], sceList[[<span class="dv">8</span>]], sceList[[<span class="dv">9</span>]], sceList[[<span class="dv">10</span>]], sceList[[<span class="dv">11</span>]], </span>
<span id="cb21-105"><a href="#cb21-105" aria-hidden="true" tabindex="-1"></a>  sceList[[<span class="dv">12</span>]], sceList[[<span class="dv">13</span>]], sceList[[<span class="dv">14</span>]], sceList[[<span class="dv">5</span>]], sceList[[<span class="dv">16</span>]]),</span>
<span id="cb21-106"><a href="#cb21-106" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Sample names</span></span>
<span id="cb21-107"><a href="#cb21-107" aria-hidden="true" tabindex="-1"></a>  <span class="at">add.cell.ids =</span> folders,  </span>
<span id="cb21-108"><a href="#cb21-108" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"scRNA"</span></span>
<span id="cb21-109"><a href="#cb21-109" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-110"><a href="#cb21-110" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sce.all, <span class="fu">here</span>(<span class="st">"projects"</span>, <span class="st">"2023_scRNA_Seurat"</span>, <span class="st">"sce.all.rds"</span>))</span>
<span id="cb21-111"><a href="#cb21-111" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-112"><a href="#cb21-112" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Filter</span></span>
<span id="cb21-113"><a href="#cb21-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-116"><a href="#cb21-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-117"><a href="#cb21-117" aria-hidden="true" tabindex="-1"></a>sce.all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">"projects"</span>, <span class="st">"2023_scRNA_Seurat"</span>, <span class="st">"sce.all.rds"</span>))</span>
<span id="cb21-118"><a href="#cb21-118" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sce.all<span class="sc">@</span>meta.data)</span>
<span id="cb21-119"><a href="#cb21-119" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce.all<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident)</span>
<span id="cb21-120"><a href="#cb21-120" aria-hidden="true" tabindex="-1"></a><span class="co"># Mitochandrial genes</span></span>
<span id="cb21-121"><a href="#cb21-121" aria-hidden="true" tabindex="-1"></a>sce.all[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sce.all, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb21-122"><a href="#cb21-122" aria-hidden="true" tabindex="-1"></a><span class="co"># Ribonucleoprotein </span></span>
<span id="cb21-123"><a href="#cb21-123" aria-hidden="true" tabindex="-1"></a>sce.all[[<span class="st">"percent.rp"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(sce.all, <span class="at">pattern =</span> <span class="st">"^RP"</span>)</span>
<span id="cb21-124"><a href="#cb21-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-125"><a href="#cb21-125" aria-hidden="true" tabindex="-1"></a><span class="co"># Specific gene set</span></span>
<span id="cb21-126"><a href="#cb21-126" aria-hidden="true" tabindex="-1"></a>HB.genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"HBA1"</span>,<span class="st">"HBA2"</span>,<span class="st">"HBB"</span>,<span class="st">"HBD"</span>,<span class="st">"HBE1"</span>,<span class="st">"HBG1"</span>,<span class="st">"HBG2"</span>,<span class="st">"HBM"</span>,<span class="st">"HBQ1"</span>,<span class="st">"HBZ"</span>) </span>
<span id="cb21-127"><a href="#cb21-127" aria-hidden="true" tabindex="-1"></a><span class="co"># Red blood cells genes</span></span>
<span id="cb21-128"><a href="#cb21-128" aria-hidden="true" tabindex="-1"></a>sce.all[[<span class="st">"percent.HB"</span>]]<span class="ot">&lt;-</span><span class="fu">PercentageFeatureSet</span>(sce.all, <span class="at">features =</span> HB.genes)</span>
<span id="cb21-129"><a href="#cb21-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-130"><a href="#cb21-130" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sce.all<span class="sc">@</span>meta.data)</span>
<span id="cb21-131"><a href="#cb21-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-132"><a href="#cb21-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-135"><a href="#cb21-135" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-136"><a href="#cb21-136" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb21-137"><a href="#cb21-137" aria-hidden="true" tabindex="-1"></a>    sce.all, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">500</span> <span class="sc">&amp;</span></span>
<span id="cb21-138"><a href="#cb21-138" aria-hidden="true" tabindex="-1"></a>        nFeature_RNA <span class="sc">&lt;</span> <span class="dv">5000</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">30</span></span>
<span id="cb21-139"><a href="#cb21-139" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-140"><a href="#cb21-140" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sce, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>), <span class="at">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb21-141"><a href="#cb21-141" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-142"><a href="#cb21-142" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Seurat workflow</span></span>
<span id="cb21-143"><a href="#cb21-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-146"><a href="#cb21-146" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-147"><a href="#cb21-147" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sce)</span>
<span id="cb21-148"><a href="#cb21-148" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(sce, <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb21-149"><a href="#cb21-149" aria-hidden="true" tabindex="-1"></a>all.genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce)</span>
<span id="cb21-150"><a href="#cb21-150" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(sce, <span class="at">features =</span> all.genes)</span>
<span id="cb21-151"><a href="#cb21-151" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(sce, <span class="at">npcs =</span> <span class="dv">50</span>)</span>
<span id="cb21-152"><a href="#cb21-152" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sce, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb21-153"><a href="#cb21-153" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sce, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb21-154"><a href="#cb21-154" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sce, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb21-155"><a href="#cb21-155" aria-hidden="true" tabindex="-1"></a><span class="co"># sce &lt;- RunTSNE(sce, dims = 1:30)</span></span>
<span id="cb21-156"><a href="#cb21-156" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-157"><a href="#cb21-157" aria-hidden="true" tabindex="-1"></a><span class="fu">#### UMAP</span></span>
<span id="cb21-158"><a href="#cb21-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-161"><a href="#cb21-161" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-162"><a href="#cb21-162" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename sample</span></span>
<span id="cb21-163"><a href="#cb21-163" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"K16733"</span>] <span class="ot">&lt;-</span> <span class="st">"P01"</span></span>
<span id="cb21-164"><a href="#cb21-164" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00006"</span>] <span class="ot">&lt;-</span> <span class="st">"P02"</span></span>
<span id="cb21-165"><a href="#cb21-165" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T2"</span>] <span class="ot">&lt;-</span> <span class="st">"P03"</span></span>
<span id="cb21-166"><a href="#cb21-166" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T3"</span>] <span class="ot">&lt;-</span> <span class="st">"P04"</span></span>
<span id="cb21-167"><a href="#cb21-167" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T4"</span>] <span class="ot">&lt;-</span> <span class="st">"P05"</span></span>
<span id="cb21-168"><a href="#cb21-168" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T5"</span>] <span class="ot">&lt;-</span> <span class="st">"P06"</span></span>
<span id="cb21-169"><a href="#cb21-169" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T6"</span>] <span class="ot">&lt;-</span> <span class="st">"P07"</span></span>
<span id="cb21-170"><a href="#cb21-170" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T8"</span>] <span class="ot">&lt;-</span> <span class="st">"P08"</span></span>
<span id="cb21-171"><a href="#cb21-171" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T9"</span>] <span class="ot">&lt;-</span> <span class="st">"P09"</span></span>
<span id="cb21-172"><a href="#cb21-172" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"T10"</span>] <span class="ot">&lt;-</span> <span class="st">"P10"</span></span>
<span id="cb21-173"><a href="#cb21-173" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00008"</span>] <span class="ot">&lt;-</span> <span class="st">"MET01"</span></span>
<span id="cb21-174"><a href="#cb21-174" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00013"</span>] <span class="ot">&lt;-</span> <span class="st">"MET02"</span></span>
<span id="cb21-175"><a href="#cb21-175" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00014"</span>] <span class="ot">&lt;-</span> <span class="st">"MET03"</span></span>
<span id="cb21-176"><a href="#cb21-176" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00016"</span>] <span class="ot">&lt;-</span> <span class="st">"MET04"</span></span>
<span id="cb21-177"><a href="#cb21-177" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00019"</span>] <span class="ot">&lt;-</span> <span class="st">"MET05"</span></span>
<span id="cb21-178"><a href="#cb21-178" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample[sce<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident <span class="sc">==</span> <span class="st">"Y00027"</span>] <span class="ot">&lt;-</span> <span class="st">"MET06"</span></span>
<span id="cb21-179"><a href="#cb21-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-180"><a href="#cb21-180" aria-hidden="true" tabindex="-1"></a><span class="co"># Add group information</span></span>
<span id="cb21-181"><a href="#cb21-181" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>meta.data<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">ifelse</span>( <span class="fu">grepl</span>(<span class="st">"MET"</span>,sce<span class="sc">@</span>meta.data<span class="sc">$</span>sample ) ,<span class="st">"MET"</span> ,<span class="st">"PT"</span> )</span>
<span id="cb21-182"><a href="#cb21-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-183"><a href="#cb21-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-186"><a href="#cb21-186" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-187"><a href="#cb21-187" aria-hidden="true" tabindex="-1"></a><span class="co"># Global</span></span>
<span id="cb21-188"><a href="#cb21-188" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb21-189"><a href="#cb21-189" aria-hidden="true" tabindex="-1"></a>  sce, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,<span class="at">repel =</span> <span class="cn">TRUE</span></span>
<span id="cb21-190"><a href="#cb21-190" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-191"><a href="#cb21-191" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample</span></span>
<span id="cb21-192"><a href="#cb21-192" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb21-193"><a href="#cb21-193" aria-hidden="true" tabindex="-1"></a>  sce, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"sample"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,</span>
<span id="cb21-194"><a href="#cb21-194" aria-hidden="true" tabindex="-1"></a>  <span class="at">repel =</span> <span class="cn">TRUE</span></span>
<span id="cb21-195"><a href="#cb21-195" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-196"><a href="#cb21-196" aria-hidden="true" tabindex="-1"></a><span class="co"># group</span></span>
<span id="cb21-197"><a href="#cb21-197" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb21-198"><a href="#cb21-198" aria-hidden="true" tabindex="-1"></a>  sce, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"group"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,</span>
<span id="cb21-199"><a href="#cb21-199" aria-hidden="true" tabindex="-1"></a>  <span class="at">repel =</span> <span class="cn">TRUE</span></span>
<span id="cb21-200"><a href="#cb21-200" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-201"><a href="#cb21-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-202"><a href="#cb21-202" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span>p3</span>
<span id="cb21-203"><a href="#cb21-203" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-204"><a href="#cb21-204" aria-hidden="true" tabindex="-1"></a><span class="fu">### Harmony to remove batch effect</span></span>
<span id="cb21-205"><a href="#cb21-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-206"><a href="#cb21-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-209"><a href="#cb21-209" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-210"><a href="#cb21-210" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> sce  <span class="sc">|&gt;</span>  <span class="fu">RunHarmony</span>(<span class="st">"sample"</span>, <span class="at">plot_convergence =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-211"><a href="#cb21-211" aria-hidden="true" tabindex="-1"></a>sce2</span>
<span id="cb21-212"><a href="#cb21-212" aria-hidden="true" tabindex="-1"></a><span class="co"># Same workflow</span></span>
<span id="cb21-213"><a href="#cb21-213" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> sce2  <span class="sc">|&gt;</span>  </span>
<span id="cb21-214"><a href="#cb21-214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)  <span class="sc">|&gt;</span>  </span>
<span id="cb21-215"><a href="#cb21-215" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindNeighbors</span>(<span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)  <span class="sc">|&gt;</span>  </span>
<span id="cb21-216"><a href="#cb21-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindClusters</span>(<span class="at">resolution =</span> <span class="fl">0.5</span>)  <span class="sc">|&gt;</span>  </span>
<span id="cb21-217"><a href="#cb21-217" aria-hidden="true" tabindex="-1"></a>  <span class="fu">identity</span>()</span>
<span id="cb21-218"><a href="#cb21-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-219"><a href="#cb21-219" aria-hidden="true" tabindex="-1"></a>p11 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,<span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-220"><a href="#cb21-220" aria-hidden="true" tabindex="-1"></a>p22 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"sample"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,<span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-221"><a href="#cb21-221" aria-hidden="true" tabindex="-1"></a>p33 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">group.by =</span> <span class="st">"group"</span>, <span class="at">pt.size=</span><span class="fl">0.5</span>, <span class="at">label =</span> F,<span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-222"><a href="#cb21-222" aria-hidden="true" tabindex="-1"></a>p11 <span class="sc">+</span> p22 <span class="sc">+</span>p33</span>
<span id="cb21-223"><a href="#cb21-223" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-224"><a href="#cb21-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-225"><a href="#cb21-225" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualize marker genes</span></span>
<span id="cb21-226"><a href="#cb21-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-229"><a href="#cb21-229" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-230"><a href="#cb21-230" aria-hidden="true" tabindex="-1"></a>Marker  <span class="ot">&lt;-</span>  <span class="fu">list</span>(</span>
<span id="cb21-231"><a href="#cb21-231" aria-hidden="true" tabindex="-1"></a>  <span class="at">Epi =</span> <span class="fu">c</span>(<span class="st">"EPCAM"</span>),</span>
<span id="cb21-232"><a href="#cb21-232" aria-hidden="true" tabindex="-1"></a>  <span class="at">Endo =</span> <span class="fu">c</span>(<span class="st">"PECAM1"</span>,<span class="st">"PLVAP"</span>),</span>
<span id="cb21-233"><a href="#cb21-233" aria-hidden="true" tabindex="-1"></a>  <span class="at">Fibroblast =</span> <span class="fu">c</span>(<span class="st">"COL3A1"</span>,<span class="st">"COL1A1"</span>,<span class="st">"COL1A2"</span>),</span>
<span id="cb21-234"><a href="#cb21-234" aria-hidden="true" tabindex="-1"></a>  <span class="at">IM =</span> <span class="fu">c</span>(<span class="st">"PTPRC"</span>),</span>
<span id="cb21-235"><a href="#cb21-235" aria-hidden="true" tabindex="-1"></a>  <span class="at">B =</span> <span class="fu">c</span>(<span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD19"</span>),</span>
<span id="cb21-236"><a href="#cb21-236" aria-hidden="true" tabindex="-1"></a>  <span class="at">T =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD8A"</span>,<span class="st">"CD4"</span>),</span>
<span id="cb21-237"><a href="#cb21-237" aria-hidden="true" tabindex="-1"></a>  <span class="at">Myeloid =</span> <span class="fu">c</span>(<span class="st">"C1QA"</span>,<span class="st">"C1QB"</span>,<span class="st">"CD163"</span>,<span class="st">"CD1C"</span>)</span>
<span id="cb21-238"><a href="#cb21-238" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-239"><a href="#cb21-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-240"><a href="#cb21-240" aria-hidden="true" tabindex="-1"></a>Marker2 <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb21-241"><a href="#cb21-241" aria-hidden="true" tabindex="-1"></a>    <span class="st">"EPCAM"</span>,</span>
<span id="cb21-242"><a href="#cb21-242" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PECAM1"</span>,<span class="st">"PLVAP"</span>,</span>
<span id="cb21-243"><a href="#cb21-243" aria-hidden="true" tabindex="-1"></a>    <span class="st">"COL3A1"</span>,<span class="st">"COL1A1"</span>,<span class="st">"COL1A2"</span>,</span>
<span id="cb21-244"><a href="#cb21-244" aria-hidden="true" tabindex="-1"></a>    <span class="st">'PTPRC'</span>,</span>
<span id="cb21-245"><a href="#cb21-245" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD79A"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD19"</span>,</span>
<span id="cb21-246"><a href="#cb21-246" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD3D"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD8A"</span>,<span class="st">"CD4"</span>,</span>
<span id="cb21-247"><a href="#cb21-247" aria-hidden="true" tabindex="-1"></a>    <span class="st">"C1QA"</span>,<span class="st">"C1QB"</span>,<span class="st">"CD163"</span>,<span class="st">"CD1C"</span></span>
<span id="cb21-248"><a href="#cb21-248" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-249"><a href="#cb21-249" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-250"><a href="#cb21-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-253"><a href="#cb21-253" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-254"><a href="#cb21-254" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(sce2, <span class="at">features =</span> Marker2, <span class="at">group.by =</span> <span class="st">"RNA_snn_res.0.5"</span>)</span>
<span id="cb21-255"><a href="#cb21-255" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-256"><a href="#cb21-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-259"><a href="#cb21-259" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-260"><a href="#cb21-260" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(sce2, <span class="at">features =</span> Marker2, <span class="at">pt.size =</span> <span class="dv">0</span>, <span class="at">ncol =</span> <span class="dv">5</span>)</span>
<span id="cb21-261"><a href="#cb21-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-262"><a href="#cb21-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-263"><a href="#cb21-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-266"><a href="#cb21-266" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-267"><a href="#cb21-267" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(</span>
<span id="cb21-268"><a href="#cb21-268" aria-hidden="true" tabindex="-1"></a>    sce2,</span>
<span id="cb21-269"><a href="#cb21-269" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">c</span>(</span>
<span id="cb21-270"><a href="#cb21-270" aria-hidden="true" tabindex="-1"></a>        <span class="st">"EPCAM"</span>, <span class="st">"PECAM1"</span>, <span class="st">"COL3A1"</span>, <span class="st">'PTPRC'</span>,</span>
<span id="cb21-271"><a href="#cb21-271" aria-hidden="true" tabindex="-1"></a>        <span class="st">"CD79A"</span>, <span class="st">"CD79B"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD3E"</span>, <span class="st">"C1QA"</span>, <span class="st">"C1QB"</span></span>
<span id="cb21-272"><a href="#cb21-272" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-273"><a href="#cb21-273" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-274"><a href="#cb21-274" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-275"><a href="#cb21-275" aria-hidden="true" tabindex="-1"></a><span class="fu">## Annotate clusters</span></span>
<span id="cb21-276"><a href="#cb21-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-277"><a href="#cb21-277" aria-hidden="true" tabindex="-1"></a><span class="fu">### Using Vector</span></span>
<span id="cb21-278"><a href="#cb21-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-281"><a href="#cb21-281" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-282"><a href="#cb21-282" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb21-283"><a href="#cb21-283" aria-hidden="true" tabindex="-1"></a>  <span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Myeloid'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'T'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'Epi'</span>,<span class="st">'Endo'</span>,<span class="st">'un'</span>,<span class="st">'Epi'</span>,<span class="st">'Epi'</span>,<span class="st">'Fibroblast'</span>,<span class="st">'un'</span>,<span class="st">'Fibroblast'</span></span>
<span id="cb21-284"><a href="#cb21-284" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-285"><a href="#cb21-285" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new.cluster.ids) <span class="ot">&lt;-</span> <span class="fu">levels</span>(sce2)</span>
<span id="cb21-286"><a href="#cb21-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-287"><a href="#cb21-287" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(sce2, new.cluster.ids)</span>
<span id="cb21-288"><a href="#cb21-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-289"><a href="#cb21-289" aria-hidden="true" tabindex="-1"></a><span class="co"># Add to metadata, for </span></span>
<span id="cb21-290"><a href="#cb21-290" aria-hidden="true" tabindex="-1"></a>sce2<span class="sc">@</span>meta.data<span class="sc">$</span>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">Idents</span>(sce2)</span>
<span id="cb21-291"><a href="#cb21-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-292"><a href="#cb21-292" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span>      <span class="fu">NoLegend</span>()</span>
<span id="cb21-293"><a href="#cb21-293" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-294"><a href="#cb21-294" aria-hidden="true" tabindex="-1"></a><span class="fu">### Directly assign</span></span>
<span id="cb21-295"><a href="#cb21-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-298"><a href="#cb21-298" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-299"><a href="#cb21-299" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(sce2) <span class="ot">&lt;-</span> <span class="st">"seurat_clusters"</span></span>
<span id="cb21-300"><a href="#cb21-300" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(sce2,</span>
<span id="cb21-301"><a href="#cb21-301" aria-hidden="true" tabindex="-1"></a>    <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-302"><a href="#cb21-302" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-303"><a href="#cb21-303" aria-hidden="true" tabindex="-1"></a>    <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Myeloid"</span>,</span>
<span id="cb21-304"><a href="#cb21-304" aria-hidden="true" tabindex="-1"></a>    <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"Fibroblast"</span>,</span>
<span id="cb21-305"><a href="#cb21-305" aria-hidden="true" tabindex="-1"></a>    <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-306"><a href="#cb21-306" aria-hidden="true" tabindex="-1"></a>    <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-307"><a href="#cb21-307" aria-hidden="true" tabindex="-1"></a>    <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"Fibroblast"</span>,</span>
<span id="cb21-308"><a href="#cb21-308" aria-hidden="true" tabindex="-1"></a>    <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-309"><a href="#cb21-309" aria-hidden="true" tabindex="-1"></a>    <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"T"</span>,</span>
<span id="cb21-310"><a href="#cb21-310" aria-hidden="true" tabindex="-1"></a>    <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-311"><a href="#cb21-311" aria-hidden="true" tabindex="-1"></a>    <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"Fibroblast"</span>,</span>
<span id="cb21-312"><a href="#cb21-312" aria-hidden="true" tabindex="-1"></a>    <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-313"><a href="#cb21-313" aria-hidden="true" tabindex="-1"></a>    <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"Endo"</span>,</span>
<span id="cb21-314"><a href="#cb21-314" aria-hidden="true" tabindex="-1"></a>    <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"un"</span>,</span>
<span id="cb21-315"><a href="#cb21-315" aria-hidden="true" tabindex="-1"></a>    <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-316"><a href="#cb21-316" aria-hidden="true" tabindex="-1"></a>    <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Epi"</span>,</span>
<span id="cb21-317"><a href="#cb21-317" aria-hidden="true" tabindex="-1"></a>    <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Fibroblast"</span>,</span>
<span id="cb21-318"><a href="#cb21-318" aria-hidden="true" tabindex="-1"></a>    <span class="st">"17"</span> <span class="ot">=</span> <span class="st">"un"</span>,</span>
<span id="cb21-319"><a href="#cb21-319" aria-hidden="true" tabindex="-1"></a>    <span class="st">"18"</span> <span class="ot">=</span> <span class="st">"Fibroblast"</span></span>
<span id="cb21-320"><a href="#cb21-320" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-321"><a href="#cb21-321" aria-hidden="true" tabindex="-1"></a>sce2<span class="sc">@</span>meta.data<span class="sc">$</span>celltype <span class="ot">&lt;-</span> <span class="fu">Idents</span>(sce2)</span>
<span id="cb21-322"><a href="#cb21-322" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb21-323"><a href="#cb21-323" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb21-324"><a href="#cb21-324" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-325"><a href="#cb21-325" aria-hidden="true" tabindex="-1"></a><span class="fu">### Add annotation to meta data</span></span>
<span id="cb21-326"><a href="#cb21-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-329"><a href="#cb21-329" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-330"><a href="#cb21-330" aria-hidden="true" tabindex="-1"></a>sce2<span class="sc">$</span>Anno <span class="ot">&lt;-</span> <span class="st">"NA"</span></span>
<span id="cb21-331"><a href="#cb21-331" aria-hidden="true" tabindex="-1"></a>celltype <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb21-332"><a href="#cb21-332" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Myeloid'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'T'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'Epi'</span>, <span class="st">'Endo'</span>, <span class="st">'un'</span>, <span class="st">'Epi'</span>, <span class="st">'Epi'</span>, <span class="st">'Fibroblast'</span>, <span class="st">'un'</span>, <span class="st">'Fibroblast'</span></span>
<span id="cb21-333"><a href="#cb21-333" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-334"><a href="#cb21-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-335"><a href="#cb21-335" aria-hidden="true" tabindex="-1"></a><span class="co"># Note:cluster start from 0</span></span>
<span id="cb21-336"><a href="#cb21-336" aria-hidden="true" tabindex="-1"></a><span class="co"># For loop to add</span></span>
<span id="cb21-337"><a href="#cb21-337" aria-hidden="true" tabindex="-1"></a>sub_length <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(sce2<span class="sc">$</span>seurat_clusters)) <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb21-338"><a href="#cb21-338" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">0</span><span class="sc">:</span>sub_length) {</span>
<span id="cb21-339"><a href="#cb21-339" aria-hidden="true" tabindex="-1"></a>    sce2<span class="sc">$</span>Anno[sce2<span class="sc">$</span>seurat_clusters <span class="sc">==</span> i] <span class="ot">=</span> celltype[i <span class="sc">+</span> <span class="dv">1</span>]</span>
<span id="cb21-340"><a href="#cb21-340" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-341"><a href="#cb21-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-342"><a href="#cb21-342" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP</span></span>
<span id="cb21-343"><a href="#cb21-343" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(</span>
<span id="cb21-344"><a href="#cb21-344" aria-hidden="true" tabindex="-1"></a>    sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">'Anno'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb21-345"><a href="#cb21-345" aria-hidden="true" tabindex="-1"></a>    <span class="fu">NoLegend</span>()</span>
<span id="cb21-346"><a href="#cb21-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-347"><a href="#cb21-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-348"><a href="#cb21-348" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(</span>
<span id="cb21-349"><a href="#cb21-349" aria-hidden="true" tabindex="-1"></a>  sce2, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span></span>
<span id="cb21-350"><a href="#cb21-350" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb21-351"><a href="#cb21-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb21-352"><a href="#cb21-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-353"><a href="#cb21-353" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sce2<span class="sc">@</span>meta.data)</span>
<span id="cb21-354"><a href="#cb21-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-355"><a href="#cb21-355" aria-hidden="true" tabindex="-1"></a><span class="co"># save(sce2, here("learn", "2023_scRNA", "sce.anno.RData"))</span></span>
<span id="cb21-356"><a href="#cb21-356" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2026 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a>, <a href="https://www.python.org/"><i class="fa-brands fa-python" aria-label="python"></i></a> and <a href="https://quarto.org/">Quarto</a></span></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="../../../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>