<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Guorui Zhong">
<meta name="description" content="Use lightweight and open-source software with R programming language and the Bioconductor suite of packages to perfrm RNA-seq data analysis.">
<title>General workflow for RNAseq data analysis with R â€“ Guorui Zhong</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting-ed09741bc9a4bc7923efc7a27dab3f48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap-baabc0bdab06a695b25d56c023bbe1ef.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
<script src="../../../../site_libs/htmlwidgets-1.6.2/htmlwidgets.js"></script><script src="../../../../site_libs/plotly-binding-4.10.3/plotly.js"></script><script src="../../../../site_libs/typedarray-0.1/typedarray.min.js"></script><script src="../../../../site_libs/jquery-3.5.1/jquery.min.js"></script><link href="../../../../site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="../../../../site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><link href="../../../../site_libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="../../../../site_libs/plotly-main-2.11.1/plotly-latest.min.js"></script>
</head>
<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../wiki/index.html"> 
<span class="menu-text">Wiki</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../programming/linux/index.html"> 
<span class="menu-text">Linux</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../programming/python/index.html"> 
<span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../programming/r/index.html"> 
<span class="menu-text">R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html"> 
<span class="menu-text">Keyboard</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://scholar.google.com/citations?user=g0MfTPMAAAAJ&amp;hl=en"> <i class="bi bi-journal-check" role="img" aria-label="journal-check">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/guorui-zhong-754761185/"> <i class="bi bi-linkedin" role="img" aria-label="linkedin">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/guoruizhong"> <i class="bi bi-github" role="img" aria-label="github">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">General workflow for RNAseq data analysis with R</h1>
                  <div>
        <div class="description">
          Use lightweight and open-source software with R programming language and the Bioconductor suite of packages to perfrm RNA-seq data analysis.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">r</div>
                <div class="quarto-category">rnaseq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://grgrzhong.github.io/">Guorui Zhong</a> <a href="https://orcid.org/0000-0002-1955-2847" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Monday, September 11, 2023</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Wednesday, January 10, 2024</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#step1-check-the-quality-of-raw-reads" id="toc-step1-check-the-quality-of-raw-reads" class="nav-link active" data-scroll-target="#step1-check-the-quality-of-raw-reads">Step1: Check the quality of raw reads</a></li>
  <li><a href="#step2-mapping-reads-with-kallisto-pseudoaligment" id="toc-step2-mapping-reads-with-kallisto-pseudoaligment" class="nav-link" data-scroll-target="#step2-mapping-reads-with-kallisto-pseudoaligment">Step2: Mapping reads with Kallisto pseudoaligment</a></li>
  <li>
<a href="#step3-import-data-into-r" id="toc-step3-import-data-into-r" class="nav-link" data-scroll-target="#step3-import-data-into-r">Step3: Import data into R</a>
  <ul>
<li><a href="#load-packages" id="toc-load-packages" class="nav-link" data-scroll-target="#load-packages">Load packages</a></li>
  <li><a href="#import-study-design-files" id="toc-import-study-design-files" class="nav-link" data-scroll-target="#import-study-design-files">Import study design files</a></li>
  <li><a href="#get-organsim-specific-annotations" id="toc-get-organsim-specific-annotations" class="nav-link" data-scroll-target="#get-organsim-specific-annotations">Get organsim specific annotations</a></li>
  <li><a href="#load-transcript-counts" id="toc-load-transcript-counts" class="nav-link" data-scroll-target="#load-transcript-counts">Load transcript counts</a></li>
  </ul>
</li>
  <li><a href="#step4-wrangling-gene-expression" id="toc-step4-wrangling-gene-expression" class="nav-link" data-scroll-target="#step4-wrangling-gene-expression">Step4: Wrangling gene expression</a></li>
  <li>
<a href="#step5-data-exploration-with-multivariate-analysis" id="toc-step5-data-exploration-with-multivariate-analysis" class="nav-link" data-scroll-target="#step5-data-exploration-with-multivariate-analysis">Step5: Data exploration with multivariate analysis</a>
  <ul>
<li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering">Hierarchical clustering</a></li>
  <li><a href="#principal-compoent-analysis" id="toc-principal-compoent-analysis" class="nav-link" data-scroll-target="#principal-compoent-analysis">Principal compoent analysis</a></li>
  </ul>
</li>
  <li>
<a href="#step6-differential-gene-expression" id="toc-step6-differential-gene-expression" class="nav-link" data-scroll-target="#step6-differential-gene-expression">Step6: Differential gene expression</a>
  <ul>
<li><a href="#set-up-design-matrix" id="toc-set-up-design-matrix" class="nav-link" data-scroll-target="#set-up-design-matrix">Set up design matrix</a></li>
  <li><a href="#model-mean-variance-trend-and-fit-linear-model-to-data" id="toc-model-mean-variance-trend-and-fit-linear-model-to-data" class="nav-link" data-scroll-target="#model-mean-variance-trend-and-fit-linear-model-to-data">Model mean-variance trend and fit linear model to data</a></li>
  </ul>
</li>
  <li><a href="#step7-module-identification" id="toc-step7-module-identification" class="nav-link" data-scroll-target="#step7-module-identification">Step7: Module identification</a></li>
  <li>
<a href="#step8-functional-enrichment-analysis" id="toc-step8-functional-enrichment-analysis" class="nav-link" data-scroll-target="#step8-functional-enrichment-analysis">Step8: Functional enrichment analysis</a>
  <ul>
<li><a href="#go-enrichment-analysis" id="toc-go-enrichment-analysis" class="nav-link" data-scroll-target="#go-enrichment-analysis">Go enrichment analysis</a></li>
  <li><a href="#perform-gsea-using-clusterprofiler" id="toc-perform-gsea-using-clusterprofiler" class="nav-link" data-scroll-target="#perform-gsea-using-clusterprofiler">Perform GSEA using clusterProfiler</a></li>
  <li><a href="#competitive-gsea-using-camera" id="toc-competitive-gsea-using-camera" class="nav-link" data-scroll-target="#competitive-gsea-using-camera">Competitive GSEA using CAMERA</a></li>
  <li><a href="#self-contained-gsea-using-roast" id="toc-self-contained-gsea-using-roast" class="nav-link" data-scroll-target="#self-contained-gsea-using-roast">Self-contained GSEA using ROAST</a></li>
  <li><a href="#single-sample-gsea-using-the-gsva-package" id="toc-single-sample-gsea-using-the-gsva-package" class="nav-link" data-scroll-target="#single-sample-gsea-using-the-gsva-package">Single sample GSEA using the GSVA package</a></li>
  </ul>
</li>
  <li><a href="#reference" id="toc-reference" class="nav-link" data-scroll-target="#reference">Reference</a></li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo">SessionInfo</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><p><img src="workflow.png" class="img-fluid"></p>
<section id="step1-check-the-quality-of-raw-reads" class="level2"><h2 class="anchored" data-anchor-id="step1-check-the-quality-of-raw-reads">Step1: Check the quality of raw reads</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">### go to row reads folder</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/../..DIY_Transcriptomics/data/fastq</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">### activate env</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> active rnaseq</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">### use fastqc to check the quality of fastq files</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="pp">*</span>.gz <span class="at">-t</span> 4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="step2-mapping-reads-with-kallisto-pseudoaligment" class="level2"><h2 class="anchored" data-anchor-id="step2-mapping-reads-with-kallisto-pseudoaligment">Step2: Mapping reads with Kallisto pseudoaligment</h2>
<ul>
<li><p>Download reference transciptome files from <a href="https://asia.ensembl.org/info/data/ftp/index.html">here</a></p></li>
<li><p>Build an index from reference fasta file</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> index <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index Homo_sapiens.GRCh38.cdna.all.fa.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Map reads to the indexed reference host transcriptome</li>
</ul>
<p><img src="map_single_end.png" class="img-fluid"></p>
<p><img src="map_paired_end.png" class="img-fluid"></p>
<ul>
<li>Summarize output with Multiqc in a single summary html</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">-d</span> .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">### first the healthy subjects (HS)</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> HS01 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668755.fastq.gz <span class="op">&amp;&gt;</span> HS01.log</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> HS02 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668756.fastq.gz <span class="op">&amp;&gt;</span> HS02.log</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> HS03 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668757.fastq.gz <span class="op">&amp;&gt;</span> HS03.log</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> HS04 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668758.fastq.gz <span class="op">&amp;&gt;</span> HS04.log</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> HS05 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668759.fastq.gz <span class="op">&amp;&gt;</span> HS05.log</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">### then the cutaneous leishmaniasis (CL) patients</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> CL08 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668769.fastq.gz <span class="op">&amp;&gt;</span> CL08.log</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> CL10 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668771.fastq.gz <span class="op">&amp;&gt;</span> CL10.log</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> CL11 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668772.fastq.gz <span class="op">&amp;&gt;</span> CL11.log</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> CL12 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668773.fastq.gz <span class="op">&amp;&gt;</span> CL12.log</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> Homo_sapiens.GRCh38.cdna.all.index <span class="at">-o</span> CL13 <span class="at">-t</span> 4 <span class="at">--single</span> <span class="at">-l</span> 250 <span class="at">-s</span> 30 SRR8668774.fastq.gz <span class="op">&amp;&gt;</span> CL13.log</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co">### Summarize outputs</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">-d</span> .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="step3-import-data-into-r" class="level2"><h2 class="anchored" data-anchor-id="step3-import-data-into-r">Step3: Import data into R</h2>
<section id="load-packages" class="level3"><h3 class="anchored" data-anchor-id="load-packages">Load packages</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rhdf5</span><span class="op">)</span> <span class="co"># provides functions for handling hdf5 file formats (kallisto outputs bootstraps in this format)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span> <span class="co"># Hadley Wickham's collection of R packages for data science, which we will use throughout the course</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tximport</span><span class="op">)</span> <span class="co"># package for getting Kallisto results into R</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ensembldb</span><span class="op">)</span> <span class="co"># helps deal with ensembl</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span> <span class="co"># replace with your organism-specific database package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span> <span class="co"># file path</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/milesmcbain/datapasta">datapasta</a></span><span class="op">)</span> <span class="co"># paste data into R from clipboard</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">edgeR</span><span class="op">)</span> <span class="co"># well known package for differential expression analysis, but we only use for the DGEList object and for normalization methods</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats">matrixStats</a></span><span class="op">)</span> <span class="co"># easily calculate stats on rows or columns of a data matrix</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span> <span class="co"># allows you to combine multiple plots in one figure</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT">DT</a></span><span class="op">)</span> <span class="co"># for making interactive tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span> <span class="co"># for making interactive plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gt.rstudio.com">gt</a></span><span class="op">)</span> <span class="co"># A layered 'grammar of tables' - think ggplot, but for tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">limma</span><span class="op">)</span> <span class="co"># venerable package for differential gene expression using linear modeling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">edgeR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span> <span class="co"># need colors to make heatmaps</span></span>
<span><span class="co"># library(gameofthrones) #because...why not.  Install using 'devtools::install_github("aljrico/gameofthrones")'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://talgalili.github.io/heatmaply/">heatmaply</a></span><span class="op">)</span> <span class="co">#for making interactive heatmaps using plotly</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots">gplots</a></span><span class="op">)</span></span>
<span><span class="co"># library(d3heatmap) # for making interactive heatmaps using D3</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span> <span class="co"># functions and methods for Gene Set Enrichment Analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span> <span class="co"># base functions for bioconductor; required by GSEABase</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GSVA</span><span class="op">)</span> <span class="co"># Gene Set Variation Analysis, a non-parametric and unsupervised method for estimating variation of gene set enrichment across samples.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gprofiler2</span><span class="op">)</span> <span class="co"># tools for accessing the GO enrichment results using g:Profiler web resources</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span> <span class="co"># provides a suite of tools for functional enrichment analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/">msigdbr</a></span><span class="op">)</span> <span class="co"># access to msigdb collections directly within R</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span> <span class="co"># great for making the standard GSEA enrichment plots</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="import-study-design-files" class="level3"><h3 class="anchored" data-anchor-id="import-study-design-files">Import study design files</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### load study design</span></span>
<span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span></span>
<span>    <span class="fu">here</span><span class="op">(</span><span class="st">"learn"</span>, <span class="st">"DIY_Transcriptomics"</span>, <span class="st">"config"</span>, <span class="st">"studydesign.txt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">targets</span></span>
<span><span class="co">## # A tibble: 10 Ã— 3</span></span>
<span><span class="co">##    sample sra_accession group  </span></span>
<span><span class="co">##    &lt;chr&gt;  &lt;chr&gt;         &lt;chr&gt;  </span></span>
<span><span class="co">##  1 HS01   SRR8668755    healthy</span></span>
<span><span class="co">##  2 HS02   SRR8668756    healthy</span></span>
<span><span class="co">##  3 HS03   SRR8668757    healthy</span></span>
<span><span class="co">##  4 HS04   SRR8668758    healthy</span></span>
<span><span class="co">##  5 HS05   SRR8668759    healthy</span></span>
<span><span class="co">##  6 CL08   SRR8668769    disease</span></span>
<span><span class="co">##  7 CL10   SRR8668771    disease</span></span>
<span><span class="co">##  8 CL11   SRR8668772    disease</span></span>
<span><span class="co">##  9 CL12   SRR8668773    disease</span></span>
<span><span class="co">## 10 CL13   SRR8668774    disease</span></span>
<span></span>
<span><span class="co">### create file paths for the abundance files</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span></span>
<span>    <span class="st">"learn"</span>, <span class="st">"DIY_Transcriptomics"</span>, <span class="st">"data"</span>, <span class="st">"fastq"</span>, <span class="va">targets</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"abundance.tsv"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">path</span></span>
<span><span class="co">##  [1] "learn/DIY_Transcriptomics/data/fastq/HS01/abundance.tsv"</span></span>
<span><span class="co">##  [2] "learn/DIY_Transcriptomics/data/fastq/HS02/abundance.tsv"</span></span>
<span><span class="co">##  [3] "learn/DIY_Transcriptomics/data/fastq/HS03/abundance.tsv"</span></span>
<span><span class="co">##  [4] "learn/DIY_Transcriptomics/data/fastq/HS04/abundance.tsv"</span></span>
<span><span class="co">##  [5] "learn/DIY_Transcriptomics/data/fastq/HS05/abundance.tsv"</span></span>
<span><span class="co">##  [6] "learn/DIY_Transcriptomics/data/fastq/CL08/abundance.tsv"</span></span>
<span><span class="co">##  [7] "learn/DIY_Transcriptomics/data/fastq/CL10/abundance.tsv"</span></span>
<span><span class="co">##  [8] "learn/DIY_Transcriptomics/data/fastq/CL11/abundance.tsv"</span></span>
<span><span class="co">##  [9] "learn/DIY_Transcriptomics/data/fastq/CL12/abundance.tsv"</span></span>
<span><span class="co">## [10] "learn/DIY_Transcriptomics/data/fastq/CL13/abundance.tsv"</span></span>
<span></span>
<span><span class="co">### make sure path is correct</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="get-organsim-specific-annotations" class="level3"><h3 class="anchored" data-anchor-id="get-organsim-specific-annotations">Get organsim specific annotations</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tx</span> <span class="op">&lt;-</span> <span class="fu">transcripts</span><span class="op">(</span></span>
<span>    <span class="va">EnsDb.Hsapiens.v86</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tx_id"</span>, <span class="st">"gene_name"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="co">### need to change first column name to 'target_id'</span></span>
<span>    <span class="fu">rename</span><span class="op">(</span>target_id <span class="op">=</span> <span class="va">tx_id</span><span class="op">)</span></span>
<span><span class="va">tx</span></span>
<span><span class="co">## # A tibble: 216,741 Ã— 7</span></span>
<span><span class="co">##    seqnames start   end width strand target_id       gene_name</span></span>
<span><span class="co">##    &lt;fct&gt;    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;fct&gt;  &lt;chr&gt;           &lt;chr&gt;    </span></span>
<span><span class="co">##  1 1        11869 14409  2541 +      ENST00000456328 DDX11L1  </span></span>
<span><span class="co">##  2 1        12010 13670  1661 +      ENST00000450305 DDX11L1  </span></span>
<span><span class="co">##  3 1        14404 29570 15167 -      ENST00000488147 WASH7P   </span></span>
<span><span class="co">##  4 1        17369 17436    68 -      ENST00000619216 MIR6859-1</span></span>
<span><span class="co">##  5 1        29554 31097  1544 +      ENST00000473358 MIR1302-2</span></span>
<span><span class="co">##  6 1        30267 31109   843 +      ENST00000469289 MIR1302-2</span></span>
<span><span class="co">##  7 1        30366 30503   138 +      ENST00000607096 MIR1302-2</span></span>
<span><span class="co">##  8 1        34554 36081  1528 -      ENST00000417324 FAM138A  </span></span>
<span><span class="co">##  9 1        35245 36073   829 -      ENST00000461467 FAM138A  </span></span>
<span><span class="co">## 10 1        52473 53312   840 +      ENST00000606857 OR4G4P   </span></span>
<span><span class="co">## # â„¹ 216,731 more rows</span></span>
<span></span>
<span><span class="co">### move transcrip ID the first column in the dataframe</span></span>
<span><span class="va">tx</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">tx</span>, <span class="st">"target_id"</span>, <span class="st">"gene_name"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="load-transcript-counts" class="level3"><h3 class="anchored" data-anchor-id="load-transcript-counts">Load transcript counts</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txi_gene</span> <span class="op">&lt;-</span> <span class="fu">tximport</span><span class="op">(</span></span>
<span>    <span class="va">path</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"kallisto"</span>,</span>
<span>    tx2gene <span class="op">=</span> <span class="va">tx</span>,</span>
<span>    txOut <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># How does the result change if this =FALSE vs =TRUE?</span></span>
<span>    countsFromAbundance <span class="op">=</span> <span class="st">"lengthScaledTPM"</span>,</span>
<span>    ignoreTxVersion <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### take a look at the type of object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">txi_gene</span><span class="op">)</span></span>
<span><span class="co">## [1] "list"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">txi_gene</span><span class="op">)</span></span>
<span><span class="co">## [1] "abundance"           "counts"              "length"             </span></span>
<span><span class="co">## [4] "countsFromAbundance"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="step4-wrangling-gene-expression" class="level2"><h2 class="anchored" data-anchor-id="step4-wrangling-gene-expression">Step4: Wrangling gene expression</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### examine the data</span></span>
<span><span class="va">tpm</span> <span class="op">&lt;-</span> <span class="va">txi_gene</span><span class="op">$</span><span class="va">abundance</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">txi_gene</span><span class="op">$</span><span class="va">counts</span></span>
<span></span>
<span><span class="co">### generate summary stats</span></span>
<span><span class="va">tpm_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/transform.html">transform</a></span><span class="op">(</span></span>
<span>    <span class="va">tpm</span>,</span>
<span>    SD <span class="op">=</span> <span class="fu">rowSds</span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span>,</span>
<span>    AVG <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span>,</span>
<span>    MED <span class="op">=</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### look at the summary data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tpm_stats</span><span class="op">)</span></span>
<span><span class="co">##                  X1          X2         X3           X4         X5          X6</span></span>
<span><span class="co">## 5S_rRNA   0.0000000   0.0000000   0.000000   0.00000000   0.000000   0.0000000</span></span>
<span><span class="co">## A1BG      6.3823340   6.7399420   3.842043   6.70710000   8.591181  14.6144150</span></span>
<span><span class="co">## A1CF      0.0757872   0.0616752   0.214226   0.00859446   0.000000   0.0265036</span></span>
<span><span class="co">## A2M     162.5622990 341.8312630 295.814803 261.47230600 169.453100 345.6983730</span></span>
<span><span class="co">## A2ML1   109.6624900  80.6018680  90.356840  64.11152700  67.035650  52.5762220</span></span>
<span><span class="co">## A2MP1     0.2237170   0.3143760   0.390022   0.15622800   0.122260   0.2700100</span></span>
<span><span class="co">##                 X7          X8           X9         X10           SD</span></span>
<span><span class="co">## 5S_rRNA   0.000000   0.0000000   0.00000000   0.0000000   0.00000000</span></span>
<span><span class="co">## A1BG     18.182700  12.8662800  12.05296100   9.8268802   4.41172300</span></span>
<span><span class="co">## A1CF      0.072777   0.1359934   0.03722713   0.0581254   0.06414927</span></span>
<span><span class="co">## A2M     643.554440 341.8972850 324.80208900 415.9663840 135.78415021</span></span>
<span><span class="co">## A2ML1     4.401309   5.1981760   1.75721000  15.1189910  39.78108982</span></span>
<span><span class="co">## A2MP1     1.087410   0.7737610   0.68834900   0.6378640   0.31619850</span></span>
<span><span class="co">##                  AVG         MED</span></span>
<span><span class="co">## 5S_rRNA   0.00000000   0.0000000</span></span>
<span><span class="co">## A1BG      9.98058361   9.2090306</span></span>
<span><span class="co">## A1CF      0.06909094   0.0599003</span></span>
<span><span class="co">## A2M     330.30523424 333.3166760</span></span>
<span><span class="co">## A2ML1    49.08202830  58.3438745</span></span>
<span><span class="co">## A2MP1     0.46639970   0.3521990</span></span>
<span></span>
<span><span class="co">### visualize it using scatter plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">tpm_stats</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">SD</span>, y <span class="op">=</span> <span class="va">MED</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">18</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="va">lm</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hex</span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="st">"Median"</span>, x <span class="op">=</span> <span class="st">"Standard deviation"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Transcripts per million (TPM)"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"unfiltered, non-normalized data"</span>,</span>
<span>        caption <span class="op">=</span> <span class="st">"DIYtranscriptomics - Spring 2020"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_dark</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### capture sample labels from study design</span></span>
<span><span class="va">sampleLabels</span> <span class="op">&lt;-</span> <span class="va">targets</span><span class="op">$</span><span class="va">sample</span></span>
<span></span>
<span><span class="co">### make DGElist from the counts</span></span>
<span><span class="va">DGEList</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span><span class="va">txi_gene</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">DGEList</span></span>
<span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##             Sample1     Sample2     Sample3     Sample4  Sample5     Sample6</span></span>
<span><span class="co">## 5S_rRNA     0.00000     0.00000     0.00000    0.000000   0.0000     0.00000</span></span>
<span><span class="co">## A1BG      184.19070   245.13169   101.73418   79.204797  13.3335   682.72256</span></span>
<span><span class="co">## A1CF       22.94477    23.53171    59.50815    1.064719   0.0000    12.98875</span></span>
<span><span class="co">## A2M     11901.13136 31538.08430 19870.31859 7832.900974 667.1460 40967.58721</span></span>
<span><span class="co">## A2ML1   15162.86062 14045.04908 11463.06052 3627.335147 498.4618 11767.57706</span></span>
<span><span class="co">##            Sample7     Sample8     Sample9    Sample10</span></span>
<span><span class="co">## 5S_rRNA     0.0000     0.00000     0.00000     0.00000</span></span>
<span><span class="co">## A1BG      391.5812   326.10430   397.97768   499.96534</span></span>
<span><span class="co">## A1CF       16.4421    36.15938    12.89508    31.02343</span></span>
<span><span class="co">## A2M     35158.3894 21982.61779 27205.95758 53686.20520</span></span>
<span><span class="co">## A2ML1     454.1300   631.23251   277.98636  3685.37686</span></span>
<span><span class="co">## 35366 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group lib.size norm.factors</span></span>
<span><span class="co">## Sample1      1 42300696            1</span></span>
<span><span class="co">## Sample2      1 51775632            1</span></span>
<span><span class="co">## Sample3      1 40485664            1</span></span>
<span><span class="co">## Sample4      1 16661634            1</span></span>
<span><span class="co">## Sample5      1  2191577            1</span></span>
<span><span class="co">## Sample6      1 55389922            1</span></span>
<span><span class="co">## Sample7      1 28800520            1</span></span>
<span><span class="co">## Sample8      1 31632164            1</span></span>
<span><span class="co">## Sample9      1 39464569            1</span></span>
<span><span class="co">## Sample10     1 67827185            1</span></span>
<span></span>
<span><span class="co">### get counts per million</span></span>
<span><span class="va">cpm</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">DGEList</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">cpm</span><span class="op">)</span></span>
<span><span class="co">##  Sample1  Sample2  Sample3  Sample4  Sample5  Sample6  Sample7  Sample8 </span></span>
<span><span class="co">##    1e+06    1e+06    1e+06    1e+06    1e+06    1e+06    1e+06    1e+06 </span></span>
<span><span class="co">##  Sample9 Sample10 </span></span>
<span><span class="co">##    1e+06    1e+06</span></span>
<span></span>
<span><span class="va">log2_cpm</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">DGEList</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">### connvert data matrix to dataframe</span></span>
<span><span class="va">log2_cpm_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">log2_cpm</span>, rownames <span class="op">=</span> <span class="st">"geneID"</span><span class="op">)</span></span>
<span><span class="co">### add sample names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">log2_cpm_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"geneID"</span>, <span class="va">sampleLabels</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### pivot from wide to long </span></span>
<span><span class="va">log2_cpm_df_pivot</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">log2_cpm_df</span>, </span>
<span>  cols <span class="op">=</span> <span class="va">HS01</span><span class="op">:</span><span class="va">CL13</span>, </span>
<span>  names_to <span class="op">=</span> <span class="st">"samples"</span>, </span>
<span>  values_to <span class="op">=</span> <span class="st">"expression"</span> </span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_df_pivot</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, y <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_violin</span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">stat_summary</span><span class="op">(</span>fun <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>        geom <span class="op">=</span> <span class="st">"point"</span>,</span>
<span>        shape <span class="op">=</span> <span class="fl">95</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="st">"log2 expression"</span>, x <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Log2 Counts per Million (CPM)"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"unfiltered, non-normalized"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>How many genes or transcripts have no read counts at all?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### filter data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">DGEList</span><span class="op">$</span><span class="va">counts</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">==</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">## 29373  5998</span></span>
<span></span>
<span><span class="co">### how many genes had more than 1 CPM (TRUE) in at least 3 samples</span></span>
<span></span>
<span><span class="co">### set cutoff</span></span>
<span><span class="va">keepers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">cpm</span><span class="op">&gt;</span><span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="co"># user defined</span></span>
<span><span class="va">DGEList_filtered</span> <span class="op">&lt;-</span> <span class="va">DGEList</span><span class="op">[</span><span class="va">keepers</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">log2_cpm_filtered</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">DGEList_filtered</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">log2_cpm_filtered_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">log2_cpm_filtered</span>, rownames <span class="op">=</span> <span class="st">"geneID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">log2_cpm_filtered_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"geneID"</span>, <span class="va">sampleLabels</span><span class="op">)</span></span>
<span><span class="va">log2_cpm_filtered_df_pivot</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_filtered_df</span>,</span>
<span>    cols <span class="op">=</span> <span class="va">HS01</span><span class="op">:</span><span class="va">CL13</span>,</span>
<span>    names_to <span class="op">=</span> <span class="st">"samples"</span>,</span>
<span>    values_to <span class="op">=</span> <span class="st">"expression"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_filtered_df_pivot</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, y <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_violin</span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">stat_summary</span><span class="op">(</span></span>
<span>        fun <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>        geom <span class="op">=</span> <span class="st">"point"</span>,</span>
<span>        shape <span class="op">=</span> <span class="fl">95</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"log2 expression"</span>, x <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Log2 Counts per Million (CPM)"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"filtered, non-normalized"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### normalize data</span></span>
<span><span class="va">DGEList_filtered_norm</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span></span>
<span>    <span class="va">DGEList_filtered</span>, method <span class="op">=</span> <span class="st">"TMM"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">log2_cpm_filtered_norm</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span></span>
<span>    <span class="va">DGEList_filtered_norm</span>, log <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">log2_cpm_filtered_norm_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_filtered_norm</span>, rownames <span class="op">=</span> <span class="st">"geneID"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">log2_cpm_filtered_norm_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"geneID"</span>, <span class="va">sampleLabels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">log2_cpm_filtered_norm_df_pivot</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_filtered_norm_df</span>,</span>
<span>    cols <span class="op">=</span> <span class="va">HS01</span><span class="op">:</span><span class="va">CL13</span>,</span>
<span>    names_to <span class="op">=</span> <span class="st">"samples"</span>,</span>
<span>    values_to <span class="op">=</span> <span class="st">"expression"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span></span>
<span>    <span class="va">log2_cpm_filtered_norm_df_pivot</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, y <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_violin</span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">stat_summary</span><span class="op">(</span></span>
<span>        fun <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>        geom <span class="op">=</span> <span class="st">"point"</span>,</span>
<span>        shape <span class="op">=</span> <span class="fl">95</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="st">"log2 expression"</span>, x <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Log2 Counts per Million (CPM)"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"filtered, TMM normalized"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### combine plots</span></span>
<span><span class="fu">plot_grid</span><span class="op">(</span></span>
<span>  <span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, </span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span><span class="op">)</span>, </span>
<span>  label_size <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="step5-data-exploration-with-multivariate-analysis" class="level2"><h2 class="anchored" data-anchor-id="step5-data-exploration-with-multivariate-analysis">Step5: Data exploration with multivariate analysis</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Identify variables of interest in study design file</span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">targets</span><span class="op">$</span><span class="va">group</span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">group</span></span>
<span><span class="co">##  [1] healthy healthy healthy healthy healthy disease disease disease disease</span></span>
<span><span class="co">## [10] disease</span></span>
<span><span class="co">## Levels: disease healthy</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="hierarchical-clustering" class="level3"><h3 class="anchored" data-anchor-id="hierarchical-clustering">Hierarchical clustering</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### calculate distance  methods (e.g. switch from 'maximum' to 'euclidean')</span></span>
<span><span class="va">distance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">log2_cpm_filtered_norm</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"maximum"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">### other distance methods are "euclidean", maximum", "manhattan", "canberra", "binary" or "minkowski"</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">distance</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span> </span>
<span><span class="co">### other agglomeration methods are "ward.D", "ward.D2", "single", "complete", "average", "mcquitty", "median", or "centroid"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">clusters</span>, labels<span class="op">=</span><span class="va">sampleLabels</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="principal-compoent-analysis" class="level3"><h3 class="anchored" data-anchor-id="principal-compoent-analysis">Principal compoent analysis</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">log2_cpm_filtered_norm</span><span class="op">)</span>,</span>
<span>    scale. <span class="op">=</span> <span class="cn">FALSE</span>, retx <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### prints variance summary for all principal components.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">)</span></span>
<span><span class="co">## Importance of components:</span></span>
<span><span class="co">##                             PC1     PC2      PC3      PC4      PC5      PC6</span></span>
<span><span class="co">## Standard deviation     113.9821 49.8074 33.42376 27.08082 26.40296 22.23750</span></span>
<span><span class="co">## Proportion of Variance   0.6656  0.1271  0.05723  0.03757  0.03571  0.02533</span></span>
<span><span class="co">## Cumulative Proportion    0.6656  0.7927  0.84993  0.88750  0.92322  0.94855</span></span>
<span><span class="co">##                             PC7      PC8      PC9      PC10</span></span>
<span><span class="co">## Standard deviation     20.50645 17.71945 16.42366 1.816e-13</span></span>
<span><span class="co">## Proportion of Variance  0.02154  0.01609  0.01382 0.000e+00</span></span>
<span><span class="co">## Cumulative Proportion   0.97010  0.98618  1.00000 1.000e+00</span></span>
<span></span>
<span><span class="co">### look at the PCA result (pca.res)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">)</span></span>
<span><span class="co">## [1] "center"   "rotation" "scale"    "sdev"     "x"</span></span>
<span></span>
<span><span class="co">### $rotation shows you how much each gene influenced each PC (called 'scores')</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">$</span><span class="va">rotation</span><span class="op">)</span> </span>
<span><span class="co">##                 PC1          PC2           PC3          PC4           PC5</span></span>
<span><span class="co">## A1BG    0.006058601 -0.005265865  0.0030322868  0.008759217  0.0021903806</span></span>
<span><span class="co">## A2M     0.004897345  0.002459393 -0.0029437295  0.010128989  0.0010754828</span></span>
<span><span class="co">## A2ML1  -0.015426376  0.002131922  0.0236833694  0.012776801  0.0151330784</span></span>
<span><span class="co">## A2MP1   0.008361769  0.005416383 -0.0080073403 -0.004989098  0.0019137300</span></span>
<span><span class="co">## A4GALT -0.004861196 -0.010071400 -0.0040747825  0.004365382 -0.0073287027</span></span>
<span><span class="co">## AAAS    0.002247684 -0.005546829 -0.0003891962  0.005401109 -0.0008667994</span></span>
<span><span class="co">##                 PC6           PC7         PC8          PC9        PC10</span></span>
<span><span class="co">## A1BG    0.002600255  1.165903e-02 0.002109927  0.006017461 -0.62642253</span></span>
<span><span class="co">## A2M    -0.003082754 -2.988929e-03 0.007583352 -0.002958490 -0.07649096</span></span>
<span><span class="co">## A2ML1  -0.013272183 -1.389944e-02 0.008501332  0.020687796  0.29435235</span></span>
<span><span class="co">## A2MP1  -0.002415560  1.626299e-03 0.011281299 -0.005577339 -0.50451276</span></span>
<span><span class="co">## A4GALT -0.009698589  6.199904e-05 0.012467687  0.004110126 -0.31572438</span></span>
<span><span class="co">## AAAS    0.006991981  3.783003e-03 0.004628266  0.003568768 -0.14061416</span></span>
<span></span>
<span><span class="co">### 'x' shows you how much each sample influenced each PC (called 'loadings')</span></span>
<span><span class="va">pca_res</span><span class="op">$</span><span class="va">x</span> </span>
<span><span class="co">##                 PC1          PC2        PC3        PC4         PC5        PC6</span></span>
<span><span class="co">## Sample1  -116.29139   38.2334807  21.154790 -31.204512  24.3342557  -5.032445</span></span>
<span><span class="co">## Sample2   -98.08443   28.5182187   6.488451  20.455248 -24.7150117  -5.625522</span></span>
<span><span class="co">## Sample3   -98.57599   46.3377421 -21.880766 -22.638077  26.9619112   8.312240</span></span>
<span><span class="co">## Sample4  -102.83232   27.1997919 -10.521681  36.884584 -34.6489813  11.837625</span></span>
<span><span class="co">## Sample5  -119.48548 -130.0448432  -8.771308  -8.078419   0.7747427  -2.609032</span></span>
<span><span class="co">## Sample6    76.83934  -13.0282654  65.541177  34.668976  32.4966464   9.756467</span></span>
<span><span class="co">## Sample7   114.80802    0.9427947 -66.411954  27.226075  30.8708143  -6.756527</span></span>
<span><span class="co">## Sample8   122.14072   -0.8992447  -2.600281 -31.000051 -26.0328136  19.278692</span></span>
<span><span class="co">## Sample9   129.23356   -4.8764679   3.983424 -16.285254 -13.4446636  25.667565</span></span>
<span><span class="co">## Sample10   92.24799    7.6167930  13.018146 -10.028570 -16.5969001 -54.829063</span></span>
<span><span class="co">##                 PC7         PC8         PC9          PC10</span></span>
<span><span class="co">## Sample1   37.289161  -8.2755495   9.2338103  6.432316e-13</span></span>
<span><span class="co">## Sample2    9.263960  37.6085903 -14.4523481  7.064106e-14</span></span>
<span><span class="co">## Sample3  -40.584484   0.6580365  -8.3486393  3.455381e-13</span></span>
<span><span class="co">## Sample4   -3.089166 -28.1333136  11.4331050 -5.619926e-14</span></span>
<span><span class="co">## Sample5   -2.014396  -0.2124505  -0.6860841 -1.205842e-13</span></span>
<span><span class="co">## Sample6  -10.563273   2.1965073   5.3204553 -6.485607e-14</span></span>
<span><span class="co">## Sample7   15.193876   4.0838551   4.1812220 -2.588419e-13</span></span>
<span><span class="co">## Sample8   -5.936520  15.6564254  28.6671255 -1.908206e-13</span></span>
<span><span class="co">## Sample9   11.836127 -13.5624080 -32.5109827 -3.271781e-13</span></span>
<span><span class="co">## Sample10 -11.395285 -10.0196929  -2.8376640  1.788202e-13</span></span>
<span></span>
<span><span class="co">### A screeplot is a standard way to view eigenvalues for each PCA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/screeplot.html">screeplot</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co">### sdev^2 captures these eigenvalues from the PCA result</span></span>
<span><span class="va">pc_var</span> <span class="op">&lt;-</span> <span class="va">pca_res</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span> </span>
<span></span>
<span><span class="co">### use these eigenvalues to calculate the percentage variance explained by each PC</span></span>
<span><span class="va">pc_per</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">pc_var</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">pc_var</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span> </span>
<span><span class="va">pc_per</span></span>
<span><span class="co">##  [1] 66.6 12.7  5.7  3.8  3.6  2.5  2.2  1.6  1.4  0.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Can we figure out the identify of outlier?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize PCA result </span></span>
<span><span class="va">pca_res_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">pca_res</span><span class="op">$</span><span class="va">x</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">bind_cols</span><span class="op">(</span><span class="va">targets</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pca_plot1</span> <span class="op">&lt;-</span> <span class="va">pca_res_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span></span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, label <span class="op">=</span> <span class="va">sample</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">stat_ellipse</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="va">pc_per</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"%"</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="va">pc_per</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"%"</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA plot"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_plot1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Another way to view PCA laodings to understand impact of each sample on each pricipal component</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pca_res_df</span> <span class="op">&lt;-</span> <span class="va">pca_res</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">add_column</span><span class="op">(</span></span>
<span>        sample <span class="op">=</span> <span class="va">sampleLabels</span>,</span>
<span>        group <span class="op">=</span> <span class="va">group</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">### dataframe to be pivoted  </span></span>
<span><span class="va">pca_pivot</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">pca_res_df</span>, </span>
<span>  cols <span class="op">=</span> <span class="va">PC1</span><span class="op">:</span><span class="va">PC4</span>, </span>
<span>  names_to <span class="op">=</span> <span class="st">"PC"</span>, </span>
<span>  values_to <span class="op">=</span> <span class="st">"loadings"</span></span>
<span><span class="op">)</span> </span>
<span><span class="co">### plot</span></span>
<span><span class="va">pca_plot2</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">pca_pivot</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">loadings</span>, fill <span class="op">=</span> <span class="va">group</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">PC</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"PCA 'small multiples' plot"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pca_plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### combine plot</span></span>
<span><span class="co"># plot_grid(pca_plot1, pca_plot2)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># make columns comparing each of the averages above that you're interested in</span></span>
<span><span class="va">mydata_df</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span></span>
<span>  <span class="va">log2_cpm_filtered_norm_df</span>,</span>
<span>  healthy.AVG <span class="op">=</span> <span class="op">(</span><span class="va">HS01</span> <span class="op">+</span> <span class="va">HS02</span> <span class="op">+</span> <span class="va">HS03</span> <span class="op">+</span> <span class="va">HS04</span> <span class="op">+</span> <span class="va">HS05</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span>, </span>
<span>  disease.AVG <span class="op">=</span> <span class="op">(</span><span class="va">CL08</span> <span class="op">+</span> <span class="va">CL10</span> <span class="op">+</span> <span class="va">CL11</span> <span class="op">+</span> <span class="va">CL12</span> <span class="op">+</span> <span class="va">CL13</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span>,</span>
<span>  LogFC <span class="op">=</span> <span class="op">(</span><span class="va">disease.AVG</span> <span class="op">-</span> <span class="va">healthy.AVG</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">mutate_if</span><span class="op">(</span><span class="va">is.numeric</span>, <span class="va">round</span>, <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### sort data</span></span>
<span><span class="va">mydata_sort</span> <span class="op">&lt;-</span> <span class="va">mydata_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">LogFC</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">geneID</span>, <span class="va">LogFC</span><span class="op">)</span></span>
<span><span class="va">mydata_sort</span></span>
<span><span class="co">## # A tibble: 15,765 Ã— 2</span></span>
<span><span class="co">##    geneID    LogFC</span></span>
<span><span class="co">##    &lt;chr&gt;     &lt;dbl&gt;</span></span>
<span><span class="co">##  1 IGLC3     10.6 </span></span>
<span><span class="co">##  2 MMP1      10.4 </span></span>
<span><span class="co">##  3 IGKV1-5   10.2 </span></span>
<span><span class="co">##  4 IGLV3-1    9.96</span></span>
<span><span class="co">##  5 IGLV3-19   9.9 </span></span>
<span><span class="co">##  6 IGLC1      9.88</span></span>
<span><span class="co">##  7 IGKV1D-39  9.87</span></span>
<span><span class="co">##  8 IGLL5      9.85</span></span>
<span><span class="co">##  9 IGKV1D-33  9.83</span></span>
<span><span class="co">## 10 IGLV2-8    9.65</span></span>
<span><span class="co">## # â„¹ 15,755 more rows</span></span>
<span></span>
<span><span class="co">### filter data</span></span>
<span><span class="va">mydata_filter</span> <span class="op">&lt;-</span> <span class="va">mydata_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">geneID</span> <span class="op">==</span> <span class="st">"MMP1"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"GZMB"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"IL1B"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"GNLY"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"IFNG"</span></span>
<span>        <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"CCL4"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"PRF1"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"APOBEC3A"</span> <span class="op">|</span> <span class="va">geneID</span> <span class="op">==</span> <span class="st">"UNC13A"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">geneID</span>, <span class="va">healthy.AVG</span>, <span class="va">disease.AVG</span>, <span class="va">LogFC</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">LogFC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mydata_filter</span></span>
<span><span class="co">## # A tibble: 9 Ã— 4</span></span>
<span><span class="co">##   geneID   healthy.AVG disease.AVG LogFC</span></span>
<span><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">## 1 MMP1            1.44       11.9  10.4 </span></span>
<span><span class="co">## 2 IFNG           -3.65        4.42  8.07</span></span>
<span><span class="co">## 3 CCL4           -1.75        6.14  7.89</span></span>
<span><span class="co">## 4 GZMB           -0.58        6.47  7.05</span></span>
<span><span class="co">## 5 GNLY            0.98        7.17  6.19</span></span>
<span><span class="co">## 6 IL1B            0.63        6.48  5.85</span></span>
<span><span class="co">## 7 PRF1            1.03        6.64  5.61</span></span>
<span><span class="co">## 8 APOBEC3A        0.27        5.6   5.33</span></span>
<span><span class="co">## 9 UNC13A         -0.78        1.95  2.73</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Produce publication-quality tables using `gt` package</span></span>
<span><span class="va">mydata_filter</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">gt</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">fmt_number</span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, decimals <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tab_header</span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="fu">md</span><span class="op">(</span><span class="st">"**Regulators of skin pathogenesis**"</span><span class="op">)</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="fu">md</span><span class="op">(</span><span class="st">"*during cutaneous leishmaniasis*"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tab_footnote</span><span class="op">(</span></span>
<span>        footnote <span class="op">=</span> <span class="st">"Deletion or blockaid ameliorates disease in mice"</span>,</span>
<span>        locations <span class="op">=</span> <span class="fu">cells_body</span><span class="op">(</span>columns <span class="op">=</span> <span class="va">geneID</span>, rows <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tab_footnote</span><span class="op">(</span></span>
<span>        footnote <span class="op">=</span> <span class="st">"Associated with treatment failure in multiple studies"</span>,</span>
<span>        locations <span class="op">=</span> <span class="fu">cells_body</span><span class="op">(</span>columns <span class="op">=</span> <span class="va">geneID</span>, rows <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tab_footnote</span><span class="op">(</span></span>
<span>        footnote <span class="op">=</span> <span class="st">"Implicated in parasite control"</span>,</span>
<span>        locations <span class="op">=</span> <span class="fu">cells_body</span><span class="op">(</span>columns <span class="op">=</span> <span class="va">geneID</span>, rows <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tab_source_note</span><span class="op">(</span></span>
<span>        source_note <span class="op">=</span> <span class="fu">md</span><span class="op">(</span><span class="st">"Reference: Amorim *et al*., (2019). DOI: 10.1126/scitranslmed.aar3619"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="prjdyjrjkq" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#prjdyjrjkq table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#prjdyjrjkq thead, #prjdyjrjkq tbody, #prjdyjrjkq tfoot, #prjdyjrjkq tr, #prjdyjrjkq td, #prjdyjrjkq th {
  border-style: none;
}

#prjdyjrjkq p {
  margin: 0;
  padding: 0;
}

#prjdyjrjkq .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#prjdyjrjkq .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#prjdyjrjkq .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#prjdyjrjkq .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#prjdyjrjkq .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#prjdyjrjkq .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#prjdyjrjkq .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#prjdyjrjkq .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#prjdyjrjkq .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#prjdyjrjkq .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#prjdyjrjkq .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#prjdyjrjkq .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#prjdyjrjkq .gt_spanner_row {
  border-bottom-style: hidden;
}

#prjdyjrjkq .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#prjdyjrjkq .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#prjdyjrjkq .gt_from_md > :first-child {
  margin-top: 0;
}

#prjdyjrjkq .gt_from_md > :last-child {
  margin-bottom: 0;
}

#prjdyjrjkq .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#prjdyjrjkq .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#prjdyjrjkq .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#prjdyjrjkq .gt_row_group_first td {
  border-top-width: 2px;
}

#prjdyjrjkq .gt_row_group_first th {
  border-top-width: 2px;
}

#prjdyjrjkq .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#prjdyjrjkq .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#prjdyjrjkq .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#prjdyjrjkq .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#prjdyjrjkq .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#prjdyjrjkq .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#prjdyjrjkq .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#prjdyjrjkq .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#prjdyjrjkq .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#prjdyjrjkq .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#prjdyjrjkq .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#prjdyjrjkq .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#prjdyjrjkq .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#prjdyjrjkq .gt_left {
  text-align: left;
}

#prjdyjrjkq .gt_center {
  text-align: center;
}

#prjdyjrjkq .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#prjdyjrjkq .gt_font_normal {
  font-weight: normal;
}

#prjdyjrjkq .gt_font_bold {
  font-weight: bold;
}

#prjdyjrjkq .gt_font_italic {
  font-style: italic;
}

#prjdyjrjkq .gt_super {
  font-size: 65%;
}

#prjdyjrjkq .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#prjdyjrjkq .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#prjdyjrjkq .gt_indent_1 {
  text-indent: 5px;
}

#prjdyjrjkq .gt_indent_2 {
  text-indent: 10px;
}

#prjdyjrjkq .gt_indent_3 {
  text-indent: 15px;
}

#prjdyjrjkq .gt_indent_4 {
  text-indent: 20px;
}

#prjdyjrjkq .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="gt_heading">
<td colspan="4" class="gt_heading gt_title gt_font_normal" style=""><strong>Regulators of skin pathogenesis</strong></td>
    </tr>
<tr class="gt_heading">
<td colspan="4" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border" style=""><em>during cutaneous leishmaniasis</em></td>
    </tr>
<tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="geneID">geneID</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="healthy.AVG">healthy.AVG</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="disease.AVG">disease.AVG</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="LogFC">LogFC</th>
    </tr>
</thead>
<tbody class="gt_table_body">
<tr>
<td headers="geneID" class="gt_row gt_left">MMP1</td>
<td headers="healthy.AVG" class="gt_row gt_right">1.4</td>
<td headers="disease.AVG" class="gt_row gt_right">11.9</td>
<td headers="LogFC" class="gt_row gt_right">10.4</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">IFNG<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1,2</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">âˆ’3.6</td>
<td headers="disease.AVG" class="gt_row gt_right">4.4</td>
<td headers="LogFC" class="gt_row gt_right">8.1</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">CCL4<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">âˆ’1.8</td>
<td headers="disease.AVG" class="gt_row gt_right">6.1</td>
<td headers="LogFC" class="gt_row gt_right">7.9</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">GZMB<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">âˆ’0.6</td>
<td headers="disease.AVG" class="gt_row gt_right">6.5</td>
<td headers="LogFC" class="gt_row gt_right">7.0</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">GNLY<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">1.0</td>
<td headers="disease.AVG" class="gt_row gt_right">7.2</td>
<td headers="LogFC" class="gt_row gt_right">6.2</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">IL1B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>3,1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">0.6</td>
<td headers="disease.AVG" class="gt_row gt_right">6.5</td>
<td headers="LogFC" class="gt_row gt_right">5.8</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">PRF1<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>3,1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">1.0</td>
<td headers="disease.AVG" class="gt_row gt_right">6.6</td>
<td headers="LogFC" class="gt_row gt_right">5.6</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">APOBEC3A<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">0.3</td>
<td headers="disease.AVG" class="gt_row gt_right">5.6</td>
<td headers="LogFC" class="gt_row gt_right">5.3</td>
</tr>
<tr>
<td headers="geneID" class="gt_row gt_left">UNC13A<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
<td headers="healthy.AVG" class="gt_row gt_right">âˆ’0.8</td>
<td headers="disease.AVG" class="gt_row gt_right">1.9</td>
<td headers="LogFC" class="gt_row gt_right">2.7</td>
</tr>
</tbody>
<tfoot class="gt_sourcenotes"><tr>
<td class="gt_sourcenote" colspan="4">Reference: Amorim <em>et al</em>., (2019). DOI: 10.1126/scitranslmed.aar3619</td>
    </tr></tfoot>
<tfoot class="gt_footnotes">
<tr>
<td class="gt_footnote" colspan="4">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span> Associated with treatment failure in multiple studies</td>
    </tr>
<tr>
<td class="gt_footnote" colspan="4">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>2</sup></span> Implicated in parasite control</td>
    </tr>
<tr>
<td class="gt_footnote" colspan="4">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>3</sup></span> Deletion or blockaid ameliorates disease in mice</td>
    </tr>
</tfoot>
</table>
</div>
</div>
</div>
<ul>
<li>Make an interactive table using the <code>DT</code> package</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># datatable(</span></span>
<span><span class="co">#     mydata_df[, c(1, 12:14)],</span></span>
<span><span class="co">#     extensions = c('KeyTable', "FixedHeader"),</span></span>
<span><span class="co">#     filter = 'top',</span></span>
<span><span class="co">#     options = list(keys = TRUE,</span></span>
<span><span class="co">#         searchHighlight = TRUE,</span></span>
<span><span class="co">#         pageLength = 10,</span></span>
<span><span class="co">#         # dom = "Blfrtip",</span></span>
<span><span class="co">#         # buttons = c("copy", "csv", "excel"),</span></span>
<span><span class="co">#         lengthMenu = c("10", "25", "50", "100"))</span></span>
<span><span class="co"># )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Make an interactive scatter plot with <code>plotly</code>
</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myplot</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">mydata_df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">healthy.AVG</span>, y<span class="op">=</span><span class="va">disease.AVG</span>, </span>
<span>      text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Symbol:"</span>, <span class="va">geneID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>shape<span class="op">=</span><span class="fl">16</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"disease vs. healthy"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myplot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ggplotly(myplot)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="step6-differential-gene-expression" class="level2"><h2 class="anchored" data-anchor-id="step6-differential-gene-expression">Step6: Differential gene expression</h2>
<section id="set-up-design-matrix" class="level3"><h3 class="anchored" data-anchor-id="set-up-design-matrix">Set up design matrix</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">targets</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">group</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># NOTE: if you need a paired analysis (a.k.a.'blocking' design) or have a batch effect, the following design is useful</span></span>
<span><span class="co"># design &lt;- model.matrix(~block + treatment)</span></span>
<span><span class="co"># this is just an example. 'block' and 'treatment' would need to be objects in your environment</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="model-mean-variance-trend-and-fit-linear-model-to-data" class="level3"><h3 class="anchored" data-anchor-id="model-mean-variance-trend-and-fit-linear-model-to-data">Model mean-variance trend and fit linear model to data</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Use VOOM function from Limma package to model the mean-variance relationship</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DGEList_filtered_norm</span><span class="op">)</span></span>
<span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##            Sample1     Sample2     Sample3    Sample4     Sample5     Sample6</span></span>
<span><span class="co">## A1BG     184.19070   245.13169   101.73418   79.20480  13.3334957   682.72256</span></span>
<span><span class="co">## A2M    11901.13136 31538.08430 19870.31859 7832.90097 667.1460374 40967.58721</span></span>
<span><span class="co">## A2ML1  15162.86062 14045.04908 11463.06052 3627.33515 498.4617540 11767.57706</span></span>
<span><span class="co">## A2MP1     29.31478    51.91491    46.89141    8.37674   0.8615391    57.27197</span></span>
<span><span class="co">## A4GALT   128.32345   355.39448   131.42493   75.43359  30.6454026   112.79123</span></span>
<span><span class="co">## AAAS     269.07806   457.62142   257.49394  147.56154  27.9749726   636.62246</span></span>
<span><span class="co">##            Sample7     Sample8    Sample9   Sample10</span></span>
<span><span class="co">## A1BG     391.58119   326.10430   397.9777   499.9653</span></span>
<span><span class="co">## A2M    35158.38935 21982.61779 27205.9576 53686.2052</span></span>
<span><span class="co">## A2ML1    454.12997   631.23251   277.9864  3685.3769</span></span>
<span><span class="co">## A2MP1    106.33010    89.04506   103.1984   147.3506</span></span>
<span><span class="co">## A4GALT    87.64872    88.03357    62.6173   230.0622</span></span>
<span><span class="co">## AAAS     382.02897   412.70476   446.8081   509.4904</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##          group lib.size norm.factors</span></span>
<span><span class="co">## Sample1      1 42300696    0.9838341</span></span>
<span><span class="co">## Sample2      1 51775632    1.0056785</span></span>
<span><span class="co">## Sample3      1 40485664    1.1055948</span></span>
<span><span class="co">## Sample4      1 16661634    1.0332639</span></span>
<span><span class="co">## Sample5      1  2191577    1.0154852</span></span>
<span><span class="co">## Sample6      1 55389922    0.8810033</span></span>
<span><span class="co">## Sample7      1 28800520    1.0117840</span></span>
<span><span class="co">## Sample8      1 31632164    1.0093077</span></span>
<span><span class="co">## Sample9      1 39464569    0.9639270</span></span>
<span><span class="co">## Sample10     1 67827185    1.0046274</span></span>
<span><span class="va">v_DEGList_filtered_norm</span> <span class="op">&lt;-</span> <span class="fu">voom</span><span class="op">(</span><span class="va">DGEList_filtered_norm</span>, <span class="va">design</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># fit a linear model to your data</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">v_DEGList_filtered_norm</span>, <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### contrast matrix</span></span>
<span><span class="va">contrast_matrix</span> <span class="op">&lt;-</span> <span class="fu">makeContrasts</span><span class="op">(</span>infection <span class="op">=</span> <span class="va">disease</span> <span class="op">-</span> <span class="va">healthy</span>,</span>
<span>                                 levels<span class="op">=</span><span class="va">design</span><span class="op">)</span></span>
<span><span class="co">### extract the linear model fit</span></span>
<span><span class="va">fits</span> <span class="op">&lt;-</span> <span class="fu">contrasts.fit</span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrast_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### get bayesian stats for the linear model fit</span></span>
<span><span class="va">ebFit</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fits</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#write.fit(ebFit, file="lmfit_results.txt")</span></span>
<span></span>
<span><span class="co">### top table to view DEGs</span></span>
<span><span class="va">TopHits</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">ebFit</span>, adjust <span class="op">=</span><span class="st">"BH"</span>, coef<span class="op">=</span><span class="fl">1</span>, number<span class="op">=</span><span class="fl">40000</span>, sort.by<span class="op">=</span><span class="st">"logFC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### convert to a tibble</span></span>
<span><span class="va">TopHits_df</span> <span class="op">&lt;-</span> <span class="va">TopHits</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"geneID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">TopHits_df</span></span>
<span><span class="co">## # A tibble: 15,765 Ã— 7</span></span>
<span><span class="co">##    geneID    logFC AveExpr     t      P.Value  adj.P.Val     B</span></span>
<span><span class="co">##    &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">##  1 MMP1      10.8    6.65  13.4  0.0000000182 0.00000123 10.0 </span></span>
<span><span class="co">##  2 IGLC3     10.3    3.49   7.50 0.00000815   0.0000778   4.06</span></span>
<span><span class="co">##  3 IGKV1D-33 10.2    1.12   9.29 0.000000913  0.0000160   6.11</span></span>
<span><span class="co">##  4 IGKV2D-28 10.1    0.898  9.50 0.000000723  0.0000137   6.32</span></span>
<span><span class="co">##  5 IGLV3-25  10.0    0.242  6.89 0.0000186    0.000147    3.27</span></span>
<span><span class="co">##  6 IGLV3-1   10.0    1.75  11.8  0.0000000715 0.00000281  8.36</span></span>
<span><span class="co">##  7 IGKV1-5    9.93   3.36  13.8  0.0000000124 0.00000101 10.0 </span></span>
<span><span class="co">##  8 IGHV1-3    9.89   0.447  6.24 0.0000468    0.000302    2.37</span></span>
<span><span class="co">##  9 IGLV3-19   9.84   2.18   6.93 0.0000175    0.000140    3.33</span></span>
<span><span class="co">## 10 IGHV3-21   9.80   1.42   9.00 0.00000128   0.0000200   5.80</span></span>
<span><span class="co">## # â„¹ 15,755 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Limma outputs
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>TopTable (from Limma) outputs a few different stats:</li>
<li>logFC, AveExpr, and P.Value should be self-explanatory</li>
<li>adj.P.Val is your adjusted P value, also known as an FDR (if BH method was used for multiple testing correction)</li>
<li>B statistic is the log-odds that that gene is differentially expressed. If B = 1.5, then log odds is e^1.5, where e is eulerâ€™s constant (approx. 2.718). So, the odds of differential expression os about 4.8 to 1</li>
<li>t statistic is ratio of the logFC to the standard error (where the error has been moderated across all genesâ€¦because of Bayesian approach)</li>
</ul>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### volcano plot</span></span>
<span><span class="va">vplot</span> <span class="op">&lt;-</span> <span class="va">TopHits_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span></span>
<span>        <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span>, x <span class="op">=</span> <span class="va">logFC</span>, text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Symbol:"</span>, <span class="va">geneID</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"longdash"</span>, colour <span class="op">=</span> <span class="st">"grey"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="st">"longdash"</span>, colour <span class="op">=</span> <span class="st">"#BE684D"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, linetype <span class="op">=</span> <span class="st">"longdash"</span>, colour <span class="op">=</span> <span class="st">"#2C467A"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># annotate("rect", xmin = 1, xmax = 12, ymin = -log10(0.01), ymax = 7.5, alpha=.2, fill="#BE684D") +</span></span>
<span>    <span class="co"># annotate("rect", xmin = -1, xmax = -12, ymin = -log10(0.01), ymax = 7.5, alpha=.2, fill="#2C467A") +</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano plot"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"Cutaneous leishmaniasis"</span>,</span>
<span>        caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"produced on "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vplot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### pull out DEGs and make venn diagram</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">decideTests</span><span class="op">(</span></span>
<span>    <span class="va">ebFit</span>, method <span class="op">=</span> <span class="st">"global"</span>, adjust.method <span class="op">=</span> <span class="st">"BH"</span>, p.value <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="co">## TestResults matrix</span></span>
<span><span class="co">##         Contrasts</span></span>
<span><span class="co">##          infection</span></span>
<span><span class="co">##   A1BG           0</span></span>
<span><span class="co">##   A2M            0</span></span>
<span><span class="co">##   A2ML1         -1</span></span>
<span><span class="co">##   A2MP1          0</span></span>
<span><span class="co">##   A4GALT         0</span></span>
<span><span class="co">##   AAAS           0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="co">##        infection</span></span>
<span><span class="co">## Down        1012</span></span>
<span><span class="co">## NotSig     13446</span></span>
<span><span class="co">## Up          1307</span></span>
<span><span class="fu">vennDiagram</span><span class="op">(</span><span class="va">results</span>, include<span class="op">=</span><span class="st">"up"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### retrieve expression data for DEGs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span><span class="op">)</span></span>
<span><span class="co">##           Sample1     Sample2    Sample3    Sample4    Sample5   Sample6</span></span>
<span><span class="co">## A1BG    2.1498709 2.237982296 1.19157184  2.2109278  2.6359548 3.8074408</span></span>
<span><span class="co">## A2M     8.1597752 9.242459704 8.79420169  8.8297563  8.2288005 9.7134419</span></span>
<span><span class="co">## A2ML1   8.5092054 8.075454826 8.00060559  7.7192264  7.8086463 7.9138199</span></span>
<span><span class="co">## A2MP1  -0.4811408 0.009534871 0.08239155 -0.9556369 -0.7089007 0.2435267</span></span>
<span><span class="co">## A4GALT  1.6301549 2.772935178 1.55941146  2.1409994  3.8068082 1.2151214</span></span>
<span><span class="co">## AAAS    2.6954627 3.137215647 2.52703139  3.1043863  3.6774835 3.7066558</span></span>
<span><span class="co">##          Sample7  Sample8   Sample9 Sample10</span></span>
<span><span class="co">## A1BG    3.750084 3.354713 3.3888732 2.876674</span></span>
<span><span class="co">## A2M    10.236678 9.427422 9.4821796 9.621825</span></span>
<span><span class="co">## A2ML1   3.963625 4.306483 2.8719810 5.757340</span></span>
<span><span class="co">## A2MP1   1.874250 1.487855 1.4467684 1.117544</span></span>
<span><span class="co">## A4GALT  1.596943 1.471466 0.7304817 1.758560</span></span>
<span><span class="co">## AAAS    3.714501 3.694027 3.5556433 2.903874</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sampleLabels</span></span>
<span></span>
<span><span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span><span class="op">[</span><span class="va">results</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">!=</span><span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">diffGenes</span><span class="op">)</span></span>
<span><span class="co">##             HS01     HS02     HS03     HS04     HS05     CL08       CL10</span></span>
<span><span class="co">## A2ML1   8.509205 8.075455 8.000606 7.719226 7.808646 7.913820  3.9636246</span></span>
<span><span class="co">## AADAC   3.783598 3.145125 3.421001 3.846583 4.124090 2.241617 -3.5057299</span></span>
<span><span class="co">## AADACL2 6.884775 7.025967 6.096531 6.700546 6.104662 5.215819 -5.8649243</span></span>
<span><span class="co">## AARD    2.817882 1.719728 2.562007 2.090062 2.784928 1.028134  0.2785801</span></span>
<span><span class="co">## ABCA10  4.429922 4.330043 6.261231 3.871370 2.785400 2.261820  2.9336235</span></span>
<span><span class="co">## ABCA12  8.609129 7.810834 7.884251 7.603805 7.963791 7.338358  3.3247149</span></span>
<span><span class="co">##              CL11       CL12       CL13</span></span>
<span><span class="co">## A2ML1   4.3064835  2.8719810  5.7573398</span></span>
<span><span class="co">## AADAC   0.1777154 -0.9997370 -0.5843452</span></span>
<span><span class="co">## AADACL2 2.9288016  2.0251967  3.5424241</span></span>
<span><span class="co">## AARD    0.2142361 -0.4255113  0.1288021</span></span>
<span><span class="co">## ABCA10  2.4838643  2.4430992  0.6882190</span></span>
<span><span class="co">## ABCA12  5.2844659  4.7055495  6.6999992</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">diffGenes</span><span class="op">)</span></span>
<span><span class="co">## [1] 2319   10</span></span>
<span></span>
<span><span class="co">#convert DEGs to a dataframe using as_tibble</span></span>
<span><span class="va">diffGenes_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">diffGenes</span>, rownames <span class="op">=</span> <span class="st">"geneID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### create interactive tables to display DEGs</span></span>
<span><span class="co"># datatable(diffGenes_df,</span></span>
<span><span class="co">#           extensions = c('KeyTable', "FixedHeader"),</span></span>
<span><span class="co">#           caption = 'Table 1: DEGs in cutaneous leishmaniasis',</span></span>
<span><span class="co">#           options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 10, lengthMenu = c("10", "25", "50", "100"))) |&gt;</span></span>
<span><span class="co">#   formatRound(columns=c(2:11), digits=2)</span></span>
<span></span>
<span><span class="co">### write DEGs to a file</span></span>
<span><span class="co"># write_tsv(diffGenes_df, "DiffGenes.txt") </span></span>
<span><span class="co">#NOTE: this .txt file can be directly used for input into other clustering or network analysis tools (e.g., String, Clust (https://github.com/BaselAbujamous/clust, etc.)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="step7-module-identification" class="level2"><h2 class="anchored" data-anchor-id="step7-module-identification">Step7: Module identification</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">heatcolors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"RdBu"</span>, n<span class="op">=</span><span class="fl">11</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#cluster rows by pearson correlation</span></span>
<span><span class="va">clustRows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">diffGenes</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  method<span class="op">=</span><span class="st">"complete"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#cluster columns by pearson correlation</span></span>
<span><span class="va">clustColumns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">diffGenes</span>, method<span class="op">=</span><span class="st">"spearman"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  method<span class="op">=</span><span class="st">"complete"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">module_assign</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">clustRows</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">module_color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">module_assign</span><span class="op">)</span><span class="op">)</span>, start<span class="op">=</span><span class="fl">0.1</span>, end<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">module_color</span> <span class="op">&lt;-</span> <span class="va">module_color</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">module_assign</span><span class="op">)</span><span class="op">]</span> </span>
<span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span></span>
<span>    <span class="va">diffGenes</span>,</span>
<span>    Rowv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">clustRows</span><span class="op">)</span>,</span>
<span>    Colv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">clustColumns</span><span class="op">)</span>,</span>
<span>    RowSideColors <span class="op">=</span> <span class="va">module_color</span>,</span>
<span>    col <span class="op">=</span> <span class="va">heatcolors</span>, scale <span class="op">=</span> <span class="st">'row'</span>, labRow <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    density.info <span class="op">=</span> <span class="st">"none"</span>, trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    cexRow <span class="op">=</span> <span class="fl">1</span>, cexCol <span class="op">=</span> <span class="fl">1</span>, margins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="va">modulePick</span> <span class="op">&lt;-</span> <span class="fl">2</span> </span>
<span><span class="va">Module_up</span> <span class="op">&lt;-</span> <span class="va">diffGenes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">module_assign</span><span class="op">[</span><span class="va">module_assign</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">modulePick</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span> </span>
<span><span class="va">hrsub_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">Module_up</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"complete"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span></span>
<span>    <span class="va">Module_up</span>, </span>
<span>    Rowv<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hrsub_up</span><span class="op">)</span>, </span>
<span>    Colv<span class="op">=</span><span class="cn">NA</span>, </span>
<span>    labRow <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    col<span class="op">=</span><span class="va">heatcolors</span>, scale<span class="op">=</span><span class="st">"row"</span>, </span>
<span>    density.info<span class="op">=</span><span class="st">"none"</span>, trace<span class="op">=</span><span class="st">"none"</span>, </span>
<span>    RowSideColors<span class="op">=</span><span class="va">module_color</span><span class="op">[</span><span class="va">module_assign</span><span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span><span class="va">modulePick</span><span class="op">]</span>, margins<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>,<span class="fl">20</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-23-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="va">modulePick</span> <span class="op">&lt;-</span> <span class="fl">1</span> </span>
<span><span class="va">Module_down</span> <span class="op">&lt;-</span> <span class="va">diffGenes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">module_assign</span><span class="op">[</span><span class="va">module_assign</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">modulePick</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span> </span>
<span><span class="va">hrsub_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">Module_down</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"complete"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">Module_down</span>, </span>
<span>          Rowv<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hrsub_down</span><span class="op">)</span>, </span>
<span>          Colv<span class="op">=</span><span class="cn">NA</span>, </span>
<span>          labRow <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>          col<span class="op">=</span><span class="va">heatcolors</span>, scale<span class="op">=</span><span class="st">"row"</span>, </span>
<span>          density.info<span class="op">=</span><span class="st">"none"</span>, trace<span class="op">=</span><span class="st">"none"</span>, </span>
<span>          RowSideColors<span class="op">=</span><span class="va">module_color</span><span class="op">[</span><span class="va">module_assign</span><span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span><span class="va">modulePick</span><span class="op">]</span>, margins<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-23-3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="step8-functional-enrichment-analysis" class="level2"><h2 class="anchored" data-anchor-id="step8-functional-enrichment-analysis">Step8: Functional enrichment analysis</h2>
<section id="go-enrichment-analysis" class="level3"><h3 class="anchored" data-anchor-id="go-enrichment-analysis">Go enrichment analysis</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### carry out GO enrichment using gProfiler2</span></span>
<span><span class="co">#### use topTable result to pick the top genes for carrying out a Gene Ontology (GO) enrichment analysis</span></span>
<span><span class="va">myTopHits</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span></span>
<span>    <span class="va">ebFit</span>,</span>
<span>    adjust <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>    coef <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    number <span class="op">=</span> <span class="fl">50</span>,</span>
<span>    sort.by <span class="op">=</span> <span class="st">"logFC"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### use the 'gost' function from the gprofiler2 package to run GO enrichment analysis</span></span>
<span><span class="va">gost_res</span> <span class="op">&lt;-</span> <span class="fu">gost</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">myTopHits</span><span class="op">)</span>,</span>
<span>    organism <span class="op">=</span> <span class="st">"hsapiens"</span>,</span>
<span>    correction_method <span class="op">=</span> <span class="st">"fdr"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### produce an interactive manhattan plot of enriched GO terms</span></span>
<span><span class="fu">gostplot</span><span class="op">(</span><span class="va">gost_res</span>, interactive <span class="op">=</span> <span class="cn">TRUE</span>, capped <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="plotly html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-9c621b3af0998a7149cb" style="width:100%;height:480px;"></div>
<script type="application/json" data-for="htmlwidget-9c621b3af0998a7149cb">{"x":{"data":[{"x":[155.31927977489644,151.91524195481168,155.3171388706071],"y":[2.6262762222965401,2.6262762222965401,2.4046201641637848],"text":["CORUM:6821 (2) <br> TTR-RBP complex <br> 2.364e-03","CORUM:1703 (1) <br> IGHM-VPREB1-IGLL1 complex <br> 2.364e-03","CORUM:6820 (5) <br> APOL1 complex B (APOL1, APOA1, HPR, FN1, IGHM) <br> 3.939e-03"],"key":["CORUM:6821","CORUM:1703","CORUM:6820"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(102,170,0,1)","opacity":0.80000000000000004,"size":[8.9679083027707129,3.7795275590551185,11.685516855496353],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(102,170,0,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"CORUM","legendgroup":"CORUM","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[40.705042417465691,44.625976580097401,40.970433157731165,68.973436751548391,68.892107331144473,52.495668128130561],"y":[17,17,17,11.718899065595979,2.5354130387224485,1.599332363568277],"text":["GO:0002250 (718) <br> adaptive immune response <br> 2.046e-43","GO:0006955 (1951) <br> immune response <br> 1.928e-33","GO:0002376 (2730) <br> immune system process <br> 8.766e-30","GO:0050896 (8955) <br> response to stimulus <br> 1.910e-12","GO:0050853 (70) <br> B cell receptor signaling pathway <br> 2.915e-03","GO:0019730 (127) <br> antimicrobial humoral response <br> 2.516e-02"],"key":["GO:0002250","GO:0006955","GO:0002376","GO:0050896","GO:0050853","GO:0019730"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(255,153,0,1)","opacity":0.80000000000000004,"size":[19.760953100387198,20.932598402907679,21.308796689486559,22.578712678924681,16.624603506234173,17.495593428699955],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(255,153,0,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"GO:BP","legendgroup":"GO:BP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[29.936074977920477,28.608841973088278,34.221753164491531,34.176798498198835,29.092639810333562,28.68162571851456,29.589281837948192,35.024515062575531,29.604266726712428,29.730567932010974,33.371895901719952,33.329081933822138,32.222340863663682,31.059941635238062,36.075597974466831,34.129703133511242,34.091170562403207,34.136125228695917,34.131843831906124,32.027537309728629,34.112577546352114,30.820183415010312,34.138265927090799,34.097592657587882,34.106155451167439,34.118999641536789,34.125421736721449,34.11471824474701,34.121140339931671],"y":[17,17,17,17,17,17,17,13.524418818442603,12.108036873239982,10.096786498495034,7.4580961601000864,7.4580961601000864,7.4580961601000864,7.4580961601000864,7.4580961601000864,6.8159446966590576,6.4449386478270601,4.2945557783861537,4.2945557783861537,3.6407037117615433,3.6407037117615433,3.4525663106858997,2.1618761887207678,2.1618761887207678,1.8789978799541966,1.5958466894761949,1.5467838723111729,1.5467838723111729,1.5467838723111729],"text":["GO:0019814 (149) <br> immunoglobulin complex <br> 1.009e-70","GO:0005576 (4206) <br> extracellular region <br> 6.987e-31","GO:0072562 (136) <br> blood microparticle <br> 1.710e-21","GO:0071944 (6202) <br> cell periphery <br> 6.350e-20","GO:0005886 (5717) <br> plasma membrane <br> 2.115e-18","GO:0005615 (3284) <br> extracellular space <br> 8.161e-18","GO:0009897 (427) <br> external side of plasma membrane <br> 2.701e-17","GO:0098552 (656) <br> side of membrane <br> 2.989e-14","GO:0009986 (955) <br> cell surface <br> 7.798e-13","GO:0016020 (9853) <br> membrane <br> 8.002e-11","GO:0070062 (2109) <br> extracellular exosome <br> 3.483e-08","GO:0065010 (2134) <br> extracellular membrane-bounded organelle <br> 3.483e-08","GO:0043230 (2134) <br> extracellular organelle <br> 3.483e-08","GO:0032991 (8527) <br> protein-containing complex <br> 3.483e-08","GO:1903561 (2133) <br> extracellular vesicle <br> 3.483e-08","GO:0071753 (4) <br> IgM immunoglobulin complex <br> 1.528e-07","GO:0071735 (5) <br> IgG immunoglobulin complex <br> 3.590e-07","GO:0071756 (3) <br> pentameric IgM immunoglobulin complex <br> 5.075e-05","GO:0071754 (3) <br> IgM immunoglobulin complex, circulating <br> 5.075e-05","GO:0042571 (6) <br> immunoglobulin complex, circulating <br> 2.287e-04","GO:0071745 (6) <br> IgA immunoglobulin complex <br> 2.287e-04","GO:0031982 (3976) <br> vesicle <br> 3.527e-04","GO:0071757 (1) <br> hexameric IgM immunoglobulin complex <br> 6.888e-03","GO:0071738 (1) <br> IgD immunoglobulin complex <br> 6.888e-03","GO:0071742 (2) <br> IgE immunoglobulin complex <br> 1.321e-02","GO:0071748 (4) <br> monomeric IgA immunoglobulin complex <br> 2.536e-02","GO:0071751 (5) <br> secretory IgA immunoglobulin complex <br> 2.839e-02","GO:0071746 (5) <br> IgA immunoglobulin complex, circulating <br> 2.839e-02","GO:0071749 (5) <br> polymeric IgA immunoglobulin complex <br> 2.839e-02"],"key":["GO:0019814","GO:0005576","GO:0072562","GO:0071944","GO:0005886","GO:0005615","GO:0009897","GO:0098552","GO:0009986","GO:0016020","GO:0070062","GO:0065010","GO:0043230","GO:0032991","GO:1903561","GO:0071753","GO:0071735","GO:0071756","GO:0071754","GO:0042571","GO:0071745","GO:0031982","GO:0071757","GO:0071738","GO:0071742","GO:0071748","GO:0071751","GO:0071746","GO:0071749"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(16,150,24,1)","opacity":0.80000000000000004,"size":[17.719933765215238,21.781214866745234,17.59218484123225,22.195368953169552,22.109308347681086,21.512285374899808,19.116518170560521,19.650844469559107,20.103856236279107,22.677165354330711,21.020527996395611,21.033795239590564,21.033795239590564,22.528057466073552,21.033267735673274,11.117005973573116,11.685516855496353,10.311454741100846,10.311454741100846,12.121311987659505,12.121311987659505,21.720451562965863,3.7795275590551185,3.7795275590551185,8.9679083027707129,11.117005973573116,11.685516855496353,11.685516855496353,11.685516855496353],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(16,150,24,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"GO:CC","legendgroup":"GO:CC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[2.7833720675371505,17.650318464020884,13.320368484163904,15.099008342424485,5.253348477203466],"y":[17,3.0124337423067518,2.4506641412764192,2.1341913567288198,1.3409707333403511],"text":["GO:0003823 (115) <br> antigen binding <br> 1.567e-44","GO:0048248 (5) <br> CXCR3 chemokine receptor binding <br> 9.718e-04","GO:0034987 (11) <br> immunoglobulin receptor binding <br> 3.543e-03","GO:0045236 (18) <br> CXCR chemokine receptor binding <br> 7.342e-03","GO:0008009 (50) <br> chemokine activity <br> 4.561e-02"],"key":["GO:0003823","GO:0048248","GO:0034987","GO:0045236","GO:0008009"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(220,57,18,1)","opacity":0.80000000000000004,"size":[17.354348527787963,11.685516855496353,13.429680893642971,14.374405574627291,16.105460718562572],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(220,57,18,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"GO:MF","legendgroup":"GO:MF","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[100.73341931506239,100.9070815720401,100.88564178722804,100.73556329354361,100.73770727202481,101.15578307586],"y":[2.3683916268072629,2.2613324725606234,2.2613324725606234,2.2613324725606234,1.8222592992503441,1.7342700263050239],"text":["KEGG:04060 (293) <br> Cytokine-cytokine receptor interaction <br> 4.282e-03","KEGG:04657 (91) <br> IL-17 signaling pathway <br> 5.479e-03","KEGG:04620 (102) <br> Toll-like receptor signaling pathway <br> 5.479e-03","KEGG:04061 (98) <br> Viral protein interaction with cytokine and cytokine receptor <br> 5.479e-03","KEGG:04062 (190) <br> Chemokine signaling pathway <br> 1.506e-02","KEGG:05171 (231) <br> Coronavirus disease - COVID-19 <br> 1.844e-02"],"key":["KEGG:04060","KEGG:04657","KEGG:04620","KEGG:04061","KEGG:04062","KEGG:05171"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(221,68,119,1)","opacity":0.80000000000000004,"size":[18.632038292976183,17.015283753483406,17.181653674821646,17.123564576575358,18.054511771768507,18.317875012744647],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(221,68,119,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"KEGG","legendgroup":"KEGG","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[104.31755628925704,105.37734357291312,104.39463172806838,104.09489391046867,106.0046520054611,107.83519367723072,105.37948455732456,107.58683948550525,103.8615266096232,107.99362652367628,107.83305269281931,104.347530071017,105.38162554173599,106.24016029071799,107.04731141382578,107.66819689313945,105.37306160409028,105.37092061967886,103.85724464080033,106.23801930630657,105.46726491819305,104.00711354960019,107.18861638497994,104.22977592838855,108.29550532568742,105.37520258850171,107.04945239823721,106.23587832189516,105.94256345752973,105.46512393378161,107.84161663046501,105.83123226813555,105.96825527046683,108.95706950881819,103.79087412404611,108.95278753999533,106.0132159431068,105.075464770902,105.94042247311829,107.80307891105932,104.2661726633828],"y":[17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,2.2207478758153774,1.6034774000123493],"text":["REAC:R-HSA-173623 (61) <br> Classical antibody-mediated complement activation <br> 2.077e-53","REAC:R-HSA-2029481 (67) <br> FCGR activation <br> 2.297e-52","REAC:R-HSA-166786 (68) <br> Creation of C4 and C2 activators <br> 2.477e-52","REAC:R-HSA-5690714 (60) <br> CD22 mediated BCR regulation <br> 2.982e-51","REAC:R-HSA-166663 (76) <br> Initial triggering of complement <br> 5.195e-51","REAC:R-HSA-2029485 (81) <br> Role of phospholipids in phagocytosis <br> 3.178e-50","REAC:R-HSA-9664323 (93) <br> FCGR3A-mediated IL10 synthesis <br> 1.878e-48","REAC:R-HSA-977606 (101) <br> Regulation of Complement cascade <br> 1.959e-47","REAC:R-HSA-983695 (84) <br> Antigen activates B Cell Receptor (BCR) leading to generation of second messengers <br> 3.866e-47","REAC:R-HSA-2168880 (69) <br> Scavenging of heme from plasma <br> 5.721e-47","REAC:R-HSA-2730905 (69) <br> Role of LAT2/NTAL/LAB on calcium mobilization <br> 5.721e-47","REAC:R-HSA-166658 (111) <br> Complement cascade <br> 2.137e-46","REAC:R-HSA-9664422 (113) <br> FCGR3A-mediated phagocytosis <br> 2.889e-46","REAC:R-HSA-9664417 (113) <br> Leishmania phagocytosis <br> 2.889e-46","REAC:R-HSA-9664407 (113) <br> Parasite infection <br> 2.889e-46","REAC:R-HSA-2029482 (116) <br> Regulation of actin dynamics for phagocytic cup formation <br> 5.832e-46","REAC:R-HSA-2871796 (83) <br> FCERI mediated MAPK activation <br> 6.875e-45","REAC:R-HSA-2871809 (83) <br> FCERI mediated Ca+2 mobilization <br> 6.875e-45","REAC:R-HSA-9662851 (132) <br> Anti-inflammatory response favouring Leishmania parasite infection <br> 1.973e-44","REAC:R-HSA-9664433 (132) <br> Leishmania parasite growth and survival <br> 1.973e-44","REAC:R-HSA-2029480 (140) <br> Fcgamma receptor (FCGR) dependent phagocytosis <br> 1.014e-43","REAC:R-HSA-2173782 (97) <br> Binding and Uptake of Ligands by Scavenger Receptors <br> 4.207e-43","REAC:R-HSA-9679191 (150) <br> Potential therapeutics for SARS <br> 1.832e-40","REAC:R-HSA-202733 (190) <br> Cell surface interactions at the vascular wall <br> 4.785e-40","REAC:R-HSA-983705 (163) <br> Signaling by the B Cell Receptor (BCR) <br> 1.594e-39","REAC:R-HSA-2871837 (136) <br> FCERI mediated NF-kB activation <br> 2.971e-39","REAC:R-HSA-9824443 (215) <br> Parasitic Infection Pathways <br> 1.248e-38","REAC:R-HSA-9658195 (215) <br> Leishmania infection <br> 1.248e-38","REAC:R-HSA-198933 (181) <br> Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell <br> 4.296e-36","REAC:R-HSA-2454202 (182) <br> Fc epsilon receptor (FCERI) signaling <br> 4.780e-36","REAC:R-HSA-9679506 (459) <br> SARS-CoV Infections <br> 6.450e-28","REAC:R-HSA-109582 (672) <br> Hemostasis <br> 1.775e-25","REAC:R-HSA-5663205 (999) <br> Infectious disease <br> 1.245e-22","REAC:R-HSA-9824446 (809) <br> Viral Infection Pathways <br> 8.361e-22","REAC:R-HSA-1280218 (809) <br> Adaptive Immune System <br> 8.361e-22","REAC:R-HSA-5653656 (718) <br> Vesicle-mediated transport <br> 1.707e-21","REAC:R-HSA-168249 (1100) <br> Innate Immune System <br> 5.114e-20","REAC:R-HSA-1643685 (1768) <br> Disease <br> 1.594e-17","REAC:R-HSA-168256 (2051) <br> Immune System <br> 3.458e-17","REAC:R-HSA-6809583 (1) <br> Retinoid metabolism disease events <br> 6.015e-03","REAC:R-HSA-380108 (57) <br> Chemokine receptors bind chemokines <br> 2.492e-02"],"key":["REAC:R-HSA-173623","REAC:R-HSA-2029481","REAC:R-HSA-166786","REAC:R-HSA-5690714","REAC:R-HSA-166663","REAC:R-HSA-2029485","REAC:R-HSA-9664323","REAC:R-HSA-977606","REAC:R-HSA-983695","REAC:R-HSA-2168880","REAC:R-HSA-2730905","REAC:R-HSA-166658","REAC:R-HSA-9664422","REAC:R-HSA-9664417","REAC:R-HSA-9664407","REAC:R-HSA-2029482","REAC:R-HSA-2871796","REAC:R-HSA-2871809","REAC:R-HSA-9662851","REAC:R-HSA-9664433","REAC:R-HSA-2029480","REAC:R-HSA-2173782","REAC:R-HSA-9679191","REAC:R-HSA-202733","REAC:R-HSA-983705","REAC:R-HSA-2871837","REAC:R-HSA-9824443","REAC:R-HSA-9658195","REAC:R-HSA-198933","REAC:R-HSA-2454202","REAC:R-HSA-9679506","REAC:R-HSA-109582","REAC:R-HSA-5663205","REAC:R-HSA-9824446","REAC:R-HSA-1280218","REAC:R-HSA-5653656","REAC:R-HSA-168249","REAC:R-HSA-1643685","REAC:R-HSA-168256","REAC:R-HSA-6809583","REAC:R-HSA-380108"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(51,102,204,1)","opacity":0.80000000000000004,"size":[16.414845775301554,16.558214613980976,16.580707444250166,16.389417683497474,16.748328670165268,16.843381923146577,17.047140132513317,17.167371200103453,16.897327557022052,16.602833306375246,16.602833306375246,17.30361286285148,17.32922894620004,17.32922894620004,17.32922894620004,17.366727849814772,16.879587308642606,16.879587308642606,17.550152874418341,17.550152874418341,17.632876281676548,17.108631019856755,17.729248020920775,18.054511771768507,17.844469050608055,17.59218484123225,18.221684151271067,18.221684151271067,17.98834742391934,17.995875078024451,19.207743110577862,19.680299899162929,20.157348803880588,19.905291932643788,19.905291932643788,19.760953100387198,20.271143076507045,20.820734342661488,20.98909146093353,3.7795275590551185,16.310181392258002],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(51,102,204,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"REAC","legendgroup":"REAC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[198.55152814351155,198.9393579468462,199.12148796498681,198.95649959561237,199.48146258907641,199.17719832347686,199.35718563552166,198.99935371752781,198.65652074220438,198.24512117181624,199.07863384307134,198.87079135178152,199.99142663987004,198.82793722986608,198.49796049111728,198.73794357384369,199.04220783944325,199.44503658544832,199.24362221244579,199.13005878936985,198.8343653481534,199.45146470373558,199.89500486556034,198.63509368124664,199.62288119139734,199.75144355714363,199.80072579734636,199.86072156802797,198.56866979227775,198.27083364496551],"y":[2.8078885861687253,2.5200441180456949,2.5200441180456949,2.1399595865885841,1.8953620494176195,1.8953620494176195,1.8953620494176195,1.6511579661900624,1.5147783179892784,1.5147783179892784,1.4465097057229968,1.4465097057229968,1.4465097057229968,1.4465097057229968,1.4459489421984575,1.4347264106025992,1.4347264106025992,1.4347264106025992,1.4347264106025992,1.4347264106025992,1.3772381076819589,1.3696142275167016,1.3503035023717875,1.3298331939424508,1.3298331939424508,1.3298331939424508,1.3298331939424508,1.3298331939424508,1.3298331939424508,1.3298331939424508],"text":["WP:WP5095 (127) <br> Overview of proinflammatory and profibrotic mediators <br> 1.556e-03","WP:WP619 (37) <br> Type II interferon signaling <br> 3.020e-03","WP:WP5088 (33) <br> Prostaglandin signaling <br> 3.020e-03","WP:WP5039 (66) <br> SARS-CoV-2 innate immunity evasion and cell-specific immune response <br> 7.245e-03","WP:WP2431 (116) <br> Spinal cord injury <br> 1.272e-02","WP:WP75 (103) <br> Toll-like receptor signaling pathway <br> 1.272e-02","WP:WP5322 (116) <br> CKAP4 signaling pathway map <br> 1.272e-02","WP:WP3929 (165) <br> Chemokine signaling pathway <br> 2.233e-02","WP:WP3934 (10) <br> Leptin and adiponectin <br> 3.056e-02","WP:WP5115 (217) <br> Network map of SARS-CoV-2 signaling pathway <br> 3.056e-02","WP:WP2435 (16) <br> Quercetin and Nf-kB / AP-1 induced apoptosis <br> 3.577e-02","WP:WP5400 (14) <br> 6q16 copy number variation <br> 3.577e-02","WP:WP4891 (15) <br> COVID-19 adverse outcome pathway <br> 3.577e-02","WP:WP3935 (17) <br> Leptin-insulin signaling overlap <br> 3.577e-02","WP:WP5413 (290) <br> IL-24 Signaling pathway <br> 3.581e-02","WP:WP4483 (25) <br> Relationship between inflammation, COX-2 and EGFR <br> 3.675e-02","WP:WP3599 (22) <br> Transcription factor regulation in adipogenesis <br> 3.675e-02","WP:WP5284 (25) <br> Cell interactions of the pancreatic cancer microenvironment <br> 3.675e-02","WP:WP2895 (25) <br> Differentiation of white and brown adipocyte <br> 3.675e-02","WP:WP4537 (22) <br> Hippo-YAP signaling <br> 3.675e-02","WP:WP129 (30) <br> Matrix metalloproteinases <br> 4.195e-02","WP:WP5113 (32) <br> Antiviral and anti-inflammatory effects of Nrf2 on SARS-CoV-2 pathway <br> 4.270e-02","WP:WP3617 (35) <br> Photodynamic therapy-induced NF-kB survival signaling <br> 4.464e-02","WP:WP5283 (46) <br> Chronic hyperglycemia impairment of neuron function <br> 4.679e-02","WP:WP5347 (45) <br> IL-26 signaling pathways <br> 4.679e-02","WP:WP4559 (48) <br> Interactions between immune cells and microRNAs in tumor microenvironment <br> 4.679e-02","WP:WP716 (42) <br> Vitamin A and carotenoid metabolism <br> 4.679e-02","WP:WP4136 (43) <br> Fibrin complement receptor 3 signaling pathway <br> 4.679e-02","WP:WP2828 (40) <br> Bladder cancer <br> 4.679e-02","WP:WP1971 (48) <br> Integrated cancer pathway <br> 4.679e-02"],"key":["WP:WP5095","WP:WP619","WP:WP5088","WP:WP5039","WP:WP2431","WP:WP75","WP:WP5322","WP:WP3929","WP:WP3934","WP:WP5115","WP:WP2435","WP:WP5400","WP:WP4891","WP:WP3935","WP:WP5413","WP:WP4483","WP:WP3599","WP:WP5284","WP:WP2895","WP:WP4537","WP:WP129","WP:WP5113","WP:WP3617","WP:WP5283","WP:WP5347","WP:WP4559","WP:WP716","WP:WP4136","WP:WP2828","WP:WP1971"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,153,198,1)","opacity":0.80000000000000004,"size":[17.495593428699955,15.621605526743602,15.432489933172905,16.5353429714215,17.366727849814772,17.195781823273968,17.366727849814772,17.861295873015131,13.235951909580116,18.234128393371993,14.156289046486304,13.903326714182342,14.034806173846196,14.269123212349129,18.6185768464894,14.960284846058837,14.736021255617715,14.960284846058837,14.960284846058837,14.736021255617715,15.2725704202277,15.381099595154099,15.530131005135631,15.973394555191513,15.938343791671642,16.040981456476214,15.827655299285238,15.865520289428702,15.748761075898194,16.040981456476214],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(0,153,198,1)"}},"hoveron":"points","set":"SharedDatab8a98bf9","name":"WP","legendgroup":"WP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[2,26.046954040382204],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(220,57,18,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[28.187124389294823,36.863374983786585],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(16,150,24,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[39.003545332699204,98.063686281291879],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(255,153,0,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[100.2038566302045,101.4087725366423],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(221,68,119,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[103.54894288555494,109.17759090319512],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(51,102,204,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[111.31776125210774,136.24432530589303],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(85,116,166,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[138.38449565480565,144.07734878291322],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(34,170,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[146.21751913182584,148.05592546154176],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(102,51,204,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[150.19609581045438,156.43897271823249],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(102,170,0,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[158.57914306714511,196.05352587660511],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(153,0,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[198.19369622551773,200.00214017034889],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(0,153,198,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[180],"y":[16.199999999999999],"text":"values above this threshold are capped","hovertext":"","textfont":{"size":7.559055118110237,"color":"rgba(190,190,190,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,210],"y":[16,16],"text":"","type":"scatter","mode":"lines","line":{"width":0.75590551181102372,"color":"rgba(190,190,190,1)","dash":"dash"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":29.283520132835207,"r":6.6417600664176026,"b":55.878166138820198,"l":61.037775010377771},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0,210],"tickmode":"array","ticktext":["GO:MF","GO:CC","GO:BP","KEGG","REAC","TF","MIRNA","HPA","CORUM","HP","WP"],"tickvals":[14.023477020191102,32.525249686540704,68.533615806995542,100.8063145834234,106.36326689437503,123.78104327900039,141.23092221885943,147.13672229668379,153.31753426434344,177.31633447187511,199.09791819793332],"categoryorder":"array","categoryarray":["GO:MF","GO:CC","GO:BP","KEGG","REAC","TF","MIRNA","HPA","CORUM","HP","WP"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.6529680365296811,"tickwidth":0,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":10.62681610626816},"tickangle":-45,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.13283520132835203,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":"","hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1,18.5],"tickmode":"array","ticktext":["0","2","4","6","8","10","12","14",">16"],"tickvals":[0,2,4,6,8,10,12,14,16],"categoryorder":"array","categoryarray":["0","2","4","6","8","10","12","14",">16"],"nticks":null,"ticks":"outside","tickcolor":"rgba(190,190,190,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"annotations":[{"text":"-log10(p-adj)","x":0,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":13.283520132835205},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis","xshift":-37.393109173931101},{"text":"query_1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(169,169,169,1)","family":"","size":13.283520132835205},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"}],"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":24.973017849730184,"yanchor":1,"ysizemode":"pixel"}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498}},"hovermode":"closest","barmode":"relative","dragmode":"zoom"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"ff7e47873a36":{"x":{},"y":{},"text":{},"colour":{},"size":{},"alpha":{},"type":"scatter"},"ff7e7efed3e9":{"x":{},"xend":{},"y":{},"yend":{}},"ff7ec0aa722":{"x":{},"xend":{},"y":{},"yend":{}},"ff7ef62af5b":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e17e68741":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e1fb9c8a9":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e5e24df84":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e46ce8365":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e20148034":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e21ece660":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e4808ba06":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e34ed04dc":{"x":{},"xend":{},"y":{},"yend":{}},"ff7e34da22a9":{"x":{},"y":{}},"ff7e5d19a05a":{"yintercept":{}}},"cur_data":"ff7e47873a36","visdat":{"ff7e47873a36":["function (y) ","x"],"ff7e7efed3e9":["function (y) ","x"],"ff7ec0aa722":["function (y) ","x"],"ff7ef62af5b":["function (y) ","x"],"ff7e17e68741":["function (y) ","x"],"ff7e1fb9c8a9":["function (y) ","x"],"ff7e5e24df84":["function (y) ","x"],"ff7e46ce8365":["function (y) ","x"],"ff7e20148034":["function (y) ","x"],"ff7e21ece660":["function (y) ","x"],"ff7e4808ba06":["function (y) ","x"],"ff7e34ed04dc":["function (y) ","x"],"ff7e34da22a9":["function (y) ","x"],"ff7e5d19a05a":["function (y) ","x"]},"highlight":{"on":"plotly_click","off":"plotly_doubleclick","persistent":false,"dynamic":false,"color":null,"selectize":false,"defaultValues":null,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0,"ctGroups":["SharedDatab8a98bf9"]},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co">### produce a publication quality static manhattan plot with specific GO terms highlighted</span></span>
<span><span class="va">mygostplot</span> <span class="op">&lt;-</span> <span class="fu">gostplot</span><span class="op">(</span></span>
<span>  <span class="va">gost_res</span>, </span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  capped <span class="op">=</span> <span class="cn">TRUE</span> <span class="co">#set interactive=FALSE to get plot for publications</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="co">### rerun the above gostplot function with 'interactive=F' and save to an object 'mygostplot'</span></span>
<span><span class="fu">publish_gostplot</span><span class="op">(</span></span>
<span>  <span class="va">mygostplot</span>, <span class="co">#your static gostplot from above</span></span>
<span>  highlight_terms <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GO:0034987"</span><span class="op">)</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  width <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  height <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co">### generate a table of  gost results</span></span>
<span><span class="fu">publish_gosttable</span><span class="op">(</span></span>
<span>  <span class="va">gost_res</span>,</span>
<span>  highlight_terms <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_colors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"source"</span>, <span class="st">"term_name"</span>, <span class="st">"term_size"</span>, <span class="st">"intersection_size"</span><span class="op">)</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ggplot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-4.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># now repeat the above steps using only genes from a single module from the step 6 script, by using `rownames(myModule)`</span></span>
<span><span class="co"># what is value in breaking up DEGs into modules for functional enrichment analysis?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="perform-gsea-using-clusterprofiler" class="level3"><h3 class="anchored" data-anchor-id="perform-gsea-using-clusterprofiler">Perform GSEA using clusterProfiler</h3>
<p>There are a few ways to get msigDB collections into R:</p>
<ul>
<li>Option1: download directly from msigdb and load from your computer. can use the â€˜read.gmtâ€™ function from clusterProfiler package to create a dataframe. alternatively, you can read in using â€˜getGmtâ€™ function from GSEABase package if you need a GeneSetCollection object</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">c2cp</span> <span class="op">&lt;-</span> <span class="fu">read.gmt</span><span class="op">(</span><span class="st">"/Users/.../MSigDB/c2.cp.v7.1.symbols.gmt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Option2: use the msigdb package to access up-to-date collections. this option has the additional advantage of providing access to species-specific collections are also retrieved as tibbles</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">msigdbr_species</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 20 Ã— 2</span></span>
<span><span class="co">##    species_name                    species_common_name                          </span></span>
<span><span class="co">##    &lt;chr&gt;                           &lt;chr&gt;                                        </span></span>
<span><span class="co">##  1 Anolis carolinensis             Carolina anole, green anole                  </span></span>
<span><span class="co">##  2 Bos taurus                      bovine, cattle, cow, dairy cow, domestic catâ€¦</span></span>
<span><span class="co">##  3 Caenorhabditis elegans          &lt;NA&gt;                                         </span></span>
<span><span class="co">##  4 Canis lupus familiaris          dog, dogs                                    </span></span>
<span><span class="co">##  5 Danio rerio                     leopard danio, zebra danio, zebra fish, zebrâ€¦</span></span>
<span><span class="co">##  6 Drosophila melanogaster         fruit fly                                    </span></span>
<span><span class="co">##  7 Equus caballus                  domestic horse, equine, horse                </span></span>
<span><span class="co">##  8 Felis catus                     cat, cats, domestic cat                      </span></span>
<span><span class="co">##  9 Gallus gallus                   bantam, chicken, chickens, Gallus domesticus </span></span>
<span><span class="co">## 10 Homo sapiens                    human                                        </span></span>
<span><span class="co">## 11 Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monkâ€¦</span></span>
<span><span class="co">## 12 Monodelphis domestica           gray short-tailed opossum                    </span></span>
<span><span class="co">## 13 Mus musculus                    house mouse, mouse                           </span></span>
<span><span class="co">## 14 Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, plaâ€¦</span></span>
<span><span class="co">## 15 Pan troglodytes                 chimpanzee                                   </span></span>
<span><span class="co">## 16 Rattus norvegicus               brown rat, Norway rat, rat, rats             </span></span>
<span><span class="co">## 17 Saccharomyces cerevisiae        baker's yeast, brewer's yeast, S. cerevisiae </span></span>
<span><span class="co">## 18 Schizosaccharomyces pombe 972h- &lt;NA&gt;                                         </span></span>
<span><span class="co">## 19 Sus scrofa                      pig, pigs, swine, wild boar                  </span></span>
<span><span class="co">## 20 Xenopus tropicalis              tropical clawed frog, western clawed frog</span></span>
<span><span class="co">### Gets all collections/signatures with human gene IDs</span></span>
<span><span class="va">hs_gsea</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">### Take a look at the categories and subcategories of signatures available to you</span></span>
<span><span class="va">hs_gsea</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">distinct</span><span class="op">(</span><span class="va">gs_cat</span>, <span class="va">gs_subcat</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gs_cat</span>, <span class="va">gs_subcat</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 23 Ã— 2</span></span>
<span><span class="co">##    gs_cat gs_subcat        </span></span>
<span><span class="co">##    &lt;chr&gt;  &lt;chr&gt;            </span></span>
<span><span class="co">##  1 C1     ""               </span></span>
<span><span class="co">##  2 C2     "CGP"            </span></span>
<span><span class="co">##  3 C2     "CP"             </span></span>
<span><span class="co">##  4 C2     "CP:BIOCARTA"    </span></span>
<span><span class="co">##  5 C2     "CP:KEGG"        </span></span>
<span><span class="co">##  6 C2     "CP:PID"         </span></span>
<span><span class="co">##  7 C2     "CP:REACTOME"    </span></span>
<span><span class="co">##  8 C2     "CP:WIKIPATHWAYS"</span></span>
<span><span class="co">##  9 C3     "MIR:MIRDB"      </span></span>
<span><span class="co">## 10 C3     "MIR:MIR_Legacy" </span></span>
<span><span class="co">## # â„¹ 13 more rows</span></span>
<span></span>
<span><span class="co">### Choose a specific msigdb collection/subcollection</span></span>
<span><span class="va">hs_gsea_c2</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span></span>
<span>  <span class="co">### change depending on species your data came from</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, </span>
<span>  category <span class="op">=</span> <span class="st">"C2"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="co">###just get the columns corresponding to signature name and gene symbols of genes in each signature </span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">### Pull out just the columns corresponding to gene symbols and LogFC for at least one pairwise comparison for the enrichment analysis</span></span>
<span><span class="va">mydata_df_sub</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">mydata_df</span>, <span class="va">geneID</span>, <span class="va">LogFC</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Construct a named vector</span></span>
<span><span class="va">mydata_gsea</span> <span class="op">&lt;-</span> <span class="va">mydata_df_sub</span><span class="op">$</span><span class="va">LogFC</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mydata_gsea</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">mydata_df_sub</span><span class="op">$</span><span class="va">geneID</span><span class="op">)</span></span>
<span><span class="va">mydata_gsea</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">mydata_gsea</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Run GSEA using the 'GSEA' function from clusterProfiler</span></span>
<span><span class="va">myGSEA_res</span> <span class="op">&lt;-</span> <span class="fu">GSEA</span><span class="op">(</span><span class="va">mydata_gsea</span>, TERM2GENE<span class="op">=</span><span class="va">hs_gsea_c2</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">myGSEA_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">myGSEA_res</span><span class="op">@</span><span class="va">result</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### View results as an interactive table</span></span>
<span><span class="co"># datatable(</span></span>
<span><span class="co">#     myGSEA_df,</span></span>
<span><span class="co">#     extensions = c('KeyTable', "FixedHeader"),</span></span>
<span><span class="co">#     caption = 'Signatures enriched in leishmaniasis',</span></span>
<span><span class="co">#     options = list(</span></span>
<span><span class="co">#         keys = TRUE,</span></span>
<span><span class="co">#         searchHighlight = TRUE,</span></span>
<span><span class="co">#         pageLength = 10,</span></span>
<span><span class="co">#         lengthMenu = c("10", "25", "50", "100")</span></span>
<span><span class="co">#     )</span></span>
<span><span class="co"># ) |&gt;</span></span>
<span><span class="co">#     formatRound(columns = c(2:10), digits = 2)</span></span>
<span></span>
<span><span class="co">### Create enrichment plots using the enrichplot package</span></span>
<span><span class="fu">gseaplot2</span><span class="op">(</span></span>
<span>    <span class="va">myGSEA_res</span>,</span>
<span>    <span class="co">### can choose multiple signatures to overlay in this plot</span></span>
<span>    geneSetID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">47</span>, <span class="fl">262</span><span class="op">)</span>, </span>
<span>    <span class="co">### can set this to FALSE for a cleaner plot</span></span>
<span>    pvalue_table <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    <span class="co">### can also turn off this title</span></span>
<span>    title <span class="op">=</span> <span class="va">myGSEA_res</span><span class="op">$</span><span class="va">Description</span><span class="op">[</span><span class="fl">47</span><span class="op">]</span></span>
<span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co">### Add a variable to this result that matches enrichment direction with phenotype</span></span>
<span><span class="va">myGSEA_df</span> <span class="op">&lt;-</span> <span class="va">myGSEA_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span></span>
<span>        phenotype <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>            <span class="va">NES</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"disease"</span>,</span>
<span>            <span class="va">NES</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"healthy"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co">### Create 'bubble plot' to summarize y signatures across x phenotypes</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">myGSEA_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="op">]</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">phenotype</span>, y<span class="op">=</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>size<span class="op">=</span><span class="va">setSize</span>, color <span class="op">=</span> <span class="va">NES</span>, alpha<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.adjust</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_gradient</span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue"</span>, high<span class="op">=</span><span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-25-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="competitive-gsea-using-camera" class="level3"><h3 class="anchored" data-anchor-id="competitive-gsea-using-camera">Competitive GSEA using CAMERA</h3>
<p>For competitive tests the null hypothesis is that genes in the set are, at most, as often differentially expressed as genes outside the set.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Create a few signatures to test in our enrichment analysis</span></span>
<span><span class="co">### if your own data is from mice, wrap this in 'toupper()' to make gene symbols all caps</span></span>
<span><span class="va">mySig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">myTopHits</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">mySig2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span></span>
<span>  <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  size <span class="op">=</span> <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">collection</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>real <span class="op">=</span> <span class="va">mySig</span>, fake <span class="op">=</span> <span class="va">mySig2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###Test for enrichment using CAMERA</span></span>
<span><span class="va">camera_res</span> <span class="op">&lt;-</span> <span class="fu">camera</span><span class="op">(</span><span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span>, <span class="va">collection</span>, <span class="va">design</span>, <span class="va">contrast_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> </span>
<span><span class="va">camera_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">camera_res</span>, rownames <span class="op">=</span> <span class="st">"setName"</span><span class="op">)</span></span>
<span><span class="va">camera_df</span></span>
<span><span class="co">## # A tibble: 2 Ã— 5</span></span>
<span><span class="co">##   setName NGenes Direction   PValue      FDR</span></span>
<span><span class="co">##   &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span><span class="co">## 1 real        50 Up        5.85e-20 1.17e-19</span></span>
<span><span class="co">## 2 fake        50 Up        9.57e- 1 9.57e- 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="self-contained-gsea-using-roast" class="level3"><h3 class="anchored" data-anchor-id="self-contained-gsea-using-roast">Self-contained GSEA using ROAST</h3>
<p>For self-contained the null hypothesis is that no genes in the set are differentially expressed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">mroast</span><span class="op">(</span></span>
<span>  <span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span>, </span>
<span>  <span class="va">collection</span>, <span class="va">design</span>, </span>
<span>  <span class="co">#mroast adjusts for multiple testing</span></span>
<span>  contrast<span class="op">=</span><span class="fl">1</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="co">##      NGenes PropDown PropUp Direction PValue   FDR PValue.Mixed FDR.Mixed</span></span>
<span><span class="co">## real     50     0.06   0.94        Up  5e-04 5e-04        5e-04     5e-04</span></span>
<span><span class="co">## fake     50     0.04   0.86        Up  5e-04 5e-04        5e-04     5e-04</span></span>
<span></span>
<span><span class="co">### Repeat with an actual gene set collection</span></span>
<span><span class="co">### camera requires collections to be presented as a list, rather than a tibble, so we must read in our signatures using the 'getGmt' function</span></span>
<span><span class="va">broadSet_C2_ALL</span> <span class="op">&lt;-</span> <span class="fu">getGmt</span><span class="op">(</span></span>
<span>    <span class="fu">here</span><span class="op">(</span><span class="st">"learn"</span>, <span class="st">"DIY_Transcriptomics"</span>, <span class="st">"data"</span>, <span class="st">"c2.all.v7.2.symbols.gmt"</span><span class="op">)</span>,</span>
<span>    geneIdType <span class="op">=</span> <span class="fu">SymbolIdentifier</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Extract as a list</span></span>
<span><span class="va">broadSet_C2_ALL</span> <span class="op">&lt;-</span> <span class="fu">geneIds</span><span class="op">(</span><span class="va">broadSet_C2_ALL</span><span class="op">)</span></span>
<span><span class="va">camera_res</span> <span class="op">&lt;-</span> <span class="fu">camera</span><span class="op">(</span></span>
<span>    <span class="va">v_DEGList_filtered_norm</span><span class="op">$</span><span class="va">E</span>,</span>
<span>    <span class="va">broadSet_C2_ALL</span>, <span class="va">design</span>,</span>
<span>    <span class="va">contrast_matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">camera_df</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">camera_res</span>, rownames <span class="op">=</span> <span class="st">"setName"</span><span class="op">)</span></span>
<span><span class="va">camera_df</span></span>
<span><span class="co">## # A tibble: 6,229 Ã— 5</span></span>
<span><span class="co">##    setName                                    NGenes Direction   PValue      FDR</span></span>
<span><span class="co">##    &lt;chr&gt;                                       &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span><span class="co">##  1 MCLACHLAN_DENTAL_CARIES_UP                    245 Up        2.80e-33 1.74e-29</span></span>
<span><span class="co">##  2 POOLA_INVASIVE_BREAST_CANCER_UP               275 Up        2.59e-32 8.06e-29</span></span>
<span><span class="co">##  3 WIELAND_UP_BY_HBV_INFECTION                    99 Up        2.10e-31 4.36e-28</span></span>
<span><span class="co">##  4 REACTOME_IMMUNOREGULATORY_INTERACTIONS_BEâ€¦    158 Up        9.85e-31 1.53e-27</span></span>
<span><span class="co">##  5 REACTOME_FCGR_ACTIVATION                       62 Up        1.88e-29 2.34e-26</span></span>
<span><span class="co">##  6 REACTOME_CD22_MEDIATED_BCR_REGULATION          53 Up        2.62e-27 2.72e-24</span></span>
<span><span class="co">##  7 REACTOME_CREATION_OF_C4_AND_C2_ACTIVATORS      58 Up        1.70e-26 1.52e-23</span></span>
<span><span class="co">##  8 REACTOME_INITIAL_TRIGGERING_OF_COMPLEMENT      66 Up        2.81e-26 2.19e-23</span></span>
<span><span class="co">##  9 WALLACE_PROSTATE_CANCER_RACE_UP               277 Up        6.49e-26 4.49e-23</span></span>
<span><span class="co">## 10 FLECHNER_BIOPSY_KIDNEY_TRANSPLANT_REJECTEâ€¦     86 Up        4.32e-25 2.69e-22</span></span>
<span><span class="co">## # â„¹ 6,219 more rows</span></span>
<span></span>
<span><span class="co">### Filter based on FDR and display as interactive table</span></span>
<span><span class="va">camera_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">camera_df</span>, <span class="va">FDR</span><span class="op">&lt;=</span><span class="fl">0.01</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># datatable(</span></span>
<span><span class="co">#     camera_df,</span></span>
<span><span class="co">#     extensions = c('KeyTable', "FixedHeader"),</span></span>
<span><span class="co">#     caption = 'Signatures enriched in leishmaniasis',</span></span>
<span><span class="co">#     options = list(</span></span>
<span><span class="co">#         keys = TRUE,</span></span>
<span><span class="co">#         searchHighlight = TRUE,</span></span>
<span><span class="co">#         pageLength = 10,</span></span>
<span><span class="co">#         lengthMenu = c("10", "25", "50", "100")</span></span>
<span><span class="co">#     )</span></span>
<span><span class="co"># ) |&gt;</span></span>
<span><span class="co">#     formatRound(columns = c(2, 4, 5), digits = 2)</span></span>
<span></span>
<span><span class="co">### Add a variable that maps up/down regulated pathways with phenotype</span></span>
<span><span class="va">camera_df</span> <span class="op">&lt;-</span> <span class="va">camera_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span></span>
<span>        phenotype <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>            <span class="va">Direction</span> <span class="op">==</span> <span class="st">"Up"</span> <span class="op">~</span> <span class="st">"disease"</span>,</span>
<span>            <span class="va">Direction</span> <span class="op">==</span> <span class="st">"Down"</span> <span class="op">~</span> <span class="st">"healthy"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co">### Filtering to return anything that has 'CD8' or 'CYTOTOX' in the name of the signature</span></span>
<span><span class="va">camera_df_sub</span> <span class="op">&lt;-</span> <span class="va">camera_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu">str_detect</span><span class="op">(</span><span class="va">setName</span>, <span class="st">"CD8|CYTOTOX"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Graph camera results as bubble chart </span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">camera_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>,<span class="op">]</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">phenotype</span>, y<span class="op">=</span><span class="va">setName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>size<span class="op">=</span><span class="va">NGenes</span>, color <span class="op">=</span> <span class="va">Direction</span>, alpha<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="single-sample-gsea-using-the-gsva-package" class="level3"><h3 class="anchored" data-anchor-id="single-sample-gsea-using-the-gsva-package">Single sample GSEA using the GSVA package</h3>
<p>The GSVA package offers a different way of approaching functional enrichment analysis.</p>
<ul>
<li>In contrast to most GSE methods, GSVA performs a change in coordinate systems,transforming the data from a gene by sample matrix to a gene set (signature) by sample matrix. This allows for the evaluation of pathway enrichment for each sample.</li>
<li>The method is both non-parametric and unsupervised bypasses the conventional approach of explicitly modeling phenotypes within enrichment scoring algorithms.</li>
<li>Focus is therefore placed on the RELATIVE enrichment of pathways across the sample space rather than the absolute enrichment with respect to a phenotype. However, with data with a moderate to small sample size (&lt; 30), other GSE methods that explicitly include the phenotype in their model are more likely to provide greater statistical power to detect functional enrichment.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Be aware that if you choose a large MsigDB file here, this step may take a while</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>GSVA_res_C2CP <span class="ot">&lt;-</span> <span class="fu">gsva</span>(</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    v_DEGList_filtered_norm<span class="sc">$</span>E,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="do">### signatures</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>    broadSet_C2_ALL,</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    <span class="do">### criteria for filtering gene sets</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.sz =</span> <span class="dv">5</span>, <span class="at">max.sz =</span> <span class="dv">500</span>,</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">mx.diff =</span> <span class="cn">FALSE</span>,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>    <span class="do">### options for method are "gsva", ssgsea', "zscore" or "plage"</span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"gsva"</span></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Estimating GSVA scores for 5919 gene sets.</span></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Estimating ECDFs with Gaussian kernels</span></span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">1</span>%</span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=</span>                                                                     <span class="er">|</span>   <span class="dv">1</span>%</span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=</span>                                                                     <span class="er">|</span>   <span class="dv">2</span>%</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==</span>                                                                    <span class="er">|</span>   <span class="dv">2</span>%</span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==</span>                                                                    <span class="er">|</span>   <span class="dv">3</span>%</span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==</span>                                                                    <span class="er">|</span>   <span class="dv">4</span>%</span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===</span>                                                                   <span class="er">|</span>   <span class="dv">4</span>%</span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===</span>                                                                   <span class="er">|</span>   <span class="dv">5</span>%</span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====</span>                                                                  <span class="er">|</span>   <span class="dv">5</span>%</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====</span>                                                                  <span class="er">|</span>   <span class="dv">6</span>%</span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====</span>                                                                 <span class="er">|</span>   <span class="dv">6</span>%</span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====</span>                                                                 <span class="er">|</span>   <span class="dv">7</span>%</span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====</span>                                                                 <span class="er">|</span>   <span class="dv">8</span>%</span>
<span id="cb43-43"><a href="#cb43-43" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-44"><a href="#cb43-44" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======</span>                                                                <span class="er">|</span>   <span class="dv">8</span>%</span>
<span id="cb43-45"><a href="#cb43-45" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-46"><a href="#cb43-46" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======</span>                                                                <span class="er">|</span>   <span class="dv">9</span>%</span>
<span id="cb43-47"><a href="#cb43-47" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-48"><a href="#cb43-48" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======</span>                                                               <span class="er">|</span>   <span class="dv">9</span>%</span>
<span id="cb43-49"><a href="#cb43-49" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-50"><a href="#cb43-50" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======</span>                                                               <span class="er">|</span>  <span class="dv">10</span>%</span>
<span id="cb43-51"><a href="#cb43-51" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-52"><a href="#cb43-52" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======</span>                                                               <span class="er">|</span>  <span class="dv">11</span>%</span>
<span id="cb43-53"><a href="#cb43-53" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-54"><a href="#cb43-54" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========</span>                                                              <span class="er">|</span>  <span class="dv">11</span>%</span>
<span id="cb43-55"><a href="#cb43-55" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-56"><a href="#cb43-56" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========</span>                                                              <span class="er">|</span>  <span class="dv">12</span>%</span>
<span id="cb43-57"><a href="#cb43-57" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-58"><a href="#cb43-58" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========</span>                                                             <span class="er">|</span>  <span class="dv">12</span>%</span>
<span id="cb43-59"><a href="#cb43-59" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-60"><a href="#cb43-60" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========</span>                                                             <span class="er">|</span>  <span class="dv">13</span>%</span>
<span id="cb43-61"><a href="#cb43-61" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-62"><a href="#cb43-62" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========</span>                                                             <span class="er">|</span>  <span class="dv">14</span>%</span>
<span id="cb43-63"><a href="#cb43-63" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-64"><a href="#cb43-64" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========</span>                                                            <span class="er">|</span>  <span class="dv">14</span>%</span>
<span id="cb43-65"><a href="#cb43-65" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-66"><a href="#cb43-66" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========</span>                                                            <span class="er">|</span>  <span class="dv">15</span>%</span>
<span id="cb43-67"><a href="#cb43-67" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-68"><a href="#cb43-68" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========</span>                                                           <span class="er">|</span>  <span class="dv">15</span>%</span>
<span id="cb43-69"><a href="#cb43-69" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-70"><a href="#cb43-70" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========</span>                                                           <span class="er">|</span>  <span class="dv">16</span>%</span>
<span id="cb43-71"><a href="#cb43-71" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-72"><a href="#cb43-72" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============</span>                                                          <span class="er">|</span>  <span class="dv">16</span>%</span>
<span id="cb43-73"><a href="#cb43-73" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-74"><a href="#cb43-74" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============</span>                                                          <span class="er">|</span>  <span class="dv">17</span>%</span>
<span id="cb43-75"><a href="#cb43-75" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-76"><a href="#cb43-76" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============</span>                                                          <span class="er">|</span>  <span class="dv">18</span>%</span>
<span id="cb43-77"><a href="#cb43-77" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-78"><a href="#cb43-78" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============</span>                                                         <span class="er">|</span>  <span class="dv">18</span>%</span>
<span id="cb43-79"><a href="#cb43-79" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-80"><a href="#cb43-80" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============</span>                                                         <span class="er">|</span>  <span class="dv">19</span>%</span>
<span id="cb43-81"><a href="#cb43-81" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-82"><a href="#cb43-82" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============</span>                                                        <span class="er">|</span>  <span class="dv">19</span>%</span>
<span id="cb43-83"><a href="#cb43-83" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-84"><a href="#cb43-84" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============</span>                                                        <span class="er">|</span>  <span class="dv">20</span>%</span>
<span id="cb43-85"><a href="#cb43-85" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-86"><a href="#cb43-86" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============</span>                                                        <span class="er">|</span>  <span class="dv">21</span>%</span>
<span id="cb43-87"><a href="#cb43-87" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-88"><a href="#cb43-88" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============</span>                                                       <span class="er">|</span>  <span class="dv">21</span>%</span>
<span id="cb43-89"><a href="#cb43-89" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-90"><a href="#cb43-90" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============</span>                                                       <span class="er">|</span>  <span class="dv">22</span>%</span>
<span id="cb43-91"><a href="#cb43-91" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-92"><a href="#cb43-92" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================</span>                                                      <span class="er">|</span>  <span class="dv">22</span>%</span>
<span id="cb43-93"><a href="#cb43-93" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-94"><a href="#cb43-94" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================</span>                                                      <span class="er">|</span>  <span class="dv">23</span>%</span>
<span id="cb43-95"><a href="#cb43-95" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-96"><a href="#cb43-96" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================</span>                                                      <span class="er">|</span>  <span class="dv">24</span>%</span>
<span id="cb43-97"><a href="#cb43-97" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-98"><a href="#cb43-98" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================</span>                                                     <span class="er">|</span>  <span class="dv">24</span>%</span>
<span id="cb43-99"><a href="#cb43-99" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-100"><a href="#cb43-100" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================</span>                                                     <span class="er">|</span>  <span class="dv">25</span>%</span>
<span id="cb43-101"><a href="#cb43-101" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-102"><a href="#cb43-102" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================</span>                                                    <span class="er">|</span>  <span class="dv">25</span>%</span>
<span id="cb43-103"><a href="#cb43-103" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-104"><a href="#cb43-104" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================</span>                                                    <span class="er">|</span>  <span class="dv">26</span>%</span>
<span id="cb43-105"><a href="#cb43-105" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-106"><a href="#cb43-106" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================</span>                                                   <span class="er">|</span>  <span class="dv">26</span>%</span>
<span id="cb43-107"><a href="#cb43-107" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-108"><a href="#cb43-108" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================</span>                                                   <span class="er">|</span>  <span class="dv">27</span>%</span>
<span id="cb43-109"><a href="#cb43-109" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-110"><a href="#cb43-110" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================</span>                                                   <span class="er">|</span>  <span class="dv">28</span>%</span>
<span id="cb43-111"><a href="#cb43-111" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-112"><a href="#cb43-112" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================</span>                                                  <span class="er">|</span>  <span class="dv">28</span>%</span>
<span id="cb43-113"><a href="#cb43-113" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-114"><a href="#cb43-114" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================</span>                                                  <span class="er">|</span>  <span class="dv">29</span>%</span>
<span id="cb43-115"><a href="#cb43-115" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-116"><a href="#cb43-116" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================</span>                                                 <span class="er">|</span>  <span class="dv">29</span>%</span>
<span id="cb43-117"><a href="#cb43-117" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-118"><a href="#cb43-118" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================</span>                                                 <span class="er">|</span>  <span class="dv">30</span>%</span>
<span id="cb43-119"><a href="#cb43-119" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-120"><a href="#cb43-120" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================</span>                                                 <span class="er">|</span>  <span class="dv">31</span>%</span>
<span id="cb43-121"><a href="#cb43-121" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-122"><a href="#cb43-122" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================</span>                                                <span class="er">|</span>  <span class="dv">31</span>%</span>
<span id="cb43-123"><a href="#cb43-123" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-124"><a href="#cb43-124" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================</span>                                                <span class="er">|</span>  <span class="dv">32</span>%</span>
<span id="cb43-125"><a href="#cb43-125" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-126"><a href="#cb43-126" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================</span>                                               <span class="er">|</span>  <span class="dv">32</span>%</span>
<span id="cb43-127"><a href="#cb43-127" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-128"><a href="#cb43-128" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================</span>                                               <span class="er">|</span>  <span class="dv">33</span>%</span>
<span id="cb43-129"><a href="#cb43-129" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-130"><a href="#cb43-130" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================</span>                                               <span class="er">|</span>  <span class="dv">34</span>%</span>
<span id="cb43-131"><a href="#cb43-131" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-132"><a href="#cb43-132" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================</span>                                              <span class="er">|</span>  <span class="dv">34</span>%</span>
<span id="cb43-133"><a href="#cb43-133" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-134"><a href="#cb43-134" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================</span>                                              <span class="er">|</span>  <span class="dv">35</span>%</span>
<span id="cb43-135"><a href="#cb43-135" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-136"><a href="#cb43-136" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================</span>                                             <span class="er">|</span>  <span class="dv">35</span>%</span>
<span id="cb43-137"><a href="#cb43-137" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-138"><a href="#cb43-138" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================</span>                                             <span class="er">|</span>  <span class="dv">36</span>%</span>
<span id="cb43-139"><a href="#cb43-139" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-140"><a href="#cb43-140" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================</span>                                            <span class="er">|</span>  <span class="dv">36</span>%</span>
<span id="cb43-141"><a href="#cb43-141" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-142"><a href="#cb43-142" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================</span>                                            <span class="er">|</span>  <span class="dv">37</span>%</span>
<span id="cb43-143"><a href="#cb43-143" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-144"><a href="#cb43-144" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================</span>                                            <span class="er">|</span>  <span class="dv">38</span>%</span>
<span id="cb43-145"><a href="#cb43-145" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-146"><a href="#cb43-146" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================</span>                                           <span class="er">|</span>  <span class="dv">38</span>%</span>
<span id="cb43-147"><a href="#cb43-147" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-148"><a href="#cb43-148" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================</span>                                           <span class="er">|</span>  <span class="dv">39</span>%</span>
<span id="cb43-149"><a href="#cb43-149" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-150"><a href="#cb43-150" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================</span>                                          <span class="er">|</span>  <span class="dv">39</span>%</span>
<span id="cb43-151"><a href="#cb43-151" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-152"><a href="#cb43-152" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================</span>                                          <span class="er">|</span>  <span class="dv">40</span>%</span>
<span id="cb43-153"><a href="#cb43-153" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-154"><a href="#cb43-154" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================</span>                                          <span class="er">|</span>  <span class="dv">41</span>%</span>
<span id="cb43-155"><a href="#cb43-155" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-156"><a href="#cb43-156" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================</span>                                         <span class="er">|</span>  <span class="dv">41</span>%</span>
<span id="cb43-157"><a href="#cb43-157" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-158"><a href="#cb43-158" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================</span>                                         <span class="er">|</span>  <span class="dv">42</span>%</span>
<span id="cb43-159"><a href="#cb43-159" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-160"><a href="#cb43-160" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================</span>                                        <span class="er">|</span>  <span class="dv">42</span>%</span>
<span id="cb43-161"><a href="#cb43-161" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-162"><a href="#cb43-162" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================</span>                                        <span class="er">|</span>  <span class="dv">43</span>%</span>
<span id="cb43-163"><a href="#cb43-163" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-164"><a href="#cb43-164" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================</span>                                        <span class="er">|</span>  <span class="dv">44</span>%</span>
<span id="cb43-165"><a href="#cb43-165" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-166"><a href="#cb43-166" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================</span>                                       <span class="er">|</span>  <span class="dv">44</span>%</span>
<span id="cb43-167"><a href="#cb43-167" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-168"><a href="#cb43-168" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================</span>                                       <span class="er">|</span>  <span class="dv">45</span>%</span>
<span id="cb43-169"><a href="#cb43-169" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-170"><a href="#cb43-170" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================</span>                                      <span class="er">|</span>  <span class="dv">45</span>%</span>
<span id="cb43-171"><a href="#cb43-171" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-172"><a href="#cb43-172" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================</span>                                      <span class="er">|</span>  <span class="dv">46</span>%</span>
<span id="cb43-173"><a href="#cb43-173" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-174"><a href="#cb43-174" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================</span>                                     <span class="er">|</span>  <span class="dv">46</span>%</span>
<span id="cb43-175"><a href="#cb43-175" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-176"><a href="#cb43-176" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================</span>                                     <span class="er">|</span>  <span class="dv">47</span>%</span>
<span id="cb43-177"><a href="#cb43-177" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-178"><a href="#cb43-178" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================</span>                                     <span class="er">|</span>  <span class="dv">48</span>%</span>
<span id="cb43-179"><a href="#cb43-179" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-180"><a href="#cb43-180" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================</span>                                    <span class="er">|</span>  <span class="dv">48</span>%</span>
<span id="cb43-181"><a href="#cb43-181" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-182"><a href="#cb43-182" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================</span>                                    <span class="er">|</span>  <span class="dv">49</span>%</span>
<span id="cb43-183"><a href="#cb43-183" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-184"><a href="#cb43-184" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================</span>                                   <span class="er">|</span>  <span class="dv">49</span>%</span>
<span id="cb43-185"><a href="#cb43-185" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-186"><a href="#cb43-186" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================</span>                                   <span class="er">|</span>  <span class="dv">50</span>%</span>
<span id="cb43-187"><a href="#cb43-187" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-188"><a href="#cb43-188" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================</span>                                   <span class="er">|</span>  <span class="dv">51</span>%</span>
<span id="cb43-189"><a href="#cb43-189" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-190"><a href="#cb43-190" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================</span>                                  <span class="er">|</span>  <span class="dv">51</span>%</span>
<span id="cb43-191"><a href="#cb43-191" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-192"><a href="#cb43-192" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================</span>                                  <span class="er">|</span>  <span class="dv">52</span>%</span>
<span id="cb43-193"><a href="#cb43-193" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-194"><a href="#cb43-194" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================</span>                                 <span class="er">|</span>  <span class="dv">52</span>%</span>
<span id="cb43-195"><a href="#cb43-195" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-196"><a href="#cb43-196" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================</span>                                 <span class="er">|</span>  <span class="dv">53</span>%</span>
<span id="cb43-197"><a href="#cb43-197" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-198"><a href="#cb43-198" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================</span>                                 <span class="er">|</span>  <span class="dv">54</span>%</span>
<span id="cb43-199"><a href="#cb43-199" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-200"><a href="#cb43-200" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================</span>                                <span class="er">|</span>  <span class="dv">54</span>%</span>
<span id="cb43-201"><a href="#cb43-201" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-202"><a href="#cb43-202" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================</span>                                <span class="er">|</span>  <span class="dv">55</span>%</span>
<span id="cb43-203"><a href="#cb43-203" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-204"><a href="#cb43-204" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================================</span>                               <span class="er">|</span>  <span class="dv">55</span>%</span>
<span id="cb43-205"><a href="#cb43-205" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-206"><a href="#cb43-206" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================================</span>                               <span class="er">|</span>  <span class="dv">56</span>%</span>
<span id="cb43-207"><a href="#cb43-207" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-208"><a href="#cb43-208" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================</span>                              <span class="er">|</span>  <span class="dv">56</span>%</span>
<span id="cb43-209"><a href="#cb43-209" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-210"><a href="#cb43-210" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================</span>                              <span class="er">|</span>  <span class="dv">57</span>%</span>
<span id="cb43-211"><a href="#cb43-211" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-212"><a href="#cb43-212" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================</span>                              <span class="er">|</span>  <span class="dv">58</span>%</span>
<span id="cb43-213"><a href="#cb43-213" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-214"><a href="#cb43-214" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================================</span>                             <span class="er">|</span>  <span class="dv">58</span>%</span>
<span id="cb43-215"><a href="#cb43-215" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-216"><a href="#cb43-216" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================================</span>                             <span class="er">|</span>  <span class="dv">59</span>%</span>
<span id="cb43-217"><a href="#cb43-217" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-218"><a href="#cb43-218" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================</span>                            <span class="er">|</span>  <span class="dv">59</span>%</span>
<span id="cb43-219"><a href="#cb43-219" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-220"><a href="#cb43-220" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================</span>                            <span class="er">|</span>  <span class="dv">60</span>%</span>
<span id="cb43-221"><a href="#cb43-221" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-222"><a href="#cb43-222" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================</span>                            <span class="er">|</span>  <span class="dv">61</span>%</span>
<span id="cb43-223"><a href="#cb43-223" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-224"><a href="#cb43-224" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================================</span>                           <span class="er">|</span>  <span class="dv">61</span>%</span>
<span id="cb43-225"><a href="#cb43-225" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-226"><a href="#cb43-226" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================================</span>                           <span class="er">|</span>  <span class="dv">62</span>%</span>
<span id="cb43-227"><a href="#cb43-227" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-228"><a href="#cb43-228" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================================</span>                          <span class="er">|</span>  <span class="dv">62</span>%</span>
<span id="cb43-229"><a href="#cb43-229" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-230"><a href="#cb43-230" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================================</span>                          <span class="er">|</span>  <span class="dv">63</span>%</span>
<span id="cb43-231"><a href="#cb43-231" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-232"><a href="#cb43-232" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================================</span>                          <span class="er">|</span>  <span class="dv">64</span>%</span>
<span id="cb43-233"><a href="#cb43-233" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-234"><a href="#cb43-234" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================================</span>                         <span class="er">|</span>  <span class="dv">64</span>%</span>
<span id="cb43-235"><a href="#cb43-235" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-236"><a href="#cb43-236" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================================</span>                         <span class="er">|</span>  <span class="dv">65</span>%</span>
<span id="cb43-237"><a href="#cb43-237" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-238"><a href="#cb43-238" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================================</span>                        <span class="er">|</span>  <span class="dv">65</span>%</span>
<span id="cb43-239"><a href="#cb43-239" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-240"><a href="#cb43-240" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================================</span>                        <span class="er">|</span>  <span class="dv">66</span>%</span>
<span id="cb43-241"><a href="#cb43-241" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-242"><a href="#cb43-242" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================</span>                       <span class="er">|</span>  <span class="dv">66</span>%</span>
<span id="cb43-243"><a href="#cb43-243" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-244"><a href="#cb43-244" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================</span>                       <span class="er">|</span>  <span class="dv">67</span>%</span>
<span id="cb43-245"><a href="#cb43-245" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-246"><a href="#cb43-246" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================</span>                       <span class="er">|</span>  <span class="dv">68</span>%</span>
<span id="cb43-247"><a href="#cb43-247" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-248"><a href="#cb43-248" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================================</span>                      <span class="er">|</span>  <span class="dv">68</span>%</span>
<span id="cb43-249"><a href="#cb43-249" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-250"><a href="#cb43-250" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================================</span>                      <span class="er">|</span>  <span class="dv">69</span>%</span>
<span id="cb43-251"><a href="#cb43-251" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-252"><a href="#cb43-252" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================</span>                     <span class="er">|</span>  <span class="dv">69</span>%</span>
<span id="cb43-253"><a href="#cb43-253" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-254"><a href="#cb43-254" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================</span>                     <span class="er">|</span>  <span class="dv">70</span>%</span>
<span id="cb43-255"><a href="#cb43-255" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-256"><a href="#cb43-256" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================</span>                     <span class="er">|</span>  <span class="dv">71</span>%</span>
<span id="cb43-257"><a href="#cb43-257" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-258"><a href="#cb43-258" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================================</span>                    <span class="er">|</span>  <span class="dv">71</span>%</span>
<span id="cb43-259"><a href="#cb43-259" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-260"><a href="#cb43-260" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================================</span>                    <span class="er">|</span>  <span class="dv">72</span>%</span>
<span id="cb43-261"><a href="#cb43-261" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-262"><a href="#cb43-262" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================================</span>                   <span class="er">|</span>  <span class="dv">72</span>%</span>
<span id="cb43-263"><a href="#cb43-263" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-264"><a href="#cb43-264" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================================</span>                   <span class="er">|</span>  <span class="dv">73</span>%</span>
<span id="cb43-265"><a href="#cb43-265" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-266"><a href="#cb43-266" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================================</span>                   <span class="er">|</span>  <span class="dv">74</span>%</span>
<span id="cb43-267"><a href="#cb43-267" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-268"><a href="#cb43-268" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================================</span>                  <span class="er">|</span>  <span class="dv">74</span>%</span>
<span id="cb43-269"><a href="#cb43-269" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-270"><a href="#cb43-270" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================================</span>                  <span class="er">|</span>  <span class="dv">75</span>%</span>
<span id="cb43-271"><a href="#cb43-271" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-272"><a href="#cb43-272" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================================</span>                 <span class="er">|</span>  <span class="dv">75</span>%</span>
<span id="cb43-273"><a href="#cb43-273" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-274"><a href="#cb43-274" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================================</span>                 <span class="er">|</span>  <span class="dv">76</span>%</span>
<span id="cb43-275"><a href="#cb43-275" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-276"><a href="#cb43-276" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================</span>                <span class="er">|</span>  <span class="dv">76</span>%</span>
<span id="cb43-277"><a href="#cb43-277" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-278"><a href="#cb43-278" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================</span>                <span class="er">|</span>  <span class="dv">77</span>%</span>
<span id="cb43-279"><a href="#cb43-279" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-280"><a href="#cb43-280" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================</span>                <span class="er">|</span>  <span class="dv">78</span>%</span>
<span id="cb43-281"><a href="#cb43-281" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-282"><a href="#cb43-282" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================================================</span>               <span class="er">|</span>  <span class="dv">78</span>%</span>
<span id="cb43-283"><a href="#cb43-283" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-284"><a href="#cb43-284" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=======================================================</span>               <span class="er">|</span>  <span class="dv">79</span>%</span>
<span id="cb43-285"><a href="#cb43-285" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-286"><a href="#cb43-286" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================================</span>              <span class="er">|</span>  <span class="dv">79</span>%</span>
<span id="cb43-287"><a href="#cb43-287" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-288"><a href="#cb43-288" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================================</span>              <span class="er">|</span>  <span class="dv">80</span>%</span>
<span id="cb43-289"><a href="#cb43-289" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-290"><a href="#cb43-290" aria-hidden="true" tabindex="-1"></a>  <span class="er">|========================================================</span>              <span class="er">|</span>  <span class="dv">81</span>%</span>
<span id="cb43-291"><a href="#cb43-291" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-292"><a href="#cb43-292" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================================================</span>             <span class="er">|</span>  <span class="dv">81</span>%</span>
<span id="cb43-293"><a href="#cb43-293" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-294"><a href="#cb43-294" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=========================================================</span>             <span class="er">|</span>  <span class="dv">82</span>%</span>
<span id="cb43-295"><a href="#cb43-295" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-296"><a href="#cb43-296" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================================</span>            <span class="er">|</span>  <span class="dv">82</span>%</span>
<span id="cb43-297"><a href="#cb43-297" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-298"><a href="#cb43-298" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================================</span>            <span class="er">|</span>  <span class="dv">83</span>%</span>
<span id="cb43-299"><a href="#cb43-299" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-300"><a href="#cb43-300" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==========================================================</span>            <span class="er">|</span>  <span class="dv">84</span>%</span>
<span id="cb43-301"><a href="#cb43-301" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-302"><a href="#cb43-302" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================================================</span>           <span class="er">|</span>  <span class="dv">84</span>%</span>
<span id="cb43-303"><a href="#cb43-303" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-304"><a href="#cb43-304" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===========================================================</span>           <span class="er">|</span>  <span class="dv">85</span>%</span>
<span id="cb43-305"><a href="#cb43-305" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-306"><a href="#cb43-306" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================================================</span>          <span class="er">|</span>  <span class="dv">85</span>%</span>
<span id="cb43-307"><a href="#cb43-307" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-308"><a href="#cb43-308" aria-hidden="true" tabindex="-1"></a>  <span class="er">|============================================================</span>          <span class="er">|</span>  <span class="dv">86</span>%</span>
<span id="cb43-309"><a href="#cb43-309" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-310"><a href="#cb43-310" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================================================</span>         <span class="er">|</span>  <span class="dv">86</span>%</span>
<span id="cb43-311"><a href="#cb43-311" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-312"><a href="#cb43-312" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================================================</span>         <span class="er">|</span>  <span class="dv">87</span>%</span>
<span id="cb43-313"><a href="#cb43-313" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-314"><a href="#cb43-314" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=============================================================</span>         <span class="er">|</span>  <span class="dv">88</span>%</span>
<span id="cb43-315"><a href="#cb43-315" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-316"><a href="#cb43-316" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================================================</span>        <span class="er">|</span>  <span class="dv">88</span>%</span>
<span id="cb43-317"><a href="#cb43-317" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-318"><a href="#cb43-318" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==============================================================</span>        <span class="er">|</span>  <span class="dv">89</span>%</span>
<span id="cb43-319"><a href="#cb43-319" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-320"><a href="#cb43-320" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================================</span>       <span class="er">|</span>  <span class="dv">89</span>%</span>
<span id="cb43-321"><a href="#cb43-321" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-322"><a href="#cb43-322" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================================</span>       <span class="er">|</span>  <span class="dv">90</span>%</span>
<span id="cb43-323"><a href="#cb43-323" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-324"><a href="#cb43-324" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===============================================================</span>       <span class="er">|</span>  <span class="dv">91</span>%</span>
<span id="cb43-325"><a href="#cb43-325" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-326"><a href="#cb43-326" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================================================</span>      <span class="er">|</span>  <span class="dv">91</span>%</span>
<span id="cb43-327"><a href="#cb43-327" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-328"><a href="#cb43-328" aria-hidden="true" tabindex="-1"></a>  <span class="er">|================================================================</span>      <span class="er">|</span>  <span class="dv">92</span>%</span>
<span id="cb43-329"><a href="#cb43-329" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-330"><a href="#cb43-330" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================================</span>     <span class="er">|</span>  <span class="dv">92</span>%</span>
<span id="cb43-331"><a href="#cb43-331" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-332"><a href="#cb43-332" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================================</span>     <span class="er">|</span>  <span class="dv">93</span>%</span>
<span id="cb43-333"><a href="#cb43-333" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-334"><a href="#cb43-334" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=================================================================</span>     <span class="er">|</span>  <span class="dv">94</span>%</span>
<span id="cb43-335"><a href="#cb43-335" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-336"><a href="#cb43-336" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================================================</span>    <span class="er">|</span>  <span class="dv">94</span>%</span>
<span id="cb43-337"><a href="#cb43-337" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-338"><a href="#cb43-338" aria-hidden="true" tabindex="-1"></a>  <span class="er">|==================================================================</span>    <span class="er">|</span>  <span class="dv">95</span>%</span>
<span id="cb43-339"><a href="#cb43-339" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-340"><a href="#cb43-340" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================================================</span>   <span class="er">|</span>  <span class="dv">95</span>%</span>
<span id="cb43-341"><a href="#cb43-341" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-342"><a href="#cb43-342" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================================================</span>   <span class="er">|</span>  <span class="dv">96</span>%</span>
<span id="cb43-343"><a href="#cb43-343" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-344"><a href="#cb43-344" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================================================</span>  <span class="er">|</span>  <span class="dv">96</span>%</span>
<span id="cb43-345"><a href="#cb43-345" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-346"><a href="#cb43-346" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================================================</span>  <span class="er">|</span>  <span class="dv">97</span>%</span>
<span id="cb43-347"><a href="#cb43-347" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-348"><a href="#cb43-348" aria-hidden="true" tabindex="-1"></a>  <span class="er">|====================================================================</span>  <span class="er">|</span>  <span class="dv">98</span>%</span>
<span id="cb43-349"><a href="#cb43-349" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-350"><a href="#cb43-350" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================================================</span> <span class="er">|</span>  <span class="dv">98</span>%</span>
<span id="cb43-351"><a href="#cb43-351" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-352"><a href="#cb43-352" aria-hidden="true" tabindex="-1"></a>  <span class="er">|=====================================================================</span> <span class="er">|</span>  <span class="dv">99</span>%</span>
<span id="cb43-353"><a href="#cb43-353" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-354"><a href="#cb43-354" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span>  <span class="dv">99</span>%</span>
<span id="cb43-355"><a href="#cb43-355" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb43-356"><a href="#cb43-356" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Apply linear model to GSVA result. Now using Limma to find significantly enriched gene sets in the same way you did to find diffGenes. This means youâ€™ll be using topTable, decideTests, etc.</li>
<li>note that you need to reference your design and contrast matrix here.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit_C2CP</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">GSVA_res_C2CP</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">ebfit_C2CP</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit_C2CP</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Use topTable and decideTests functions to identify the differentially enriched gene sets</span></span>
<span><span class="va">topPaths_C2CP</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">ebfit_C2CP</span>, adjust <span class="op">=</span><span class="st">"BH"</span>, coef<span class="op">=</span><span class="fl">1</span>, number<span class="op">=</span><span class="fl">50</span>, sort.by<span class="op">=</span><span class="st">"logFC"</span><span class="op">)</span></span>
<span><span class="va">res_C2CP</span> <span class="op">&lt;-</span> <span class="fu">decideTests</span><span class="op">(</span><span class="va">ebfit_C2CP</span>, method<span class="op">=</span><span class="st">"global"</span>, adjust.method<span class="op">=</span><span class="st">"BH"</span>, p.value<span class="op">=</span><span class="fl">0.05</span>, lfc<span class="op">=</span><span class="fl">0.58</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### The summary of the decideTests result shows how many sets were enriched in induced and repressed genes in all sample types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">res_C2CP</span><span class="op">)</span></span>
<span><span class="co">##        disease healthy</span></span>
<span><span class="co">## Down        12      42</span></span>
<span><span class="co">## NotSig    5851    5864</span></span>
<span><span class="co">## Up          56      13</span></span>
<span></span>
<span><span class="co"># Pull out the gene sets that are differentially enriched between groups</span></span>
<span><span class="va">diffSets_C2CP</span> <span class="op">&lt;-</span> <span class="va">GSVA_res_C2CP</span><span class="op">[</span><span class="va">res_C2CP</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">!=</span><span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">diffSets_C2CP</span><span class="op">)</span></span>
<span><span class="co">##                                                  HS01       HS02       HS03</span></span>
<span><span class="co">## ST_IL_13_PATHWAY                           -0.7173858 -0.7147717 -0.7164712</span></span>
<span><span class="co">## WATANABE_ULCERATIVE_COLITIS_WITH_CANCER_DN -0.6188325 -0.5710992 -0.4131349</span></span>
<span><span class="co">## FARMER_BREAST_CANCER_CLUSTER_1             -0.6886216 -0.5784651 -0.5593251</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX     -0.6707584 -0.5317945 -0.5360570</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_TAMOXIFEN                -0.6717424 -0.5377614 -0.5172726</span></span>
<span><span class="co">## KAUFFMANN_MELANOMA_RELAPSE_DN               0.5698331  0.7398284  0.6624807</span></span>
<span><span class="co">##                                                  HS04       HS05       CL08</span></span>
<span><span class="co">## ST_IL_13_PATHWAY                           -0.5968289 -0.4130711  0.8128353</span></span>
<span><span class="co">## WATANABE_ULCERATIVE_COLITIS_WITH_CANCER_DN -0.6270975 -0.5185058  0.2930852</span></span>
<span><span class="co">## FARMER_BREAST_CANCER_CLUSTER_1             -0.6325020 -0.4714246  0.5083211</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX     -0.6457163 -0.5252770  0.5529664</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_TAMOXIFEN                -0.6692727 -0.5128578  0.5408430</span></span>
<span><span class="co">## KAUFFMANN_MELANOMA_RELAPSE_DN               0.4426951  0.6730961 -0.6256108</span></span>
<span><span class="co">##                                                  CL10       CL11       CL12</span></span>
<span><span class="co">## ST_IL_13_PATHWAY                            0.6256980  0.5670608  0.6341371</span></span>
<span><span class="co">## WATANABE_ULCERATIVE_COLITIS_WITH_CANCER_DN  0.5838839  0.6787619  0.7118710</span></span>
<span><span class="co">## FARMER_BREAST_CANCER_CLUSTER_1              0.5593513  0.6517882  0.7463300</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX      0.4914990  0.7723819  0.6409015</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_TAMOXIFEN                 0.5064169  0.7568069  0.6405129</span></span>
<span><span class="co">## KAUFFMANN_MELANOMA_RELAPSE_DN              -0.6277048 -0.6895166 -0.8205882</span></span>
<span><span class="co">##                                                  CL13</span></span>
<span><span class="co">## ST_IL_13_PATHWAY                            0.6455147</span></span>
<span><span class="co">## WATANABE_ULCERATIVE_COLITIS_WITH_CANCER_DN  0.6389032</span></span>
<span><span class="co">## FARMER_BREAST_CANCER_CLUSTER_1              0.5802493</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_EXTRACELLULAR_MATRIX      0.6434423</span></span>
<span><span class="co">## BOWIE_RESPONSE_TO_TAMOXIFEN                 0.6232986</span></span>
<span><span class="co">## KAUFFMANN_MELANOMA_RELAPSE_DN              -0.3907835</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">diffSets_C2CP</span><span class="op">)</span></span>
<span><span class="co">## [1] 68 10</span></span>
<span></span>
<span><span class="co">### Make a heatmap of differentially enriched gene sets </span></span>
<span><span class="va">hr_C2CP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">diffSets_C2CP</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="co">### cluster rows by pearson correlation</span></span>
<span>    method <span class="op">=</span> <span class="st">"complete"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">hc_C2CP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">diffSets_C2CP</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="co">### cluster columns by spearman correlation</span></span>
<span>    method <span class="op">=</span> <span class="st">"complete"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Cut the resulting tree and create color vector for clusters.  Vary the cut height to give more or fewer clusters, or you the 'k' argument to force n number of clusters</span></span>
<span><span class="va">mycl_C2CP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">hr_C2CP</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mycolhc_C2CP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">mycl_C2CP</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    start <span class="op">=</span> <span class="fl">0.1</span>, end <span class="op">=</span> <span class="fl">0.9</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">mycolhc_C2CP</span> <span class="op">&lt;-</span> <span class="va">mycolhc_C2CP</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">mycl_C2CP</span><span class="op">)</span><span class="op">]</span> </span>
<span></span>
<span><span class="co">### assign your favorite heatmap color scheme. Some useful examples: colorpanel(40, "darkblue", "yellow", "white"); heat.colors(75); cm.colors(75); rainbow(75); redgreen(75); library(RColorBrewer); rev(brewer.pal(9,"Blues")[-1]). Type demo.col(20) to see more color schemes.</span></span>
<span><span class="va">myheatcol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"yellow"</span>,<span class="st">"white"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Plot the hclust results as a heatmap</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span></span>
<span>    <span class="va">diffSets_C2CP</span>,</span>
<span>    Rowv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hr_C2CP</span><span class="op">)</span>,</span>
<span>    Colv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hc_C2CP</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="va">myheatcol</span>, scale <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>    density.info <span class="op">=</span> <span class="st">"none"</span>, trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    <span class="co">### Creates heatmap for entire data set where the obtained clusters are indicated in the color bar.</span></span>
<span>    cexRow <span class="op">=</span> <span class="fl">0.9</span>, cexCol <span class="op">=</span> <span class="fl">1</span>, margins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">14</span><span class="op">)</span></span>
<span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="reference" class="level2"><h2 class="anchored" data-anchor-id="reference">Reference</h2>
<ul>
<li><a href="https://diytranscriptomics.com/">DIY Transcriptomics</a></li>
<li><a href="https://protocols.hostmicrobe.org/">protocols</a></li>
</ul></section><section id="sessioninfo" class="level2"><h2 class="anchored" data-anchor-id="sessioninfo">SessionInfo</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.1.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Singapore</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] enrichplot_1.20.0         msigdbr_7.5.1            </span></span>
<span><span class="co">##  [3] clusterProfiler_4.8.2     gprofiler2_0.2.2         </span></span>
<span><span class="co">##  [5] GSVA_1.48.3               GSEABase_1.62.0          </span></span>
<span><span class="co">##  [7] graph_1.78.0              annotate_1.80.0          </span></span>
<span><span class="co">##  [9] XML_3.99-0.15             gplots_3.1.3             </span></span>
<span><span class="co">## [11] heatmaply_1.5.0           viridis_0.6.4            </span></span>
<span><span class="co">## [13] viridisLite_0.4.2         RColorBrewer_1.1-3       </span></span>
<span><span class="co">## [15] gt_0.10.0                 plotly_4.10.3            </span></span>
<span><span class="co">## [17] DT_0.30                   cowplot_1.1.1            </span></span>
<span><span class="co">## [19] matrixStats_1.1.0         edgeR_4.0.1              </span></span>
<span><span class="co">## [21] limma_3.58.1              datapasta_3.1.0          </span></span>
<span><span class="co">## [23] here_1.0.1                EnsDb.Hsapiens.v86_2.99.0</span></span>
<span><span class="co">## [25] ensembldb_2.24.1          AnnotationFilter_1.24.0  </span></span>
<span><span class="co">## [27] GenomicFeatures_1.52.2    AnnotationDbi_1.64.1     </span></span>
<span><span class="co">## [29] Biobase_2.62.0            GenomicRanges_1.52.1     </span></span>
<span><span class="co">## [31] GenomeInfoDb_1.38.0       IRanges_2.36.0           </span></span>
<span><span class="co">## [33] S4Vectors_0.40.1          BiocGenerics_0.48.1      </span></span>
<span><span class="co">## [35] tximport_1.28.0           lubridate_1.9.3          </span></span>
<span><span class="co">## [37] forcats_1.0.0             stringr_1.5.1            </span></span>
<span><span class="co">## [39] dplyr_1.1.4               purrr_1.0.2              </span></span>
<span><span class="co">## [41] readr_2.1.4               tidyr_1.3.0              </span></span>
<span><span class="co">## [43] tibble_3.2.1              ggplot2_3.4.4            </span></span>
<span><span class="co">## [45] tidyverse_2.0.0           rhdf5_2.44.0             </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.3                    ProtGenerics_1.32.0        </span></span>
<span><span class="co">##   [3] bitops_1.0-7                HDO.db_0.99.1              </span></span>
<span><span class="co">##   [5] httr_1.4.7                  webshot_0.5.5              </span></span>
<span><span class="co">##   [7] tools_4.3.1                 utf8_1.2.4                 </span></span>
<span><span class="co">##   [9] R6_2.5.1                    HDF5Array_1.28.1           </span></span>
<span><span class="co">##  [11] mgcv_1.9-0                  lazyeval_0.2.2             </span></span>
<span><span class="co">##  [13] rhdf5filters_1.12.1         withr_2.5.2                </span></span>
<span><span class="co">##  [15] prettyunits_1.2.0           gridExtra_2.3              </span></span>
<span><span class="co">##  [17] cli_3.6.1                   TSP_1.2-4                  </span></span>
<span><span class="co">##  [19] scatterpie_0.2.1            sass_0.4.7                 </span></span>
<span><span class="co">##  [21] labeling_0.4.3              commonmark_1.9.0           </span></span>
<span><span class="co">##  [23] Rsamtools_2.16.0            yulab.utils_0.1.0          </span></span>
<span><span class="co">##  [25] gson_0.1.0                  DOSE_3.26.2                </span></span>
<span><span class="co">##  [27] RSQLite_2.3.3               generics_0.1.3             </span></span>
<span><span class="co">##  [29] gridGraphics_0.5-1          BiocIO_1.10.0              </span></span>
<span><span class="co">##  [31] crosstalk_1.2.0             vroom_1.6.4                </span></span>
<span><span class="co">##  [33] gtools_3.9.5                dendextend_1.17.1          </span></span>
<span><span class="co">##  [35] GO.db_3.17.0                Matrix_1.6-3               </span></span>
<span><span class="co">##  [37] fansi_1.0.5                 abind_1.4-5                </span></span>
<span><span class="co">##  [39] lifecycle_1.0.4             yaml_2.3.7                 </span></span>
<span><span class="co">##  [41] SummarizedExperiment_1.30.2 qvalue_2.32.0              </span></span>
<span><span class="co">##  [43] BiocFileCache_2.8.0         grid_4.3.1                 </span></span>
<span><span class="co">##  [45] blob_1.2.4                  promises_1.2.1             </span></span>
<span><span class="co">##  [47] crayon_1.5.2                lattice_0.22-5             </span></span>
<span><span class="co">##  [49] beachmat_2.16.0             KEGGREST_1.42.0            </span></span>
<span><span class="co">##  [51] pillar_1.9.0                knitr_1.45                 </span></span>
<span><span class="co">##  [53] fgsea_1.26.0                rjson_0.2.21               </span></span>
<span><span class="co">##  [55] codetools_0.2-19            fastmatch_1.1-4            </span></span>
<span><span class="co">##  [57] glue_1.6.2                  downloader_0.4             </span></span>
<span><span class="co">##  [59] ggfun_0.1.3                 data.table_1.14.8          </span></span>
<span><span class="co">##  [61] vctrs_0.6.4                 png_0.1-8                  </span></span>
<span><span class="co">##  [63] treeio_1.24.3               gtable_0.3.4               </span></span>
<span><span class="co">##  [65] assertthat_0.2.1            cachem_1.0.8               </span></span>
<span><span class="co">##  [67] xfun_0.41                   mime_0.12                  </span></span>
<span><span class="co">##  [69] S4Arrays_1.0.6              tidygraph_1.2.3            </span></span>
<span><span class="co">##  [71] SingleCellExperiment_1.22.0 seriation_1.4.2            </span></span>
<span><span class="co">##  [73] iterators_1.0.14            statmod_1.5.0              </span></span>
<span><span class="co">##  [75] ellipsis_0.3.2              nlme_3.1-163               </span></span>
<span><span class="co">##  [77] ggtree_3.8.2                bit64_4.0.5                </span></span>
<span><span class="co">##  [79] progress_1.2.2              filelock_1.0.2             </span></span>
<span><span class="co">##  [81] rprojroot_2.0.4             irlba_2.3.5.1              </span></span>
<span><span class="co">##  [83] KernSmooth_2.23-22          colorspace_2.1-0           </span></span>
<span><span class="co">##  [85] DBI_1.1.3                   tidyselect_1.2.0           </span></span>
<span><span class="co">##  [87] bit_4.0.5                   compiler_4.3.1             </span></span>
<span><span class="co">##  [89] curl_5.1.0                  xml2_1.3.5                 </span></span>
<span><span class="co">##  [91] DelayedArray_0.26.7         shadowtext_0.1.2           </span></span>
<span><span class="co">##  [93] rtracklayer_1.60.1          scales_1.2.1               </span></span>
<span><span class="co">##  [95] caTools_1.18.2              hexbin_1.28.3              </span></span>
<span><span class="co">##  [97] rappdirs_0.3.3              digest_0.6.33              </span></span>
<span><span class="co">##  [99] rmarkdown_2.25              ca_0.71.1                  </span></span>
<span><span class="co">## [101] XVector_0.42.0              htmltools_0.5.7            </span></span>
<span><span class="co">## [103] pkgconfig_2.0.3             sparseMatrixStats_1.12.2   </span></span>
<span><span class="co">## [105] MatrixGenerics_1.14.0       dbplyr_2.4.0               </span></span>
<span><span class="co">## [107] fastmap_1.1.1               rlang_1.1.2                </span></span>
<span><span class="co">## [109] htmlwidgets_1.6.2           shiny_1.8.0                </span></span>
<span><span class="co">## [111] DelayedMatrixStats_1.22.6   farver_2.1.1               </span></span>
<span><span class="co">## [113] jsonlite_1.8.7              BiocParallel_1.36.0        </span></span>
<span><span class="co">## [115] GOSemSim_2.26.1             BiocSingular_1.16.0        </span></span>
<span><span class="co">## [117] RCurl_1.98-1.13             magrittr_2.0.3             </span></span>
<span><span class="co">## [119] GenomeInfoDbData_1.2.11     ggplotify_0.1.2            </span></span>
<span><span class="co">## [121] patchwork_1.1.3             Rhdf5lib_1.22.1            </span></span>
<span><span class="co">## [123] munsell_0.5.0               Rcpp_1.0.11                </span></span>
<span><span class="co">## [125] ape_5.7-1                   babelgene_22.9             </span></span>
<span><span class="co">## [127] stringi_1.8.1               ggraph_2.1.0               </span></span>
<span><span class="co">## [129] zlibbioc_1.48.0             MASS_7.3-60                </span></span>
<span><span class="co">## [131] plyr_1.8.9                  parallel_4.3.1             </span></span>
<span><span class="co">## [133] ggrepel_0.9.4               Biostrings_2.70.1          </span></span>
<span><span class="co">## [135] graphlayouts_1.0.2          splines_4.3.1              </span></span>
<span><span class="co">## [137] hms_1.1.3                   locfit_1.5-9.8             </span></span>
<span><span class="co">## [139] igraph_1.5.1                markdown_1.11              </span></span>
<span><span class="co">## [141] reshape2_1.4.4              biomaRt_2.56.1             </span></span>
<span><span class="co">## [143] ScaledMatrix_1.8.1          evaluate_0.23              </span></span>
<span><span class="co">## [145] httpuv_1.6.12               tzdb_0.4.0                 </span></span>
<span><span class="co">## [147] foreach_1.5.2               tweenr_2.0.2               </span></span>
<span><span class="co">## [149] polyclip_1.10-6             ggforce_0.4.1              </span></span>
<span><span class="co">## [151] rsvd_1.0.5                  xtable_1.8-4               </span></span>
<span><span class="co">## [153] restfulr_0.0.15             tidytree_0.4.5             </span></span>
<span><span class="co">## [155] later_1.3.1                 aplot_0.2.2                </span></span>
<span><span class="co">## [157] memoise_2.0.1               registry_0.5-1             </span></span>
<span><span class="co">## [159] GenomicAlignments_1.36.0    timechange_0.2.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/grgrzhong\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2024 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a> and <a href="https://quarto.org/">Quarto</a></span></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>