<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta charset="UTF-8">
<meta name="generator" content="quarto-1.8.26">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Guorui Zhong">
<title>Seurat V5 | 10X course example – Guorui Zhong</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting-1b3db88def35042d172274863c1cdcf0.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap-a59be64ae26c4866fb77368000ae33d0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
</head>
<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../wiki/index.html"> 
<span class="menu-text">Wiki</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../linux/index.html"> 
<span class="menu-text">Linux</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../python/index.html"> 
<span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../r/index.html"> 
<span class="menu-text">R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html"> 
<span class="menu-text">Keyboard</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Seurat V5 | 10X course example</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">seurat</div>
                <div class="quarto-category">scrna</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://grgrzhong.github.io/">Guorui Zhong</a> <a href="https://orcid.org/0000-0002-1955-2847" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Monday, March 18, 2024</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Sunday, January 26, 2025</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#load-packages-and-data" id="toc-load-packages-and-data" class="nav-link active" data-scroll-target="#load-packages-and-data">Load packages and data</a></li>
  <li><a href="#explore-qc-metrics" id="toc-explore-qc-metrics" class="nav-link" data-scroll-target="#explore-qc-metrics">Explore QC metrics</a></li>
  <li><a href="#setting-qc-cutoff-and-filtering" id="toc-setting-qc-cutoff-and-filtering" class="nav-link" data-scroll-target="#setting-qc-cutoff-and-filtering">Setting QC Cutoff and Filtering</a></li>
  <li><a href="#normalisation-selection-and-scaling" id="toc-normalisation-selection-and-scaling" class="nav-link" data-scroll-target="#normalisation-selection-and-scaling">Normalisation, Selection and Scaling</a></li>
  <li>
<a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">Dimensionality reduction</a>
  <ul class="collapse">
<li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#tsne" id="toc-tsne" class="nav-link" data-scroll-target="#tsne">tSNE</a></li>
  </ul>
</li>
  <li><a href="#defining-cell-clusters" id="toc-defining-cell-clusters" class="nav-link" data-scroll-target="#defining-cell-clusters">Defining Cell Clusters</a></li>
  <li><a href="#find-markers" id="toc-find-markers" class="nav-link" data-scroll-target="#find-markers">Find Markers</a></li>
  <li><a href="#automated-cell-type-annotation" id="toc-automated-cell-type-annotation" class="nav-link" data-scroll-target="#automated-cell-type-annotation">Automated Cell Type Annotation</a></li>
  <li><a href="#reference" id="toc-reference" class="nav-link" data-scroll-target="#reference">Reference</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="load-packages-and-data" class="level2"><h2 class="anchored" data-anchor-id="load-packages-and-data">Load packages and data</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://lce.biohpc.swmed.edu/scina/">SCINA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/get_theme.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="st">"projects/babraham_training_bioinformatics/10X_scRNAseq_Course"</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">Read10X_h5</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">CreateSeuratObject</span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">data</span>,</span>
<span>    project <span class="op">=</span> <span class="st">"course"</span>,</span>
<span>    min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    min.features <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="explore-qc-metrics" class="level2"><h2 class="anchored" data-anchor-id="explore-qc-metrics">Explore QC metrics</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Amount of MT genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">percent_MT</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">data</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">percent_MT</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Amount of Ribosomal genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^RP[LS]"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">percent_Ribosomal</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">data</span>, pattern <span class="op">=</span> <span class="st">"^RP[LS]"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">percent_Ribosomal</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Percentage of Largest Gene</span></span>
<span><span class="va">data_nomalat</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"MALAT1"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span></span>
<span>    <span class="va">data_nomalat</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">layers</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>    <span class="fl">2</span>,</span>
<span>    <span class="va">max</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span></span>
<span>    <span class="va">data_nomalat</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">layers</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>    <span class="fl">2</span>,</span>
<span>    <span class="va">which.max</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_nomalat</span><span class="op">)</span><span class="op">[</span><span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_index</span><span class="op">]</span></span>
<span></span>
<span><span class="va">data_nomalat</span><span class="op">$</span><span class="va">percent.Largest.Gene</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_count</span> <span class="op">/</span></span>
<span>    <span class="va">data_nomalat</span><span class="op">$</span><span class="va">nCount_RNA</span></span>
<span></span>
<span><span class="va">data_nomalat</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">largest_gene</span> <span class="op">&lt;-</span> <span class="va">data_nomalat</span><span class="op">$</span><span class="va">largest_gene</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">percent.Largest.Gene</span> <span class="op">&lt;-</span> <span class="va">data_nomalat</span><span class="op">$</span><span class="va">percent.Largest.Gene</span></span>
<span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">data_nomalat</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### For some metrics it’s better to view on a log scale.</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    layer <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"nCount_RNA"</span>, <span class="st">"percent_MT"</span>, <span class="st">"percent_Ribosomal"</span>, <span class="st">"percent.Largest.Gene"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    log <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">FeatureScatter</span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, </span>
<span>    feature2 <span class="op">=</span> <span class="st">"percent.Largest.Gene"</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## ggplot</span></span>
<span><span class="va">qc_metrics</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    rownames <span class="op">=</span> <span class="st">"Cell.Barcode"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">percent_MT</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">nCount_RNA</span>, <span class="va">nFeature_RNA</span>, colour <span class="op">=</span> <span class="va">percent_MT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green2"</span>, <span class="st">"red"</span>, <span class="st">"yellow"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Example of plotting QC metrics"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">750</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log scale</span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">percent_MT</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">nCount_RNA</span>, <span class="va">nFeature_RNA</span>, colour <span class="op">=</span> <span class="va">percent_MT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green2"</span>, <span class="st">"red"</span>, <span class="st">"yellow"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Example of plotting QC metrics"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">750</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Complexity </span></span>
<span><span class="va">qc_metrics</span> <span class="op">&lt;-</span> <span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>complexity<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nFeature_RNA</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nCount_RNA</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">complexity.lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">complexity.lm</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">&lt;-</span> <span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    complexity_diff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nFeature_RNA</span><span class="op">)</span> <span class="op">-</span> <span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span><span class="op">*</span><span class="va">complexity.lm</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">+</span><span class="va">complexity.lm</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">complexity_diff</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>fill<span class="op">=</span><span class="st">"yellow"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">complexity_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">$</span><span class="va">complexity_diff</span><span class="op">)</span>,<span class="fl">0</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">qc_metrics</span><span class="op">$</span><span class="va">complexity_diff</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>complexity_diff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/replace.html">replace</a></span><span class="op">(</span><span class="va">complexity_diff</span>,<span class="va">complexity_diff</span><span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.1</span>,<span class="op">-</span><span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nCount_RNA</span><span class="op">)</span>, y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nFeature_RNA</span><span class="op">)</span>, colour<span class="op">=</span><span class="va">complexity_diff</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span><span class="op">(</span>slope<span class="op">=</span><span class="va">complexity.lm</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, intercept <span class="op">=</span> <span class="va">complexity.lm</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient2</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"blue2"</span>,mid<span class="op">=</span><span class="st">"grey"</span>,high<span class="op">=</span><span class="st">"red2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">complexity_diff</span>, y<span class="op">=</span><span class="va">percent.Largest.Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Largest gene</span></span>
<span><span class="va">largest_gene_list</span> <span class="op">&lt;-</span> <span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">largest_gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">largest_genes_to_plot</span> <span class="op">&lt;-</span> <span class="va">largest_gene_list</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">140</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">largest_gene</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">largest_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">largest_genes_to_plot</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>      largest_gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">largest_gene</span>, levels <span class="op">=</span> <span class="va">largest_genes_to_plot</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">largest_gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nCount_RNA</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">nFeature_RNA</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">largest_gene</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span></span>
<span>      values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Set1"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">largest_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">largest_genes_to_plot</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>      largest_gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">largest_gene</span>, levels <span class="op">=</span> <span class="va">largest_genes_to_plot</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">largest_gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">complexity_diff</span>, y <span class="op">=</span> <span class="va">percent.Largest.Gene</span>, colour <span class="op">=</span> <span class="va">largest_gene</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Set1"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">percent_MT</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">complexity_diff</span>, y <span class="op">=</span> <span class="va">percent.Largest.Gene</span>, colour <span class="op">=</span> <span class="va">percent_MT</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"grey"</span>, high <span class="op">=</span> <span class="st">"red2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">percent_Ribosomal</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">complexity_diff</span>, y <span class="op">=</span> <span class="va">percent.Largest.Gene</span>, colour <span class="op">=</span> <span class="va">percent_Ribosomal</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"grey"</span>, high <span class="op">=</span> <span class="st">"red2"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="setting-qc-cutoff-and-filtering" class="level2"><h2 class="anchored" data-anchor-id="setting-qc-cutoff-and-filtering">Setting QC Cutoff and Filtering</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">percent_MT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.5</span>, fill <span class="op">=</span> <span class="st">"yellow"</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Distribution of Percentage Mitochondrion"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_metrics</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">percent.Largest.Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.7</span>, fill <span class="op">=</span> <span class="st">"yellow"</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Distribution of Percentage Largest Gene"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co">### Filtering</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">750</span> <span class="op">&amp;</span></span>
<span>        <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2000</span> <span class="op">&amp;</span></span>
<span>        <span class="va">percent_MT</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">&amp;</span></span>
<span>        <span class="va">percent.Largest.Gene</span> <span class="op">&lt;</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="normalisation-selection-and-scaling" class="level2"><h2 class="anchored" data-anchor-id="normalisation-selection-and-scaling">Normalisation, Selection and Scaling</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Normalization</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">data</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">layers</span><span class="op">$</span><span class="va">data</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">gene_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gene_expression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">gene_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">gene_expression</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene_expression</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="st">"GAPDH"</span>, <span class="op">]</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">layers</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span>, fill <span class="op">=</span> <span class="st">"yellow"</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"GAPDH expression"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">layers</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span></span>
<span>        cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        names_to <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>        values_to <span class="op">=</span> <span class="st">"expression"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell</span>, y <span class="op">=</span> <span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_summary.html">stat_summary</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"crossbar"</span>, fun.data <span class="op">=</span> <span class="va">mean_sdl</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Cell Cycle Scoring</span></span>
<span><span class="va">cc.genes.updated.2019</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">CellCycleScoring</span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    s.features <span class="op">=</span> <span class="va">cc.genes.updated.2019</span><span class="op">$</span><span class="va">s.genes</span>,</span>
<span>    g2m.features <span class="op">=</span> <span class="va">cc.genes.updated.2019</span><span class="op">$</span><span class="va">g2m.genes</span>,</span>
<span>    set.ident <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Phase</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">S.Score</span>, y <span class="op">=</span> <span class="va">G2M.Score</span>, color <span class="op">=</span> <span class="va">Phase</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>, <span class="fl">0.15</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>, <span class="fl">0.15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Gene selection</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>    nfeatures <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">variance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span></span>
<span>    <span class="fu">HVFInfo</span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, rownames <span class="op">=</span> <span class="st">"Gene"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">variance_data</span> <span class="op">&lt;-</span> <span class="va">variance_data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        hypervariable <span class="op">=</span> <span class="va">Gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">variance_data</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">variance_data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">variance</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">hypervariable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Scaling</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="dimensionality-reduction" class="level2"><h2 class="anchored" data-anchor-id="dimensionality-reduction">Dimensionality reduction</h2>
<section id="pca" class="level3"><h3 class="anchored" data-anchor-id="pca">PCA</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Calculate all PCs</span></span>
<span><span class="co">### list of genes which are most highly and lowly weighted in the different PCs</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### See if the cell cycle is having a big effect on the clusters we’re picking out</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span></span>
<span>    <span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">"largest_gene"</span>,</span>
<span>    label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label.size <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### This nicely shows us the power, but also the limitations of PCA in that we </span></span>
<span><span class="co">### can see that not all of the useful information is captured in the first two </span></span>
<span><span class="co">### principal components.</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### How far down the set of PCs do we need to go to capture all of the </span></span>
<span><span class="co">### biologically relevant information.</span></span>
<span><span class="fu">ElbowPlot</span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Plots of PCA weightings for the most highly and lowly weighted genes, </span></span>
<span><span class="co">### shown against the set of cells which are most highly influenced by the PC. </span></span>
<span><span class="co">### The idea is that as long as we’re seeing clear structure in one of these </span></span>
<span><span class="co">### plots then we’re still adding potentially useful information to the analysis.</span></span>
<span><span class="fu">DimHeatmap</span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, cells <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="tsne" class="level3"><h3 class="anchored" data-anchor-id="tsne">tSNE</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Setting this to a low value will help resolve small clusters, </span></span>
<span><span class="co">### but at the expense of large clusters becoming more diffuse</span></span>
<span><span class="va">saved_seed</span> <span class="op">&lt;-</span> <span class="fl">8482</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">saved_seed</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">RunTSNE</span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, seed.use <span class="op">=</span> <span class="va">saved_seed</span>, perplexity <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"tSNE with Perplexity 10"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">RunTSNE</span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, seed.use <span class="op">=</span> <span class="va">saved_seed</span>, perplexity <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"tSNE with Perplexity 200"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">RunTSNE</span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, seed.use <span class="op">=</span> <span class="va">saved_seed</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"tSNE with default Perplexity (30)"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="defining-cell-clusters" class="level2"><h2 class="anchored" data-anchor-id="defining-cell-clusters">Defining Cell Clusters</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Finds the ‘k’ nearest neighbours to each cell and makes this into a graph</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Get another sparse matrix of distances.</span></span>
<span><span class="va">data</span><span class="op">@</span><span class="va">graphs</span><span class="op">$</span><span class="va">RNA_snn</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Segment the graph, </span></span>
<span><span class="co">### Larger resolution give larger clusters, smaller values gives smaller clusters.</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span><span class="va">data</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">seurat_clusters</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Some of the clusters don't resolve very well in PC1 vs PC2</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"PC1 vs PC2 with Clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Some of the clusters which are overlaid in PC1 start to separate in other PCs</span></span>
<span><span class="co">### These differences represent a small proportion of the overall variance</span></span>
<span><span class="co">### but can be important in resolving changes.</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">9</span><span class="op">)</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"PC4 vs PC9 with Clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### tSNE plot showing all of the information across the PCs used is preserved</span></span>
<span><span class="co">### and we see the overall similarity of the cells</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Check the clusters to see if they are influenced by any of the QC metrics</span></span>
<span><span class="co">### We can see that some of the clusters are skewed in one or more of the</span></span>
<span><span class="co">### metrics we’ve calculated so we will want to take note of this.</span></span>
<span><span class="co">### Some of these skews could be biological in nature,</span></span>
<span><span class="co">### but they could be noise coming from the data.</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"nCount_RNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Cluster 8, 9, 11, and 12 could be GEMs where two or more cells were captured</span></span>
<span><span class="co">### Since they all usually high coverage and diversity</span></span>
<span><span class="co">### Theya are also small and tightly clustered away from the main groups of points</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"percent_MT"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"MALAT1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>,features<span class="op">=</span><span class="st">"percent.Largest.Gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Which largest gene</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">largest_gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">seurat_clusters</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Cell cycle</span></span>
<span><span class="va">data</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">Phase</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">seurat_clusters</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>percent <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_clusters</span>, y <span class="op">=</span> <span class="va">percent</span>, fill <span class="op">=</span> <span class="va">Phase</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Percentage of cell cycle phases per cluster"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### That’s already quite nice for explaining some of the functionality of the </span></span>
<span><span class="co">### clusters, but there’s more in there than just the behaviour of the most </span></span>
<span><span class="co">### expressed gene</span></span>
<span><span class="va">data</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">tsne</span><span class="op">@</span><span class="va">cell.embeddings</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/add_column.html">add_column</a></span><span class="op">(</span></span>
<span>      seurat_clusters <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">seurat_clusters</span>, largest_gene <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">largest_gene</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">largest_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">largest_genes_to_plot</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tSNE_1</span>, y <span class="op">=</span> <span class="va">tSNE_2</span>, colour <span class="op">=</span> <span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">largest_gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="find-markers" class="level2"><h2 class="anchored" data-anchor-id="find-markers">Find Markers</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Evaluate the cluster by identifying genes whose expression definies each</span></span>
<span><span class="co">### cluster which has been identified</span></span>
<span></span>
<span><span class="co">### Find genes appears to be upregulated in a specific cluster compared to all</span></span>
<span><span class="co">### cells not in that cluster</span></span>
<span><span class="fu">FindMarkers</span><span class="op">(</span><span class="va">data</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">### Show the expression levels of these genes in the cells in each cluster</span></span>
<span><span class="co">### VCAN gene is more highly expressed in cluster 0 than any of the other </span></span>
<span><span class="co">### clusters, but we can also see that it is also reasonably highly expressed </span></span>
<span><span class="co">### in clusters 9 and 12. These were both clusters we suspected of being </span></span>
<span><span class="co">### multiple cells though</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"VCAN"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### FindMarkers function on all of the clusters</span></span>
<span><span class="va">cluster_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="st">"seurat_clusters"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">data</span>, ident.1 <span class="op">=</span> <span class="va">x</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Adds the cluster number to the results of FindMarkers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span></span>
<span>    <span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">cluster_markers</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">cluster_markers</span><span class="op">[[</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cluster_markers</span><span class="op">[[</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="va">cluster_markers</span><span class="op">[[</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;&lt;-</span> <span class="va">x</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Generate tibble and sort by FDR to put the most significant ones first</span></span>
<span><span class="va">cluster_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">cluster_markers</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p_val_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_markers</span></span>
<span></span>
<span><span class="co">### Extract the most upregulated gene from each cluster</span></span>
<span><span class="va">best_wilcox_gene_per_cluster</span> <span class="op">&lt;-</span> <span class="va">cluster_markers</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### We can see that for some clusters (eg Cluster 8 - CDKN1C) We really do have</span></span>
<span><span class="co">### a gene which can uniquely predict, but for many others (eg cluster 7 IL7R)</span></span>
<span><span class="co">### we have a hit which also picks up other clusters (clusters 1, 3 and 4 in this case).</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="va">best_wilcox_gene_per_cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Clean this up for any individual cluster by using the roc analysis.</span></span>
<span><span class="fu">FindMarkers</span><span class="op">(</span></span>
<span>    <span class="va">data</span>, ident.1 <span class="op">=</span> <span class="fl">7</span>, ident.2 <span class="op">=</span> <span class="fl">4</span>, test.use <span class="op">=</span> <span class="st">"roc"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">### Slightly better job at separating cluster 5 from cluster 4, but it also</span></span>
<span><span class="co">### comes up all over the place in other clusters</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"LTB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### This could actually be a better option to use as a marker for this cluster.</span></span>
<span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="st">"TPT1"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="automated-cell-type-annotation" class="level2"><h2 class="anchored" data-anchor-id="automated-cell-type-annotation">Automated Cell Type Annotation</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scina_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">data</span>, layer <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_signatures.RData"</span>, package <span class="op">=</span> <span class="st">"SCINA"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">signatures</span></span>
<span></span>
<span><span class="va">scina_results</span> <span class="op">&lt;-</span> <span class="fu">SCINA</span><span class="op">(</span></span>
<span>    <span class="va">scina_data</span>,</span>
<span>    <span class="va">signatures</span>,</span>
<span>    max_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    convergence_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    convergence_rate <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>    sensitivity_cutoff <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>    rm_overlap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    allow_unknown <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">scina_labels</span> <span class="op">&lt;-</span> <span class="va">scina_results</span><span class="op">$</span><span class="va">cell_labels</span></span>
<span></span>
<span><span class="co">### plot out the tsne spread coloured by the automatic annotation</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span></span>
<span>    <span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">"scina_labels"</span>, label.size <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Relate this to the clusters which we automatically detected.</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>    cluster <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">seurat_clusters</span>,</span>
<span>    cell_type <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">scina_labels</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span>, <span class="va">cell_type</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        percent <span class="op">=</span> <span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Cluster"</span>, <span class="va">cluster</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="va">percent</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_radial.html">coord_polar</a></span><span class="op">(</span><span class="st">"y"</span>, start <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Colouring by genes</span></span>
<span><span class="co">### Some of these genes very specifically isolate to their own cluster, but for</span></span>
<span><span class="co">### others we see expression which is more widely spread over a number of clusters</span></span>
<span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="va">best_wilcox_gene_per_cluster</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">data</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">tsne</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># data@reductions$tsne@cell.embeddings[, ] |&gt;</span></span>
<span><span class="co">#     as_tibble(rownames = "barcode") |&gt;</span></span>
<span><span class="co">#     mutate(barcode = paste0(barcode, "-1")) |&gt;</span></span>
<span><span class="co">#     write_csv("for_loupe_import.csv")</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="reference" class="level2"><h2 class="anchored" data-anchor-id="reference">Reference</h2>
<ol type="1">
<li><p><a href="https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html#Gene_Selection">Seurat Example</a></p></li>
<li><p><a href="https://hbctraining.github.io/scRNA-seq_online/schedule/links-to-lessons.html">Single-cell RNA-seq data analysis workshop</a></p></li>
</ol></section><section id="session-info" class="level2"><h2 class="anchored" data-anchor-id="session-info">Session info</h2>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div>


<!-- -->

</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/grgrzhong\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb26" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Seurat V5 | 10X course example"</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> 2024-03-18</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="an">date-modified:</span><span class="co"> last-modified</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co">    - seurat</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="co">    - scrna</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="co"># image:</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="co"># description: </span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co">    freeze: true</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="co">    eval: false</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include=FALSE}</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   fig.width = 6,</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   fig.height = 6 * 0.618,</span></span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   fig.retina = 3,</span></span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">dev =</span> <span class="st">"ragg_png"</span>,</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">fig.align =</span> <span class="st">"center"</span>,</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">out.width =</span> <span class="st">"90%"</span>,</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a><span class="co">#   cache.extra = 1234,  # Change number to invalidate cache</span></span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">collapse =</span> <span class="cn">TRUE</span></span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">digits =</span> <span class="dv">4</span>,</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">width =</span> <span class="dv">300</span>,</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">dplyr.summarise.inform =</span> <span class="cn">FALSE</span></span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a><span class="fu">## Load packages and data</span></span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SCINA)</span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span>
<span id="cb26-46"><a href="#cb26-46" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-47"><a href="#cb26-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-48"><a href="#cb26-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-51"><a href="#cb26-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-52"><a href="#cb26-52" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"projects/babraham_training_bioinformatics/10X_scRNAseq_Course"</span>)</span>
<span id="cb26-53"><a href="#cb26-53" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">Read10X_h5</span>(<span class="fu">here</span>(dir, <span class="st">"data/filtered_feature_bc_matrix.h5"</span>))</span>
<span id="cb26-54"><a href="#cb26-54" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb26-55"><a href="#cb26-55" aria-hidden="true" tabindex="-1"></a>    <span class="at">counts =</span> data,</span>
<span id="cb26-56"><a href="#cb26-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">project =</span> <span class="st">"course"</span>,</span>
<span id="cb26-57"><a href="#cb26-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.cells =</span> <span class="dv">3</span>,</span>
<span id="cb26-58"><a href="#cb26-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.features =</span> <span class="dv">200</span></span>
<span id="cb26-59"><a href="#cb26-59" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-60"><a href="#cb26-60" aria-hidden="true" tabindex="-1"></a>data</span>
<span id="cb26-61"><a href="#cb26-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-62"><a href="#cb26-62" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(data))</span>
<span id="cb26-63"><a href="#cb26-63" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-64"><a href="#cb26-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-65"><a href="#cb26-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-66"><a href="#cb26-66" aria-hidden="true" tabindex="-1"></a><span class="fu">## Explore QC metrics</span></span>
<span id="cb26-67"><a href="#cb26-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-70"><a href="#cb26-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-71"><a href="#cb26-71" aria-hidden="true" tabindex="-1"></a><span class="do">### Amount of MT genes</span></span>
<span id="cb26-72"><a href="#cb26-72" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">"^MT-"</span>, <span class="fu">rownames</span>(data), <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb26-73"><a href="#cb26-73" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>percent_MT <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(data, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb26-74"><a href="#cb26-74" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data<span class="sc">$</span>percent_MT)</span>
<span id="cb26-75"><a href="#cb26-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-76"><a href="#cb26-76" aria-hidden="true" tabindex="-1"></a><span class="do">### Amount of Ribosomal genes</span></span>
<span id="cb26-77"><a href="#cb26-77" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">"^RP[LS]"</span>, <span class="fu">rownames</span>(data), <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb26-78"><a href="#cb26-78" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>percent_Ribosomal <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(data, <span class="at">pattern =</span> <span class="st">"^RP[LS]"</span>)</span>
<span id="cb26-79"><a href="#cb26-79" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data<span class="sc">$</span>percent_Ribosomal)</span>
<span id="cb26-80"><a href="#cb26-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-81"><a href="#cb26-81" aria-hidden="true" tabindex="-1"></a><span class="do">### Percentage of Largest Gene</span></span>
<span id="cb26-82"><a href="#cb26-82" aria-hidden="true" tabindex="-1"></a>data_nomalat <span class="ot">&lt;-</span> data[<span class="fu">rownames</span>(data) <span class="sc">!=</span> <span class="st">"MALAT1"</span>, ]</span>
<span id="cb26-83"><a href="#cb26-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-84"><a href="#cb26-84" aria-hidden="true" tabindex="-1"></a>data_nomalat<span class="sc">$</span>largest_count <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb26-85"><a href="#cb26-85" aria-hidden="true" tabindex="-1"></a>    data_nomalat<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts,</span>
<span id="cb26-86"><a href="#cb26-86" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span>,</span>
<span id="cb26-87"><a href="#cb26-87" aria-hidden="true" tabindex="-1"></a>    max</span>
<span id="cb26-88"><a href="#cb26-88" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-89"><a href="#cb26-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-90"><a href="#cb26-90" aria-hidden="true" tabindex="-1"></a>data_nomalat<span class="sc">$</span>largest_index <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb26-91"><a href="#cb26-91" aria-hidden="true" tabindex="-1"></a>    data_nomalat<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>counts,</span>
<span id="cb26-92"><a href="#cb26-92" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span>,</span>
<span id="cb26-93"><a href="#cb26-93" aria-hidden="true" tabindex="-1"></a>    which.max</span>
<span id="cb26-94"><a href="#cb26-94" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-95"><a href="#cb26-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-96"><a href="#cb26-96" aria-hidden="true" tabindex="-1"></a>data_nomalat<span class="sc">$</span>largest_gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(data_nomalat)[data_nomalat<span class="sc">$</span>largest_index]</span>
<span id="cb26-97"><a href="#cb26-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-98"><a href="#cb26-98" aria-hidden="true" tabindex="-1"></a>data_nomalat<span class="sc">$</span>percent.Largest.Gene <span class="ot">&lt;-</span> <span class="dv">100</span> <span class="sc">*</span> data_nomalat<span class="sc">$</span>largest_count <span class="sc">/</span></span>
<span id="cb26-99"><a href="#cb26-99" aria-hidden="true" tabindex="-1"></a>    data_nomalat<span class="sc">$</span>nCount_RNA</span>
<span id="cb26-100"><a href="#cb26-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-101"><a href="#cb26-101" aria-hidden="true" tabindex="-1"></a>data_nomalat[[]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ]</span>
<span id="cb26-102"><a href="#cb26-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-103"><a href="#cb26-103" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>largest_gene <span class="ot">&lt;-</span> data_nomalat<span class="sc">$</span>largest_gene</span>
<span id="cb26-104"><a href="#cb26-104" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>percent.Largest.Gene <span class="ot">&lt;-</span> data_nomalat<span class="sc">$</span>percent.Largest.Gene</span>
<span id="cb26-105"><a href="#cb26-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-106"><a href="#cb26-106" aria-hidden="true" tabindex="-1"></a>data[[]][<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ]</span>
<span id="cb26-107"><a href="#cb26-107" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(data_nomalat)</span>
<span id="cb26-108"><a href="#cb26-108" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-109"><a href="#cb26-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-112"><a href="#cb26-112" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-113"><a href="#cb26-113" aria-hidden="true" tabindex="-1"></a><span class="do">### For some metrics it’s better to view on a log scale.</span></span>
<span id="cb26-114"><a href="#cb26-114" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(</span>
<span id="cb26-115"><a href="#cb26-115" aria-hidden="true" tabindex="-1"></a>    data,</span>
<span id="cb26-116"><a href="#cb26-116" aria-hidden="true" tabindex="-1"></a>    <span class="at">layer =</span> <span class="st">"counts"</span>,</span>
<span id="cb26-117"><a href="#cb26-117" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">c</span>(</span>
<span id="cb26-118"><a href="#cb26-118" aria-hidden="true" tabindex="-1"></a>        <span class="st">"nCount_RNA"</span>, <span class="st">"percent_MT"</span>, <span class="st">"percent_Ribosomal"</span>, <span class="st">"percent.Largest.Gene"</span></span>
<span id="cb26-119"><a href="#cb26-119" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb26-120"><a href="#cb26-120" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb26-121"><a href="#cb26-121" aria-hidden="true" tabindex="-1"></a>    <span class="at">log =</span> <span class="cn">TRUE</span></span>
<span id="cb26-122"><a href="#cb26-122" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-123"><a href="#cb26-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-124"><a href="#cb26-124" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(</span>
<span id="cb26-125"><a href="#cb26-125" aria-hidden="true" tabindex="-1"></a>    data,</span>
<span id="cb26-126"><a href="#cb26-126" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, </span>
<span id="cb26-127"><a href="#cb26-127" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature2 =</span> <span class="st">"percent.Largest.Gene"</span></span>
<span id="cb26-128"><a href="#cb26-128" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-129"><a href="#cb26-129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-130"><a href="#cb26-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-133"><a href="#cb26-133" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-134"><a href="#cb26-134" aria-hidden="true" tabindex="-1"></a><span class="do">## ggplot</span></span>
<span id="cb26-135"><a href="#cb26-135" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">as_tibble</span>(</span>
<span id="cb26-136"><a href="#cb26-136" aria-hidden="true" tabindex="-1"></a>    data[[]],</span>
<span id="cb26-137"><a href="#cb26-137" aria-hidden="true" tabindex="-1"></a>    <span class="at">rownames =</span> <span class="st">"Cell.Barcode"</span></span>
<span id="cb26-138"><a href="#cb26-138" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-139"><a href="#cb26-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-140"><a href="#cb26-140" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(qc_metrics)</span>
<span id="cb26-141"><a href="#cb26-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-142"><a href="#cb26-142" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-143"><a href="#cb26-143" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(percent_MT) <span class="sc">|&gt;</span></span>
<span id="cb26-144"><a href="#cb26-144" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(nCount_RNA, nFeature_RNA, <span class="at">colour =</span> percent_MT)) <span class="sc">+</span></span>
<span id="cb26-145"><a href="#cb26-145" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-146"><a href="#cb26-146" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green2"</span>, <span class="st">"red"</span>, <span class="st">"yellow"</span>)) <span class="sc">+</span></span>
<span id="cb26-147"><a href="#cb26-147" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Example of plotting QC metrics"</span>) <span class="sc">+</span></span>
<span id="cb26-148"><a href="#cb26-148" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">750</span>) <span class="sc">+</span></span>
<span id="cb26-149"><a href="#cb26-149" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2000</span>)</span>
<span id="cb26-150"><a href="#cb26-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-151"><a href="#cb26-151" aria-hidden="true" tabindex="-1"></a><span class="do">## Log scale</span></span>
<span id="cb26-152"><a href="#cb26-152" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-153"><a href="#cb26-153" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(percent_MT) <span class="sc">|&gt;</span></span>
<span id="cb26-154"><a href="#cb26-154" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(nCount_RNA, nFeature_RNA, <span class="at">colour =</span> percent_MT)) <span class="sc">+</span></span>
<span id="cb26-155"><a href="#cb26-155" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb26-156"><a href="#cb26-156" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green2"</span>, <span class="st">"red"</span>, <span class="st">"yellow"</span>)) <span class="sc">+</span></span>
<span id="cb26-157"><a href="#cb26-157" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Example of plotting QC metrics"</span>) <span class="sc">+</span></span>
<span id="cb26-158"><a href="#cb26-158" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">750</span>) <span class="sc">+</span></span>
<span id="cb26-159"><a href="#cb26-159" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2000</span>) <span class="sc">+</span></span>
<span id="cb26-160"><a href="#cb26-160" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb26-161"><a href="#cb26-161" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>()</span>
<span id="cb26-162"><a href="#cb26-162" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-163"><a href="#cb26-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-166"><a href="#cb26-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-167"><a href="#cb26-167" aria-hidden="true" tabindex="-1"></a><span class="do">### Complexity </span></span>
<span id="cb26-168"><a href="#cb26-168" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="ot">&lt;-</span> qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-169"><a href="#cb26-169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">complexity=</span><span class="fu">log10</span>(nFeature_RNA) <span class="sc">/</span> <span class="fu">log10</span>(nCount_RNA))</span>
<span id="cb26-170"><a href="#cb26-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-171"><a href="#cb26-171" aria-hidden="true" tabindex="-1"></a>complexity.lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="fu">log10</span>(qc_metrics<span class="sc">$</span>nFeature_RNA)<span class="sc">~</span><span class="fu">log10</span>(qc_metrics<span class="sc">$</span>nCount_RNA)) </span>
<span id="cb26-172"><a href="#cb26-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-173"><a href="#cb26-173" aria-hidden="true" tabindex="-1"></a>complexity.lm</span>
<span id="cb26-174"><a href="#cb26-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-175"><a href="#cb26-175" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="ot">&lt;-</span> qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-176"><a href="#cb26-176" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb26-177"><a href="#cb26-177" aria-hidden="true" tabindex="-1"></a>    <span class="at">complexity_diff =</span> <span class="fu">log10</span>(nFeature_RNA) <span class="sc">-</span> ((<span class="fu">log10</span>(qc_metrics<span class="sc">$</span>nCount_RNA)<span class="sc">*</span>complexity.lm<span class="sc">$</span>coefficients[<span class="dv">2</span>])<span class="sc">+</span>complexity.lm<span class="sc">$</span>coefficients[<span class="dv">1</span>])</span>
<span id="cb26-178"><a href="#cb26-178" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb26-179"><a href="#cb26-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-180"><a href="#cb26-180" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-181"><a href="#cb26-181" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>complexity_diff)) <span class="sc">+</span></span>
<span id="cb26-182"><a href="#cb26-182" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill=</span><span class="st">"yellow"</span>) </span>
<span id="cb26-183"><a href="#cb26-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-184"><a href="#cb26-184" aria-hidden="true" tabindex="-1"></a>complexity_scale <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="fu">c</span>(<span class="fu">max</span>(qc_metrics<span class="sc">$</span>complexity_diff),<span class="dv">0</span><span class="sc">-</span><span class="fu">min</span>(qc_metrics<span class="sc">$</span>complexity_diff)))</span>
<span id="cb26-185"><a href="#cb26-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-186"><a href="#cb26-186" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-187"><a href="#cb26-187" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">complexity_diff=</span><span class="fu">replace</span>(complexity_diff,complexity_diff<span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">0.1</span>,<span class="sc">-</span><span class="fl">0.1</span>)) <span class="sc">|&gt;</span></span>
<span id="cb26-188"><a href="#cb26-188" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">log10</span>(nCount_RNA), <span class="at">y=</span><span class="fu">log10</span>(nFeature_RNA), <span class="at">colour=</span>complexity_diff)) <span class="sc">+</span></span>
<span id="cb26-189"><a href="#cb26-189" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb26-190"><a href="#cb26-190" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope=</span>complexity.lm<span class="sc">$</span>coefficients[<span class="dv">2</span>], <span class="at">intercept =</span> complexity.lm<span class="sc">$</span>coefficients[<span class="dv">1</span>]) <span class="sc">+</span></span>
<span id="cb26-191"><a href="#cb26-191" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient2</span>(<span class="at">low=</span><span class="st">"blue2"</span>,<span class="at">mid=</span><span class="st">"grey"</span>,<span class="at">high=</span><span class="st">"red2"</span>)</span>
<span id="cb26-192"><a href="#cb26-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-193"><a href="#cb26-193" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-194"><a href="#cb26-194" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>complexity_diff, <span class="at">y=</span>percent.Largest.Gene)) <span class="sc">+</span></span>
<span id="cb26-195"><a href="#cb26-195" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb26-196"><a href="#cb26-196" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-197"><a href="#cb26-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-200"><a href="#cb26-200" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-201"><a href="#cb26-201" aria-hidden="true" tabindex="-1"></a><span class="do">### Largest gene</span></span>
<span id="cb26-202"><a href="#cb26-202" aria-hidden="true" tabindex="-1"></a>largest_gene_list <span class="ot">&lt;-</span> qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-203"><a href="#cb26-203" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(largest_gene) <span class="sc">|&gt;</span></span>
<span id="cb26-204"><a href="#cb26-204" aria-hidden="true" tabindex="-1"></a>    <span class="fu">count</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-205"><a href="#cb26-205" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span>
<span id="cb26-206"><a href="#cb26-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-207"><a href="#cb26-207" aria-hidden="true" tabindex="-1"></a>largest_genes_to_plot <span class="ot">&lt;-</span> largest_gene_list <span class="sc">|&gt;</span></span>
<span id="cb26-208"><a href="#cb26-208" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(n <span class="sc">&gt;</span> <span class="dv">140</span>) <span class="sc">|&gt;</span></span>
<span id="cb26-209"><a href="#cb26-209" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(largest_gene)</span>
<span id="cb26-210"><a href="#cb26-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-211"><a href="#cb26-211" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-212"><a href="#cb26-212" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(largest_gene <span class="sc">%in%</span> largest_genes_to_plot) <span class="sc">|&gt;</span></span>
<span id="cb26-213"><a href="#cb26-213" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb26-214"><a href="#cb26-214" aria-hidden="true" tabindex="-1"></a>      <span class="at">largest_gene =</span> <span class="fu">factor</span>(largest_gene, <span class="at">levels =</span> largest_genes_to_plot)</span>
<span id="cb26-215"><a href="#cb26-215" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-216"><a href="#cb26-216" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(largest_gene) <span class="sc">|&gt;</span></span>
<span id="cb26-217"><a href="#cb26-217" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(</span>
<span id="cb26-218"><a href="#cb26-218" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(nCount_RNA), <span class="at">y =</span> <span class="fu">log10</span>(nFeature_RNA), <span class="at">colour =</span> largest_gene)</span>
<span id="cb26-219"><a href="#cb26-219" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb26-220"><a href="#cb26-220" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-221"><a href="#cb26-221" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_manual</span>(</span>
<span id="cb26-222"><a href="#cb26-222" aria-hidden="true" tabindex="-1"></a>      <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Set1"</span>))</span>
<span id="cb26-223"><a href="#cb26-223" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb26-224"><a href="#cb26-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-225"><a href="#cb26-225" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-226"><a href="#cb26-226" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(largest_gene <span class="sc">%in%</span> largest_genes_to_plot) <span class="sc">|&gt;</span></span>
<span id="cb26-227"><a href="#cb26-227" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb26-228"><a href="#cb26-228" aria-hidden="true" tabindex="-1"></a>      <span class="at">largest_gene =</span> <span class="fu">factor</span>(largest_gene, <span class="at">levels =</span> largest_genes_to_plot)</span>
<span id="cb26-229"><a href="#cb26-229" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-230"><a href="#cb26-230" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(largest_gene) <span class="sc">|&gt;</span></span>
<span id="cb26-231"><a href="#cb26-231" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(</span>
<span id="cb26-232"><a href="#cb26-232" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x =</span> complexity_diff, <span class="at">y =</span> percent.Largest.Gene, <span class="at">colour =</span> largest_gene)</span>
<span id="cb26-233"><a href="#cb26-233" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb26-234"><a href="#cb26-234" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-235"><a href="#cb26-235" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Set1"</span>)))</span>
<span id="cb26-236"><a href="#cb26-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-237"><a href="#cb26-237" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-238"><a href="#cb26-238" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(percent_MT) <span class="sc">|&gt;</span></span>
<span id="cb26-239"><a href="#cb26-239" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(</span>
<span id="cb26-240"><a href="#cb26-240" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x =</span> complexity_diff, <span class="at">y =</span> percent.Largest.Gene, <span class="at">colour =</span> percent_MT)</span>
<span id="cb26-241"><a href="#cb26-241" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb26-242"><a href="#cb26-242" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-243"><a href="#cb26-243" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">"grey"</span>, <span class="at">high =</span> <span class="st">"red2"</span>)</span>
<span id="cb26-244"><a href="#cb26-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-245"><a href="#cb26-245" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-246"><a href="#cb26-246" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(percent_Ribosomal) <span class="sc">|&gt;</span></span>
<span id="cb26-247"><a href="#cb26-247" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(</span>
<span id="cb26-248"><a href="#cb26-248" aria-hidden="true" tabindex="-1"></a>      <span class="fu">aes</span>(<span class="at">x =</span> complexity_diff, <span class="at">y =</span> percent.Largest.Gene, <span class="at">colour =</span> percent_Ribosomal)</span>
<span id="cb26-249"><a href="#cb26-249" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb26-250"><a href="#cb26-250" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-251"><a href="#cb26-251" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">"grey"</span>, <span class="at">high =</span> <span class="st">"red2"</span>)</span>
<span id="cb26-252"><a href="#cb26-252" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-253"><a href="#cb26-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-254"><a href="#cb26-254" aria-hidden="true" tabindex="-1"></a><span class="fu">## Setting QC Cutoff and Filtering</span></span>
<span id="cb26-255"><a href="#cb26-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-258"><a href="#cb26-258" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-259"><a href="#cb26-259" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-260"><a href="#cb26-260" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(percent_MT)) <span class="sc">+</span></span>
<span id="cb26-261"><a href="#cb26-261" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.5</span>, <span class="at">fill =</span> <span class="st">"yellow"</span>, <span class="at">colour =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb26-262"><a href="#cb26-262" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Distribution of Percentage Mitochondrion"</span>) <span class="sc">+</span></span>
<span id="cb26-263"><a href="#cb26-263" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">10</span>)</span>
<span id="cb26-264"><a href="#cb26-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-265"><a href="#cb26-265" aria-hidden="true" tabindex="-1"></a>qc_metrics <span class="sc">|&gt;</span></span>
<span id="cb26-266"><a href="#cb26-266" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(percent.Largest.Gene)) <span class="sc">+</span></span>
<span id="cb26-267"><a href="#cb26-267" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.7</span>, <span class="at">fill =</span> <span class="st">"yellow"</span>, <span class="at">colour =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb26-268"><a href="#cb26-268" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Distribution of Percentage Largest Gene"</span>) <span class="sc">+</span></span>
<span id="cb26-269"><a href="#cb26-269" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">10</span>) <span class="do">### Filtering</span></span>
<span id="cb26-270"><a href="#cb26-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-271"><a href="#cb26-271" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb26-272"><a href="#cb26-272" aria-hidden="true" tabindex="-1"></a>    data,</span>
<span id="cb26-273"><a href="#cb26-273" aria-hidden="true" tabindex="-1"></a>    nFeature_RNA <span class="sc">&gt;</span> <span class="dv">750</span> <span class="sc">&amp;</span></span>
<span id="cb26-274"><a href="#cb26-274" aria-hidden="true" tabindex="-1"></a>        nFeature_RNA <span class="sc">&lt;</span> <span class="dv">2000</span> <span class="sc">&amp;</span></span>
<span id="cb26-275"><a href="#cb26-275" aria-hidden="true" tabindex="-1"></a>        percent_MT <span class="sc">&lt;</span> <span class="dv">10</span> <span class="sc">&amp;</span></span>
<span id="cb26-276"><a href="#cb26-276" aria-hidden="true" tabindex="-1"></a>        percent.Largest.Gene <span class="sc">&lt;</span> <span class="dv">10</span></span>
<span id="cb26-277"><a href="#cb26-277" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-278"><a href="#cb26-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-279"><a href="#cb26-279" aria-hidden="true" tabindex="-1"></a>data</span>
<span id="cb26-280"><a href="#cb26-280" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-281"><a href="#cb26-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-282"><a href="#cb26-282" aria-hidden="true" tabindex="-1"></a><span class="fu">## Normalisation, Selection and Scaling</span></span>
<span id="cb26-283"><a href="#cb26-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-286"><a href="#cb26-286" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-287"><a href="#cb26-287" aria-hidden="true" tabindex="-1"></a><span class="do">### Normalization</span></span>
<span id="cb26-288"><a href="#cb26-288" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(data, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>)</span>
<span id="cb26-289"><a href="#cb26-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-290"><a href="#cb26-290" aria-hidden="true" tabindex="-1"></a>gene_expression <span class="ot">&lt;-</span> <span class="fu">apply</span>(data<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data, <span class="dv">1</span>, mean)</span>
<span id="cb26-291"><a href="#cb26-291" aria-hidden="true" tabindex="-1"></a>gene_expression <span class="ot">&lt;-</span> <span class="fu">names</span>(gene_expression) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(data)</span>
<span id="cb26-292"><a href="#cb26-292" aria-hidden="true" tabindex="-1"></a>gene_expression <span class="ot">&lt;-</span> <span class="fu">sort</span>(gene_expression, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb26-293"><a href="#cb26-293" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gene_expression, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb26-294"><a href="#cb26-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-295"><a href="#cb26-295" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(data[<span class="st">"GAPDH"</span>, ]<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb26-296"><a href="#cb26-296" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.05</span>, <span class="at">fill =</span> <span class="st">"yellow"</span>, <span class="at">colour =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb26-297"><a href="#cb26-297" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"GAPDH expression"</span>)</span>
<span id="cb26-298"><a href="#cb26-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-299"><a href="#cb26-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-300"><a href="#cb26-300" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(</span>
<span id="cb26-301"><a href="#cb26-301" aria-hidden="true" tabindex="-1"></a>    data<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>layers<span class="sc">$</span>data[, <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]</span>
<span id="cb26-302"><a href="#cb26-302" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb26-303"><a href="#cb26-303" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb26-304"><a href="#cb26-304" aria-hidden="true" tabindex="-1"></a>        <span class="at">cols =</span> <span class="fu">everything</span>(),</span>
<span id="cb26-305"><a href="#cb26-305" aria-hidden="true" tabindex="-1"></a>        <span class="at">names_to =</span> <span class="st">"cell"</span>,</span>
<span id="cb26-306"><a href="#cb26-306" aria-hidden="true" tabindex="-1"></a>        <span class="at">values_to =</span> <span class="st">"expression"</span></span>
<span id="cb26-307"><a href="#cb26-307" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-308"><a href="#cb26-308" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> cell, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb26-309"><a href="#cb26-309" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_summary</span>(<span class="at">geom =</span> <span class="st">"crossbar"</span>, <span class="at">fun.data =</span> mean_sdl)</span>
<span id="cb26-310"><a href="#cb26-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-311"><a href="#cb26-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-314"><a href="#cb26-314" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-315"><a href="#cb26-315" aria-hidden="true" tabindex="-1"></a><span class="do">### Cell Cycle Scoring</span></span>
<span id="cb26-316"><a href="#cb26-316" aria-hidden="true" tabindex="-1"></a>cc.genes.updated<span class="fl">.2019</span></span>
<span id="cb26-317"><a href="#cb26-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-318"><a href="#cb26-318" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(</span>
<span id="cb26-319"><a href="#cb26-319" aria-hidden="true" tabindex="-1"></a>    data,</span>
<span id="cb26-320"><a href="#cb26-320" aria-hidden="true" tabindex="-1"></a>    <span class="at">s.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>s.genes,</span>
<span id="cb26-321"><a href="#cb26-321" aria-hidden="true" tabindex="-1"></a>    <span class="at">g2m.features =</span> cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>g2m.genes,</span>
<span id="cb26-322"><a href="#cb26-322" aria-hidden="true" tabindex="-1"></a>    <span class="at">set.ident =</span> <span class="cn">TRUE</span></span>
<span id="cb26-323"><a href="#cb26-323" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-324"><a href="#cb26-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-325"><a href="#cb26-325" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(data[[]])</span>
<span id="cb26-326"><a href="#cb26-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-327"><a href="#cb26-327" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(data[[]]) <span class="sc">|&gt;</span></span>
<span id="cb26-328"><a href="#cb26-328" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(Phase)) <span class="sc">+</span> <span class="fu">geom_bar</span>()</span>
<span id="cb26-329"><a href="#cb26-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-330"><a href="#cb26-330" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(data[[]]) <span class="sc">|&gt;</span></span>
<span id="cb26-331"><a href="#cb26-331" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> S.Score, <span class="at">y =</span> G2M.Score, <span class="at">color =</span> Phase)) <span class="sc">+</span></span>
<span id="cb26-332"><a href="#cb26-332" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-333"><a href="#cb26-333" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.15</span>, <span class="fl">0.15</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.15</span>, <span class="fl">0.15</span>))</span>
<span id="cb26-334"><a href="#cb26-334" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-335"><a href="#cb26-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-336"><a href="#cb26-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-339"><a href="#cb26-339" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-340"><a href="#cb26-340" aria-hidden="true" tabindex="-1"></a><span class="do">### Gene selection</span></span>
<span id="cb26-341"><a href="#cb26-341" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(</span>
<span id="cb26-342"><a href="#cb26-342" aria-hidden="true" tabindex="-1"></a>    data,</span>
<span id="cb26-343"><a href="#cb26-343" aria-hidden="true" tabindex="-1"></a>    <span class="at">selection.method =</span> <span class="st">"vst"</span>,</span>
<span id="cb26-344"><a href="#cb26-344" aria-hidden="true" tabindex="-1"></a>    <span class="at">nfeatures =</span> <span class="dv">500</span></span>
<span id="cb26-345"><a href="#cb26-345" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-346"><a href="#cb26-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-347"><a href="#cb26-347" aria-hidden="true" tabindex="-1"></a>variance_data <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb26-348"><a href="#cb26-348" aria-hidden="true" tabindex="-1"></a>    <span class="fu">HVFInfo</span>(data), <span class="at">rownames =</span> <span class="st">"Gene"</span></span>
<span id="cb26-349"><a href="#cb26-349" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-350"><a href="#cb26-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-351"><a href="#cb26-351" aria-hidden="true" tabindex="-1"></a>variance_data <span class="ot">&lt;-</span> variance_data <span class="sc">|&gt;</span></span>
<span id="cb26-352"><a href="#cb26-352" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb26-353"><a href="#cb26-353" aria-hidden="true" tabindex="-1"></a>        <span class="at">hypervariable =</span> Gene <span class="sc">%in%</span> <span class="fu">VariableFeatures</span>(data)</span>
<span id="cb26-354"><a href="#cb26-354" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb26-355"><a href="#cb26-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-356"><a href="#cb26-356" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(variance_data, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb26-357"><a href="#cb26-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-358"><a href="#cb26-358" aria-hidden="true" tabindex="-1"></a>variance_data <span class="sc">|&gt;</span></span>
<span id="cb26-359"><a href="#cb26-359" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="fu">log</span>(mean), <span class="fu">log</span>(variance), <span class="at">color =</span> hypervariable)) <span class="sc">+</span></span>
<span id="cb26-360"><a href="#cb26-360" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-361"><a href="#cb26-361" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span>
<span id="cb26-362"><a href="#cb26-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-363"><a href="#cb26-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-364"><a href="#cb26-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-367"><a href="#cb26-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-368"><a href="#cb26-368" aria-hidden="true" tabindex="-1"></a><span class="do">### Scaling</span></span>
<span id="cb26-369"><a href="#cb26-369" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(data, <span class="at">features =</span> <span class="fu">rownames</span>(data))</span>
<span id="cb26-370"><a href="#cb26-370" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-371"><a href="#cb26-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-372"><a href="#cb26-372" aria-hidden="true" tabindex="-1"></a><span class="fu">## Dimensionality reduction</span></span>
<span id="cb26-373"><a href="#cb26-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-374"><a href="#cb26-374" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA</span></span>
<span id="cb26-377"><a href="#cb26-377" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-378"><a href="#cb26-378" aria-hidden="true" tabindex="-1"></a><span class="do">### Calculate all PCs</span></span>
<span id="cb26-379"><a href="#cb26-379" aria-hidden="true" tabindex="-1"></a><span class="do">### list of genes which are most highly and lowly weighted in the different PCs</span></span>
<span id="cb26-380"><a href="#cb26-380" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(data, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(data))</span>
<span id="cb26-381"><a href="#cb26-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-382"><a href="#cb26-382" aria-hidden="true" tabindex="-1"></a><span class="do">### See if the cell cycle is having a big effect on the clusters we’re picking out</span></span>
<span id="cb26-383"><a href="#cb26-383" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb26-384"><a href="#cb26-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-385"><a href="#cb26-385" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(</span>
<span id="cb26-386"><a href="#cb26-386" aria-hidden="true" tabindex="-1"></a>    data, <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb26-387"><a href="#cb26-387" aria-hidden="true" tabindex="-1"></a>    <span class="at">group.by =</span> <span class="st">"largest_gene"</span>,</span>
<span id="cb26-388"><a href="#cb26-388" aria-hidden="true" tabindex="-1"></a>    <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb26-389"><a href="#cb26-389" aria-hidden="true" tabindex="-1"></a>    <span class="at">label.size =</span> <span class="dv">3</span></span>
<span id="cb26-390"><a href="#cb26-390" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb26-391"><a href="#cb26-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-392"><a href="#cb26-392" aria-hidden="true" tabindex="-1"></a><span class="do">### This nicely shows us the power, but also the limitations of PCA in that we </span></span>
<span id="cb26-393"><a href="#cb26-393" aria-hidden="true" tabindex="-1"></a><span class="do">### can see that not all of the useful information is captured in the first two </span></span>
<span id="cb26-394"><a href="#cb26-394" aria-hidden="true" tabindex="-1"></a><span class="do">### principal components.</span></span>
<span id="cb26-395"><a href="#cb26-395" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>))</span>
<span id="cb26-396"><a href="#cb26-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-397"><a href="#cb26-397" aria-hidden="true" tabindex="-1"></a><span class="do">### How far down the set of PCs do we need to go to capture all of the </span></span>
<span id="cb26-398"><a href="#cb26-398" aria-hidden="true" tabindex="-1"></a><span class="do">### biologically relevant information.</span></span>
<span id="cb26-399"><a href="#cb26-399" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(data)</span>
<span id="cb26-400"><a href="#cb26-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-401"><a href="#cb26-401" aria-hidden="true" tabindex="-1"></a><span class="do">### Plots of PCA weightings for the most highly and lowly weighted genes, </span></span>
<span id="cb26-402"><a href="#cb26-402" aria-hidden="true" tabindex="-1"></a><span class="do">### shown against the set of cells which are most highly influenced by the PC. </span></span>
<span id="cb26-403"><a href="#cb26-403" aria-hidden="true" tabindex="-1"></a><span class="do">### The idea is that as long as we’re seeing clear structure in one of these </span></span>
<span id="cb26-404"><a href="#cb26-404" aria-hidden="true" tabindex="-1"></a><span class="do">### plots then we’re still adding potentially useful information to the analysis.</span></span>
<span id="cb26-405"><a href="#cb26-405" aria-hidden="true" tabindex="-1"></a><span class="fu">DimHeatmap</span>(data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">cells =</span> <span class="dv">500</span>)</span>
<span id="cb26-406"><a href="#cb26-406" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-407"><a href="#cb26-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-408"><a href="#cb26-408" aria-hidden="true" tabindex="-1"></a><span class="fu">### tSNE</span></span>
<span id="cb26-409"><a href="#cb26-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-412"><a href="#cb26-412" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-413"><a href="#cb26-413" aria-hidden="true" tabindex="-1"></a><span class="do">### Setting this to a low value will help resolve small clusters, </span></span>
<span id="cb26-414"><a href="#cb26-414" aria-hidden="true" tabindex="-1"></a><span class="do">### but at the expense of large clusters becoming more diffuse</span></span>
<span id="cb26-415"><a href="#cb26-415" aria-hidden="true" tabindex="-1"></a>saved_seed <span class="ot">&lt;-</span> <span class="dv">8482</span></span>
<span id="cb26-416"><a href="#cb26-416" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(saved_seed)</span>
<span id="cb26-417"><a href="#cb26-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-418"><a href="#cb26-418" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">seed.use =</span> saved_seed, <span class="at">perplexity =</span> <span class="dv">10</span>)</span>
<span id="cb26-419"><a href="#cb26-419" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">pt.size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-420"><a href="#cb26-420" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"tSNE with Perplexity 10"</span>)</span>
<span id="cb26-421"><a href="#cb26-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-422"><a href="#cb26-422" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">seed.use =</span> saved_seed, <span class="at">perplexity =</span> <span class="dv">200</span>)</span>
<span id="cb26-423"><a href="#cb26-423" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">pt.size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-424"><a href="#cb26-424" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"tSNE with Perplexity 200"</span>)</span>
<span id="cb26-425"><a href="#cb26-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-426"><a href="#cb26-426" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">seed.use =</span> saved_seed)</span>
<span id="cb26-427"><a href="#cb26-427" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">pt.size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-428"><a href="#cb26-428" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"tSNE with default Perplexity (30)"</span>)</span>
<span id="cb26-429"><a href="#cb26-429" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-430"><a href="#cb26-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-431"><a href="#cb26-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-432"><a href="#cb26-432" aria-hidden="true" tabindex="-1"></a><span class="fu">## Defining Cell Clusters</span></span>
<span id="cb26-433"><a href="#cb26-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-436"><a href="#cb26-436" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-437"><a href="#cb26-437" aria-hidden="true" tabindex="-1"></a><span class="do">### Finds the ‘k’ nearest neighbours to each cell and makes this into a graph</span></span>
<span id="cb26-438"><a href="#cb26-438" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(data, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb26-439"><a href="#cb26-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-440"><a href="#cb26-440" aria-hidden="true" tabindex="-1"></a><span class="do">### Get another sparse matrix of distances.</span></span>
<span id="cb26-441"><a href="#cb26-441" aria-hidden="true" tabindex="-1"></a>data<span class="sc">@</span>graphs<span class="sc">$</span>RNA_snn[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb26-442"><a href="#cb26-442" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-443"><a href="#cb26-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-444"><a href="#cb26-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-447"><a href="#cb26-447" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-448"><a href="#cb26-448" aria-hidden="true" tabindex="-1"></a><span class="do">### Segment the graph, </span></span>
<span id="cb26-449"><a href="#cb26-449" aria-hidden="true" tabindex="-1"></a><span class="do">### Larger resolution give larger clusters, smaller values gives smaller clusters.</span></span>
<span id="cb26-450"><a href="#cb26-450" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(data, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb26-451"><a href="#cb26-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-452"><a href="#cb26-452" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data<span class="sc">$</span>seurat_clusters, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb26-453"><a href="#cb26-453" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-454"><a href="#cb26-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-457"><a href="#cb26-457" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-458"><a href="#cb26-458" aria-hidden="true" tabindex="-1"></a><span class="do">### Some of the clusters don't resolve very well in PC1 vs PC2</span></span>
<span id="cb26-459"><a href="#cb26-459" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb26-460"><a href="#cb26-460" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"PC1 vs PC2 with Clusters"</span>)</span>
<span id="cb26-461"><a href="#cb26-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-462"><a href="#cb26-462" aria-hidden="true" tabindex="-1"></a><span class="do">### Some of the clusters which are overlaid in PC1 start to separate in other PCs</span></span>
<span id="cb26-463"><a href="#cb26-463" aria-hidden="true" tabindex="-1"></a><span class="do">### These differences represent a small proportion of the overall variance</span></span>
<span id="cb26-464"><a href="#cb26-464" aria-hidden="true" tabindex="-1"></a><span class="do">### but can be important in resolving changes.</span></span>
<span id="cb26-465"><a href="#cb26-465" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">9</span>), <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb26-466"><a href="#cb26-466" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"PC4 vs PC9 with Clusters"</span>)</span>
<span id="cb26-467"><a href="#cb26-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-468"><a href="#cb26-468" aria-hidden="true" tabindex="-1"></a><span class="do">### tSNE plot showing all of the information across the PCs used is preserved</span></span>
<span id="cb26-469"><a href="#cb26-469" aria-hidden="true" tabindex="-1"></a><span class="do">### and we see the overall similarity of the cells</span></span>
<span id="cb26-470"><a href="#cb26-470" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(data, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">pt.size =</span> <span class="dv">1</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">7</span>)</span>
<span id="cb26-471"><a href="#cb26-471" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-472"><a href="#cb26-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-475"><a href="#cb26-475" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-476"><a href="#cb26-476" aria-hidden="true" tabindex="-1"></a><span class="do">### Check the clusters to see if they are influenced by any of the QC metrics</span></span>
<span id="cb26-477"><a href="#cb26-477" aria-hidden="true" tabindex="-1"></a><span class="do">### We can see that some of the clusters are skewed in one or more of the</span></span>
<span id="cb26-478"><a href="#cb26-478" aria-hidden="true" tabindex="-1"></a><span class="do">### metrics we’ve calculated so we will want to take note of this.</span></span>
<span id="cb26-479"><a href="#cb26-479" aria-hidden="true" tabindex="-1"></a><span class="do">### Some of these skews could be biological in nature,</span></span>
<span id="cb26-480"><a href="#cb26-480" aria-hidden="true" tabindex="-1"></a><span class="do">### but they could be noise coming from the data.</span></span>
<span id="cb26-481"><a href="#cb26-481" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"nCount_RNA"</span>)</span>
<span id="cb26-482"><a href="#cb26-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-483"><a href="#cb26-483" aria-hidden="true" tabindex="-1"></a><span class="do">### Cluster 8, 9, 11, and 12 could be GEMs where two or more cells were captured</span></span>
<span id="cb26-484"><a href="#cb26-484" aria-hidden="true" tabindex="-1"></a><span class="do">### Since they all usually high coverage and diversity</span></span>
<span id="cb26-485"><a href="#cb26-485" aria-hidden="true" tabindex="-1"></a><span class="do">### Theya are also small and tightly clustered away from the main groups of points</span></span>
<span id="cb26-486"><a href="#cb26-486" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"nFeature_RNA"</span>)</span>
<span id="cb26-487"><a href="#cb26-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-488"><a href="#cb26-488" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"percent_MT"</span>)</span>
<span id="cb26-489"><a href="#cb26-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-490"><a href="#cb26-490" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"MALAT1"</span>)</span>
<span id="cb26-491"><a href="#cb26-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-492"><a href="#cb26-492" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data,<span class="at">features=</span><span class="st">"percent.Largest.Gene"</span>)</span>
<span id="cb26-493"><a href="#cb26-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-494"><a href="#cb26-494" aria-hidden="true" tabindex="-1"></a><span class="do">### Which largest gene</span></span>
<span id="cb26-495"><a href="#cb26-495" aria-hidden="true" tabindex="-1"></a>data[[]] <span class="sc">|&gt;</span></span>
<span id="cb26-496"><a href="#cb26-496" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seurat_clusters, largest_gene) <span class="sc">|&gt;</span></span>
<span id="cb26-497"><a href="#cb26-497" aria-hidden="true" tabindex="-1"></a>    <span class="fu">count</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-498"><a href="#cb26-498" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="fu">desc</span>(n)) <span class="sc">|&gt;</span></span>
<span id="cb26-499"><a href="#cb26-499" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seurat_clusters) <span class="sc">|&gt;</span></span>
<span id="cb26-500"><a href="#cb26-500" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>) <span class="sc">|&gt;</span></span>
<span id="cb26-501"><a href="#cb26-501" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-502"><a href="#cb26-502" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(seurat_clusters, <span class="fu">desc</span>(n))</span>
<span id="cb26-503"><a href="#cb26-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-504"><a href="#cb26-504" aria-hidden="true" tabindex="-1"></a><span class="do">### Cell cycle</span></span>
<span id="cb26-505"><a href="#cb26-505" aria-hidden="true" tabindex="-1"></a>data<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb26-506"><a href="#cb26-506" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seurat_clusters, Phase) <span class="sc">|&gt;</span></span>
<span id="cb26-507"><a href="#cb26-507" aria-hidden="true" tabindex="-1"></a>    <span class="fu">count</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-508"><a href="#cb26-508" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seurat_clusters) <span class="sc">|&gt;</span></span>
<span id="cb26-509"><a href="#cb26-509" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">percent =</span> <span class="dv">100</span> <span class="sc">*</span> n <span class="sc">/</span> <span class="fu">sum</span>(n)) <span class="sc">|&gt;</span></span>
<span id="cb26-510"><a href="#cb26-510" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-511"><a href="#cb26-511" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> seurat_clusters, <span class="at">y =</span> percent, <span class="at">fill =</span> Phase)) <span class="sc">+</span></span>
<span id="cb26-512"><a href="#cb26-512" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb26-513"><a href="#cb26-513" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Percentage of cell cycle phases per cluster"</span>)</span>
<span id="cb26-514"><a href="#cb26-514" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-515"><a href="#cb26-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-516"><a href="#cb26-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-519"><a href="#cb26-519" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-520"><a href="#cb26-520" aria-hidden="true" tabindex="-1"></a><span class="do">### That’s already quite nice for explaining some of the functionality of the </span></span>
<span id="cb26-521"><a href="#cb26-521" aria-hidden="true" tabindex="-1"></a><span class="do">### clusters, but there’s more in there than just the behaviour of the most </span></span>
<span id="cb26-522"><a href="#cb26-522" aria-hidden="true" tabindex="-1"></a><span class="do">### expressed gene</span></span>
<span id="cb26-523"><a href="#cb26-523" aria-hidden="true" tabindex="-1"></a>data<span class="sc">@</span>reductions<span class="sc">$</span>tsne<span class="sc">@</span>cell.embeddings <span class="sc">|&gt;</span></span>
<span id="cb26-524"><a href="#cb26-524" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-525"><a href="#cb26-525" aria-hidden="true" tabindex="-1"></a>    <span class="fu">add_column</span>(</span>
<span id="cb26-526"><a href="#cb26-526" aria-hidden="true" tabindex="-1"></a>      <span class="at">seurat_clusters =</span> data<span class="sc">$</span>seurat_clusters, <span class="at">largest_gene =</span> data<span class="sc">$</span>largest_gene</span>
<span id="cb26-527"><a href="#cb26-527" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-528"><a href="#cb26-528" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(largest_gene <span class="sc">%in%</span> largest_genes_to_plot) <span class="sc">|&gt;</span></span>
<span id="cb26-529"><a href="#cb26-529" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> tSNE_1, <span class="at">y =</span> tSNE_2, <span class="at">colour =</span> seurat_clusters)) <span class="sc">+</span></span>
<span id="cb26-530"><a href="#cb26-530" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-531"><a href="#cb26-531" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(largest_gene))</span>
<span id="cb26-532"><a href="#cb26-532" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-533"><a href="#cb26-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-534"><a href="#cb26-534" aria-hidden="true" tabindex="-1"></a><span class="fu">## Find Markers</span></span>
<span id="cb26-535"><a href="#cb26-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-538"><a href="#cb26-538" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-539"><a href="#cb26-539" aria-hidden="true" tabindex="-1"></a><span class="do">### Evaluate the cluster by identifying genes whose expression definies each</span></span>
<span id="cb26-540"><a href="#cb26-540" aria-hidden="true" tabindex="-1"></a><span class="do">### cluster which has been identified</span></span>
<span id="cb26-541"><a href="#cb26-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-542"><a href="#cb26-542" aria-hidden="true" tabindex="-1"></a><span class="do">### Find genes appears to be upregulated in a specific cluster compared to all</span></span>
<span id="cb26-543"><a href="#cb26-543" aria-hidden="true" tabindex="-1"></a><span class="do">### cells not in that cluster</span></span>
<span id="cb26-544"><a href="#cb26-544" aria-hidden="true" tabindex="-1"></a><span class="fu">FindMarkers</span>(data, <span class="at">ident.1 =</span> <span class="dv">0</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span>
<span id="cb26-545"><a href="#cb26-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-546"><a href="#cb26-546" aria-hidden="true" tabindex="-1"></a><span class="do">### Show the expression levels of these genes in the cells in each cluster</span></span>
<span id="cb26-547"><a href="#cb26-547" aria-hidden="true" tabindex="-1"></a><span class="do">### VCAN gene is more highly expressed in cluster 0 than any of the other </span></span>
<span id="cb26-548"><a href="#cb26-548" aria-hidden="true" tabindex="-1"></a><span class="do">### clusters, but we can also see that it is also reasonably highly expressed </span></span>
<span id="cb26-549"><a href="#cb26-549" aria-hidden="true" tabindex="-1"></a><span class="do">### in clusters 9 and 12. These were both clusters we suspected of being </span></span>
<span id="cb26-550"><a href="#cb26-550" aria-hidden="true" tabindex="-1"></a><span class="do">### multiple cells though</span></span>
<span id="cb26-551"><a href="#cb26-551" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"VCAN"</span>)</span>
<span id="cb26-552"><a href="#cb26-552" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-553"><a href="#cb26-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-556"><a href="#cb26-556" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-557"><a href="#cb26-557" aria-hidden="true" tabindex="-1"></a><span class="do">### FindMarkers function on all of the clusters</span></span>
<span id="cb26-558"><a href="#cb26-558" aria-hidden="true" tabindex="-1"></a>cluster_markers <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb26-559"><a href="#cb26-559" aria-hidden="true" tabindex="-1"></a>    <span class="fu">levels</span>(data[[<span class="st">"seurat_clusters"</span>]][[<span class="dv">1</span>]]),</span>
<span id="cb26-560"><a href="#cb26-560" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">FindMarkers</span>(data, <span class="at">ident.1 =</span> x, <span class="at">min.pct =</span> <span class="fl">0.25</span>)</span>
<span id="cb26-561"><a href="#cb26-561" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-562"><a href="#cb26-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-563"><a href="#cb26-563" aria-hidden="true" tabindex="-1"></a><span class="do">### Adds the cluster number to the results of FindMarkers</span></span>
<span id="cb26-564"><a href="#cb26-564" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(</span>
<span id="cb26-565"><a href="#cb26-565" aria-hidden="true" tabindex="-1"></a>    <span class="dv">0</span><span class="sc">:</span>(<span class="fu">length</span>(cluster_markers) <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb26-566"><a href="#cb26-566" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) {</span>
<span id="cb26-567"><a href="#cb26-567" aria-hidden="true" tabindex="-1"></a>        cluster_markers[[x <span class="sc">+</span> <span class="dv">1</span>]]<span class="sc">$</span>gene <span class="ot">&lt;&lt;-</span> <span class="fu">rownames</span>(cluster_markers[[x <span class="sc">+</span> <span class="dv">1</span>]])</span>
<span id="cb26-568"><a href="#cb26-568" aria-hidden="true" tabindex="-1"></a>        cluster_markers[[x <span class="sc">+</span> <span class="dv">1</span>]]<span class="sc">$</span>cluster <span class="ot">&lt;&lt;-</span> x</span>
<span id="cb26-569"><a href="#cb26-569" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb26-570"><a href="#cb26-570" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-571"><a href="#cb26-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-572"><a href="#cb26-572" aria-hidden="true" tabindex="-1"></a><span class="do">### Generate tibble and sort by FDR to put the most significant ones first</span></span>
<span id="cb26-573"><a href="#cb26-573" aria-hidden="true" tabindex="-1"></a>cluster_markers <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">do.call</span>(rbind, cluster_markers)) <span class="sc">|&gt;</span></span>
<span id="cb26-574"><a href="#cb26-574" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(p_val_adj)</span>
<span id="cb26-575"><a href="#cb26-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-576"><a href="#cb26-576" aria-hidden="true" tabindex="-1"></a>cluster_markers</span>
<span id="cb26-577"><a href="#cb26-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-578"><a href="#cb26-578" aria-hidden="true" tabindex="-1"></a><span class="do">### Extract the most upregulated gene from each cluster</span></span>
<span id="cb26-579"><a href="#cb26-579" aria-hidden="true" tabindex="-1"></a>best_wilcox_gene_per_cluster <span class="ot">&lt;-</span> cluster_markers <span class="sc">|&gt;</span></span>
<span id="cb26-580"><a href="#cb26-580" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">|&gt;</span></span>
<span id="cb26-581"><a href="#cb26-581" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb26-582"><a href="#cb26-582" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(gene)</span>
<span id="cb26-583"><a href="#cb26-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-584"><a href="#cb26-584" aria-hidden="true" tabindex="-1"></a><span class="do">### We can see that for some clusters (eg Cluster 8 - CDKN1C) We really do have</span></span>
<span id="cb26-585"><a href="#cb26-585" aria-hidden="true" tabindex="-1"></a><span class="do">### a gene which can uniquely predict, but for many others (eg cluster 7 IL7R)</span></span>
<span id="cb26-586"><a href="#cb26-586" aria-hidden="true" tabindex="-1"></a><span class="do">### we have a hit which also picks up other clusters (clusters 1, 3 and 4 in this case).</span></span>
<span id="cb26-587"><a href="#cb26-587" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> best_wilcox_gene_per_cluster)</span>
<span id="cb26-588"><a href="#cb26-588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-589"><a href="#cb26-589" aria-hidden="true" tabindex="-1"></a><span class="do">### Clean this up for any individual cluster by using the roc analysis.</span></span>
<span id="cb26-590"><a href="#cb26-590" aria-hidden="true" tabindex="-1"></a><span class="fu">FindMarkers</span>(</span>
<span id="cb26-591"><a href="#cb26-591" aria-hidden="true" tabindex="-1"></a>    data, <span class="at">ident.1 =</span> <span class="dv">7</span>, <span class="at">ident.2 =</span> <span class="dv">4</span>, <span class="at">test.use =</span> <span class="st">"roc"</span>, <span class="at">only.pos =</span> <span class="cn">TRUE</span></span>
<span id="cb26-592"><a href="#cb26-592" aria-hidden="true" tabindex="-1"></a>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span>
<span id="cb26-593"><a href="#cb26-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-594"><a href="#cb26-594" aria-hidden="true" tabindex="-1"></a><span class="do">### Slightly better job at separating cluster 5 from cluster 4, but it also</span></span>
<span id="cb26-595"><a href="#cb26-595" aria-hidden="true" tabindex="-1"></a><span class="do">### comes up all over the place in other clusters</span></span>
<span id="cb26-596"><a href="#cb26-596" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"LTB"</span>)</span>
<span id="cb26-597"><a href="#cb26-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-598"><a href="#cb26-598" aria-hidden="true" tabindex="-1"></a><span class="do">### This could actually be a better option to use as a marker for this cluster.</span></span>
<span id="cb26-599"><a href="#cb26-599" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(data, <span class="at">features =</span> <span class="st">"TPT1"</span>)</span>
<span id="cb26-600"><a href="#cb26-600" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-601"><a href="#cb26-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-602"><a href="#cb26-602" aria-hidden="true" tabindex="-1"></a><span class="fu">## Automated Cell Type Annotation</span></span>
<span id="cb26-605"><a href="#cb26-605" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-606"><a href="#cb26-606" aria-hidden="true" tabindex="-1"></a>scina_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">GetAssayData</span>(data, <span class="at">layer =</span> <span class="st">"data"</span>))</span>
<span id="cb26-607"><a href="#cb26-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-608"><a href="#cb26-608" aria-hidden="true" tabindex="-1"></a>signatures <span class="ot">&lt;-</span> <span class="fu">get</span>(</span>
<span id="cb26-609"><a href="#cb26-609" aria-hidden="true" tabindex="-1"></a>    <span class="fu">load</span>(<span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"example_signatures.RData"</span>, <span class="at">package =</span> <span class="st">"SCINA"</span>))</span>
<span id="cb26-610"><a href="#cb26-610" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-611"><a href="#cb26-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-612"><a href="#cb26-612" aria-hidden="true" tabindex="-1"></a>signatures</span>
<span id="cb26-613"><a href="#cb26-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-614"><a href="#cb26-614" aria-hidden="true" tabindex="-1"></a>scina_results <span class="ot">&lt;-</span> <span class="fu">SCINA</span>(</span>
<span id="cb26-615"><a href="#cb26-615" aria-hidden="true" tabindex="-1"></a>    scina_data,</span>
<span id="cb26-616"><a href="#cb26-616" aria-hidden="true" tabindex="-1"></a>    signatures,</span>
<span id="cb26-617"><a href="#cb26-617" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_iter =</span> <span class="dv">100</span>,</span>
<span id="cb26-618"><a href="#cb26-618" aria-hidden="true" tabindex="-1"></a>    <span class="at">convergence_n =</span> <span class="dv">10</span>,</span>
<span id="cb26-619"><a href="#cb26-619" aria-hidden="true" tabindex="-1"></a>    <span class="at">convergence_rate =</span> <span class="fl">0.999</span>,</span>
<span id="cb26-620"><a href="#cb26-620" aria-hidden="true" tabindex="-1"></a>    <span class="at">sensitivity_cutoff =</span> <span class="fl">0.9</span>,</span>
<span id="cb26-621"><a href="#cb26-621" aria-hidden="true" tabindex="-1"></a>    <span class="at">rm_overlap =</span> <span class="cn">TRUE</span>,</span>
<span id="cb26-622"><a href="#cb26-622" aria-hidden="true" tabindex="-1"></a>    <span class="at">allow_unknown =</span> <span class="cn">TRUE</span></span>
<span id="cb26-623"><a href="#cb26-623" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-624"><a href="#cb26-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-625"><a href="#cb26-625" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>scina_labels <span class="ot">&lt;-</span> scina_results<span class="sc">$</span>cell_labels</span>
<span id="cb26-626"><a href="#cb26-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-627"><a href="#cb26-627" aria-hidden="true" tabindex="-1"></a><span class="do">### plot out the tsne spread coloured by the automatic annotation</span></span>
<span id="cb26-628"><a href="#cb26-628" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(</span>
<span id="cb26-629"><a href="#cb26-629" aria-hidden="true" tabindex="-1"></a>    data, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">pt.size =</span> <span class="dv">1</span>, <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb26-630"><a href="#cb26-630" aria-hidden="true" tabindex="-1"></a>    <span class="at">group.by =</span> <span class="st">"scina_labels"</span>, <span class="at">label.size =</span> <span class="dv">5</span></span>
<span id="cb26-631"><a href="#cb26-631" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-632"><a href="#cb26-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-633"><a href="#cb26-633" aria-hidden="true" tabindex="-1"></a><span class="do">### Relate this to the clusters which we automatically detected.</span></span>
<span id="cb26-634"><a href="#cb26-634" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb26-635"><a href="#cb26-635" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster =</span> data<span class="sc">$</span>seurat_clusters,</span>
<span id="cb26-636"><a href="#cb26-636" aria-hidden="true" tabindex="-1"></a>    <span class="at">cell_type =</span> data<span class="sc">$</span>scina_labels</span>
<span id="cb26-637"><a href="#cb26-637" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb26-638"><a href="#cb26-638" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster, cell_type) <span class="sc">|&gt;</span></span>
<span id="cb26-639"><a href="#cb26-639" aria-hidden="true" tabindex="-1"></a>    <span class="fu">count</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-640"><a href="#cb26-640" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">|&gt;</span></span>
<span id="cb26-641"><a href="#cb26-641" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb26-642"><a href="#cb26-642" aria-hidden="true" tabindex="-1"></a>        <span class="at">percent =</span> (<span class="dv">100</span> <span class="sc">*</span> n) <span class="sc">/</span> <span class="fu">sum</span>(n)</span>
<span id="cb26-643"><a href="#cb26-643" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-644"><a href="#cb26-644" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="sc">|&gt;</span></span>
<span id="cb26-645"><a href="#cb26-645" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb26-646"><a href="#cb26-646" aria-hidden="true" tabindex="-1"></a>        <span class="at">cluster =</span> <span class="fu">paste</span>(<span class="st">"Cluster"</span>, cluster)</span>
<span id="cb26-647"><a href="#cb26-647" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb26-648"><a href="#cb26-648" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> percent, <span class="at">fill =</span> cell_type)) <span class="sc">+</span></span>
<span id="cb26-649"><a href="#cb26-649" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb26-650"><a href="#cb26-650" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb26-651"><a href="#cb26-651" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(cluster)) <span class="sc">+</span></span>
<span id="cb26-652"><a href="#cb26-652" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb26-653"><a href="#cb26-653" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb26-654"><a href="#cb26-654" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="cn">NULL</span>)</span>
<span id="cb26-655"><a href="#cb26-655" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-656"><a href="#cb26-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-659"><a href="#cb26-659" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-660"><a href="#cb26-660" aria-hidden="true" tabindex="-1"></a><span class="do">### Colouring by genes</span></span>
<span id="cb26-661"><a href="#cb26-661" aria-hidden="true" tabindex="-1"></a><span class="do">### Some of these genes very specifically isolate to their own cluster, but for</span></span>
<span id="cb26-662"><a href="#cb26-662" aria-hidden="true" tabindex="-1"></a><span class="do">### others we see expression which is more widely spread over a number of clusters</span></span>
<span id="cb26-663"><a href="#cb26-663" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(data, <span class="at">features =</span> best_wilcox_gene_per_cluster, <span class="at">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb26-664"><a href="#cb26-664" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-665"><a href="#cb26-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-668"><a href="#cb26-668" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-669"><a href="#cb26-669" aria-hidden="true" tabindex="-1"></a>data<span class="sc">@</span>reductions<span class="sc">$</span>tsne<span class="sc">@</span>cell.embeddings[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ]</span>
<span id="cb26-670"><a href="#cb26-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-671"><a href="#cb26-671" aria-hidden="true" tabindex="-1"></a><span class="co"># data@reductions$tsne@cell.embeddings[, ] |&gt;</span></span>
<span id="cb26-672"><a href="#cb26-672" aria-hidden="true" tabindex="-1"></a><span class="co">#     as_tibble(rownames = "barcode") |&gt;</span></span>
<span id="cb26-673"><a href="#cb26-673" aria-hidden="true" tabindex="-1"></a><span class="co">#     mutate(barcode = paste0(barcode, "-1")) |&gt;</span></span>
<span id="cb26-674"><a href="#cb26-674" aria-hidden="true" tabindex="-1"></a><span class="co">#     write_csv("for_loupe_import.csv")</span></span>
<span id="cb26-675"><a href="#cb26-675" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb26-676"><a href="#cb26-676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-677"><a href="#cb26-677" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reference</span></span>
<span id="cb26-678"><a href="#cb26-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-679"><a href="#cb26-679" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="co">[</span><span class="ot">Seurat Example</span><span class="co">](https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html#Gene_Selection)</span></span>
<span id="cb26-680"><a href="#cb26-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-681"><a href="#cb26-681" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="co">[</span><span class="ot">Single-cell RNA-seq data analysis workshop</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/schedule/links-to-lessons.html)</span></span>
<span id="cb26-682"><a href="#cb26-682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-683"><a href="#cb26-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-684"><a href="#cb26-684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-685"><a href="#cb26-685" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session info</span></span>
<span id="cb26-688"><a href="#cb26-688" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb26-689"><a href="#cb26-689" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb26-690"><a href="#cb26-690" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb26-691"><a href="#cb26-691" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2026 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a>, <a href="https://www.python.org/"><i class="fa-brands fa-python" aria-label="python"></i></a> and <a href="https://quarto.org/">Quarto</a></span></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="../../../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>