<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta charset="UTF-8">
<meta name="generator" content="quarto-1.8.26">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Guorui Zhong">
<title>Multiplex image-based phenotypic data analysis â€“ Guorui Zhong</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting-1b3db88def35042d172274863c1cdcf0.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap-73f67446a8635c361532e94f67b7efbe.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
</head>
<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html"> 
<span class="menu-text">Wiki</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html"> 
<span class="menu-text">VSCode</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../linux/index.html"> 
<span class="menu-text">Linux</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../python/index.html"> 
<span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../r/index.html"> 
<span class="menu-text">R</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Multiplex image-based phenotypic data analysis</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
                                <div class="quarto-categories">
                <div class="quarto-category">mIF</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://grgrzhong.github.io/">Guorui Zhong</a> <a href="https://orcid.org/0000-0002-1955-2847" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Thursday, April 18, 2024</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Thursday, May 9, 2024</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#initial-general-setup" id="toc-initial-general-setup" class="nav-link active" data-scroll-target="#initial-general-setup">Initial general setup</a></li>
  <li>
<a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation">Data preparation</a>
  <ul class="collapse">
<li><a href="#download-example-data" id="toc-download-example-data" class="nav-link" data-scroll-target="#download-example-data">Download example data</a></li>
  <li><a href="#preprocess-data" id="toc-preprocess-data" class="nav-link" data-scroll-target="#preprocess-data">Preprocess data</a></li>
  <li><a href="#read-in-images" id="toc-read-in-images" class="nav-link" data-scroll-target="#read-in-images">Read in images</a></li>
  <li><a href="#save-data" id="toc-save-data" class="nav-link" data-scroll-target="#save-data">Save data</a></li>
  </ul>
</li>
  <li>
<a href="#spillover-correction" id="toc-spillover-correction" class="nav-link" data-scroll-target="#spillover-correction">Spillover correction</a>
  <ul class="collapse">
<li><a href="#generate-the-spillover-matrix" id="toc-generate-the-spillover-matrix" class="nav-link" data-scroll-target="#generate-the-spillover-matrix">Generate the spillover matrix</a></li>
  <li><a href="#single-cell-data-compensation" id="toc-single-cell-data-compensation" class="nav-link" data-scroll-target="#single-cell-data-compensation">Single-cell data compensation</a></li>
  <li><a href="#image-compensation" id="toc-image-compensation" class="nav-link" data-scroll-target="#image-compensation">Image compensation</a></li>
  <li><a href="#write-out-compensated-images" id="toc-write-out-compensated-images" class="nav-link" data-scroll-target="#write-out-compensated-images">Write out compensated images</a></li>
  </ul>
</li>
  <li>
<a href="#image-and-cell-quality-control" id="toc-image-and-cell-quality-control" class="nav-link" data-scroll-target="#image-and-cell-quality-control">Image and cell quality control</a>
  <ul class="collapse">
<li><a href="#segmentation-quality-control" id="toc-segmentation-quality-control" class="nav-link" data-scroll-target="#segmentation-quality-control">Segmentation quality control</a></li>
  <li><a href="#image_level-quality-control" id="toc-image_level-quality-control" class="nav-link" data-scroll-target="#image_level-quality-control">Image_level quality control</a></li>
  <li><a href="#cell_level-quality-control" id="toc-cell_level-quality-control" class="nav-link" data-scroll-target="#cell_level-quality-control">Cell_level quality control</a></li>
  </ul>
</li>
  <li>
<a href="#batch-effect-correction" id="toc-batch-effect-correction" class="nav-link" data-scroll-target="#batch-effect-correction">Batch effect correction</a>
  <ul class="collapse">
<li><a href="#fastmnn-correction" id="toc-fastmnn-correction" class="nav-link" data-scroll-target="#fastmnn-correction">fastMNN correction</a></li>
  <li><a href="#harmony-correction" id="toc-harmony-correction" class="nav-link" data-scroll-target="#harmony-correction">harmony correction</a></li>
  <li><a href="#seurat-correction" id="toc-seurat-correction" class="nav-link" data-scroll-target="#seurat-correction">Seurat correction</a></li>
  </ul>
</li>
  <li>
<a href="#cell-phenotyping" id="toc-cell-phenotyping" class="nav-link" data-scroll-target="#cell-phenotyping">Cell phenotyping</a>
  <ul class="collapse">
<li><a href="#rphenograph" id="toc-rphenograph" class="nav-link" data-scroll-target="#rphenograph">Rphenograph</a></li>
  <li><a href="#shared-nearest-neighbour-graph" id="toc-shared-nearest-neighbour-graph" class="nav-link" data-scroll-target="#shared-nearest-neighbour-graph">Shared nearest neighbour graph</a></li>
  <li><a href="#self-organizing-maps" id="toc-self-organizing-maps" class="nav-link" data-scroll-target="#self-organizing-maps">Self organizing maps</a></li>
  <li><a href="#compare-between-clustering-approaches" id="toc-compare-between-clustering-approaches" class="nav-link" data-scroll-target="#compare-between-clustering-approaches">Compare between clustering approaches</a></li>
  <li><a href="#further-clustering-notes" id="toc-further-clustering-notes" class="nav-link" data-scroll-target="#further-clustering-notes">Further clustering notes</a></li>
  <li><a href="#classfication-approach" id="toc-classfication-approach" class="nav-link" data-scroll-target="#classfication-approach">Classfication approach</a></li>
  </ul>
</li>
  <li>
<a href="#single-cell-visualization" id="toc-single-cell-visualization" class="nav-link" data-scroll-target="#single-cell-visualization">Single cell visualization</a>
  <ul class="collapse">
<li><a href="#inital-setup" id="toc-inital-setup" class="nav-link" data-scroll-target="#inital-setup">Inital setup</a></li>
  <li><a href="#cell-type-level" id="toc-cell-type-level" class="nav-link" data-scroll-target="#cell-type-level">Cell-type level</a></li>
  <li><a href="#sample-level" id="toc-sample-level" class="nav-link" data-scroll-target="#sample-level">Sample level</a></li>
  <li><a href="#complexheatmap" id="toc-complexheatmap" class="nav-link" data-scroll-target="#complexheatmap">ComplexHeatmap</a></li>
  </ul>
</li>
  <li><a href="#reference" id="toc-reference" class="nav-link" data-scroll-target="#reference">Reference</a></li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo">Sessioninfo</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="initial-general-setup" class="level2"><h2 class="anchored" data-anchor-id="initial-general-setup">Initial general setup</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Required libraries</span></span>
<span><span class="co"># if (!require("BiocManager", quietly = TRUE))</span></span>
<span><span class="co">#     install.packages("BiocManager")</span></span>
<span><span class="co"># BiocManager::install(</span></span>
<span><span class="co">#     c(</span></span>
<span><span class="co">#         "rmarkdown", "bookdown", "pheatmap", "viridis", "zoo",</span></span>
<span><span class="co">#         "devtools", "testthat", "tiff", "distill", "ggrepel",</span></span>
<span><span class="co">#         "patchwork", "mclust", "RColorBrewer", "uwot", "Rtsne",</span></span>
<span><span class="co">#         "harmony", "Seurat", "SeuratObject", "cowplot", "kohonen",</span></span>
<span><span class="co">#         "caret", "randomForest", "ggridges", "cowplot",</span></span>
<span><span class="co">#         "gridGraphics", "scales", "tiff", "harmony", "Matrix",</span></span>
<span><span class="co">#         "CATALYST", "scuttle", "scater", "dittoSeq",</span></span>
<span><span class="co">#         "tidyverse", "BiocStyle", "batchelor", "bluster", "scran",</span></span>
<span><span class="co">#         "lisaClust", "spicyR", "iSEE", "imcRtools", "cytomapper",</span></span>
<span><span class="co">#         "imcdatasets", "cytoviewer"</span></span>
<span><span class="co">#     ),</span></span>
<span><span class="co">#     force = TRUE</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs">qs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/">ggridges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">imcRtools</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">cytomapper</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CATALYST</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BiocParallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.rforge.net/tiff/">tiff</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EBImage</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://mclust-org.github.io/mclust/">mclust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/immunogenomics/harmony">harmony</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BiocSingular</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.github.io/seurat-object/">SeuratObject</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Rphenograph</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/">igraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/">caret</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">lisaClust</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Project directory</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="st">"projects/2024_IMC_Profile"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">20240419</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="data-preparation" class="level2"><h2 class="anchored" data-anchor-id="data-preparation">Data preparation</h2>
<p>IMC example data is from <a href="https://zenodo.org/records/7624451">here</a></p>
<section id="download-example-data" class="level3"><h3 class="anchored" data-anchor-id="download-example-data">Download example data</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Create directory for data </span></span>
<span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/create.html">dir_create</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download sample/patient metadata information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/7575859/files/sample_metadata.csv"</span>,</span>
<span>    destfile <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/sample_metadata.csv"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download intensities</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/intensities.zip"</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/intensities.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">destfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">destfile</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">destfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download regionprops</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/regionprops.zip"</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/regionprops.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">destfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">destfile</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">destfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download neighbors</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/neighbors.zip"</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/neighbors.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">destfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">destfile</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">destfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download images</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/img.zip"</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/img.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">destfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">destfile</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">destfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download masks</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/masks_deepcell.zip"</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/masks_deepcell.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">destfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">destfile</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">destfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Download individual files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/7624451/files/panel.csv"</span>,</span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/panel.csv"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/7624451/files/images.csv"</span>,</span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/images.csv"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/7624451/files/steinbock.sh"</span>,</span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/steinbock.sh"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Files for spillover matrix estimation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/7575859/files/compensation.zip"</span>,</span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/compensation.zip"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/compensation.zip"</span><span class="op">)</span>,</span>
<span>    exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/compensation.zip"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Gated cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://zenodo.org/record/8095133/files/gated_cells.zip"</span>,</span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/gated_cells.zip"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/gated_cells.zip"</span><span class="op">)</span>,</span>
<span>    exdir <span class="op">=</span> <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/gated_cells.zip"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="preprocess-data" class="level3"><h3 class="anchored" data-anchor-id="preprocess-data">Preprocess data</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Read steinbock generated data into a SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">imcRtools</span><span class="fu">::</span><span class="fu">read_steinbock</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spe</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SpatialExperiment 
dim: 40 47859 
metadata(0):
assays(1): counts
rownames(40): MPO HistoneH3 ... DNA1 DNA2
rowData names(12): channel name ... Final.Concentration...Dilution
  uL.to.add
colnames: NULL
colData names(8): sample_id ObjectNumber ... width_px height_px
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : Pos_X Pos_Y
imgData names(1): sample_id</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Summarized pixel intensities</span></span>
<span><span class="fu">counts</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>               [,1]       [,2]      [,3]      [,4]      [,5]
MPO       0.5751064  0.4166667 0.4975494  0.890154 0.1818182
HistoneH3 3.1273082 11.3597883 2.3841440  7.712961 1.4512715
SMA       0.2600939  1.6720383 0.1535190  1.193948 0.2986703
CD16      2.0347747  2.5880536 2.2943074 15.629083 0.6084220
CD38      0.2530137  0.6826669 1.1902979  2.126060 0.2917793</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 8 columns
     sample_id ObjectNumber      area axis_major_length axis_minor_length
   &lt;character&gt;    &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt;         &lt;numeric&gt;
1 Patient1_001            1        12           7.40623           1.89529
2 Patient1_001            2        24          16.48004           1.96284
3 Patient1_001            3        17           9.85085           1.98582
4 Patient1_001            4        24           8.08290           3.91578
5 Patient1_001            5        22           8.79367           3.11653
6 Patient1_001            6        25           9.17436           3.46929
  eccentricity  width_px height_px
     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
1     0.966702       600       600
2     0.992882       600       600
3     0.979470       600       600
4     0.874818       600       600
5     0.935091       600       600
6     0.925744       600       600</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### SpatialExperiment container stores locations of all cells in the</span></span>
<span><span class="co">### spatialCoords slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">spatialCoords</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Pos_X     Pos_Y
[1,] 468.5833 0.4166667
[2,] 515.8333 0.4166667
[3,] 587.2353 0.4705882
[4,] 192.2500 1.2500000
[5,] 231.7727 0.9090909
[6,] 270.1600 1.0400000</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">colPair</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"neighborhood"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>SelfHits object with 257116 hits and 0 metadata columns:
                from        to
           &lt;integer&gt; &lt;integer&gt;
       [1]         1        27
       [2]         1        55
       [3]         2        10
       [4]         2        44
       [5]         2        81
       ...       ...       ...
  [257112]     47858     47836
  [257113]     47859     47792
  [257114]     47859     47819
  [257115]     47859     47828
  [257116]     47859     47854
  -------
  nnode: 47859</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 12 columns
              channel        name      keep   ilastik  deepcell  cellpose
          &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;
MPO               Y89         MPO         1        NA        NA        NA
HistoneH3       In113   HistoneH3         1         1         1        NA
SMA             In115         SMA         1        NA        NA        NA
CD16            Pr141        CD16         1        NA        NA        NA
CD38            Nd142        CD38         1        NA        NA        NA
HLADR           Nd143       HLADR         1        NA        NA        NA
          Tube.Number              Target Antibody.Clone Stock.Concentration
            &lt;numeric&gt;         &lt;character&gt;    &lt;character&gt;           &lt;numeric&gt;
MPO              2101 Myeloperoxidase MPO Polyclonal MPO                 500
HistoneH3        2113          Histone H3           D1H2                 500
SMA              1914                 SMA            1A4                 500
CD16             2079                CD16       EPR16784                 500
CD38             2095                CD38        EPR4106                 500
HLADR            2087              HLA-DR        TAL 1B5                 500
          Final.Concentration...Dilution   uL.to.add
                             &lt;character&gt; &lt;character&gt;
MPO                              4 ug/mL         0.8
HistoneH3                        1 ug/mL         0.2
SMA                           0.25 ug/mL        0.05
CD16                             5 ug/mL           1
CD38                           2.5 ug/mL         0.5
HLADR                            1 ug/mL         0.2</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Add additional metadata: generate unique identifiers per cell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"_"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">ObjectNumber</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read patient metadata</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/sample_metadata.csv"</span><span class="op">)</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Extract patient id and ROI id from sample name</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html">str_extract</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"Patient[1-4]"</span><span class="op">)</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">ROI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html">str_extract</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"00[1-8]"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Store cancer type in SPE object</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">indication</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">Indication</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span>, <span class="va">meta</span><span class="op">$</span><span class="va">`Sample ID`</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Patient1" "Patient2" "Patient3" "Patient4"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">ROI</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "001" "002" "003" "004" "005" "006" "007" "008"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SCCHN" "BCC"   "NSCLC" "CRC"  </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Transform counts</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoRidgePlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"CD3"</span>, group.by <span class="op">=</span> <span class="st">"patient_id"</span>, assay <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"CD3 - before transformation"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Perform counts transformation using an inverse hyperbolic sine function</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html">asinh</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoRidgePlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"CD3"</span>, group.by <span class="op">=</span> <span class="st">"patient_id"</span>, assay <span class="op">=</span> <span class="st">"exprs"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"CD3 - after transformation"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">wrap_plots</span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">plot_layout</span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.22</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0984</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Add Feature meta for easy specifies the markers of interest.</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"DNA|Histone"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Define color schemes for different metadata entries of the data</span></span>
<span><span class="va">color_vectors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ROI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">ROI</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"BrBG"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">ROI</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">patient_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"YlOrRd"</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"PuBu"</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"YlGn"</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"BuPu"</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">indication</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">color_vectors</span><span class="op">$</span><span class="va">ROI</span> <span class="op">&lt;-</span> <span class="va">ROI</span></span>
<span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span> <span class="op">&lt;-</span> <span class="va">patient_id</span></span>
<span><span class="va">color_vectors</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="va">sample_id</span></span>
<span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span> <span class="op">&lt;-</span> <span class="va">indication</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span> <span class="op">&lt;-</span> <span class="va">color_vectors</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="read-in-images" class="level3"><h3 class="anchored" data-anchor-id="read-in-images">Read in images</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">loadImages</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/img/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>All files in the provided location will be read in.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu">loadImages</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/steinbock/masks_deepcell/"</span><span class="op">)</span>, as.is <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>All files in the provided location will be read in.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Make sure that the channel order is identical between the single-cell data and the images</span></span>
<span><span class="fu">channelNames</span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span><span class="va">images</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>CytoImageList containing 14 image(s)
names(14): Patient1_001 Patient1_002 Patient1_003 Patient2_001 Patient2_002 Patient2_003 Patient2_004 Patient3_001 Patient3_002 Patient3_003 Patient4_005 Patient4_006 Patient4_007 Patient4_008 
Each image contains 40 channel(s)
channelNames(40): MPO HistoneH3 SMA CD16 CD38 HLADR CD27 CD15 CD45RA CD163 B2M CD20 CD68 Ido1 CD3 LAG3 / LAG33 CD11c PD1 PDGFRb CD7 GrzB PDL1 TCF7 CD45RO FOXP3 ICOS CD8a CarbonicAnhydrase CD33 Ki67 VISTA CD40 CD4 CD14 Ecad CD303 CD206 cleavedPARP DNA1 DNA2 </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Order of the images</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">masks</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Extract patient id from image name</span></span>
<span><span class="va">patient_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html">str_extract</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="st">"Patient[1-4]"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Retrieve cancer type per patient from metadata file</span></span>
<span><span class="va">indication</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">Indication</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">patient_id</span>, <span class="va">meta</span><span class="op">$</span><span class="va">`Sample ID`</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">### Store patient and image level information in elementMetadata</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">masks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>,</span>
<span>    patient_id <span class="op">=</span> <span class="va">patient_id</span>,</span>
<span>    indication <span class="op">=</span> <span class="va">indication</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="save-data" class="level3"><h3 class="anchored" data-anchor-id="save-data">Save data</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">images</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/images.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">masks</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/masks.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="spillover-correction" class="level2"><h2 class="anchored" data-anchor-id="spillover-correction">Spillover correction</h2>
<section id="generate-the-spillover-matrix" class="level3"><h3 class="anchored" data-anchor-id="generate-the-spillover-matrix">Generate the spillover matrix</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Create SingleCellExperiment from TXT files</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">readSCEfromTXT</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/compensation/"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Spotted channels:  Y89, In113, In115, Pr141, Nd142, Nd143, Nd144, Nd145, Nd146, Sm147, Nd148, Sm149, Nd150, Eu151, Sm152, Eu153, Sm154, Gd155, Gd156, Gd158, Tb159, Gd160, Dy161, Dy162, Dy163, Dy164, Ho165, Er166, Er167, Er168, Tm169, Er170, Yb171, Yb172, Yb173, Yb174, Lu175, Yb176
Acquired channels:  Ar80, Y89, In113, In115, Xe131, Xe134, Ba136, La138, Pr141, Nd142, Nd143, Nd144, Nd145, Nd146, Sm147, Nd148, Sm149, Nd150, Eu151, Sm152, Eu153, Sm154, Gd155, Gd156, Gd158, Tb159, Gd160, Dy161, Dy162, Dy163, Dy164, Ho165, Er166, Er167, Er168, Tm169, Er170, Yb171, Yb172, Yb173, Yb174, Lu175, Yb176, Ir191, Ir193, Pt196, Pb206
Channels spotted but not acquired:  
Channels acquired but not spotted:  Ar80, Xe131, Xe134, Ba136, La138, Ir191, Ir193, Pt196, Pb206</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html">asinh</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Quality control</span></span>
<span><span class="co">### Log10 median pixel counts per spot and channel</span></span>
<span><span class="fu">plotSpotHeatmap</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Thresholded on 200 pixel counts</span></span>
<span><span class="fu">plotSpotHeatmap</span><span class="op">(</span><span class="va">sce</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Optional pixel binning</span></span>
<span><span class="co">### Define grouping</span></span>
<span><span class="va">bin_size</span> <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu">binAcrossPixels</span><span class="op">(</span><span class="va">sce</span>, bin_size <span class="op">=</span> <span class="va">bin_size</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Log10 median pixel counts per spot and channel</span></span>
<span><span class="fu">plotSpotHeatmap</span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Thresholded on 200 pixel counts</span></span>
<span><span class="fu">plotSpotHeatmap</span><span class="op">(</span><span class="va">sce2</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Filtering incorrectly assigned pixels</span></span>
<span><span class="va">bc_key</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bc_key</span> <span class="op">&lt;-</span> <span class="va">bc_key</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">bc_key</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">assignPrelim</span><span class="op">(</span><span class="va">sce</span>, bc_key <span class="op">=</span> <span class="va">bc_key</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Debarcoding data...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> o ordering</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> o classifying events</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing deltas...</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">estCutoffs</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">applyCutoffs</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Visualize the correctly and incorrectly assigned pixels</span></span>
<span><span class="va">cur_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">bc_id</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">cur_table</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Compute the fraction of unassigned pixels per spot</span></span>
<span><span class="va">cur_table</span><span class="op">[</span><span class="st">"0"</span>, <span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">cur_table</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   113    115    141    142    143    144    145    146    147    148    149 
0.1985 0.1060 0.2575 0.3195 0.3190 0.3825 0.3545 0.4280 0.3570 0.4770 0.4200 
   150    151    152    153    154    155    156    158    159    160    161 
0.4120 0.4025 0.4050 0.4630 0.4190 0.4610 0.3525 0.4020 0.4655 0.4250 0.5595 
   162    163    164    165    166    167    168    169    170    171    172 
0.4340 0.4230 0.4390 0.4055 0.5210 0.3900 0.3285 0.3680 0.5015 0.4900 0.5650 
   173    174    175    176     89 
0.3125 0.4605 0.4710 0.2845 0.3015 </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">filterPixels</span><span class="op">(</span><span class="va">sce</span>, minevents <span class="op">=</span> <span class="fl">40</span>, correct_pixels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Compute spillover matrix</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">computeSpillmat</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">isotope_list</span> <span class="op">&lt;-</span> <span class="fu">CATALYST</span><span class="fu">::</span><span class="va">isotope_list</span></span>
<span><span class="va">isotope_list</span><span class="op">$</span><span class="va">Ar</span> <span class="op">&lt;-</span> <span class="fl">80</span></span>
<span><span class="fu">plotSpillmat</span><span class="op">(</span><span class="va">sce</span>, isotope_list <span class="op">=</span> <span class="va">isotope_list</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `guide` argument in `scale_*()` cannot be `FALSE`. This was deprecated in
ggplot2 3.3.4.
â„¹ Please use "none" instead.
â„¹ The deprecated feature was likely used in the CATALYST package.
  Please report the issue at &lt;https://github.com/HelenaLC/CATALYST/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Save spillover matrix in variable</span></span>
<span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">spillover_matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">sm</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/sm.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="single-cell-data-compensation" class="level3"><h3 class="anchored" data-anchor-id="single-cell-data-compensation">Single-cell data compensation</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">channel_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">channel</span>, <span class="st">"Di"</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">compCytof</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, <span class="va">sm</span>,</span>
<span>    transform <span class="op">=</span> <span class="cn">TRUE</span>, cofactor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    isotope_list <span class="op">=</span> <span class="va">isotope_list</span>,</span>
<span>    overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Check the effect of channel spillover compensation</span></span>
<span><span class="va">before</span> <span class="op">&lt;-</span> <span class="fu">dittoScatterPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, x.var <span class="op">=</span> <span class="st">"Ecad"</span>, y.var <span class="op">=</span> <span class="st">"CD303"</span>,</span>
<span>    assay.x <span class="op">=</span> <span class="st">"exprs"</span>, assay.y <span class="op">=</span> <span class="st">"exprs"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Before compensation"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">after</span> <span class="op">&lt;-</span> <span class="fu">dittoScatterPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, x.var <span class="op">=</span> <span class="st">"Ecad"</span>, y.var <span class="op">=</span> <span class="st">"CD303"</span>,</span>
<span>    assay.x <span class="op">=</span> <span class="st">"compexprs"</span>, assay.y <span class="op">=</span> <span class="st">"compexprs"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"After compensation"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">wrap_plots</span><span class="op">(</span><span class="va">before</span>, <span class="va">after</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"compcounts"</span><span class="op">)</span> </span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"compexprs"</span><span class="op">)</span> </span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"compcounts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"compexprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="image-compensation" class="level3"><h3 class="anchored" data-anchor-id="image-compensation">Image compensation</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/images.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">channelNames</span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">channel_name</span></span>
<span></span>
<span><span class="va">adapted_sm</span> <span class="op">&lt;-</span> <span class="fu">adaptSpillmat</span><span class="op">(</span></span>
<span>    <span class="va">sm</span>, <span class="fu">channelNames</span><span class="op">(</span><span class="va">images</span><span class="op">)</span>,</span>
<span>    isotope_list <span class="op">=</span> <span class="va">isotope_list</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">images_comp</span> <span class="op">&lt;-</span> <span class="fu">compImage</span><span class="op">(</span></span>
<span>    <span class="va">images</span>, <span class="va">adapted_sm</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Visualize the image before and after compensation</span></span>
<span><span class="co"># Before compensation</span></span>
<span><span class="fu">plotPixels</span><span class="op">(</span></span>
<span>    <span class="va">images</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, colour_by <span class="op">=</span> <span class="st">"Yb173Di"</span>,</span>
<span>    image_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"Yb173 (Ecad) - before"</span>, position <span class="op">=</span> <span class="st">"topleft"</span><span class="op">)</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">NULL</span>, bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Yb173Di <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">plotPixels</span><span class="op">(</span></span>
<span>    <span class="va">images</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, colour_by <span class="op">=</span> <span class="st">"Yb174Di"</span>,</span>
<span>    image_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"Yb174 (CD303) - before"</span>, position <span class="op">=</span> <span class="st">"topleft"</span><span class="op">)</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">NULL</span>, bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Yb174Di <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># After compensation</span></span>
<span><span class="fu">plotPixels</span><span class="op">(</span></span>
<span>    <span class="va">images_comp</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, colour_by <span class="op">=</span> <span class="st">"Yb173Di"</span>,</span>
<span>    image_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"Yb173 (Ecad) - after"</span>, position <span class="op">=</span> <span class="st">"topleft"</span><span class="op">)</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">NULL</span>, bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Yb173Di <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">plotPixels</span><span class="op">(</span></span>
<span>    <span class="va">images_comp</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, colour_by <span class="op">=</span> <span class="st">"Yb174Di"</span>,</span>
<span>    image_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"Yb174 (CD303) - after"</span>, position <span class="op">=</span> <span class="st">"topleft"</span><span class="op">)</span>,</span>
<span>    legend <span class="op">=</span> <span class="cn">NULL</span>, bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Yb174Di <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Re-set the channelNames to their biological targtes</span></span>
<span><span class="fu">channelNames</span><span class="op">(</span><span class="va">images_comp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="write-out-compensated-images" class="level3"><h3 class="anchored" data-anchor-id="write-out-compensated-images">Write out compensated images</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/create.html">dir_create</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/comp_img"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images_comp</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu">writeImage</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/array.html">as.array</a></span><span class="op">(</span><span class="va">images_comp</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="fl">16</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"/data/comp_img/"</span>, <span class="va">x</span>, <span class="st">".tiff"</span><span class="op">)</span>,</span>
<span>            bits.per.sample <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Save the compensated SpatialExperiment and CytoImageList objects</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">images_comp</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>,<span class="st">"data/images.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="image-and-cell-quality-control" class="level2"><h2 class="anchored" data-anchor-id="image-and-cell-quality-control">Image and cell quality control</h2>
<section id="segmentation-quality-control" class="level3"><h3 class="anchored" data-anchor-id="segmentation-quality-control">Segmentation quality control</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Load data: 63.01 MB</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/images.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/masks.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">lobstr</span><span class="fu">::</span><span class="fu"><a href="https://lobstr.r-lib.org/reference/obj_size.html">obj_size</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">image</span>, <span class="va">masks</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Select 3 random images</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">20220118</span><span class="op">)</span></span>
<span><span class="va">img_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Image- and channel-wise min-max normalization and scaled to a range of 0-1</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="va">images</span><span class="op">[</span><span class="va">img_ids</span><span class="op">]</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu">normalize</span><span class="op">(</span><span class="va">cur_images</span>, separateImages <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Clipping the maximum intensity to 0.2</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu">normalize</span><span class="op">(</span><span class="va">cur_images</span>, inputRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Segmentation approach here appears to correctly segment cells</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu">plotPixels</span><span class="op">(</span></span>
<span>    <span class="va">cur_images</span>,</span>
<span>    mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="va">img_ids</span><span class="op">]</span>,</span>
<span>    img_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>    missing_colour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>    colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"Ecad"</span>, <span class="st">"DNA1"</span><span class="op">)</span>,</span>
<span>    colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        CD163 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>        CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>        CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>        Ecad <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"cyan"</span><span class="op">)</span>,</span>
<span>        DNA1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    image_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        colour_by.title.cex <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>        colour_by.labels.cex <span class="op">=</span> <span class="fl">0.7</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Heatmap to observe cell segmentation quality and</span></span>
<span><span class="co">### potentially also antibody specificity issues</span></span>
<span></span>
<span><span class="co">### Sub-sample the dataset to 2000 cells</span></span>
<span><span class="va">cur_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span>, <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Epithelial cells (Ecad+) and immune cells (CD45RO+) can be differentiate</span></span>
<span><span class="co">### Some of the markers are detected in specific cells (e.g., Ki67, CD20, Ecad) ### while others are more broadly expressed across cells (e.g., HLADR, B2M, CD4).</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>, <span class="va">cur_cells</span><span class="op">]</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="st">"indication"</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="image_level-quality-control" class="level3"><h3 class="anchored" data-anchor-id="image_level-quality-control">Image_level quality control</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Average SNR versus the average signal intensity across all images</span></span>
<span><span class="va">cur_snr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">img</span> <span class="op">&lt;-</span> <span class="va">images</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">img</span>, <span class="fl">3</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ch</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="co"># Otsu threshold</span></span>
<span>        <span class="va">thres</span> <span class="op">&lt;-</span> <span class="fu">otsu</span><span class="op">(</span><span class="va">ch</span>, range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">ch</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">ch</span><span class="op">)</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fl">65536</span><span class="op">)</span></span>
<span>        <span class="co"># Signal-to-noise ratio</span></span>
<span>        <span class="va">snr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ch</span><span class="op">[</span><span class="va">ch</span> <span class="op">&gt;</span> <span class="va">thres</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ch</span><span class="op">[</span><span class="va">ch</span> <span class="op">&lt;=</span> <span class="va">thres</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="co"># Signal intensity</span></span>
<span>        <span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ch</span><span class="op">[</span><span class="va">ch</span> <span class="op">&gt;</span> <span class="va">thres</span><span class="op">]</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>snr <span class="op">=</span> <span class="va">snr</span>, ps <span class="op">=</span> <span class="va">ps</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span>  </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">x</span>,</span>
<span>               marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span>  </span>
<span>        <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">snr</span>, <span class="va">ps</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_snr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">cur_snr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_snr</span> <span class="op">|&gt;</span>  </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">marker</span>, <span class="va">name</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>log_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">name</span>, values_from <span class="op">=</span> <span class="va">log_mean</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">ps</span>, <span class="va">snr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">ps</span>, <span class="va">snr</span>, label <span class="op">=</span> <span class="va">marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Signal-to-noise ratio [log2]"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Signal intensity [log2]"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'marker'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 9 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Remove markers that have a positive signal of below 2 per image</span></span>
<span><span class="va">cur_snr</span> <span class="op">&lt;-</span> <span class="va">cur_snr</span> <span class="op">|&gt;</span>  </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">name</span>, values_from <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">ps</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">snr</span>, <span class="va">ps</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_snr</span> <span class="op">|&gt;</span>  </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">marker</span>, <span class="va">name</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>log_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">name</span>, values_from <span class="op">=</span> <span class="va">log_mean</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">ps</span>, <span class="va">snr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">ps</span>, <span class="va">snr</span>, label <span class="op">=</span> <span class="va">marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Signal-to-noise ratio [log2]"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Signal intensity [log2]"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'marker'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 7 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Compute the percentage of covered image area</span></span>
<span><span class="va">cell_density</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="co"># Compute the number of pixels covered by cells and </span></span>
<span>    <span class="co"># the total number of pixels</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>cell_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">area</span><span class="op">)</span>,</span>
<span>              no_pixels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">width_px</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">height_px</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="co"># Divide the total number of pixels </span></span>
<span>    <span class="co"># by the number of pixels covered by cells</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>covered_area <span class="op">=</span> <span class="va">cell_area</span> <span class="op">/</span> <span class="va">no_pixels</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Visualize the image area covered by cells per image</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cell_density</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">sample_id</span>,<span class="va">covered_area</span><span class="op">)</span>, <span class="va">covered_area</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"% covered area"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Normalize and clip images</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="va">images</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Patient4_005"</span>, <span class="st">"Patient4_007"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu">normalize</span><span class="op">(</span><span class="va">cur_images</span>, separateImages <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cur_images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu">normalize</span><span class="op">(</span><span class="va">cur_images</span>, inputRange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotPixels</span><span class="op">(</span><span class="va">cur_images</span>,</span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Patient4_005"</span>, <span class="st">"Patient4_007"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>           missing_colour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"Ecad"</span>, <span class="st">"DNA1"</span><span class="op">)</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>CD163 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span>,</span>
<span>                         CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>                         CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>                         Ecad <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"cyan"</span><span class="op">)</span>,</span>
<span>                         DNA1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>colour_by.title.cex <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                         colour_by.labels.cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">###  Visualize the mean marker expression per image to identify images with outlying marker expression</span></span>
<span><span class="va">image_mean</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    ids <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span>, </span>
<span>    statistics<span class="op">=</span><span class="st">"mean"</span>,</span>
<span>    use.assay.type <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">image_mean</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html">asinh</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">image_mean</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">image_mean</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"indication"</span>, <span class="st">"patient_id"</span>, <span class="st">"ROI"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span>,</span>
<span>                                      patient_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>                                      ROI <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">ROI</span></span>
<span>        <span class="op">)</span>,</span>
<span>    show_colnames <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span>                        </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="cell_level-quality-control" class="level3"><h3 class="anchored" data-anchor-id="cell_level-quality-control">Cell_level quality control</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220224</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">cur_exprs</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="va">x</span>,<span class="op">]</span></span>
<span>    <span class="va">cur_counts</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="va">x</span>,<span class="op">]</span></span>
<span>    </span>
<span>    <span class="va">cur_model</span> <span class="op">&lt;-</span> <span class="fu">Mclust</span><span class="op">(</span><span class="va">cur_exprs</span>, G <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="va">mean1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cur_counts</span><span class="op">[</span><span class="va">cur_model</span><span class="op">$</span><span class="va">classification</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">mean2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cur_counts</span><span class="op">[</span><span class="va">cur_model</span><span class="op">$</span><span class="va">classification</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">signal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">mean1</span> <span class="op">&gt;</span> <span class="va">mean2</span>, <span class="va">mean1</span>, <span class="va">mean2</span><span class="op">)</span></span>
<span>    <span class="va">noise</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">mean1</span> <span class="op">&gt;</span> <span class="va">mean2</span>, <span class="va">mean2</span>, <span class="va">mean1</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>snr <span class="op">=</span> <span class="va">signal</span><span class="op">/</span><span class="va">noise</span>, ps <span class="op">=</span> <span class="va">signal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">cur_snr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span>  </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_snr</span> <span class="op">|&gt;</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">snr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_label_repel</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">snr</span><span class="op">)</span>, label <span class="op">=</span> <span class="va">marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Signal-to-noise ratio [log2]"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Signal intensity [log2]"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 2 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Observe the distributions of cell size across the individual images</span></span>
<span><span class="fu">dittoPlot</span><span class="op">(</span><span class="va">spe</span>, var <span class="op">=</span> <span class="st">"area"</span>, </span>
<span>          group.by <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>          plots <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell area"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">area</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   5.00   47.00   70.00   76.48   98.00  466.00 </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">area</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="va">spe</span><span class="op">$</span><span class="va">area</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="co">### Absolute measure of cell density</span></span>
<span><span class="va">cell_density</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>cell_count <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>           no_pixels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">width_px</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">height_px</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cells_per_mm2 <span class="op">=</span> <span class="va">cell_count</span><span class="op">/</span><span class="op">(</span><span class="va">no_pixels</span><span class="op">/</span><span class="fl">1000000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cell_density</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">sample_id</span>,<span class="va">cells_per_mm2</span><span class="op">)</span>, <span class="va">cells_per_mm2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cells per mm2"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span> <span class="co">### Observing staining differences between samples or batches of samples</span></span>
<span><span class="fu">multi_dittoPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">"patient_id"</span>, plots <span class="op">=</span> <span class="st">"ridgeplot"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    color.panel <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0118</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0247</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0809</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0408</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.163</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0766</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.083</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0675</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.105</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0795</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0444</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.107</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0599</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0992</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0211</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.11</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0364</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0901</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.119</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0582</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0542</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0804</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.106</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0306</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0469</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0825</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0485</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0845</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.111</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.081</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0939</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0973</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.15</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.173</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0642</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0987</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0117</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: No shared levels found between `names(values)` of the manual scale and the
data's colour values.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb154"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220225</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">runUMAP</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, subset_row <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>, exprs_values <span class="op">=</span> <span class="st">"exprs"</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb159"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">runTSNE</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, subset_row <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>, exprs_values <span class="op">=</span> <span class="st">"exprs"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">reducedDims</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 9
names(9): UMAP TSNE fastMNN ... UMAP_harmony seurat UMAP_seurat</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb161"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"UMAP"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   UMAP1     UMAP2
Patient1_001_1 -4.965459 -2.914412
Patient1_001_2 -4.336567 -2.855069
Patient1_001_3 -4.357023 -2.846398
Patient1_001_4 -3.930147 -2.693578
Patient1_001_5 -6.713229 -1.826328
Patient1_001_6 -6.416659 -2.157444</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb163"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize patient id </span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb165"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"TSNE"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on TSNE"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb167"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize region of interest id</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"ROI"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">ROI</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"ROI ID on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb169"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"ROI"</span>, reduction.use <span class="op">=</span> <span class="st">"TSNE"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">ROI</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"ROI ID on TSNE"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb171"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize indication</span></span>
<span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"indication"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Indication on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb173"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p6</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"indication"</span>, reduction.use <span class="op">=</span> <span class="st">"TSNE"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Indication on TSNE"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb175"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p5</span> <span class="op">+</span> <span class="va">p6</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb176"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize marker expression</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>        <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"Ecad"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, </span>
<span>        assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_viridis</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ecad"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"E-Cadherin expression on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb178"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"CD45RO"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_viridis</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"CD45RO"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"CD45RO expression on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb180"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"Ecad"</span>, reduction.use <span class="op">=</span> <span class="st">"TSNE"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_viridis</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ecad"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Ecad expression on TSNE"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb182"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"CD45RO"</span>, reduction.use <span class="op">=</span> <span class="st">"TSNE"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_viridis</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"CD45RO"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"CD45RO expression on TSNE"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb184"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We observe a strong separation of tumor cells (Ecad+ cells) between the patients. Here, each patient was diagnosed with a different tumor type. The separation of tumor cells could be of biological origin since tumor cells tend to display differences in expression between patients and cancer types and/or of technical origin: the panel only contains a single tumor marker (E-Cadherin) and therefore slight technical differences in staining causes visible separation between cells of different patients. Nevertheless, the immune compartment (CD45RO+ cells) mix between patients and we can rule out systematic staining differences between patients.</p>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb185"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Save objects for further downstream analysis</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="batch-effect-correction" class="level2"><h2 class="anchored" data-anchor-id="batch-effect-correction">Batch effect correction</h2>
<section id="fastmnn-correction" class="level3"><h3 class="anchored" data-anchor-id="fastmnn-correction">fastMNN correction</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb186"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb187"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Perform sample correction</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220228</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu">fastMNN</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    batch <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>    auto.merge <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"exprs"</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in check_numbers(k = k, nu = nu, nv = nv, limit = min(dim(x)) - : more
singular values/vectors requested than available</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =
TRUE, : You're computing too large a percentage of total singular values, use a
standard svd instead.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb190"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Check that order of cells is the same</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Transfer the correction results to the main spe object</span></span>
<span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"fastMNN"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">out</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Quality control of correction results</span></span>
<span><span class="va">merge_info</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">$</span><span class="va">merge.info</span> </span>
<span></span>
<span><span class="co">### 1. We observe that Patient4 and Patient2 are most similar with a low batch effect. </span></span>
<span><span class="co">### 2. Merging cells of Patient3 into the combined batch of Patient1, Patient2 </span></span>
<span><span class="co">### and Patient4 resulted in the highest percentage of lost variance and the </span></span>
<span><span class="co">### detection of the largest batch effect. </span></span>
<span><span class="va">merge_info</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"right"</span>, <span class="st">"batch.size"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3 rows and 3 columns
                        left    right batch.size
                      &lt;List&gt;   &lt;List&gt;  &lt;numeric&gt;
1                   Patient4 Patient2   0.366641
2          Patient4,Patient2 Patient1   0.541466
3 Patient4,Patient2,Patient1 Patient3   0.749047</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb192"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merge_info</span><span class="op">$</span><span class="va">lost.var</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Patient1    Patient2   Patient3    Patient4
[1,] 0.000000000 0.030385015 0.00000000 0.048613071
[2,] 0.042567359 0.007911340 0.00000000 0.011963319
[3,] 0.007594552 0.003602307 0.07579009 0.006601185</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb194"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Recompute the UMAP embedding using the corrected low-dimensional </span></span>
<span><span class="co">### coordinates for each cell.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220228</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">runUMAP</span><span class="op">(</span><span class="va">spe</span>, dimred<span class="op">=</span> <span class="st">"fastMNN"</span>, name <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb199"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize patient id </span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP before correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb201"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP after correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb203"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### We observe an imperfect merging of Patient3 into all other samples. </span></span>
<span><span class="co">### This was already seen when displaying the merging information above. </span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb204"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize the expression of selected markers across all cells before and </span></span>
<span><span class="co">### after batch correction</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Ecad"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"FOXP3"</span>, <span class="st">"CD206"</span>, <span class="st">"MPO"</span>, <span class="st">"SMA"</span>, <span class="st">"Ki67"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Before correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb206"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb207"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### After correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> </span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb209"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### We observe that immune cells across patients are merged after batch </span></span>
<span><span class="co">### correction using fastMNN. However, the tumor cells of different patients </span></span>
<span><span class="co">### still cluster separately.</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span> </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="harmony-correction" class="level3"><h3 class="anchored" data-anchor-id="harmony-correction">harmony correction</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb210"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### harmony returns the corrected low-dimensional coordinates for each cell</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    subset_row <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>, </span>
<span>    exprs_values <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    ncomponents <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">230616</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">RunHarmony</span><span class="op">(</span><span class="va">spe</span>, group.by.vars <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Transposing data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Initializing state using k-means centroids initialization</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 1/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 2/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 3/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 4/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony 5/10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Harmony converged after 5 iterations</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb219"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Check that order of cells is the same</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"harmony"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">out</span>, <span class="st">"HARMONY"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb220"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220228</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"harmony"</span>, name <span class="op">=</span> <span class="st">"UMAP_harmony"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb225"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### visualize patient id</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP before correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb227"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP_harmony"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP after correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb229"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb230"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize selected marker expression</span></span>
<span><span class="co">### Before correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb233"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb234"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### After correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP_harmony"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb236"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### We observe a more aggressive merging of cells from different patients </span></span>
<span><span class="co">### compared to the results after fastMNN correction. Importantly, immune cell </span></span>
<span><span class="co">### and epithelial markers are expressed in distinct regions of the UMAP.</span></span>
<span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span> </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-26-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="seurat-correction" class="level3"><h3 class="anchored" data-anchor-id="seurat-correction">Seurat correction</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb237"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu">as.Seurat</span><span class="op">(</span><span class="va">spe</span>, counts <span class="op">=</span> <span class="st">"counts"</span>, data <span class="op">=</span> <span class="st">"exprs"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from UMAP to UMAP_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from TSNE to TSNE_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from UMAP to UMAP_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'UMAP_' taken, using 'umapmnncorrected_' instead</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from PC to PC_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from UMAP to UMAP_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'UMAP_' taken, using 'umapharmony_' instead</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'PC_' taken, using 'seurat_' instead</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Keys should be one or more alphanumeric characters followed by an
underscore, setting key from UMAP to UMAP_</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'UMAP_' taken, using 'umapseurat_' instead</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb248"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu">SplitObject</span><span class="op">(</span><span class="va">seurat_obj</span>, split.by <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Define the features used for integration and perform PCA on cells of each</span></span>
<span><span class="co">### patient individually</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="va">seurat_list</span>,</span>
<span>    FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">x</span>, features <span class="op">=</span> <span class="va">features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>        <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">x</span>, features <span class="op">=</span> <span class="va">features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, approx <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'PC_' taken, using 'pca_' instead</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Key 'PC_' taken, using 'pca_' instead
Key 'PC_' taken, using 'pca_' instead
Key 'PC_' taken, using 'pca_' instead</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb251"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">anchors</span> <span class="op">&lt;-</span> <span class="fu">FindIntegrationAnchors</span><span class="op">(</span></span>
<span>    object.list <span class="op">=</span> <span class="va">seurat_list</span>,</span>
<span>    anchor.features <span class="op">=</span> <span class="va">features</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">"rpca"</span>,</span>
<span>    k.anchor <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scaling features for provided objects</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing within dataset neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding all pairwise anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 53333 anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 47418 anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 46957 anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 55977 anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 67563 anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 52131 anchors</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb279"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">IntegrateData</span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">anchors</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging dataset 2 into 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting anchors for merged samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in irlba(A = t(x = object), nv = npcs, ...): You're computing too large
a percentage of total singular values, use a standard svd instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Integrating data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging dataset 1 into 4 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting anchors for merged samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in irlba(A = t(x = object), nv = npcs, ...): You're computing too large
a percentage of total singular values, use a standard svd instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Integrating data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging dataset 3 into 4 2 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting anchors for merged samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in irlba(A = t(x = object), nv = npcs, ...): You're computing too large
a percentage of total singular values, use a standard svd instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Integrating data</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb298"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">combined</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"integrated"</span></span>
<span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">combined</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">combined</span>, npcs <span class="op">=</span> <span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, approx <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Check that order of cells is the same</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">combined</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"seurat"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">Embeddings</span><span class="op">(</span><span class="va">combined</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb299"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220228</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"seurat"</span>, name <span class="op">=</span> <span class="st">"UMAP_seurat"</span><span class="op">)</span> </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb304"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize patient id </span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>        <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, </span>
<span>        reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP before correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb306"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>        <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"patient_id"</span>, </span>
<span>        reduction.use <span class="op">=</span> <span class="st">"UMAP_seurat"</span>, size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID on UMAP after correction"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb308"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb309"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Before correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> </span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb312"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-28-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb313"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### After correction</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu">multi_dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="va">markers</span>, reduction.use <span class="op">=</span> <span class="st">"UMAP_seurat"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, size <span class="op">=</span> <span class="fl">0.2</span>, list.out <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> </span>
<span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">plot_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb315"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Similar to the methods presented above, Seurat integrates immune cells </span></span>
<span><span class="co">### correctly. When visualizing the patient IDs, slight patient-to-patient </span></span>
<span><span class="co">### differences within tumor cells can be detected.</span></span>
<span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span> </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-28-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb316"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Save modified object</span></span>
<span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="cell-phenotyping" class="level2"><h2 class="anchored" data-anchor-id="cell-phenotyping">Cell phenotyping</h2>
<p>A common step during single-cell data analysis is the annotation of cells based on their phenotype. Defining cell phenotypes is often subjective and relies on previous biological knowledge.</p>
<p>In highly-multiplexed imaging, target proteins or molecules are manually selected based on the biological question at hand. It narrows down the feature space and facilitates the manual annotation of clusters to derive cell phenotypes.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb317"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###  Sample 2000 cells to visualize cluster membership.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220619</span><span class="op">)</span></span>
<span><span class="va">cur_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span>, <span class="fl">2000</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="rphenograph" class="level3"><h3 class="anchored" data-anchor-id="rphenograph">Rphenograph</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb318"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">230619</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">Rphenograph</span><span class="op">(</span><span class="va">mat</span>, k <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Run Rphenograph starts:
  -Input data of 47794 rows and 37 columns
  -k is set to 45</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Finding nearest neighbors...DONE ~ 57.221 s
  Compute jaccard coefficient between nearest-neighbor sets...DONE ~ 20.057 s
  Build undirected graph from the weighted links...DONE ~ 2.609 s
  Run louvain clustering on the graph ...DONE ~ 5.252 s</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Run Rphenograph DONE, totally takes 85.139s.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Return a community class
  -Modularity value: 0.8620567 
  -Number of clusters: 24</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb323"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu">membership</span><span class="op">(</span><span class="va">out</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters</span> <span class="op">&lt;-</span> <span class="va">clusters</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"pg_clusters"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Phenograph clusters on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb324"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### We can observe that some of the clusters only contain cells of a single </span></span>
<span><span class="co">### patient. This can often be observed in the tumor compartment.</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>, <span class="va">cur_cells</span><span class="op">]</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pg_clusters"</span>, <span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>        <span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-31-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb325"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Use the integrated cells in low dimensional embedding for clustering</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"fastMNN"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">230619</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">Rphenograph</span><span class="op">(</span><span class="va">mat</span>, k <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Run Rphenograph starts:
  -Input data of 47794 rows and 36 columns
  -k is set to 45</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Finding nearest neighbors...DONE ~ 51.466 s
  Compute jaccard coefficient between nearest-neighbor sets...DONE ~ 20.096 s
  Build undirected graph from the weighted links...DONE ~ 2.728 s
  Run louvain clustering on the graph ...DONE ~ 7.708 s</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Run Rphenograph DONE, totally takes 81.998s.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Return a community class
  -Modularity value: 0.8453626 
  -Number of clusters: 25</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb330"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu">membership</span><span class="op">(</span><span class="va">out</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters_corrected</span> <span class="op">&lt;-</span> <span class="va">clusters</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"pg_clusters_corrected"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Phenograph clusters on UMAP, integrated cells"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb331"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Clustering using the integrated embedding leads to clusters that contain </span></span>
<span><span class="co">### cells of different patients. </span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>, <span class="va">cur_cells</span><span class="op">]</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pg_clusters_corrected"</span>, <span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters_corrected</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-32-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="shared-nearest-neighbour-graph" class="level3"><h3 class="anchored" data-anchor-id="shared-nearest-neighbour-graph">Shared nearest neighbour graph</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb332"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">combinations</span> <span class="op">&lt;-</span> <span class="fu">clusterSweep</span><span class="op">(</span></span>
<span>    <span class="va">mat</span>,</span>
<span>    BLUSPARAM <span class="op">=</span> <span class="fu">SNNGraphParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10L</span>, <span class="fl">20L</span><span class="op">)</span>,</span>
<span>    type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rank"</span>, <span class="st">"jaccard"</span><span class="op">)</span>,</span>
<span>    cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span>RNGseed <span class="op">=</span> <span class="fl">220427</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="va">combinations</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="fu">approxSilhouette</span><span class="op">(</span><span class="va">mat</span>, <span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">width</span><span class="op">)</span>,</span>
<span>    <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span>,</span>
<span>    sil <span class="op">=</span> <span class="va">sil</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">method</span>, <span class="va">sil</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Cluster parameter combination"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Average silhouette width"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb333"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pur</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="va">combinations</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span>, </span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="fu">neighborPurity</span><span class="op">(</span><span class="va">mat</span>, <span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">purity</span><span class="op">)</span>, </span>
<span>    <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pur</span><span class="op">)</span>, pur <span class="op">=</span> <span class="va">pur</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">method</span>, <span class="va">pur</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Cluster parameter combination"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Average neighborhood purity"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-33-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb334"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220620</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>,<span class="op">]</span>, </span>
<span>    assay.type <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    BLUSPARAM <span class="op">=</span> <span class="fu">SNNGraphParam</span><span class="op">(</span>k<span class="op">=</span><span class="fl">20</span>, </span>
<span>    cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"rank"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters</span> <span class="op">&lt;-</span> <span class="va">clusters</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"nn_clusters"</span>, </span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SNN clusters on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb335"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>             genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>             assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>             heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>             annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nn_clusters"</span>, <span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>             annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                              <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-34-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb336"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220621</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    use.dimred <span class="op">=</span> <span class="st">"fastMNN"</span>, </span>
<span>    BLUSPARAM <span class="op">=</span> <span class="fu">SNNGraphParam</span><span class="op">(</span>k <span class="op">=</span> <span class="fl">20</span>, </span>
<span>    cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"rank"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters_corrected</span> <span class="op">&lt;-</span> <span class="va">clusters</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"nn_clusters_corrected"</span>, </span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SNN clusters on UMAP, integrated cells"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb337"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nn_clusters_corrected"</span>,<span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters_corrected</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                    <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-35-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="self-organizing-maps" class="level3"><h3 class="anchored" data-anchor-id="self-organizing-maps">Self organizing maps</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb338"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run FlowSOM and ConsensusClusterPlus clustering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220410</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">CATALYST</span><span class="fu">::</span><span class="fu">cluster</span><span class="op">(</span><span class="va">spe</span>, </span>
<span>               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>               maxK <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>o running FlowSOM clustering...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>o running ConsensusClusterPlus metaclustering...</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb341"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Assess cluster stability</span></span>
<span><span class="fu">delta_area</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb342"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">som_clusters</span> <span class="op">&lt;-</span> <span class="fu">cluster_ids</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"meta13"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"som_clusters"</span>, </span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SOM clusters on UMAP"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-36-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb343"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"som_clusters"</span>, <span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">som_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-36-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb344"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">kohonen</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'kohonen'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:mclust':

    map</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    map</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb348"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ConsensusClusterPlus</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Select integrated cells</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"fastMNN"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Perform SOM clustering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220410</span><span class="op">)</span></span>
<span><span class="va">som.out</span> <span class="op">&lt;-</span> <span class="fu">clusterRows</span><span class="op">(</span><span class="va">mat</span>, <span class="fu">SomParam</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, full <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Cluster the 100 SOM codes into larger clusters</span></span>
<span><span class="va">ccp</span> <span class="op">&lt;-</span> <span class="fu">ConsensusClusterPlus</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">som.out</span><span class="op">$</span><span class="va">objects</span><span class="op">$</span><span class="va">som</span><span class="op">$</span><span class="va">codes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    maxK <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    reps <span class="op">=</span> <span class="fl">100</span>, </span>
<span>    distance <span class="op">=</span> <span class="st">"euclidean"</span>, </span>
<span>    seed <span class="op">=</span> <span class="fl">220410</span>, </span>
<span>    plot <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>end fraction</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-8.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-9.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-10.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-11.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-12.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-13.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-14.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-15.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-16.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-17.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-18.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-19.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-20.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-21.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-22.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-23.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-24.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-25.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-26.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-27.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-28.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>clustered</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-29.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-30.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-31.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-32.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-33.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb379"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize delta area plot</span></span>
<span><span class="fu">CATALYST</span><span class="fu">:::</span><span class="fu">.plot_delta_area</span><span class="op">(</span><span class="va">ccp</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-34.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb380"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Link ConsensusClusterPlus clusters with SOM codes and save in object</span></span>
<span><span class="va">som.cluster</span> <span class="op">&lt;-</span> <span class="va">ccp</span><span class="op">[[</span><span class="fl">16</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"consensusClass"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">som.out</span><span class="op">$</span><span class="va">clusters</span><span class="op">]</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">som_clusters_corrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">som.cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, var <span class="op">=</span> <span class="st">"som_clusters_corrected"</span>, </span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SOM clusters on UMAP, integrated cells"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-35.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb381"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"som_clusters_corrected"</span>,<span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annot.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">dittoColors</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">som_clusters_corrected</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                              <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-37-36.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="compare-between-clustering-approaches" class="level3"><h3 class="anchored" data-anchor-id="compare-between-clustering-approaches">Compare between clustering approaches</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb382"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'gridExtra'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:EBImage':

    combine</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Biobase':

    combine</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:BiocGenerics':

    combine</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    combine</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb388"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tab1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Rphenograph"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SNN"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">tab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Rphenograph"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SOM"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">som_clusters</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">tab3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SNN"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SOM"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">som_clusters</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab1</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb389"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab2</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-38-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb390"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab3</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-38-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="further-clustering-notes" class="level3"><h3 class="anchored" data-anchor-id="further-clustering-notes">Further clustering notes</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb391"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">cluster_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters_corrected</span>,</span>
<span>                            <span class="st">"1"</span> <span class="op">=</span> <span class="st">"Tumor_proliferating"</span>,</span>
<span>                            <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Myeloid"</span>,</span>
<span>                            <span class="st">"3"</span> <span class="op">=</span> <span class="st">"Tumor"</span>,</span>
<span>                            <span class="st">"4"</span> <span class="op">=</span> <span class="st">"Tumor"</span>,</span>
<span>                            <span class="st">"5"</span> <span class="op">=</span> <span class="st">"Stroma"</span>,</span>
<span>                            <span class="st">"6"</span> <span class="op">=</span> <span class="st">"Proliferating"</span>,</span>
<span>                            <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Myeloid"</span>,</span>
<span>                            <span class="st">"8"</span> <span class="op">=</span> <span class="st">"Plasma_cell"</span>,</span>
<span>                            <span class="st">"9"</span> <span class="op">=</span> <span class="st">"CD8"</span>,</span>
<span>                            <span class="st">"10"</span> <span class="op">=</span> <span class="st">"CD4"</span>,</span>
<span>                            <span class="st">"11"</span> <span class="op">=</span> <span class="st">"Neutrophil"</span>,</span>
<span>                            <span class="st">"12"</span> <span class="op">=</span> <span class="st">"Bcell"</span>,</span>
<span>                            <span class="st">"13"</span> <span class="op">=</span> <span class="st">"Stroma"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">cluster_celltype</span> <span class="op">&lt;-</span> <span class="va">cluster_celltype</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb392"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="classfication-approach" class="level3"><h3 class="anchored" data-anchor-id="classfication-approach">Classfication approach</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb393"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Manual labeling of cells</span></span>
<span><span class="co"># if (interactive()) {</span></span>
<span></span>
<span><span class="co">#     images &lt;- qread(here(dir, "data/images.qs"))</span></span>
<span><span class="co">#     masks &lt;- qread(here(dir, "data/masks.qs"))</span></span>
<span></span>
<span><span class="co">#     cytomapperShiny(object = spe, mask = masks, image = images,</span></span>
<span><span class="co">#                     cell_id = "ObjectNumber", img_id = "sample_id")</span></span>
<span><span class="co"># }</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">### Define color vectors</span></span>
<span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#3F1B03"</span>, <span class="st">"#F4AD31"</span>, <span class="st">"#894F36"</span>, <span class="st">"#1C750C"</span>, <span class="st">"#EF8ECC"</span>,</span>
<span>        <span class="st">"#6471E2"</span>, <span class="st">"#4DB23B"</span>, <span class="st">"grey"</span>, <span class="st">"#F4800C"</span>, <span class="st">"#BF0A3D"</span>, <span class="st">"#066970"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Tumor"</span>, <span class="st">"Stroma"</span>, <span class="st">"Myeloid"</span>, <span class="st">"CD8"</span>, <span class="st">"Plasma_cell"</span>,</span>
<span>        <span class="st">"Treg"</span>, <span class="st">"CD4"</span>, <span class="st">"undefined"</span>, <span class="st">"BnTcell"</span>, <span class="st">"Bcell"</span>, <span class="st">"Neutrophil"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">celltype</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb394"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Read in and consolidate labeled data</span></span>
<span><span class="va">label_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/gated_cells"</span><span class="op">)</span>, </span>
<span>    full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">".rds$"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read in SPE objects</span></span>
<span><span class="va">spes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">label_files</span>, <span class="va">readRDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Merge SPE objects</span></span>
<span><span class="va">concat_spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="va">spes</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>'sample_id's are duplicated across 'SpatialExperiment' objects to cbind; appending sample indices.</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb396"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">filter_labels</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">object</span>, </span>
<span>                          <span class="va">label</span> <span class="op">=</span> <span class="st">"cytomapper_CellLabel"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cur_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html">unclass</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>, <span class="va">object</span><span class="op">[[</span><span class="va">label</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">cur_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cur_tab</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">cur_tab</span>, <span class="fl">1</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cur_labels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cur_tab</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">cur_labels</span> <span class="op">&lt;-</span> <span class="va">cur_labels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">cur_tab</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">cur_labels</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu">filter_labels</span><span class="op">(</span><span class="va">concat_spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_spe</span> <span class="op">&lt;-</span> <span class="va">concat_spe</span><span class="op">[</span>,<span class="va">concat_spe</span><span class="op">$</span><span class="va">cytomapper_CellLabel</span> <span class="op">!=</span> <span class="st">"Tumor"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">non_tumor_labels</span> <span class="op">&lt;-</span> <span class="fu">filter_labels</span><span class="op">(</span><span class="va">cur_spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">additional_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">non_tumor_labels</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">final_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">labels</span>, <span class="va">non_tumor_labels</span><span class="op">[</span><span class="va">additional_cells</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Transfer labels to SPE object</span></span>
<span><span class="va">spe_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"unlabeled"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">spe_labels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span><span class="va">spe_labels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">final_labels</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">final_labels</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">cell_labels</span> <span class="op">&lt;-</span> <span class="va">spe_labels</span></span>
<span></span>
<span><span class="co">### Number of cells labeled per patient</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">cell_labels</span>, <span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             
              Patient1 Patient2 Patient3 Patient4
  Bcell            152      131      234      263
  BnTcell          396       37      240     1029
  CD4               45      342      167      134
  CD8               60      497      137      128
  Myeloid          183      378      672      517
  Neutrophil        97        4       17       16
  Plasma_cell       34      536       87       59
  Stroma            84       37       85      236
  Treg             139      149       49       24
  Tumor           2342      906     1618     1133
  unlabeled       7214     9780     7826     9580</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb398"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Split between labeled and unlabeled cells</span></span>
<span><span class="va">lab_spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="va">spe</span><span class="op">$</span><span class="va">cell_labels</span> <span class="op">!=</span> <span class="st">"unlabeled"</span><span class="op">]</span></span>
<span><span class="va">unlab_spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>,<span class="va">spe</span><span class="op">$</span><span class="va">cell_labels</span> <span class="op">==</span> <span class="st">"unlabeled"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">### Randomly split into train and test data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">221029</span><span class="op">)</span></span>
<span><span class="va">trainIndex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html">createDataPartition</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">lab_spe</span><span class="op">$</span><span class="va">cell_labels</span><span class="op">)</span>, p <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span><span class="va">train_spe</span> <span class="op">&lt;-</span> <span class="va">lab_spe</span><span class="op">[</span>,<span class="va">trainIndex</span><span class="op">$</span><span class="va">Resample1</span><span class="op">]</span></span>
<span><span class="va">test_spe</span> <span class="op">&lt;-</span> <span class="va">lab_spe</span><span class="op">[</span>,<span class="op">-</span><span class="va">trainIndex</span><span class="op">$</span><span class="va">Resample1</span><span class="op">]</span></span>
<span></span>
<span><span class="co">### Define fit parameters for 5-fold cross validation</span></span>
<span><span class="va">fitControl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html">trainControl</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"cv"</span>,number <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Select the arsinh-transformed counts for training</span></span>
<span><span class="va">cur_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">train_spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">train_spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Train a random forest classifier</span></span>
<span><span class="va">rffit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html">train</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">cur_mat</span>, </span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">train_spe</span><span class="op">$</span><span class="va">cell_labels</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"rf"</span>, ntree <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    tuneLength <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    trControl <span class="op">=</span> <span class="va">fitControl</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rffit</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Random Forest 

10049 samples
   37 predictor
   10 classes: 'Bcell', 'BnTcell', 'CD4', 'CD8', 'Myeloid', 'Neutrophil', 'Plasma_cell', 'Stroma', 'Treg', 'Tumor' 

No pre-processing
Resampling: Cross-Validated (5 fold) 
Summary of sample sizes: 8040, 8039, 8038, 8038, 8041 
Resampling results across tuning parameters:

  mtry  Accuracy   Kappa    
   2    0.9643721  0.9523612
  10    0.9754201  0.9672910
  19    0.9757188  0.9676911
  28    0.9751223  0.9668928
  37    0.9735309  0.9647805

Accuracy was used to select the optimal model using the largest value.
The final value used for the model was mtry = 19.</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb400"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Classifier performance</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">rffit</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_errorbar</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rffit</span><span class="op">$</span><span class="va">results</span>,</span>
<span>                <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">Accuracy</span> <span class="op">-</span> <span class="va">AccuracySD</span>,</span>
<span>                    ymax <span class="op">=</span> <span class="va">Accuracy</span> <span class="op">+</span> <span class="va">AccuracySD</span><span class="op">)</span>,</span>
<span>                width <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb401"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize the variable importance of the classifier.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/caret/man/varImp.html">varImp</a></span><span class="op">(</span><span class="va">rffit</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-45-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb402"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Select the arsinh-transformed counts of the test data</span></span>
<span><span class="va">cur_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">test_spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">test_spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Predict the cell phenotype labels of the test data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">231019</span><span class="op">)</span></span>
<span><span class="va">cur_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">rffit</span>, newdata <span class="op">=</span> <span class="va">cur_mat</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb403"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/confusionMatrix.html">confusionMatrix</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">cur_pred</span>, </span>
<span>    reference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">cell_labels</span><span class="op">)</span>, </span>
<span>    mode <span class="op">=</span> <span class="st">"everything"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">cm</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Confusion Matrix and Statistics

             Reference
Prediction    Bcell BnTcell  CD4  CD8 Myeloid Neutrophil Plasma_cell Stroma
  Bcell         185       4    0    0       0          0           5      0
  BnTcell         3     421    2    1       0          0           1      0
  CD4             0       0  161    0       0          2           4      2
  CD8             0       0    0  197       0          0           7      0
  Myeloid         0       0    2    3     437          0           0      0
  Neutrophil      0       0    0    0       0         30           0      0
  Plasma_cell     0       0    4    1       0          0         158      0
  Stroma          0       0    2    0       0          0           0    108
  Treg            0       0    0    0       0          0           3      0
  Tumor           7       0    1    3       0          1           1      0
             Reference
Prediction    Treg Tumor
  Bcell          0     0
  BnTcell        0     0
  CD4            0     4
  CD8            1     5
  Myeloid        0     0
  Neutrophil     0     0
  Plasma_cell    0     0
  Stroma         0     0
  Treg          89     2
  Tumor          0  1488

Overall Statistics
                                          
               Accuracy : 0.9788          
                 95% CI : (0.9733, 0.9834)
    No Information Rate : 0.4481          
    P-Value [Acc &gt; NIR] : &lt; 2.2e-16       
                                          
                  Kappa : 0.9717          
                                          
 Mcnemar's Test P-Value : NA              

Statistics by Class:

                     Class: Bcell Class: BnTcell Class: CD4 Class: CD8
Sensitivity               0.94872         0.9906    0.93605    0.96098
Specificity               0.99714         0.9976    0.99622    0.99586
Pos Pred Value            0.95361         0.9836    0.93064    0.93810
Neg Pred Value            0.99683         0.9986    0.99653    0.99745
Precision                 0.95361         0.9836    0.93064    0.93810
Recall                    0.94872         0.9906    0.93605    0.96098
F1                        0.95116         0.9871    0.93333    0.94940
Prevalence                0.05830         0.1271    0.05142    0.06129
Detection Rate            0.05531         0.1259    0.04813    0.05889
Detection Prevalence      0.05800         0.1280    0.05172    0.06278
Balanced Accuracy         0.97293         0.9941    0.96613    0.97842
                     Class: Myeloid Class: Neutrophil Class: Plasma_cell
Sensitivity                  1.0000          0.909091            0.88268
Specificity                  0.9983          1.000000            0.99842
Pos Pred Value               0.9887          1.000000            0.96933
Neg Pred Value               1.0000          0.999095            0.99340
Precision                    0.9887          1.000000            0.96933
Recall                       1.0000          0.909091            0.88268
F1                           0.9943          0.952381            0.92398
Prevalence                   0.1306          0.009865            0.05351
Detection Rate               0.1306          0.008969            0.04723
Detection Prevalence         0.1321          0.008969            0.04873
Balanced Accuracy            0.9991          0.954545            0.94055
                     Class: Stroma Class: Treg Class: Tumor
Sensitivity                0.98182     0.98889       0.9927
Specificity                0.99938     0.99846       0.9930
Pos Pred Value             0.98182     0.94681       0.9913
Neg Pred Value             0.99938     0.99969       0.9940
Precision                  0.98182     0.94681       0.9913
Recall                     0.98182     0.98889       0.9927
F1                         0.98182     0.96739       0.9920
Prevalence                 0.03288     0.02691       0.4481
Detection Rate             0.03229     0.02661       0.4448
Detection Prevalence       0.03288     0.02810       0.4487
Balanced Accuracy          0.99060     0.99368       0.9928</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb405"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">byClass</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"Class: "</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">byClass</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">Specificity</span>, <span class="va">Sensitivity</span>, </span>
<span>                 size <span class="op">=</span> <span class="va">Detection.Rate</span>,</span>
<span>                 fill <span class="op">=</span> <span class="va">class</span><span class="op">)</span>,</span>
<span>             shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Sensitivity (TPR)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"1 - Specificity (FPR)"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb406"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">231019</span><span class="op">)</span></span>
<span><span class="va">cur_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">rffit</span>, </span>
<span>                    newdata <span class="op">=</span> <span class="va">cur_mat</span>, </span>
<span>                    type <span class="op">=</span> <span class="st">"prob"</span><span class="op">)</span></span>
<span><span class="va">cur_pred</span><span class="op">$</span><span class="va">truth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">cell_labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_pred</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="va">Bcell</span><span class="op">:</span><span class="va">Tumor</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">name</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">name</span><span class="op">)</span>, outlier.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">truth</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb407"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Classification of new data</span></span>
<span><span class="co">### Select the arsinh-transformed counts of the unlabeled data for prediction</span></span>
<span><span class="va">cur_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">unlab_spe</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">unlab_spe</span><span class="op">)</span><span class="op">$</span><span class="va">use_channel</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Predict the cell phenotype labels of the unlabeled data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">231014</span><span class="op">)</span></span>
<span><span class="va">cell_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span></span>
<span>    <span class="va">rffit</span>, </span>
<span>    newdata <span class="op">=</span> <span class="va">cur_mat</span>, </span>
<span>    type <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cell_class</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cur_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">cell_class</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>cell_class
      Bcell     BnTcell         CD4         CD8     Myeloid  Neutrophil 
        799        1047        3664        2728        5726         453 
Plasma_cell      Stroma        Treg       Tumor 
       3369        4669        1057       10888 </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb409"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Extract prediction probabilities for each cell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">231014</span><span class="op">)</span></span>
<span><span class="va">cell_prob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">rffit</span>, newdata <span class="op">=</span> <span class="va">cur_mat</span>, type <span class="op">=</span> <span class="st">"prob"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Distribution of maximum probabilities</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>max_prob <span class="op">=</span> <span class="fu">rowMax</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">cell_prob</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       type <span class="op">=</span> <span class="va">cell_class</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">geom_density_ridges</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">max_prob</span>, y <span class="op">=</span> <span class="va">cell_class</span>, fill <span class="op">=</span> <span class="va">cell_class</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Maximum probability"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell type"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0281</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb411"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Label undefined cells</span></span>
<span><span class="va">cell_class</span><span class="op">[</span><span class="fu">rowMax</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">cell_prob</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"undefined"</span></span>
<span></span>
<span><span class="co">### Store labels in SpatialExperiment onject</span></span>
<span><span class="va">cell_labels</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">$</span><span class="va">cell_labels</span></span>
<span><span class="va">cell_labels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">unlab_spe</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cell_class</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">cell_labels</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             
              Patient1 Patient2 Patient3 Patient4
  Bcell            179      517      422      460
  BnTcell          422      601      604     1110
  CD4              407     1383      700     1394
  CD8              510     1370      478     1156
  Myeloid         1210     1876     1599     2685
  Neutrophil       284        6       92      164
  Plasma_cell      809     2428      425      338
  Stroma           582      603      662     3169
  Treg             524      379      244      258
  Tumor           5593     3360     5793     2117
  undefined        226      274      113      268</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb413"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tab1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Rphenograph"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SNN"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tab3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SOM"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">som_clusters</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab1</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb414"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab2</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-51-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb415"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab3</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-51-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb416"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tab1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Rphenograph"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">pg_clusters_corrected</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SNN"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">nn_clusters_corrected</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tab3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"SOM"</span>, <span class="va">spe</span><span class="op">$</span><span class="va">som_clusters_corrected</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab1</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb417"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab2</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-52-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb418"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">tab3</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-52-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb419"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">qsave</span><span class="op">(</span><span class="va">spe</span>, <span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section></section><section id="single-cell-visualization" class="level2"><h2 class="anchored" data-anchor-id="single-cell-visualization">Single cell visualization</h2>
<section id="inital-setup" class="level3"><h3 class="anchored" data-anchor-id="inital-setup">Inital setup</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb420"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">qread</span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html">here</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"data/spe.qs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">reducedDims</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 9
names(9): UMAP TSNE fastMNN ... UMAP_harmony seurat UMAP_seurat</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb422"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "sample_id"              "ObjectNumber"           "area"                  
 [4] "axis_major_length"      "axis_minor_length"      "eccentricity"          
 [7] "width_px"               "height_px"              "patient_id"            
[10] "ROI"                    "indication"             "pg_clusters"           
[13] "pg_clusters_corrected"  "nn_clusters"            "cluster_id"            
[16] "som_clusters"           "som_clusters_corrected" "nn_clusters_corrected" 
[19] "cluster_celltype"       "cell_labels"            "celltype"              </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb424"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Define cell phenotype markers</span></span>
<span><span class="va">type_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Ecad"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"FOXP3"</span>, <span class="st">"CD206"</span>, <span class="st">"MPO"</span>,</span>
<span>    <span class="st">"SMA"</span>, <span class="st">"CD8a"</span>, <span class="st">"CD4"</span>, <span class="st">"HLADR"</span>, <span class="st">"CD15"</span>, <span class="st">"CD38"</span>, <span class="st">"PDGFRb"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Define cell state markers</span></span>
<span><span class="va">state_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CarbonicAnhydrase"</span>, <span class="st">"Ki67"</span>, <span class="st">"PD1"</span>, <span class="st">"GrzB"</span>, <span class="st">"PDL1"</span>,</span>
<span>    <span class="st">"ICOS"</span>, <span class="st">"TCF7"</span>, <span class="st">"VISTA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Add to spe</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">type_markers</span>, <span class="st">"type"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">state_markers</span>, <span class="st">"state"</span>,</span>
<span>        <span class="st">"other"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="cell-type-level" class="level3"><h3 class="anchored" data-anchor-id="cell-type-level">Cell-type level</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb425"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Dimensionality reduction visualization</span></span>
<span><span class="co">### Interpreting these UMAP, tSNE plots is not trivial, but local neighborhoods </span></span>
<span><span class="co">### in the plot can suggest similarity in expression for given cells.</span></span>
<span><span class="co">### tSNE/UMAP aim to preserve the distances between each cell and its neighbors in the high-dimensional space.</span></span>
<span></span>
<span><span class="co">### UMAP colored by cell type and expression - dittoDimPlot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    var <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Cell types on UMAP, integrated cells"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb427"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    var <span class="op">=</span> <span class="st">"Ecad"</span>, </span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>    colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    do.label <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_viridis</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>do.label was/were ignored for non-discrete data.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb429"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span>    <span class="co"># coord_fixed()</span></span>
<span>  </span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb430"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># UMAP colored by expression for all markers - plotReducedDim</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span></span>
<span>            <span class="va">spe</span>,</span>
<span>            dimred <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span>            colour_by <span class="op">=</span> <span class="va">x</span>,</span>
<span>            by_exprs_values <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>            point_size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>        <span class="op">)</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot_grid</span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb431"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Heatmap visualization, it is often useful to visualize single-cell </span></span>
<span><span class="co">### expression per cell type in form of a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">220818</span><span class="op">)</span></span>
<span><span class="va">cur_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span>, <span class="fl">4000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Heatmap visualization</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"indication"</span>, <span class="st">"patient_id"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span>,</span>
<span>        patient_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>        celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb432"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Visualize the mean marker expression per cell type for all cells</span></span>
<span><span class="co">### aggregate by cell type</span></span>
<span><span class="va">celltype_mean</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>,  </span>
<span>    ids <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>    statistics <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    use.assay.type <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    subset.row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### No scaling</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">celltype_mean</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>        ncells <span class="op">=</span> <span class="fu">plasma</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb433"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Scaled to max</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">celltype_mean</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    scaled.to.max <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    heatmap.colors.max.scaled <span class="op">=</span> <span class="fu">inferno</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>        ncells <span class="op">=</span> <span class="fu">plasma</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-58-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb434"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### # Z score scaled</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">celltype_mean</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>        ncells <span class="op">=</span> <span class="fu">plasma</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-58-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb435"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Violin plot visualization - plotExpression</span></span>
<span><span class="fu">plotExpression</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>    exprs_values <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    colour_by <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb437"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Scatter plot visualization</span></span>
<span><span class="fu">dittoScatterPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>,</span>
<span>    x.var <span class="op">=</span> <span class="st">"CD3"</span>,</span>
<span>    y.var <span class="op">=</span> <span class="st">"CD20"</span>,</span>
<span>    assay.x <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    assay.y <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    color.var <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Scatterplot for CD3/CD20 labelled by celltype"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb439"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### by sample_id - percentage</span></span>
<span><span class="fu">dittoBarPlot</span><span class="op">(</span><span class="va">spe</span>, var <span class="op">=</span> <span class="st">"celltype"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-61-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb441"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### by patient_id - percentage</span></span>
<span><span class="fu">dittoBarPlot</span><span class="op">(</span><span class="va">spe</span>, var <span class="op">=</span> <span class="st">"celltype"</span>, group.by <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-61-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb443"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### by patient_id - count</span></span>
<span><span class="fu">dittoBarPlot</span><span class="op">(</span><span class="va">spe</span>, scale <span class="op">=</span> <span class="st">"count"</span>,var <span class="op">=</span> <span class="st">"celltype"</span>, group.by <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-61-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb445"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### We can see that cell type frequencies change between samples/patients and that the highest proportion/counts of plasma cells and stromal cells can be observed for Patient 2 and Patient 4, respectively.</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb446"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### CATALYST-based visualization</span></span>
<span><span class="co">### Save SPE in CATALYST-compatible object with renamed colData entries and</span></span>
<span><span class="co">### new metadata information</span></span>
<span><span class="va">spe_cat</span> <span class="op">&lt;-</span> <span class="va">spe</span></span>
<span><span class="va">spe_cat</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="va">spe_cat</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span></span>
<span><span class="va">spe_cat</span><span class="op">$</span><span class="va">cluster_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Add celltype information to metadata</span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">cluster_codes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">spe_cat</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Pseudobulk-level MDS plot</span></span>
<span><span class="co">### MDS pseudobulk by cell type to highlight expression similarities between </span></span>
<span><span class="co">### cell types and subsequently for each celltype-sample-combination.</span></span>
<span><span class="fu">CATALYST</span><span class="fu">::</span><span class="fu">pbMDS</span><span class="op">(</span></span>
<span>    <span class="va">spe_cat</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"cluster_id"</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>,</span>
<span>    label_by <span class="op">=</span> <span class="st">"cluster_id"</span>,</span>
<span>    k <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb448"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### MDS pseudobulk by cell type and sample_id</span></span>
<span><span class="co">### We can see that the pseudobulk-expression profile of neutrophils seems </span></span>
<span><span class="co">### markedly distinct from the other cell types, while comparable cell types </span></span>
<span><span class="co">### such as the T cell subtypes group together. Furthermore, pseudobulk </span></span>
<span><span class="co">## cell-type profiles from SCCHN appear different from the other indications.</span></span>
<span><span class="fu">CATALYST</span><span class="fu">::</span><span class="fu">pbMDS</span><span class="op">(</span></span>
<span>    <span class="va">spe_cat</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>, </span>
<span>    k <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>    shape_by <span class="op">=</span> <span class="st">"condition"</span>, </span>
<span>    size_by <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-62-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb450"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Reduced dimension plot on CLR of proportions</span></span>
<span><span class="co">### The output plots aim to illustrate the degree of similarity between </span></span>
<span><span class="co">### cell types based on sample proportions.</span></span>
<span><span class="co">### CLR on cluster proportions across samples</span></span>
<span><span class="co">### We can again observe that neutrophils have a divergent profile also in terms of their sample proportions.</span></span>
<span><span class="fu">CATALYST</span><span class="fu">::</span><span class="fu">clrDR</span><span class="op">(</span></span>
<span>    <span class="va">spe_cat</span>, </span>
<span>    dr <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"cluster_id"</span>, </span>
<span>    k <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>    label_by <span class="op">=</span> <span class="st">"cluster_id"</span>, </span>
<span>    arrow_col <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>    point_pal <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-62-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb451"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Pseudobulk expression boxplot</span></span>
<span><span class="co">### Notably, CD15 levels are elevated in SCCHN in comparison to all other </span></span>
<span><span class="co">### indications for most cell types.</span></span>
<span><span class="fu">CATALYST</span><span class="fu">::</span><span class="fu">plotPbExprs</span><span class="op">(</span></span>
<span>    <span class="va">spe_cat</span>, </span>
<span>    k <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>    facet_by <span class="op">=</span> <span class="st">"cluster_id"</span>, </span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe_cat</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-62-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="sample-level" class="level3"><h3 class="anchored" data-anchor-id="sample-level">Sample level</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb452"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Dimensionality reduction visualization</span></span>
<span><span class="co">## UMAP colored by cell type and expression - dittoDimPlot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>,</span>
<span>    var <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Sample ID"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb454"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>,</span>
<span>    var <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Sample ID"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb456"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>,</span>
<span>    var <span class="op">=</span> <span class="st">"patient_id"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb458"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">dittoDimPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>,</span>
<span>    var <span class="op">=</span> <span class="st">"patient_id"</span>,</span>
<span>    reduction.use <span class="op">=</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    do.label <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Patient ID"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb460"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### fastMNN approach (right side of the plot) leads to mixing of cells across </span></span>
<span><span class="co">### samples/patients and thus batch effect correction.</span></span>
<span><span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb461"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Heatmap visualization to highlight biological differences across samples/patients.</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>,<span class="va">cur_cells</span><span class="op">]</span>, </span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    order.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"patient_id"</span>,<span class="st">"sample_id"</span><span class="op">)</span>,</span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, </span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="st">"indication"</span>, <span class="st">"patient_id"</span>, <span class="st">"sample_id"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span>,</span>
<span>        patient_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>        sample_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-64-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb462"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### mean expression by patient_id</span></span>
<span><span class="va">patient_mean</span> <span class="op">&lt;-</span> <span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>,  </span>
<span>    ids <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">patient_id</span>, </span>
<span>    statistics <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    use.assay.type <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    subset.row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### No scaling</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">patient_mean</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    heatmap.colors <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"patient_id"</span>,<span class="st">"indication"</span>,<span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        patient_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span>,</span>
<span>        ncells <span class="op">=</span> <span class="fu">plasma</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-64-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb463"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Max expression scaling</span></span>
<span><span class="fu">dittoHeatmap</span><span class="op">(</span></span>
<span>    <span class="va">patient_mean</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"exprs"</span>, </span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    scaled.to.max <span class="op">=</span>  <span class="cn">TRUE</span>,</span>
<span>    heatmap.colors.max.scaled <span class="op">=</span> <span class="fu">inferno</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    annot.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"patient_id"</span>,<span class="st">"indication"</span>,<span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        patient_id <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span>,</span>
<span>        indication <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span>,</span>
<span>        ncells <span class="op">=</span> <span class="fu">plasma</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-64-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb464"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Barplot visualization</span></span>
<span><span class="fu">dittoBarPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    var <span class="op">=</span> <span class="st">"patient_id"</span>, </span>
<span>    group.by <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">patient_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-65-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb466"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### </span></span>
<span><span class="fu">dittoBarPlot</span><span class="op">(</span></span>
<span>    <span class="va">spe</span>, </span>
<span>    var <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>    group.by <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-65-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="complexheatmap" class="level3"><h3 class="anchored" data-anchor-id="complexheatmap">ComplexHeatmap</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb468"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### 1. Heatmap bodies ###</span></span>
<span><span class="co">### Heatmap body color</span></span>
<span><span class="va">col_exprs</span> <span class="op">&lt;-</span> <span class="fu">colorRamp2</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#3B518BFF"</span>, <span class="st">"#20938CFF"</span>, <span class="st">"#6ACD5AFF"</span>, <span class="st">"#FDE725FF"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Create Heatmap objects</span></span>
<span><span class="co">### By cell type markers</span></span>
<span><span class="va">celltype_mean</span> <span class="op">&lt;-</span> <span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>,</span>
<span>    ids <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>    statistics <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    use.assay.type <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"type"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">h_type</span> <span class="op">&lt;-</span> <span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">celltype_mean</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"type_markers"</span>,</span>
<span>    col <span class="op">=</span> <span class="va">col_exprs</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"mean exprs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># By cell state markers</span></span>
<span><span class="va">cellstate_mean</span> <span class="op">&lt;-</span> <span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>,</span>
<span>    ids <span class="op">=</span> <span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>    statistics <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    use.assay.type <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">marker_class</span> <span class="op">==</span> <span class="st">"state"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">h_state</span> <span class="op">&lt;-</span> <span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cellstate_mean</span>, <span class="st">"exprs"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"state_markers"</span>,</span>
<span>    col <span class="op">=</span> <span class="va">col_exprs</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"mean exprs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">### 2. Heatmap annotation ###</span></span>
<span><span class="co">### 2.1  Metadata features</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">celltype_mean</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">celltype</span>, <span class="va">ncells</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Proportion of indication per celltype</span></span>
<span><span class="va">indication</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html">unclass</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">spe</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span>, margin <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Number of contributing patients per celltype</span></span>
<span><span class="va">cluster_PID</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">celltype</span>, <span class="va">patient_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n_PID</span> <span class="op">&lt;-</span> <span class="va">cluster_PID</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Freq</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"n_PID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"celltype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Create HeatmapAnnotation objects</span></span>
<span><span class="va">ha_anno</span> <span class="op">&lt;-</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span></span>
<span>    celltype <span class="op">=</span> <span class="va">anno</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>    border <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>celltype <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>,</span>
<span>    which <span class="op">=</span> <span class="st">"row"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ha_meta</span> <span class="op">&lt;-</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span></span>
<span>    n_cells <span class="op">=</span> <span class="fu">anno_barplot</span><span class="op">(</span><span class="va">anno</span><span class="op">$</span><span class="va">ncells</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    n_PID <span class="op">=</span> <span class="fu">anno_barplot</span><span class="op">(</span><span class="va">n_PID</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    indication <span class="op">=</span> <span class="fu">anno_barplot</span><span class="op">(</span><span class="va">indication</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    border <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotation_name_rot <span class="op">=</span> <span class="fl">90</span>,</span>
<span>    gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    which <span class="op">=</span> <span class="st">"row"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### 2.2 Spatial features</span></span>
<span><span class="co">### Add number of neighbors to spe object (saved in colPair)</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">n_neighbors</span> <span class="op">&lt;-</span> <span class="fu">countLnodeHits</span><span class="op">(</span><span class="fu">colPair</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"neighborhood"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Select spatial features and average over celltypes</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">area</span>, <span class="va">celltype</span>, <span class="va">n_neighbors</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="va">spatial</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">celltype</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span><span class="op">(</span>by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>celltype <span class="op">=</span> <span class="va">spatial</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"celltype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Create HeatmapAnnotation object</span></span>
<span><span class="va">ha_spatial</span> <span class="op">&lt;-</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span></span>
<span>    area <span class="op">=</span> <span class="va">spatial</span><span class="op">$</span><span class="va">area</span>,</span>
<span>    n_neighbors <span class="op">=</span> <span class="va">spatial</span><span class="op">$</span><span class="va">n_neighbors</span>,</span>
<span>    border <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    which <span class="op">=</span> <span class="st">"row"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### 3. Plot rich heatmap ###</span></span>
<span></span>
<span><span class="co">### Create HeatmapList object</span></span>
<span><span class="va">h_list</span> <span class="op">&lt;-</span> <span class="va">h_type</span> <span class="op">+</span></span>
<span>    <span class="va">h_state</span> <span class="op">+</span></span>
<span>    <span class="va">ha_anno</span> <span class="op">+</span></span>
<span>    <span class="va">ha_spatial</span> <span class="op">+</span></span>
<span>    <span class="va">ha_meta</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Heatmap/annotation names are duplicated: mean exprs

Warning: Heatmap/annotation names are duplicated: mean exprs

Warning: Heatmap/annotation names are duplicated: mean exprs

Warning: Heatmap/annotation names are duplicated: mean exprs</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb470"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Add customized legend for anno_barplot()</span></span>
<span><span class="va">lgd</span> <span class="op">&lt;-</span> <span class="fu">Legend</span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"indication"</span>,</span>
<span>    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">indication</span><span class="op">)</span>,</span>
<span>    legend_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">color_vectors</span><span class="op">$</span><span class="va">indication</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Plot</span></span>
<span><span class="fu">draw</span><span class="op">(</span><span class="va">h_list</span>, annotation_legend_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">lgd</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="reference" class="level2"><h2 class="anchored" data-anchor-id="reference">Reference</h2>
<ul>
<li><a href="https://bodenmillergroup.github.io/IMCDataAnalysis/">Analysis workflow for IMC data</a></li>
<li><a href="https://complexinterface.com/research/projects/image-analysis/">Working with CellProfiler data in R</a></li>
</ul></section><section id="sessioninfo" class="level2"><h2 class="anchored" data-anchor-id="sessioninfo">Sessioninfo</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb471"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Singapore
tzcode source: internal

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] gridExtra_2.3               ConsensusClusterPlus_1.66.0
 [3] kohonen_3.0.12              circlize_0.4.16            
 [5] ComplexHeatmap_2.18.0       lisaClust_1.10.1           
 [7] caret_6.0-94                lattice_0.22-6             
 [9] scran_1.30.2                bluster_1.12.0             
[11] Rphenograph_0.99.1          igraph_2.0.3               
[13] Seurat_5.0.3                SeuratObject_5.0.1         
[15] sp_2.1-3                    BiocSingular_1.18.0        
[17] harmony_1.2.0               Rcpp_1.0.12                
[19] scater_1.30.1               scuttle_1.12.0             
[21] batchelor_1.18.1            mclust_6.1                 
[23] tiff_0.1-12                 BiocParallel_1.36.0        
[25] pheatmap_1.0.12             CATALYST_1.26.1            
[27] dittoSeq_1.14.3             cytomapper_1.14.0          
[29] EBImage_4.44.0              imcRtools_1.8.0            
[31] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0
[33] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[35] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
[37] IRanges_2.36.0              S4Vectors_0.40.2           
[39] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[41] matrixStats_1.3.0           viridis_0.6.5              
[43] viridisLite_0.4.2           RColorBrewer_1.1-3         
[45] cowplot_1.1.3               patchwork_1.2.0            
[47] ggridges_0.5.6              ggrepel_0.9.5              
[49] lubridate_1.9.3             forcats_1.0.0              
[51] stringr_1.5.1               dplyr_1.1.4                
[53] purrr_1.0.2                 readr_2.1.5                
[55] tidyr_1.3.1                 tibble_3.2.1               
[57] ggplot2_3.5.1               tidyverse_2.0.0            
[59] qs_0.26.1                   fs_1.6.4                   
[61] here_1.0.1                 

loaded via a namespace (and not attached):
  [1] vroom_1.6.5                 nnet_7.3-19                
  [3] goftest_1.2-3               DT_0.33                    
  [5] HDF5Array_1.30.1            TH.data_1.1-2              
  [7] vctrs_0.6.5                 spatstat.random_3.2-3      
  [9] RApiSerialize_0.1.2         digest_0.6.35              
 [11] png_0.1-8                   shape_1.4.6.1              
 [13] proxy_0.4-27                spicyR_1.14.3              
 [15] deldir_2.0-4                parallelly_1.37.1          
 [17] magick_2.8.3                MASS_7.3-60.0.1            
 [19] reshape2_1.4.4              httpuv_1.6.15              
 [21] foreach_1.5.2               withr_3.0.0                
 [23] xfun_0.43                   ggpubr_0.6.0               
 [25] survival_3.5-8              memoise_2.0.1              
 [27] RTriangle_1.6-0.13          ggbeeswarm_0.7.2           
 [29] RProtoBufLib_2.14.1         drc_3.0-1                  
 [31] systemfonts_1.0.6           zoo_1.8-12                 
 [33] GlobalOptions_0.1.2         gtools_3.9.5               
 [35] pbapply_1.7-2               promises_1.3.0             
 [37] httr_1.4.7                  rstatix_0.7.2              
 [39] globals_0.16.3              fitdistrplus_1.1-11        
 [41] rhdf5filters_1.14.1         stringfish_0.16.0          
 [43] rhdf5_2.46.1                archive_1.1.7              
 [45] units_0.8-5                 miniUI_0.1.1.1             
 [47] generics_0.1.3              concaveman_1.1.0           
 [49] zlibbioc_1.48.2             ScaledMatrix_1.10.0        
 [51] ggraph_2.2.1                randomForest_4.7-1.1       
 [53] polyclip_1.10-6             GenomeInfoDbData_1.2.11    
 [55] SparseArray_1.2.4           fftwtools_0.9-11           
 [57] xtable_1.8-4                doParallel_1.0.17          
 [59] evaluate_0.23               S4Arrays_1.2.1             
 [61] hms_1.1.3                   irlba_2.3.5.1              
 [63] colorspace_2.1-0            ROCR_1.0-11                
 [65] reticulate_1.36.0           spatstat.data_3.0-4        
 [67] magrittr_2.0.3              lmtest_0.9-40              
 [69] later_1.3.2                 spatstat.geom_3.2-9        
 [71] future.apply_1.11.2         scattermore_1.2            
 [73] XML_3.99-0.16.1             RcppAnnoy_0.0.22           
 [75] class_7.3-22                svgPanZoom_0.3.4           
 [77] pillar_1.9.0                nlme_3.1-164               
 [79] iterators_1.0.14            compiler_4.3.2             
 [81] beachmat_2.18.1             RSpectra_0.16-1            
 [83] stringi_1.8.4               gower_1.0.1                
 [85] sf_1.0-16                   minqa_1.2.6                
 [87] ClassifyR_3.6.5             tensor_1.5                 
 [89] plyr_1.8.9                  crayon_1.5.2               
 [91] abind_1.4-5                 locfit_1.5-9.9             
 [93] graphlayouts_1.1.1          bit_4.0.5                  
 [95] terra_1.7-71                sandwich_3.1-0             
 [97] codetools_0.2-20            multcomp_1.4-25            
 [99] recipes_1.0.10              e1071_1.7-14               
[101] GetoptLong_1.0.5            plotly_4.10.4              
[103] MultiAssayExperiment_1.28.0 mime_0.12                  
[105] splines_4.3.2               fastDummies_1.7.3          
[107] sparseMatrixStats_1.14.0    knitr_1.45                 
[109] utf8_1.2.4                  clue_0.3-65                
[111] lme4_1.1-35.3               listenv_0.9.1              
[113] nnls_1.5                    DelayedMatrixStats_1.24.0  
[115] ggsignif_0.6.4              scam_1.2-16                
[117] Matrix_1.6-5                statmod_1.5.0              
[119] tzdb_0.4.0                  svglite_2.1.3              
[121] tweenr_2.0.3                pkgconfig_2.0.3            
[123] tools_4.3.2                 cachem_1.0.8               
[125] RhpcBLASctl_0.23-42         numDeriv_2016.8-1.1        
[127] DBI_1.2.2                   fastmap_1.1.1              
[129] rmarkdown_2.26              scales_1.3.0               
[131] ica_1.0-3                   shinydashboard_0.7.2       
[133] broom_1.0.5                 dotCall64_1.1-1            
[135] carData_3.0-5               rpart_4.1.23               
[137] RANN_2.6.1                  farver_2.1.1               
[139] mgcv_1.9-1                  tidygraph_1.3.1            
[141] yaml_2.3.8                  cli_3.6.2                  
[143] leiden_0.4.3.1              lifecycle_1.0.4            
[145] uwot_0.1.16                 mvtnorm_1.2-4              
[147] lava_1.8.0                  backports_1.4.1            
[149] cytolib_2.14.1              timechange_0.3.0           
[151] gtable_0.3.5                rjson_0.2.21               
[153] progressr_0.14.0            pROC_1.18.5                
[155] limma_3.58.1                parallel_4.3.2             
[157] edgeR_4.0.16                jsonlite_1.8.8             
[159] RcppHNSW_0.6.0              bitops_1.0-7               
[161] bit64_4.0.5                 Rtsne_0.17                 
[163] FlowSOM_2.10.0              spatstat.utils_3.0-4       
[165] BiocNeighbors_1.20.2        RcppParallel_5.1.7         
[167] flowCore_2.14.2             metapod_1.10.1             
[169] dqrng_0.3.2                 timeDate_4032.109          
[171] lazyeval_0.2.2              shiny_1.8.1.1              
[173] htmltools_0.5.8.1           sctransform_0.4.1          
[175] distances_0.1.10            glue_1.7.0                 
[177] spam_2.10-0                 ResidualMatrix_1.12.0      
[179] XVector_0.42.0              RCurl_1.98-1.14            
[181] rprojroot_2.0.4             classInt_0.4-10            
[183] jpeg_0.1-10                 boot_1.3-30                
[185] R6_2.5.1                    labeling_0.4.3             
[187] cluster_2.1.6               Rhdf5lib_1.24.2            
[189] ipred_0.9-14                nloptr_2.0.3               
[191] DelayedArray_0.28.0         tidyselect_1.2.1           
[193] vipor_0.4.7                 plotrix_3.8-4              
[195] ggforce_0.4.2               raster_3.6-26              
[197] car_3.1-2                   future_1.33.2              
[199] ModelMetrics_1.2.2.2        rsvd_1.0.5                 
[201] munsell_0.5.1               KernSmooth_2.23-22         
[203] data.table_1.15.4           htmlwidgets_1.6.4          
[205] rlang_1.1.3                 spatstat.sparse_3.0-3      
[207] spatstat.explore_3.2-7      lmerTest_3.1-3             
[209] colorRamps_2.3.4            Cairo_1.6-2                
[211] ggnewscale_0.4.10           fansi_1.0.6                
[213] hardhat_1.3.1               prodlim_2023.08.28         
[215] beeswarm_0.4.0             </code></pre>
</div>
</div>


<!-- -->

</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/grgrzhong\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb473" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb473-1"><a href="#cb473-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb473-2"><a href="#cb473-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Multiplex image-based phenotypic data analysis"</span></span>
<span id="cb473-3"><a href="#cb473-3" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> 2024-04-18</span></span>
<span id="cb473-4"><a href="#cb473-4" aria-hidden="true" tabindex="-1"></a><span class="an">date-modified:</span><span class="co"> last-modified</span></span>
<span id="cb473-5"><a href="#cb473-5" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb473-6"><a href="#cb473-6" aria-hidden="true" tabindex="-1"></a><span class="co">  - mIF</span></span>
<span id="cb473-7"><a href="#cb473-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-8"><a href="#cb473-8" aria-hidden="true" tabindex="-1"></a><span class="an">image:</span><span class="co"> image-profile.jpg</span></span>
<span id="cb473-9"><a href="#cb473-9" aria-hidden="true" tabindex="-1"></a><span class="co"># description: All the learning materials are from credited or adapted from Harvard Chan Bioinformatics Core. </span></span>
<span id="cb473-10"><a href="#cb473-10" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb473-11"><a href="#cb473-11" aria-hidden="true" tabindex="-1"></a><span class="co">  freeze: auto</span></span>
<span id="cb473-12"><a href="#cb473-12" aria-hidden="true" tabindex="-1"></a><span class="co">  eval: true</span></span>
<span id="cb473-13"><a href="#cb473-13" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb473-14"><a href="#cb473-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-15"><a href="#cb473-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-16"><a href="#cb473-16" aria-hidden="true" tabindex="-1"></a><span class="fu">## Initial general setup</span></span>
<span id="cb473-19"><a href="#cb473-19" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-20"><a href="#cb473-20" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb473-21"><a href="#cb473-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-22"><a href="#cb473-22" aria-hidden="true" tabindex="-1"></a><span class="do">### Required libraries</span></span>
<span id="cb473-23"><a href="#cb473-23" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!require("BiocManager", quietly = TRUE))</span></span>
<span id="cb473-24"><a href="#cb473-24" aria-hidden="true" tabindex="-1"></a><span class="co">#     install.packages("BiocManager")</span></span>
<span id="cb473-25"><a href="#cb473-25" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install(</span></span>
<span id="cb473-26"><a href="#cb473-26" aria-hidden="true" tabindex="-1"></a><span class="co">#     c(</span></span>
<span id="cb473-27"><a href="#cb473-27" aria-hidden="true" tabindex="-1"></a><span class="co">#         "rmarkdown", "bookdown", "pheatmap", "viridis", "zoo",</span></span>
<span id="cb473-28"><a href="#cb473-28" aria-hidden="true" tabindex="-1"></a><span class="co">#         "devtools", "testthat", "tiff", "distill", "ggrepel",</span></span>
<span id="cb473-29"><a href="#cb473-29" aria-hidden="true" tabindex="-1"></a><span class="co">#         "patchwork", "mclust", "RColorBrewer", "uwot", "Rtsne",</span></span>
<span id="cb473-30"><a href="#cb473-30" aria-hidden="true" tabindex="-1"></a><span class="co">#         "harmony", "Seurat", "SeuratObject", "cowplot", "kohonen",</span></span>
<span id="cb473-31"><a href="#cb473-31" aria-hidden="true" tabindex="-1"></a><span class="co">#         "caret", "randomForest", "ggridges", "cowplot",</span></span>
<span id="cb473-32"><a href="#cb473-32" aria-hidden="true" tabindex="-1"></a><span class="co">#         "gridGraphics", "scales", "tiff", "harmony", "Matrix",</span></span>
<span id="cb473-33"><a href="#cb473-33" aria-hidden="true" tabindex="-1"></a><span class="co">#         "CATALYST", "scuttle", "scater", "dittoSeq",</span></span>
<span id="cb473-34"><a href="#cb473-34" aria-hidden="true" tabindex="-1"></a><span class="co">#         "tidyverse", "BiocStyle", "batchelor", "bluster", "scran",</span></span>
<span id="cb473-35"><a href="#cb473-35" aria-hidden="true" tabindex="-1"></a><span class="co">#         "lisaClust", "spicyR", "iSEE", "imcRtools", "cytomapper",</span></span>
<span id="cb473-36"><a href="#cb473-36" aria-hidden="true" tabindex="-1"></a><span class="co">#         "imcdatasets", "cytoviewer"</span></span>
<span id="cb473-37"><a href="#cb473-37" aria-hidden="true" tabindex="-1"></a><span class="co">#     ),</span></span>
<span id="cb473-38"><a href="#cb473-38" aria-hidden="true" tabindex="-1"></a><span class="co">#     force = TRUE</span></span>
<span id="cb473-39"><a href="#cb473-39" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb473-40"><a href="#cb473-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-41"><a href="#cb473-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb473-42"><a href="#cb473-42" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fs)</span>
<span id="cb473-43"><a href="#cb473-43" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qs)</span>
<span id="cb473-44"><a href="#cb473-44" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb473-45"><a href="#cb473-45" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb473-46"><a href="#cb473-46" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb473-47"><a href="#cb473-47" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb473-48"><a href="#cb473-48" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb473-49"><a href="#cb473-49" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb473-50"><a href="#cb473-50" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb473-51"><a href="#cb473-51" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(imcRtools)</span>
<span id="cb473-52"><a href="#cb473-52" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cytomapper)</span>
<span id="cb473-53"><a href="#cb473-53" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dittoSeq)</span>
<span id="cb473-54"><a href="#cb473-54" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(CATALYST)</span>
<span id="cb473-55"><a href="#cb473-55" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb473-56"><a href="#cb473-56" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb473-57"><a href="#cb473-57" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tiff)</span>
<span id="cb473-58"><a href="#cb473-58" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EBImage)</span>
<span id="cb473-59"><a href="#cb473-59" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mclust)</span>
<span id="cb473-60"><a href="#cb473-60" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(batchelor)</span>
<span id="cb473-61"><a href="#cb473-61" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb473-62"><a href="#cb473-62" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb473-63"><a href="#cb473-63" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(harmony)</span>
<span id="cb473-64"><a href="#cb473-64" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocSingular)</span>
<span id="cb473-65"><a href="#cb473-65" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb473-66"><a href="#cb473-66" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratObject)</span>
<span id="cb473-67"><a href="#cb473-67" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Rphenograph)</span>
<span id="cb473-68"><a href="#cb473-68" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(igraph)</span>
<span id="cb473-69"><a href="#cb473-69" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bluster)</span>
<span id="cb473-70"><a href="#cb473-70" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb473-71"><a href="#cb473-71" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caret)</span>
<span id="cb473-72"><a href="#cb473-72" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lisaClust)</span>
<span id="cb473-73"><a href="#cb473-73" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scuttle)</span>
<span id="cb473-74"><a href="#cb473-74" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb473-75"><a href="#cb473-75" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb473-76"><a href="#cb473-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-77"><a href="#cb473-77" aria-hidden="true" tabindex="-1"></a><span class="do">### Project directory</span></span>
<span id="cb473-78"><a href="#cb473-78" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"projects/2024_IMC_Profile"</span>)</span>
<span id="cb473-79"><a href="#cb473-79" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">20240419</span>)</span>
<span id="cb473-80"><a href="#cb473-80" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-81"><a href="#cb473-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-82"><a href="#cb473-82" aria-hidden="true" tabindex="-1"></a><span class="fu">## Data preparation</span></span>
<span id="cb473-83"><a href="#cb473-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-84"><a href="#cb473-84" aria-hidden="true" tabindex="-1"></a>IMC example data is from <span class="co">[</span><span class="ot">here</span><span class="co">](https://zenodo.org/records/7624451)</span></span>
<span id="cb473-85"><a href="#cb473-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-86"><a href="#cb473-86" aria-hidden="true" tabindex="-1"></a><span class="fu">### Download example data</span></span>
<span id="cb473-89"><a href="#cb473-89" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-90"><a href="#cb473-90" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-91"><a href="#cb473-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-92"><a href="#cb473-92" aria-hidden="true" tabindex="-1"></a><span class="do">### Create directory for data </span></span>
<span id="cb473-93"><a href="#cb473-93" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(dir, <span class="st">"data/steinbock"</span>)</span>
<span id="cb473-94"><a href="#cb473-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-95"><a href="#cb473-95" aria-hidden="true" tabindex="-1"></a><span class="do">### Download sample/patient metadata information</span></span>
<span id="cb473-96"><a href="#cb473-96" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-97"><a href="#cb473-97" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/7575859/files/sample_metadata.csv"</span>,</span>
<span id="cb473-98"><a href="#cb473-98" aria-hidden="true" tabindex="-1"></a>    <span class="at">destfile =</span> <span class="fu">here</span>(dir, <span class="st">"data/sample_metadata.csv"</span>)</span>
<span id="cb473-99"><a href="#cb473-99" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-100"><a href="#cb473-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-101"><a href="#cb473-101" aria-hidden="true" tabindex="-1"></a><span class="do">### Download intensities</span></span>
<span id="cb473-102"><a href="#cb473-102" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/intensities.zip"</span></span>
<span id="cb473-103"><a href="#cb473-103" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock/intensities.zip"</span>)</span>
<span id="cb473-104"><a href="#cb473-104" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, destfile)</span>
<span id="cb473-105"><a href="#cb473-105" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(destfile, <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-106"><a href="#cb473-106" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(destfile)</span>
<span id="cb473-107"><a href="#cb473-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-108"><a href="#cb473-108" aria-hidden="true" tabindex="-1"></a><span class="do">### Download regionprops</span></span>
<span id="cb473-109"><a href="#cb473-109" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/regionprops.zip"</span></span>
<span id="cb473-110"><a href="#cb473-110" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock/regionprops.zip"</span>)</span>
<span id="cb473-111"><a href="#cb473-111" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, destfile)</span>
<span id="cb473-112"><a href="#cb473-112" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(destfile, <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-113"><a href="#cb473-113" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(destfile)</span>
<span id="cb473-114"><a href="#cb473-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-115"><a href="#cb473-115" aria-hidden="true" tabindex="-1"></a><span class="do">### Download neighbors</span></span>
<span id="cb473-116"><a href="#cb473-116" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/neighbors.zip"</span></span>
<span id="cb473-117"><a href="#cb473-117" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock/neighbors.zip"</span>)</span>
<span id="cb473-118"><a href="#cb473-118" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, destfile)</span>
<span id="cb473-119"><a href="#cb473-119" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(destfile, <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-120"><a href="#cb473-120" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(destfile)</span>
<span id="cb473-121"><a href="#cb473-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-122"><a href="#cb473-122" aria-hidden="true" tabindex="-1"></a><span class="do">### Download images</span></span>
<span id="cb473-123"><a href="#cb473-123" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/img.zip"</span></span>
<span id="cb473-124"><a href="#cb473-124" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock/img.zip"</span>)</span>
<span id="cb473-125"><a href="#cb473-125" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, destfile)</span>
<span id="cb473-126"><a href="#cb473-126" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(destfile, <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-127"><a href="#cb473-127" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(destfile)</span>
<span id="cb473-128"><a href="#cb473-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-129"><a href="#cb473-129" aria-hidden="true" tabindex="-1"></a><span class="do">### Download masks</span></span>
<span id="cb473-130"><a href="#cb473-130" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://zenodo.org/record/7624451/files/masks_deepcell.zip"</span></span>
<span id="cb473-131"><a href="#cb473-131" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock/masks_deepcell.zip"</span>)</span>
<span id="cb473-132"><a href="#cb473-132" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, destfile)</span>
<span id="cb473-133"><a href="#cb473-133" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(destfile, <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data/steinbock"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-134"><a href="#cb473-134" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(destfile)</span>
<span id="cb473-135"><a href="#cb473-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-136"><a href="#cb473-136" aria-hidden="true" tabindex="-1"></a><span class="do">### Download individual files</span></span>
<span id="cb473-137"><a href="#cb473-137" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-138"><a href="#cb473-138" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/7624451/files/panel.csv"</span>,</span>
<span id="cb473-139"><a href="#cb473-139" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/steinbock/panel.csv"</span>)</span>
<span id="cb473-140"><a href="#cb473-140" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-141"><a href="#cb473-141" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-142"><a href="#cb473-142" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/7624451/files/images.csv"</span>,</span>
<span id="cb473-143"><a href="#cb473-143" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/steinbock/images.csv"</span>)</span>
<span id="cb473-144"><a href="#cb473-144" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-145"><a href="#cb473-145" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-146"><a href="#cb473-146" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/7624451/files/steinbock.sh"</span>,</span>
<span id="cb473-147"><a href="#cb473-147" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/steinbock/steinbock.sh"</span>)</span>
<span id="cb473-148"><a href="#cb473-148" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-149"><a href="#cb473-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-150"><a href="#cb473-150" aria-hidden="true" tabindex="-1"></a><span class="do">### Files for spillover matrix estimation</span></span>
<span id="cb473-151"><a href="#cb473-151" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-152"><a href="#cb473-152" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/7575859/files/compensation.zip"</span>,</span>
<span id="cb473-153"><a href="#cb473-153" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/compensation.zip"</span>)</span>
<span id="cb473-154"><a href="#cb473-154" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-155"><a href="#cb473-155" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(</span>
<span id="cb473-156"><a href="#cb473-156" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/compensation.zip"</span>),</span>
<span id="cb473-157"><a href="#cb473-157" aria-hidden="true" tabindex="-1"></a>    <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span></span>
<span id="cb473-158"><a href="#cb473-158" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-159"><a href="#cb473-159" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="fu">here</span>(dir, <span class="st">"data/compensation.zip"</span>))</span>
<span id="cb473-160"><a href="#cb473-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-161"><a href="#cb473-161" aria-hidden="true" tabindex="-1"></a><span class="do">### Gated cells</span></span>
<span id="cb473-162"><a href="#cb473-162" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb473-163"><a href="#cb473-163" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/8095133/files/gated_cells.zip"</span>,</span>
<span id="cb473-164"><a href="#cb473-164" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/gated_cells.zip"</span>)</span>
<span id="cb473-165"><a href="#cb473-165" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-166"><a href="#cb473-166" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span>(</span>
<span id="cb473-167"><a href="#cb473-167" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/gated_cells.zip"</span>),</span>
<span id="cb473-168"><a href="#cb473-168" aria-hidden="true" tabindex="-1"></a>    <span class="at">exdir =</span> <span class="fu">here</span>(dir, <span class="st">"data"</span>), <span class="at">overwrite =</span> <span class="cn">TRUE</span></span>
<span id="cb473-169"><a href="#cb473-169" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-170"><a href="#cb473-170" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="fu">here</span>(dir, <span class="st">"data/gated_cells.zip"</span>))</span>
<span id="cb473-171"><a href="#cb473-171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-172"><a href="#cb473-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-173"><a href="#cb473-173" aria-hidden="true" tabindex="-1"></a><span class="fu">### Preprocess data</span></span>
<span id="cb473-174"><a href="#cb473-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-177"><a href="#cb473-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-178"><a href="#cb473-178" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb473-179"><a href="#cb473-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-180"><a href="#cb473-180" aria-hidden="true" tabindex="-1"></a><span class="do">### Read steinbock generated data into a SpatialExperiment object</span></span>
<span id="cb473-181"><a href="#cb473-181" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> imcRtools<span class="sc">::</span><span class="fu">read_steinbock</span>(<span class="fu">here</span>(dir, <span class="st">"data/steinbock/"</span>))</span>
<span id="cb473-182"><a href="#cb473-182" aria-hidden="true" tabindex="-1"></a>spe</span>
<span id="cb473-183"><a href="#cb473-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-184"><a href="#cb473-184" aria-hidden="true" tabindex="-1"></a><span class="do">### Summarized pixel intensities</span></span>
<span id="cb473-185"><a href="#cb473-185" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(spe)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb473-186"><a href="#cb473-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-187"><a href="#cb473-187" aria-hidden="true" tabindex="-1"></a><span class="do">### Metadata</span></span>
<span id="cb473-188"><a href="#cb473-188" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(spe))</span>
<span id="cb473-189"><a href="#cb473-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-190"><a href="#cb473-190" aria-hidden="true" tabindex="-1"></a><span class="do">### SpatialExperiment container stores locations of all cells in the</span></span>
<span id="cb473-191"><a href="#cb473-191" aria-hidden="true" tabindex="-1"></a><span class="do">### spatialCoords slot</span></span>
<span id="cb473-192"><a href="#cb473-192" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">spatialCoords</span>(spe))</span>
<span id="cb473-193"><a href="#cb473-193" aria-hidden="true" tabindex="-1"></a><span class="fu">colPair</span>(spe, <span class="st">"neighborhood"</span>)</span>
<span id="cb473-194"><a href="#cb473-194" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(spe))</span>
<span id="cb473-195"><a href="#cb473-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-196"><a href="#cb473-196" aria-hidden="true" tabindex="-1"></a><span class="do">### Add additional metadata: generate unique identifiers per cell</span></span>
<span id="cb473-197"><a href="#cb473-197" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(spe<span class="sc">$</span>sample_id, <span class="st">"_"</span>, spe<span class="sc">$</span>ObjectNumber)</span>
<span id="cb473-198"><a href="#cb473-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-199"><a href="#cb473-199" aria-hidden="true" tabindex="-1"></a><span class="do">### Read patient metadata</span></span>
<span id="cb473-200"><a href="#cb473-200" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="fu">here</span>(dir, <span class="st">"data/sample_metadata.csv"</span>), <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb473-201"><a href="#cb473-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-202"><a href="#cb473-202" aria-hidden="true" tabindex="-1"></a><span class="do">### Extract patient id and ROI id from sample name</span></span>
<span id="cb473-203"><a href="#cb473-203" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>patient_id <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(spe<span class="sc">$</span>sample_id, <span class="st">"Patient[1-4]"</span>)</span>
<span id="cb473-204"><a href="#cb473-204" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>ROI <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(spe<span class="sc">$</span>sample_id, <span class="st">"00[1-8]"</span>)</span>
<span id="cb473-205"><a href="#cb473-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-206"><a href="#cb473-206" aria-hidden="true" tabindex="-1"></a><span class="do">### Store cancer type in SPE object</span></span>
<span id="cb473-207"><a href="#cb473-207" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>indication <span class="ot">&lt;-</span> meta<span class="sc">$</span>Indication[<span class="fu">match</span>(spe<span class="sc">$</span>patient_id, meta<span class="sc">$</span><span class="st">`</span><span class="at">Sample ID</span><span class="st">`</span>)]</span>
<span id="cb473-208"><a href="#cb473-208" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(spe<span class="sc">$</span>patient_id)</span>
<span id="cb473-209"><a href="#cb473-209" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(spe<span class="sc">$</span>ROI)</span>
<span id="cb473-210"><a href="#cb473-210" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(spe<span class="sc">$</span>indication)</span>
<span id="cb473-211"><a href="#cb473-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-212"><a href="#cb473-212" aria-hidden="true" tabindex="-1"></a><span class="do">### Transform counts</span></span>
<span id="cb473-213"><a href="#cb473-213" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoRidgePlot</span>(</span>
<span id="cb473-214"><a href="#cb473-214" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"CD3"</span>, <span class="at">group.by =</span> <span class="st">"patient_id"</span>, <span class="at">assay =</span> <span class="st">"counts"</span></span>
<span id="cb473-215"><a href="#cb473-215" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-216"><a href="#cb473-216" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"CD3 - before transformation"</span>)</span>
<span id="cb473-217"><a href="#cb473-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-218"><a href="#cb473-218" aria-hidden="true" tabindex="-1"></a><span class="do">### Perform counts transformation using an inverse hyperbolic sine function</span></span>
<span id="cb473-219"><a href="#cb473-219" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(spe, <span class="st">"exprs"</span>) <span class="ot">&lt;-</span> <span class="fu">asinh</span>(<span class="fu">counts</span>(spe) <span class="sc">/</span> <span class="dv">1</span>)</span>
<span id="cb473-220"><a href="#cb473-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-221"><a href="#cb473-221" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoRidgePlot</span>(</span>
<span id="cb473-222"><a href="#cb473-222" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"CD3"</span>, <span class="at">group.by =</span> <span class="st">"patient_id"</span>, <span class="at">assay =</span> <span class="st">"exprs"</span></span>
<span id="cb473-223"><a href="#cb473-223" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-224"><a href="#cb473-224" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"CD3 - after transformation"</span>)</span>
<span id="cb473-225"><a href="#cb473-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-226"><a href="#cb473-226" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb473-227"><a href="#cb473-227" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">&amp;</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb473-228"><a href="#cb473-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-229"><a href="#cb473-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-232"><a href="#cb473-232" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-233"><a href="#cb473-233" aria-hidden="true" tabindex="-1"></a><span class="do">### Add Feature meta for easy specifies the markers of interest.</span></span>
<span id="cb473-234"><a href="#cb473-234" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"DNA|Histone"</span>, <span class="fu">rownames</span>(spe))</span>
<span id="cb473-235"><a href="#cb473-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-236"><a href="#cb473-236" aria-hidden="true" tabindex="-1"></a><span class="do">### Define color schemes for different metadata entries of the data</span></span>
<span id="cb473-237"><a href="#cb473-237" aria-hidden="true" tabindex="-1"></a>color_vectors <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb473-238"><a href="#cb473-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-239"><a href="#cb473-239" aria-hidden="true" tabindex="-1"></a>ROI <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb473-240"><a href="#cb473-240" aria-hidden="true" tabindex="-1"></a>    <span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>ROI)), <span class="at">name =</span> <span class="st">"BrBG"</span>),</span>
<span id="cb473-241"><a href="#cb473-241" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>(spe<span class="sc">$</span>ROI)</span>
<span id="cb473-242"><a href="#cb473-242" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-243"><a href="#cb473-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-244"><a href="#cb473-244" aria-hidden="true" tabindex="-1"></a>patient_id <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb473-245"><a href="#cb473-245" aria-hidden="true" tabindex="-1"></a>    <span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>patient_id)), <span class="at">name =</span> <span class="st">"Set1"</span>),</span>
<span id="cb473-246"><a href="#cb473-246" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>(spe<span class="sc">$</span>patient_id)</span>
<span id="cb473-247"><a href="#cb473-247" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-248"><a href="#cb473-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-249"><a href="#cb473-249" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb473-250"><a href="#cb473-250" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(</span>
<span id="cb473-251"><a href="#cb473-251" aria-hidden="true" tabindex="-1"></a>        <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"YlOrRd"</span>)[<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>],</span>
<span id="cb473-252"><a href="#cb473-252" aria-hidden="true" tabindex="-1"></a>        <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"PuBu"</span>)[<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>],</span>
<span id="cb473-253"><a href="#cb473-253" aria-hidden="true" tabindex="-1"></a>        <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"YlGn"</span>)[<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>],</span>
<span id="cb473-254"><a href="#cb473-254" aria-hidden="true" tabindex="-1"></a>        <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"BuPu"</span>)[<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb473-255"><a href="#cb473-255" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb473-256"><a href="#cb473-256" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>(spe<span class="sc">$</span>sample_id)</span>
<span id="cb473-257"><a href="#cb473-257" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-258"><a href="#cb473-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-259"><a href="#cb473-259" aria-hidden="true" tabindex="-1"></a>indication <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb473-260"><a href="#cb473-260" aria-hidden="true" tabindex="-1"></a>    <span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>indication)), <span class="at">name =</span> <span class="st">"Set2"</span>),</span>
<span id="cb473-261"><a href="#cb473-261" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>(spe<span class="sc">$</span>indication)</span>
<span id="cb473-262"><a href="#cb473-262" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-263"><a href="#cb473-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-264"><a href="#cb473-264" aria-hidden="true" tabindex="-1"></a>color_vectors<span class="sc">$</span>ROI <span class="ot">&lt;-</span> ROI</span>
<span id="cb473-265"><a href="#cb473-265" aria-hidden="true" tabindex="-1"></a>color_vectors<span class="sc">$</span>patient_id <span class="ot">&lt;-</span> patient_id</span>
<span id="cb473-266"><a href="#cb473-266" aria-hidden="true" tabindex="-1"></a>color_vectors<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> sample_id</span>
<span id="cb473-267"><a href="#cb473-267" aria-hidden="true" tabindex="-1"></a>color_vectors<span class="sc">$</span>indication <span class="ot">&lt;-</span> indication</span>
<span id="cb473-268"><a href="#cb473-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-269"><a href="#cb473-269" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors <span class="ot">&lt;-</span> color_vectors</span>
<span id="cb473-270"><a href="#cb473-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-271"><a href="#cb473-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-272"><a href="#cb473-272" aria-hidden="true" tabindex="-1"></a><span class="fu">### Read in images</span></span>
<span id="cb473-275"><a href="#cb473-275" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-276"><a href="#cb473-276" aria-hidden="true" tabindex="-1"></a>images <span class="ot">&lt;-</span> <span class="fu">loadImages</span>(<span class="fu">here</span>(dir, <span class="st">"data/steinbock/img/"</span>))</span>
<span id="cb473-277"><a href="#cb473-277" aria-hidden="true" tabindex="-1"></a>masks <span class="ot">&lt;-</span> <span class="fu">loadImages</span>(<span class="fu">here</span>(dir, <span class="st">"data/steinbock/masks_deepcell/"</span>), <span class="at">as.is =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-278"><a href="#cb473-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-279"><a href="#cb473-279" aria-hidden="true" tabindex="-1"></a><span class="do">### Make sure that the channel order is identical between the single-cell data and the images</span></span>
<span id="cb473-280"><a href="#cb473-280" aria-hidden="true" tabindex="-1"></a><span class="fu">channelNames</span>(images) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)</span>
<span id="cb473-281"><a href="#cb473-281" aria-hidden="true" tabindex="-1"></a>images</span>
<span id="cb473-282"><a href="#cb473-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-283"><a href="#cb473-283" aria-hidden="true" tabindex="-1"></a><span class="do">### Order of the images</span></span>
<span id="cb473-284"><a href="#cb473-284" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">names</span>(images), <span class="fu">names</span>(masks))</span>
<span id="cb473-285"><a href="#cb473-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-286"><a href="#cb473-286" aria-hidden="true" tabindex="-1"></a><span class="do">### Extract patient id from image name</span></span>
<span id="cb473-287"><a href="#cb473-287" aria-hidden="true" tabindex="-1"></a>patient_id <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(<span class="fu">names</span>(images), <span class="st">"Patient[1-4]"</span>)</span>
<span id="cb473-288"><a href="#cb473-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-289"><a href="#cb473-289" aria-hidden="true" tabindex="-1"></a><span class="do">### Retrieve cancer type per patient from metadata file</span></span>
<span id="cb473-290"><a href="#cb473-290" aria-hidden="true" tabindex="-1"></a>indication <span class="ot">&lt;-</span> meta<span class="sc">$</span>Indication[<span class="fu">match</span>(patient_id, meta<span class="sc">$</span><span class="st">`</span><span class="at">Sample ID</span><span class="st">`</span>)]</span>
<span id="cb473-291"><a href="#cb473-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-292"><a href="#cb473-292" aria-hidden="true" tabindex="-1"></a><span class="do">### Store patient and image level information in elementMetadata</span></span>
<span id="cb473-293"><a href="#cb473-293" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(images) <span class="ot">&lt;-</span> <span class="fu">mcols</span>(masks) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(</span>
<span id="cb473-294"><a href="#cb473-294" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_id =</span> <span class="fu">names</span>(images),</span>
<span id="cb473-295"><a href="#cb473-295" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient_id =</span> patient_id,</span>
<span id="cb473-296"><a href="#cb473-296" aria-hidden="true" tabindex="-1"></a>    <span class="at">indication =</span> indication</span>
<span id="cb473-297"><a href="#cb473-297" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-298"><a href="#cb473-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-299"><a href="#cb473-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-300"><a href="#cb473-300" aria-hidden="true" tabindex="-1"></a><span class="fu">### Save data</span></span>
<span id="cb473-301"><a href="#cb473-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-304"><a href="#cb473-304" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-305"><a href="#cb473-305" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-306"><a href="#cb473-306" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-307"><a href="#cb473-307" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(images, <span class="fu">here</span>(dir, <span class="st">"data/images.qs"</span>))</span>
<span id="cb473-308"><a href="#cb473-308" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(masks, <span class="fu">here</span>(dir, <span class="st">"data/masks.qs"</span>))</span>
<span id="cb473-309"><a href="#cb473-309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-310"><a href="#cb473-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-311"><a href="#cb473-311" aria-hidden="true" tabindex="-1"></a><span class="fu">## Spillover correction</span></span>
<span id="cb473-312"><a href="#cb473-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-313"><a href="#cb473-313" aria-hidden="true" tabindex="-1"></a><span class="fu">### Generate the spillover matrix</span></span>
<span id="cb473-316"><a href="#cb473-316" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-317"><a href="#cb473-317" aria-hidden="true" tabindex="-1"></a><span class="do">### Create SingleCellExperiment from TXT files</span></span>
<span id="cb473-318"><a href="#cb473-318" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">readSCEfromTXT</span>(<span class="fu">here</span>(dir, <span class="st">"data/compensation/"</span>))</span>
<span id="cb473-319"><a href="#cb473-319" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(sce, <span class="st">"exprs"</span>) <span class="ot">&lt;-</span> <span class="fu">asinh</span>(<span class="fu">counts</span>(sce)<span class="sc">/</span><span class="dv">5</span>)</span>
<span id="cb473-320"><a href="#cb473-320" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-321"><a href="#cb473-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-322"><a href="#cb473-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-325"><a href="#cb473-325" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-326"><a href="#cb473-326" aria-hidden="true" tabindex="-1"></a><span class="do">### Quality control</span></span>
<span id="cb473-327"><a href="#cb473-327" aria-hidden="true" tabindex="-1"></a><span class="do">### Log10 median pixel counts per spot and channel</span></span>
<span id="cb473-328"><a href="#cb473-328" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpotHeatmap</span>(sce)</span>
<span id="cb473-329"><a href="#cb473-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-330"><a href="#cb473-330" aria-hidden="true" tabindex="-1"></a><span class="do">### Thresholded on 200 pixel counts</span></span>
<span id="cb473-331"><a href="#cb473-331" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpotHeatmap</span>(sce, <span class="at">log =</span> <span class="cn">FALSE</span>, <span class="at">threshold =</span> <span class="dv">200</span>)</span>
<span id="cb473-332"><a href="#cb473-332" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-333"><a href="#cb473-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-334"><a href="#cb473-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-337"><a href="#cb473-337" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-338"><a href="#cb473-338" aria-hidden="true" tabindex="-1"></a><span class="do">### Optional pixel binning</span></span>
<span id="cb473-339"><a href="#cb473-339" aria-hidden="true" tabindex="-1"></a><span class="do">### Define grouping</span></span>
<span id="cb473-340"><a href="#cb473-340" aria-hidden="true" tabindex="-1"></a>bin_size <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb473-341"><a href="#cb473-341" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> <span class="fu">binAcrossPixels</span>(sce, <span class="at">bin_size =</span> bin_size)</span>
<span id="cb473-342"><a href="#cb473-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-343"><a href="#cb473-343" aria-hidden="true" tabindex="-1"></a><span class="do">### Log10 median pixel counts per spot and channel</span></span>
<span id="cb473-344"><a href="#cb473-344" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpotHeatmap</span>(sce2)</span>
<span id="cb473-345"><a href="#cb473-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-346"><a href="#cb473-346" aria-hidden="true" tabindex="-1"></a><span class="do">### Thresholded on 200 pixel counts</span></span>
<span id="cb473-347"><a href="#cb473-347" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpotHeatmap</span>(sce2, <span class="at">log =</span> <span class="cn">FALSE</span>, <span class="at">threshold =</span> <span class="dv">200</span>)</span>
<span id="cb473-348"><a href="#cb473-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-349"><a href="#cb473-349" aria-hidden="true" tabindex="-1"></a><span class="do">### Filtering incorrectly assigned pixels</span></span>
<span id="cb473-350"><a href="#cb473-350" aria-hidden="true" tabindex="-1"></a>bc_key <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">unique</span>(sce<span class="sc">$</span>sample_mass))</span>
<span id="cb473-351"><a href="#cb473-351" aria-hidden="true" tabindex="-1"></a>bc_key <span class="ot">&lt;-</span> bc_key[<span class="fu">order</span>(bc_key)]</span>
<span id="cb473-352"><a href="#cb473-352" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">assignPrelim</span>(sce, <span class="at">bc_key =</span> bc_key)</span>
<span id="cb473-353"><a href="#cb473-353" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">estCutoffs</span>(sce)</span>
<span id="cb473-354"><a href="#cb473-354" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">applyCutoffs</span>(sce)</span>
<span id="cb473-355"><a href="#cb473-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-356"><a href="#cb473-356" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the correctly and incorrectly assigned pixels</span></span>
<span id="cb473-357"><a href="#cb473-357" aria-hidden="true" tabindex="-1"></a>cur_table <span class="ot">&lt;-</span> <span class="fu">table</span>(sce<span class="sc">$</span>bc_id, sce<span class="sc">$</span>sample_mass)</span>
<span id="cb473-358"><a href="#cb473-358" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(</span>
<span id="cb473-359"><a href="#cb473-359" aria-hidden="true" tabindex="-1"></a>    <span class="fu">log10</span>(cur_table <span class="sc">+</span> <span class="dv">1</span>),</span>
<span id="cb473-360"><a href="#cb473-360" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb473-361"><a href="#cb473-361" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">FALSE</span></span>
<span id="cb473-362"><a href="#cb473-362" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-363"><a href="#cb473-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-364"><a href="#cb473-364" aria-hidden="true" tabindex="-1"></a><span class="do">### Compute the fraction of unassigned pixels per spot</span></span>
<span id="cb473-365"><a href="#cb473-365" aria-hidden="true" tabindex="-1"></a>cur_table[<span class="st">"0"</span>, ] <span class="sc">/</span> <span class="fu">colSums</span>(cur_table)</span>
<span id="cb473-366"><a href="#cb473-366" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">filterPixels</span>(sce, <span class="at">minevents =</span> <span class="dv">40</span>, <span class="at">correct_pixels =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-367"><a href="#cb473-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-368"><a href="#cb473-368" aria-hidden="true" tabindex="-1"></a><span class="do">### Compute spillover matrix</span></span>
<span id="cb473-369"><a href="#cb473-369" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">computeSpillmat</span>(sce)</span>
<span id="cb473-370"><a href="#cb473-370" aria-hidden="true" tabindex="-1"></a>isotope_list <span class="ot">&lt;-</span> CATALYST<span class="sc">::</span>isotope_list</span>
<span id="cb473-371"><a href="#cb473-371" aria-hidden="true" tabindex="-1"></a>isotope_list<span class="sc">$</span>Ar <span class="ot">&lt;-</span> <span class="dv">80</span></span>
<span id="cb473-372"><a href="#cb473-372" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSpillmat</span>(sce, <span class="at">isotope_list =</span> isotope_list)</span>
<span id="cb473-373"><a href="#cb473-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-374"><a href="#cb473-374" aria-hidden="true" tabindex="-1"></a><span class="do">### Save spillover matrix in variable</span></span>
<span id="cb473-375"><a href="#cb473-375" aria-hidden="true" tabindex="-1"></a>sm <span class="ot">&lt;-</span> <span class="fu">metadata</span>(sce)<span class="sc">$</span>spillover_matrix</span>
<span id="cb473-376"><a href="#cb473-376" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sm, <span class="fu">here</span>(dir, <span class="st">"data/sm.csv"</span>))</span>
<span id="cb473-377"><a href="#cb473-377" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-378"><a href="#cb473-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-379"><a href="#cb473-379" aria-hidden="true" tabindex="-1"></a><span class="fu">### Single-cell data compensation</span></span>
<span id="cb473-380"><a href="#cb473-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-383"><a href="#cb473-383" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-384"><a href="#cb473-384" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-385"><a href="#cb473-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-386"><a href="#cb473-386" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>channel_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rowData</span>(spe)<span class="sc">$</span>channel, <span class="st">"Di"</span>)</span>
<span id="cb473-387"><a href="#cb473-387" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">compCytof</span>(</span>
<span id="cb473-388"><a href="#cb473-388" aria-hidden="true" tabindex="-1"></a>    spe, sm,</span>
<span id="cb473-389"><a href="#cb473-389" aria-hidden="true" tabindex="-1"></a>    <span class="at">transform =</span> <span class="cn">TRUE</span>, <span class="at">cofactor =</span> <span class="dv">1</span>,</span>
<span id="cb473-390"><a href="#cb473-390" aria-hidden="true" tabindex="-1"></a>    <span class="at">isotope_list =</span> isotope_list,</span>
<span id="cb473-391"><a href="#cb473-391" aria-hidden="true" tabindex="-1"></a>    <span class="at">overwrite =</span> <span class="cn">FALSE</span></span>
<span id="cb473-392"><a href="#cb473-392" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-393"><a href="#cb473-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-394"><a href="#cb473-394" aria-hidden="true" tabindex="-1"></a><span class="do">### Check the effect of channel spillover compensation</span></span>
<span id="cb473-395"><a href="#cb473-395" aria-hidden="true" tabindex="-1"></a>before <span class="ot">&lt;-</span> <span class="fu">dittoScatterPlot</span>(</span>
<span id="cb473-396"><a href="#cb473-396" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">x.var =</span> <span class="st">"Ecad"</span>, <span class="at">y.var =</span> <span class="st">"CD303"</span>,</span>
<span id="cb473-397"><a href="#cb473-397" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.x =</span> <span class="st">"exprs"</span>, <span class="at">assay.y =</span> <span class="st">"exprs"</span></span>
<span id="cb473-398"><a href="#cb473-398" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-399"><a href="#cb473-399" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Before compensation"</span>)</span>
<span id="cb473-400"><a href="#cb473-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-401"><a href="#cb473-401" aria-hidden="true" tabindex="-1"></a>after <span class="ot">&lt;-</span> <span class="fu">dittoScatterPlot</span>(</span>
<span id="cb473-402"><a href="#cb473-402" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">x.var =</span> <span class="st">"Ecad"</span>, <span class="at">y.var =</span> <span class="st">"CD303"</span>,</span>
<span id="cb473-403"><a href="#cb473-403" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.x =</span> <span class="st">"compexprs"</span>, <span class="at">assay.y =</span> <span class="st">"compexprs"</span></span>
<span id="cb473-404"><a href="#cb473-404" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-405"><a href="#cb473-405" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"After compensation"</span>)</span>
<span id="cb473-406"><a href="#cb473-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-407"><a href="#cb473-407" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(before, after)</span>
<span id="cb473-408"><a href="#cb473-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-409"><a href="#cb473-409" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(spe, <span class="st">"counts"</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(spe, <span class="st">"compcounts"</span>) </span>
<span id="cb473-410"><a href="#cb473-410" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(spe, <span class="st">"exprs"</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(spe, <span class="st">"compexprs"</span>) </span>
<span id="cb473-411"><a href="#cb473-411" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(spe, <span class="st">"compcounts"</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(spe, <span class="st">"compexprs"</span>) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb473-412"><a href="#cb473-412" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-413"><a href="#cb473-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-414"><a href="#cb473-414" aria-hidden="true" tabindex="-1"></a><span class="fu">### Image compensation</span></span>
<span id="cb473-415"><a href="#cb473-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-418"><a href="#cb473-418" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-419"><a href="#cb473-419" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-420"><a href="#cb473-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-421"><a href="#cb473-421" aria-hidden="true" tabindex="-1"></a>images <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/images.qs"</span>))</span>
<span id="cb473-422"><a href="#cb473-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-423"><a href="#cb473-423" aria-hidden="true" tabindex="-1"></a><span class="fu">channelNames</span>(images) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>channel_name</span>
<span id="cb473-424"><a href="#cb473-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-425"><a href="#cb473-425" aria-hidden="true" tabindex="-1"></a>adapted_sm <span class="ot">&lt;-</span> <span class="fu">adaptSpillmat</span>(</span>
<span id="cb473-426"><a href="#cb473-426" aria-hidden="true" tabindex="-1"></a>    sm, <span class="fu">channelNames</span>(images),</span>
<span id="cb473-427"><a href="#cb473-427" aria-hidden="true" tabindex="-1"></a>    <span class="at">isotope_list =</span> isotope_list</span>
<span id="cb473-428"><a href="#cb473-428" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-429"><a href="#cb473-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-430"><a href="#cb473-430" aria-hidden="true" tabindex="-1"></a>images_comp <span class="ot">&lt;-</span> <span class="fu">compImage</span>(</span>
<span id="cb473-431"><a href="#cb473-431" aria-hidden="true" tabindex="-1"></a>    images, adapted_sm,</span>
<span id="cb473-432"><a href="#cb473-432" aria-hidden="true" tabindex="-1"></a>    <span class="at">BPPARAM =</span> <span class="fu">MulticoreParam</span>()</span>
<span id="cb473-433"><a href="#cb473-433" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-434"><a href="#cb473-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-435"><a href="#cb473-435" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the image before and after compensation</span></span>
<span id="cb473-436"><a href="#cb473-436" aria-hidden="true" tabindex="-1"></a><span class="co"># Before compensation</span></span>
<span id="cb473-437"><a href="#cb473-437" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPixels</span>(</span>
<span id="cb473-438"><a href="#cb473-438" aria-hidden="true" tabindex="-1"></a>    images[<span class="dv">5</span>], <span class="at">colour_by =</span> <span class="st">"Yb173Di"</span>,</span>
<span id="cb473-439"><a href="#cb473-439" aria-hidden="true" tabindex="-1"></a>    <span class="at">image_title =</span> <span class="fu">list</span>(<span class="at">text =</span> <span class="st">"Yb173 (Ecad) - before"</span>, <span class="at">position =</span> <span class="st">"topleft"</span>),</span>
<span id="cb473-440"><a href="#cb473-440" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="cn">NULL</span>, <span class="at">bcg =</span> <span class="fu">list</span>(<span class="at">Yb173Di =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">1</span>))</span>
<span id="cb473-441"><a href="#cb473-441" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-442"><a href="#cb473-442" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPixels</span>(</span>
<span id="cb473-443"><a href="#cb473-443" aria-hidden="true" tabindex="-1"></a>    images[<span class="dv">5</span>], <span class="at">colour_by =</span> <span class="st">"Yb174Di"</span>,</span>
<span id="cb473-444"><a href="#cb473-444" aria-hidden="true" tabindex="-1"></a>    <span class="at">image_title =</span> <span class="fu">list</span>(<span class="at">text =</span> <span class="st">"Yb174 (CD303) - before"</span>, <span class="at">position =</span> <span class="st">"topleft"</span>),</span>
<span id="cb473-445"><a href="#cb473-445" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="cn">NULL</span>, <span class="at">bcg =</span> <span class="fu">list</span>(<span class="at">Yb174Di =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">1</span>))</span>
<span id="cb473-446"><a href="#cb473-446" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-447"><a href="#cb473-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-448"><a href="#cb473-448" aria-hidden="true" tabindex="-1"></a><span class="co"># After compensation</span></span>
<span id="cb473-449"><a href="#cb473-449" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPixels</span>(</span>
<span id="cb473-450"><a href="#cb473-450" aria-hidden="true" tabindex="-1"></a>    images_comp[<span class="dv">5</span>], <span class="at">colour_by =</span> <span class="st">"Yb173Di"</span>,</span>
<span id="cb473-451"><a href="#cb473-451" aria-hidden="true" tabindex="-1"></a>    <span class="at">image_title =</span> <span class="fu">list</span>(<span class="at">text =</span> <span class="st">"Yb173 (Ecad) - after"</span>, <span class="at">position =</span> <span class="st">"topleft"</span>),</span>
<span id="cb473-452"><a href="#cb473-452" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="cn">NULL</span>, <span class="at">bcg =</span> <span class="fu">list</span>(<span class="at">Yb173Di =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">1</span>))</span>
<span id="cb473-453"><a href="#cb473-453" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-454"><a href="#cb473-454" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPixels</span>(</span>
<span id="cb473-455"><a href="#cb473-455" aria-hidden="true" tabindex="-1"></a>    images_comp[<span class="dv">5</span>], <span class="at">colour_by =</span> <span class="st">"Yb174Di"</span>,</span>
<span id="cb473-456"><a href="#cb473-456" aria-hidden="true" tabindex="-1"></a>    <span class="at">image_title =</span> <span class="fu">list</span>(<span class="at">text =</span> <span class="st">"Yb174 (CD303) - after"</span>, <span class="at">position =</span> <span class="st">"topleft"</span>),</span>
<span id="cb473-457"><a href="#cb473-457" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="cn">NULL</span>, <span class="at">bcg =</span> <span class="fu">list</span>(<span class="at">Yb174Di =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">1</span>))</span>
<span id="cb473-458"><a href="#cb473-458" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-459"><a href="#cb473-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-460"><a href="#cb473-460" aria-hidden="true" tabindex="-1"></a><span class="do">### Re-set the channelNames to their biological targtes</span></span>
<span id="cb473-461"><a href="#cb473-461" aria-hidden="true" tabindex="-1"></a><span class="fu">channelNames</span>(images_comp) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)</span>
<span id="cb473-462"><a href="#cb473-462" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-463"><a href="#cb473-463" aria-hidden="true" tabindex="-1"></a><span class="fu">### Write out compensated images</span></span>
<span id="cb473-464"><a href="#cb473-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-467"><a href="#cb473-467" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-468"><a href="#cb473-468" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-469"><a href="#cb473-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-470"><a href="#cb473-470" aria-hidden="true" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(<span class="fu">here</span>(dir, <span class="st">"data/comp_img"</span>))</span>
<span id="cb473-471"><a href="#cb473-471" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(</span>
<span id="cb473-472"><a href="#cb473-472" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(images_comp), <span class="cf">function</span>(x) {</span>
<span id="cb473-473"><a href="#cb473-473" aria-hidden="true" tabindex="-1"></a>        <span class="fu">writeImage</span>(<span class="fu">as.array</span>(images_comp[[x]]) <span class="sc">/</span> (<span class="dv">2</span><span class="sc">^</span><span class="dv">16</span> <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb473-474"><a href="#cb473-474" aria-hidden="true" tabindex="-1"></a>            <span class="fu">paste0</span>(dir, <span class="st">"/data/comp_img/"</span>, x, <span class="st">".tiff"</span>),</span>
<span id="cb473-475"><a href="#cb473-475" aria-hidden="true" tabindex="-1"></a>            <span class="at">bits.per.sample =</span> <span class="dv">16</span>)</span>
<span id="cb473-476"><a href="#cb473-476" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb473-477"><a href="#cb473-477" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-478"><a href="#cb473-478" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-479"><a href="#cb473-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-482"><a href="#cb473-482" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-483"><a href="#cb473-483" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-484"><a href="#cb473-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-485"><a href="#cb473-485" aria-hidden="true" tabindex="-1"></a><span class="do">### Save the compensated SpatialExperiment and CytoImageList objects</span></span>
<span id="cb473-486"><a href="#cb473-486" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-487"><a href="#cb473-487" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(images_comp, <span class="fu">here</span>(dir,<span class="st">"data/images.qs"</span>))</span>
<span id="cb473-488"><a href="#cb473-488" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-489"><a href="#cb473-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-490"><a href="#cb473-490" aria-hidden="true" tabindex="-1"></a><span class="fu">## Image and cell quality control</span></span>
<span id="cb473-491"><a href="#cb473-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-492"><a href="#cb473-492" aria-hidden="true" tabindex="-1"></a><span class="fu">### Segmentation quality control</span></span>
<span id="cb473-493"><a href="#cb473-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-496"><a href="#cb473-496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-497"><a href="#cb473-497" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb473-498"><a href="#cb473-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-499"><a href="#cb473-499" aria-hidden="true" tabindex="-1"></a><span class="do">### Load data: 63.01 MB</span></span>
<span id="cb473-500"><a href="#cb473-500" aria-hidden="true" tabindex="-1"></a>images <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/images.qs"</span>))</span>
<span id="cb473-501"><a href="#cb473-501" aria-hidden="true" tabindex="-1"></a>masks <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/masks.qs"</span>))</span>
<span id="cb473-502"><a href="#cb473-502" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-503"><a href="#cb473-503" aria-hidden="true" tabindex="-1"></a>lobstr<span class="sc">::</span><span class="fu">obj_size</span>(spe, image, masks)</span>
<span id="cb473-504"><a href="#cb473-504" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-505"><a href="#cb473-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-506"><a href="#cb473-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-509"><a href="#cb473-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-510"><a href="#cb473-510" aria-hidden="true" tabindex="-1"></a><span class="do">### Select 3 random images</span></span>
<span id="cb473-511"><a href="#cb473-511" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">20220118</span>)</span>
<span id="cb473-512"><a href="#cb473-512" aria-hidden="true" tabindex="-1"></a>img_ids <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_along</span>(images), <span class="dv">3</span>)</span>
<span id="cb473-513"><a href="#cb473-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-514"><a href="#cb473-514" aria-hidden="true" tabindex="-1"></a><span class="do">### Image- and channel-wise min-max normalization and scaled to a range of 0-1</span></span>
<span id="cb473-515"><a href="#cb473-515" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> images[img_ids]</span>
<span id="cb473-516"><a href="#cb473-516" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> cytomapper<span class="sc">::</span><span class="fu">normalize</span>(cur_images, <span class="at">separateImages =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-517"><a href="#cb473-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-518"><a href="#cb473-518" aria-hidden="true" tabindex="-1"></a><span class="do">### Clipping the maximum intensity to 0.2</span></span>
<span id="cb473-519"><a href="#cb473-519" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> cytomapper<span class="sc">::</span><span class="fu">normalize</span>(cur_images, <span class="at">inputRange =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>))</span>
<span id="cb473-520"><a href="#cb473-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-521"><a href="#cb473-521" aria-hidden="true" tabindex="-1"></a><span class="do">### Segmentation approach here appears to correctly segment cells</span></span>
<span id="cb473-522"><a href="#cb473-522" aria-hidden="true" tabindex="-1"></a>cytomapper<span class="sc">::</span><span class="fu">plotPixels</span>(</span>
<span id="cb473-523"><a href="#cb473-523" aria-hidden="true" tabindex="-1"></a>    cur_images,</span>
<span id="cb473-524"><a href="#cb473-524" aria-hidden="true" tabindex="-1"></a>    <span class="at">mask =</span> masks[img_ids],</span>
<span id="cb473-525"><a href="#cb473-525" aria-hidden="true" tabindex="-1"></a>    <span class="at">img_id =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb473-526"><a href="#cb473-526" aria-hidden="true" tabindex="-1"></a>    <span class="at">missing_colour =</span> <span class="st">"white"</span>,</span>
<span id="cb473-527"><a href="#cb473-527" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"Ecad"</span>, <span class="st">"DNA1"</span>),</span>
<span id="cb473-528"><a href="#cb473-528" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="fu">list</span>(</span>
<span id="cb473-529"><a href="#cb473-529" aria-hidden="true" tabindex="-1"></a>        <span class="at">CD163 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"yellow"</span>),</span>
<span id="cb473-530"><a href="#cb473-530" aria-hidden="true" tabindex="-1"></a>        <span class="at">CD20 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>),</span>
<span id="cb473-531"><a href="#cb473-531" aria-hidden="true" tabindex="-1"></a>        <span class="at">CD3 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"green"</span>),</span>
<span id="cb473-532"><a href="#cb473-532" aria-hidden="true" tabindex="-1"></a>        <span class="at">Ecad =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"cyan"</span>),</span>
<span id="cb473-533"><a href="#cb473-533" aria-hidden="true" tabindex="-1"></a>        <span class="at">DNA1 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>)</span>
<span id="cb473-534"><a href="#cb473-534" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb473-535"><a href="#cb473-535" aria-hidden="true" tabindex="-1"></a>    <span class="at">image_title =</span> <span class="cn">NULL</span>,</span>
<span id="cb473-536"><a href="#cb473-536" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="fu">list</span>(</span>
<span id="cb473-537"><a href="#cb473-537" aria-hidden="true" tabindex="-1"></a>        <span class="at">colour_by.title.cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb473-538"><a href="#cb473-538" aria-hidden="true" tabindex="-1"></a>        <span class="at">colour_by.labels.cex =</span> <span class="fl">0.7</span></span>
<span id="cb473-539"><a href="#cb473-539" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-540"><a href="#cb473-540" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-541"><a href="#cb473-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-542"><a href="#cb473-542" aria-hidden="true" tabindex="-1"></a><span class="do">### Heatmap to observe cell segmentation quality and</span></span>
<span id="cb473-543"><a href="#cb473-543" aria-hidden="true" tabindex="-1"></a><span class="do">### potentially also antibody specificity issues</span></span>
<span id="cb473-544"><a href="#cb473-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-545"><a href="#cb473-545" aria-hidden="true" tabindex="-1"></a><span class="do">### Sub-sample the dataset to 2000 cells</span></span>
<span id="cb473-546"><a href="#cb473-546" aria-hidden="true" tabindex="-1"></a>cur_cells <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(spe)), <span class="dv">2000</span>)</span>
<span id="cb473-547"><a href="#cb473-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-548"><a href="#cb473-548" aria-hidden="true" tabindex="-1"></a><span class="do">### Epithelial cells (Ecad+) and immune cells (CD45RO+) can be differentiate</span></span>
<span id="cb473-549"><a href="#cb473-549" aria-hidden="true" tabindex="-1"></a><span class="do">### Some of the markers are detected in specific cells (e.g., Ki67, CD20, Ecad) </span><span class="al">###</span><span class="do"> while others are more broadly expressed across cells (e.g., HLADR, B2M, CD4).</span></span>
<span id="cb473-550"><a href="#cb473-550" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-551"><a href="#cb473-551" aria-hidden="true" tabindex="-1"></a>    spe[, cur_cells],</span>
<span id="cb473-552"><a href="#cb473-552" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-553"><a href="#cb473-553" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-554"><a href="#cb473-554" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-555"><a href="#cb473-555" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-556"><a href="#cb473-556" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-557"><a href="#cb473-557" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="st">"indication"</span>,</span>
<span id="cb473-558"><a href="#cb473-558" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-559"><a href="#cb473-559" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication</span>
<span id="cb473-560"><a href="#cb473-560" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-561"><a href="#cb473-561" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-562"><a href="#cb473-562" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-563"><a href="#cb473-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-564"><a href="#cb473-564" aria-hidden="true" tabindex="-1"></a><span class="fu">### Image_level quality control</span></span>
<span id="cb473-565"><a href="#cb473-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-568"><a href="#cb473-568" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-569"><a href="#cb473-569" aria-hidden="true" tabindex="-1"></a><span class="do">### Average SNR versus the average signal intensity across all images</span></span>
<span id="cb473-570"><a href="#cb473-570" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(images), <span class="cf">function</span>(x){</span>
<span id="cb473-571"><a href="#cb473-571" aria-hidden="true" tabindex="-1"></a>    img <span class="ot">&lt;-</span> images[[x]]</span>
<span id="cb473-572"><a href="#cb473-572" aria-hidden="true" tabindex="-1"></a>    mat <span class="ot">&lt;-</span> <span class="fu">apply</span>(img, <span class="dv">3</span>, <span class="cf">function</span>(ch){</span>
<span id="cb473-573"><a href="#cb473-573" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Otsu threshold</span></span>
<span id="cb473-574"><a href="#cb473-574" aria-hidden="true" tabindex="-1"></a>        thres <span class="ot">&lt;-</span> <span class="fu">otsu</span>(ch, <span class="at">range =</span> <span class="fu">c</span>(<span class="fu">min</span>(ch), <span class="fu">max</span>(ch)), <span class="at">levels =</span> <span class="dv">65536</span>)</span>
<span id="cb473-575"><a href="#cb473-575" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Signal-to-noise ratio</span></span>
<span id="cb473-576"><a href="#cb473-576" aria-hidden="true" tabindex="-1"></a>        snr <span class="ot">&lt;-</span> <span class="fu">mean</span>(ch[ch <span class="sc">&gt;</span> thres]) <span class="sc">/</span> <span class="fu">mean</span>(ch[ch <span class="sc">&lt;=</span> thres])</span>
<span id="cb473-577"><a href="#cb473-577" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Signal intensity</span></span>
<span id="cb473-578"><a href="#cb473-578" aria-hidden="true" tabindex="-1"></a>        ps <span class="ot">&lt;-</span> <span class="fu">mean</span>(ch[ch <span class="sc">&gt;</span> thres])</span>
<span id="cb473-579"><a href="#cb473-579" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb473-580"><a href="#cb473-580" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(<span class="fu">c</span>(<span class="at">snr =</span> snr, <span class="at">ps =</span> ps))</span>
<span id="cb473-581"><a href="#cb473-581" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb473-582"><a href="#cb473-582" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>(mat) <span class="sc">|&gt;</span>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span>  </span>
<span id="cb473-583"><a href="#cb473-583" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">image =</span> x,</span>
<span id="cb473-584"><a href="#cb473-584" aria-hidden="true" tabindex="-1"></a>               <span class="at">marker =</span> <span class="fu">colnames</span>(mat)) <span class="sc">|&gt;</span>  </span>
<span id="cb473-585"><a href="#cb473-585" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(snr, ps))</span>
<span id="cb473-586"><a href="#cb473-586" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb473-587"><a href="#cb473-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-588"><a href="#cb473-588" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, cur_snr)</span>
<span id="cb473-589"><a href="#cb473-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-590"><a href="#cb473-590" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="sc">|&gt;</span>  </span>
<span id="cb473-591"><a href="#cb473-591" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(marker, name) <span class="sc">|&gt;</span> </span>
<span id="cb473-592"><a href="#cb473-592" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">log_mean =</span> <span class="fu">log2</span>(<span class="fu">mean</span>(value))) <span class="sc">|&gt;</span> </span>
<span id="cb473-593"><a href="#cb473-593" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> name, <span class="at">values_from =</span> log_mean) <span class="sc">|&gt;</span> </span>
<span id="cb473-594"><a href="#cb473-594" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb473-595"><a href="#cb473-595" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(ps, snr)) <span class="sc">+</span></span>
<span id="cb473-596"><a href="#cb473-596" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(ps, snr, <span class="at">label =</span> marker)) <span class="sc">+</span></span>
<span id="cb473-597"><a href="#cb473-597" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Signal-to-noise ratio [log2]"</span>) <span class="sc">+</span></span>
<span id="cb473-598"><a href="#cb473-598" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Signal intensity [log2]"</span>)</span>
<span id="cb473-599"><a href="#cb473-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-600"><a href="#cb473-600" aria-hidden="true" tabindex="-1"></a><span class="do">### Remove markers that have a positive signal of below 2 per image</span></span>
<span id="cb473-601"><a href="#cb473-601" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="ot">&lt;-</span> cur_snr <span class="sc">|&gt;</span>  </span>
<span id="cb473-602"><a href="#cb473-602" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> name, <span class="at">values_from =</span> value) <span class="sc">|&gt;</span> </span>
<span id="cb473-603"><a href="#cb473-603" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(ps <span class="sc">&gt;</span> <span class="dv">2</span>) <span class="sc">|&gt;</span> </span>
<span id="cb473-604"><a href="#cb473-604" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(snr, ps))</span>
<span id="cb473-605"><a href="#cb473-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-606"><a href="#cb473-606" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="sc">|&gt;</span>  </span>
<span id="cb473-607"><a href="#cb473-607" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(marker, name) <span class="sc">|&gt;</span> </span>
<span id="cb473-608"><a href="#cb473-608" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">log_mean =</span> <span class="fu">log2</span>(<span class="fu">mean</span>(value))) <span class="sc">|&gt;</span> </span>
<span id="cb473-609"><a href="#cb473-609" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> name, <span class="at">values_from =</span> log_mean) <span class="sc">|&gt;</span> </span>
<span id="cb473-610"><a href="#cb473-610" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb473-611"><a href="#cb473-611" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(ps, snr)) <span class="sc">+</span></span>
<span id="cb473-612"><a href="#cb473-612" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(ps, snr, <span class="at">label =</span> marker)) <span class="sc">+</span></span>
<span id="cb473-613"><a href="#cb473-613" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Signal-to-noise ratio [log2]"</span>) <span class="sc">+</span></span>
<span id="cb473-614"><a href="#cb473-614" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Signal intensity [log2]"</span>)</span>
<span id="cb473-615"><a href="#cb473-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-616"><a href="#cb473-616" aria-hidden="true" tabindex="-1"></a><span class="do">### Compute the percentage of covered image area</span></span>
<span id="cb473-617"><a href="#cb473-617" aria-hidden="true" tabindex="-1"></a>cell_density <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe) <span class="sc">|&gt;</span> </span>
<span id="cb473-618"><a href="#cb473-618" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb473-619"><a href="#cb473-619" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(sample_id) <span class="sc">|&gt;</span> </span>
<span id="cb473-620"><a href="#cb473-620" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Compute the number of pixels covered by cells and </span></span>
<span id="cb473-621"><a href="#cb473-621" aria-hidden="true" tabindex="-1"></a>    <span class="co"># the total number of pixels</span></span>
<span id="cb473-622"><a href="#cb473-622" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">cell_area =</span> <span class="fu">sum</span>(area),</span>
<span id="cb473-623"><a href="#cb473-623" aria-hidden="true" tabindex="-1"></a>              <span class="at">no_pixels =</span> <span class="fu">mean</span>(width_px) <span class="sc">*</span> <span class="fu">mean</span>(height_px)) <span class="sc">|&gt;</span> </span>
<span id="cb473-624"><a href="#cb473-624" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Divide the total number of pixels </span></span>
<span id="cb473-625"><a href="#cb473-625" aria-hidden="true" tabindex="-1"></a>    <span class="co"># by the number of pixels covered by cells</span></span>
<span id="cb473-626"><a href="#cb473-626" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">covered_area =</span> cell_area <span class="sc">/</span> no_pixels)</span>
<span id="cb473-627"><a href="#cb473-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-628"><a href="#cb473-628" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the image area covered by cells per image</span></span>
<span id="cb473-629"><a href="#cb473-629" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(cell_density) <span class="sc">+</span></span>
<span id="cb473-630"><a href="#cb473-630" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="fu">reorder</span>(sample_id,covered_area), covered_area)) <span class="sc">+</span> </span>
<span id="cb473-631"><a href="#cb473-631" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb473-632"><a href="#cb473-632" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">15</span>)) <span class="sc">+</span></span>
<span id="cb473-633"><a href="#cb473-633" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb473-634"><a href="#cb473-634" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ylab</span>(<span class="st">"% covered area"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">""</span>)</span>
<span id="cb473-635"><a href="#cb473-635" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-636"><a href="#cb473-636" aria-hidden="true" tabindex="-1"></a><span class="do">### Normalize and clip images</span></span>
<span id="cb473-637"><a href="#cb473-637" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> images[<span class="fu">c</span>(<span class="st">"Patient4_005"</span>, <span class="st">"Patient4_007"</span>)]</span>
<span id="cb473-638"><a href="#cb473-638" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> cytomapper<span class="sc">::</span><span class="fu">normalize</span>(cur_images, <span class="at">separateImages =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-639"><a href="#cb473-639" aria-hidden="true" tabindex="-1"></a>cur_images <span class="ot">&lt;-</span> cytomapper<span class="sc">::</span><span class="fu">normalize</span>(cur_images, <span class="at">inputRange =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>))</span>
<span id="cb473-640"><a href="#cb473-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-641"><a href="#cb473-641" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPixels</span>(cur_images,</span>
<span id="cb473-642"><a href="#cb473-642" aria-hidden="true" tabindex="-1"></a>           <span class="at">mask =</span> masks[<span class="fu">c</span>(<span class="st">"Patient4_005"</span>, <span class="st">"Patient4_007"</span>)],</span>
<span id="cb473-643"><a href="#cb473-643" aria-hidden="true" tabindex="-1"></a>           <span class="at">img_id =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb473-644"><a href="#cb473-644" aria-hidden="true" tabindex="-1"></a>           <span class="at">missing_colour =</span> <span class="st">"white"</span>,</span>
<span id="cb473-645"><a href="#cb473-645" aria-hidden="true" tabindex="-1"></a>           <span class="at">colour_by =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"Ecad"</span>, <span class="st">"DNA1"</span>),</span>
<span id="cb473-646"><a href="#cb473-646" aria-hidden="true" tabindex="-1"></a>           <span class="at">colour =</span> <span class="fu">list</span>(<span class="at">CD163 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"yellow"</span>),</span>
<span id="cb473-647"><a href="#cb473-647" aria-hidden="true" tabindex="-1"></a>                         <span class="at">CD20 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>),</span>
<span id="cb473-648"><a href="#cb473-648" aria-hidden="true" tabindex="-1"></a>                         <span class="at">CD3 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"green"</span>),</span>
<span id="cb473-649"><a href="#cb473-649" aria-hidden="true" tabindex="-1"></a>                         <span class="at">Ecad =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"cyan"</span>),</span>
<span id="cb473-650"><a href="#cb473-650" aria-hidden="true" tabindex="-1"></a>                         <span class="at">DNA1 =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>)),</span>
<span id="cb473-651"><a href="#cb473-651" aria-hidden="true" tabindex="-1"></a>           <span class="at">legend =</span> <span class="fu">list</span>(<span class="at">colour_by.title.cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb473-652"><a href="#cb473-652" aria-hidden="true" tabindex="-1"></a>                         <span class="at">colour_by.labels.cex =</span> <span class="fl">0.7</span>))</span>
<span id="cb473-653"><a href="#cb473-653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-654"><a href="#cb473-654" aria-hidden="true" tabindex="-1"></a><span class="do">###  Visualize the mean marker expression per image to identify images with outlying marker expression</span></span>
<span id="cb473-655"><a href="#cb473-655" aria-hidden="true" tabindex="-1"></a>image_mean <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">aggregateAcrossCells</span>(</span>
<span id="cb473-656"><a href="#cb473-656" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-657"><a href="#cb473-657" aria-hidden="true" tabindex="-1"></a>    <span class="at">ids =</span> spe<span class="sc">$</span>sample_id, </span>
<span id="cb473-658"><a href="#cb473-658" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics=</span><span class="st">"mean"</span>,</span>
<span id="cb473-659"><a href="#cb473-659" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.assay.type =</span> <span class="st">"counts"</span></span>
<span id="cb473-660"><a href="#cb473-660" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-661"><a href="#cb473-661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-662"><a href="#cb473-662" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(image_mean, <span class="st">"exprs"</span>) <span class="ot">&lt;-</span> <span class="fu">asinh</span>(<span class="fu">counts</span>(image_mean))</span>
<span id="cb473-663"><a href="#cb473-663" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-664"><a href="#cb473-664" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-665"><a href="#cb473-665" aria-hidden="true" tabindex="-1"></a>    image_mean, </span>
<span id="cb473-666"><a href="#cb473-666" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-667"><a href="#cb473-667" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-668"><a href="#cb473-668" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-669"><a href="#cb473-669" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"indication"</span>, <span class="st">"patient_id"</span>, <span class="st">"ROI"</span>),</span>
<span id="cb473-670"><a href="#cb473-670" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-671"><a href="#cb473-671" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication,</span>
<span id="cb473-672"><a href="#cb473-672" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">patient_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id,</span>
<span id="cb473-673"><a href="#cb473-673" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">ROI =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>ROI</span>
<span id="cb473-674"><a href="#cb473-674" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb473-675"><a href="#cb473-675" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_colnames =</span> <span class="cn">TRUE</span></span>
<span id="cb473-676"><a href="#cb473-676" aria-hidden="true" tabindex="-1"></a>)                        </span>
<span id="cb473-677"><a href="#cb473-677" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-678"><a href="#cb473-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-679"><a href="#cb473-679" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cell_level quality control</span></span>
<span id="cb473-680"><a href="#cb473-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-683"><a href="#cb473-683" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-684"><a href="#cb473-684" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220224</span>)</span>
<span id="cb473-685"><a href="#cb473-685" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(spe)), <span class="cf">function</span>(x){</span>
<span id="cb473-686"><a href="#cb473-686" aria-hidden="true" tabindex="-1"></a>    cur_exprs <span class="ot">&lt;-</span> <span class="fu">assay</span>(spe, <span class="st">"exprs"</span>)[x,]</span>
<span id="cb473-687"><a href="#cb473-687" aria-hidden="true" tabindex="-1"></a>    cur_counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(spe, <span class="st">"counts"</span>)[x,]</span>
<span id="cb473-688"><a href="#cb473-688" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-689"><a href="#cb473-689" aria-hidden="true" tabindex="-1"></a>    cur_model <span class="ot">&lt;-</span> <span class="fu">Mclust</span>(cur_exprs, <span class="at">G =</span> <span class="dv">2</span>)</span>
<span id="cb473-690"><a href="#cb473-690" aria-hidden="true" tabindex="-1"></a>    mean1 <span class="ot">&lt;-</span> <span class="fu">mean</span>(cur_counts[cur_model<span class="sc">$</span>classification <span class="sc">==</span> <span class="dv">1</span>])</span>
<span id="cb473-691"><a href="#cb473-691" aria-hidden="true" tabindex="-1"></a>    mean2 <span class="ot">&lt;-</span> <span class="fu">mean</span>(cur_counts[cur_model<span class="sc">$</span>classification <span class="sc">==</span> <span class="dv">2</span>])</span>
<span id="cb473-692"><a href="#cb473-692" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-693"><a href="#cb473-693" aria-hidden="true" tabindex="-1"></a>    signal <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(mean1 <span class="sc">&gt;</span> mean2, mean1, mean2)</span>
<span id="cb473-694"><a href="#cb473-694" aria-hidden="true" tabindex="-1"></a>    noise <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(mean1 <span class="sc">&gt;</span> mean2, mean2, mean1)</span>
<span id="cb473-695"><a href="#cb473-695" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-696"><a href="#cb473-696" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">c</span>(<span class="at">snr =</span> signal<span class="sc">/</span>noise, <span class="at">ps =</span> signal))</span>
<span id="cb473-697"><a href="#cb473-697" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb473-698"><a href="#cb473-698" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-699"><a href="#cb473-699" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="ot">&lt;-</span> <span class="fu">t</span>(mat) <span class="sc">|&gt;</span>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span>  </span>
<span id="cb473-700"><a href="#cb473-700" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">marker =</span> <span class="fu">rownames</span>(spe))</span>
<span id="cb473-701"><a href="#cb473-701" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-702"><a href="#cb473-702" aria-hidden="true" tabindex="-1"></a>cur_snr <span class="sc">|&gt;</span>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb473-703"><a href="#cb473-703" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="fu">log2</span>(ps), <span class="fu">log2</span>(snr))) <span class="sc">+</span></span>
<span id="cb473-704"><a href="#cb473-704" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(<span class="fu">log2</span>(ps), <span class="fu">log2</span>(snr), <span class="at">label =</span> marker)) <span class="sc">+</span></span>
<span id="cb473-705"><a href="#cb473-705" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"Signal-to-noise ratio [log2]"</span>) <span class="sc">+</span></span>
<span id="cb473-706"><a href="#cb473-706" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Signal intensity [log2]"</span>)</span>
<span id="cb473-707"><a href="#cb473-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-708"><a href="#cb473-708" aria-hidden="true" tabindex="-1"></a><span class="do">### Observe the distributions of cell size across the individual images</span></span>
<span id="cb473-709"><a href="#cb473-709" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoPlot</span>(spe, <span class="at">var =</span> <span class="st">"area"</span>, </span>
<span id="cb473-710"><a href="#cb473-710" aria-hidden="true" tabindex="-1"></a>          <span class="at">group.by =</span> <span class="st">"sample_id"</span>, </span>
<span id="cb473-711"><a href="#cb473-711" aria-hidden="true" tabindex="-1"></a>          <span class="at">plots =</span> <span class="st">"boxplot"</span>) <span class="sc">+</span></span>
<span id="cb473-712"><a href="#cb473-712" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ylab</span>(<span class="st">"Cell area"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">""</span>)</span>
<span id="cb473-713"><a href="#cb473-713" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(spe<span class="sc">$</span>area)</span>
<span id="cb473-714"><a href="#cb473-714" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(spe<span class="sc">$</span>area <span class="sc">&lt;</span> <span class="dv">5</span>)</span>
<span id="cb473-715"><a href="#cb473-715" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-716"><a href="#cb473-716" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>area <span class="sc">&gt;=</span> <span class="dv">5</span>]</span>
<span id="cb473-717"><a href="#cb473-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-718"><a href="#cb473-718" aria-hidden="true" tabindex="-1"></a><span class="do">### Absolute measure of cell density</span></span>
<span id="cb473-719"><a href="#cb473-719" aria-hidden="true" tabindex="-1"></a>cell_density <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe) <span class="sc">|&gt;</span> </span>
<span id="cb473-720"><a href="#cb473-720" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb473-721"><a href="#cb473-721" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(sample_id) <span class="sc">|&gt;</span> </span>
<span id="cb473-722"><a href="#cb473-722" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">cell_count =</span> <span class="fu">n</span>(),</span>
<span id="cb473-723"><a href="#cb473-723" aria-hidden="true" tabindex="-1"></a>           <span class="at">no_pixels =</span> <span class="fu">mean</span>(width_px) <span class="sc">*</span> <span class="fu">mean</span>(height_px)) <span class="sc">|&gt;</span> </span>
<span id="cb473-724"><a href="#cb473-724" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">cells_per_mm2 =</span> cell_count<span class="sc">/</span>(no_pixels<span class="sc">/</span><span class="dv">1000000</span>))</span>
<span id="cb473-725"><a href="#cb473-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-726"><a href="#cb473-726" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(cell_density) <span class="sc">+</span></span>
<span id="cb473-727"><a href="#cb473-727" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="fu">reorder</span>(sample_id,cells_per_mm2), cells_per_mm2)) <span class="sc">+</span> </span>
<span id="cb473-728"><a href="#cb473-728" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> </span>
<span id="cb473-729"><a href="#cb473-729" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">8</span>)) <span class="sc">+</span></span>
<span id="cb473-730"><a href="#cb473-730" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cells per mm2"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">""</span>)</span>
<span id="cb473-731"><a href="#cb473-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-732"><a href="#cb473-732" aria-hidden="true" tabindex="-1"></a> <span class="do">### Observing staining differences between samples or batches of samples</span></span>
<span id="cb473-733"><a href="#cb473-733" aria-hidden="true" tabindex="-1"></a><span class="fu">multi_dittoPlot</span>(</span>
<span id="cb473-734"><a href="#cb473-734" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-735"><a href="#cb473-735" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-736"><a href="#cb473-736" aria-hidden="true" tabindex="-1"></a>    <span class="at">group.by =</span> <span class="st">"patient_id"</span>, <span class="at">plots =</span> <span class="st">"ridgeplot"</span>, </span>
<span id="cb473-737"><a href="#cb473-737" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-738"><a href="#cb473-738" aria-hidden="true" tabindex="-1"></a>    <span class="at">color.panel =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id</span>
<span id="cb473-739"><a href="#cb473-739" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-740"><a href="#cb473-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-741"><a href="#cb473-741" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220225</span>)</span>
<span id="cb473-742"><a href="#cb473-742" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runUMAP</span>(</span>
<span id="cb473-743"><a href="#cb473-743" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">subset_row =</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel, <span class="at">exprs_values =</span> <span class="st">"exprs"</span></span>
<span id="cb473-744"><a href="#cb473-744" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-745"><a href="#cb473-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-746"><a href="#cb473-746" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runTSNE</span>(</span>
<span id="cb473-747"><a href="#cb473-747" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">subset_row =</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel, <span class="at">exprs_values =</span> <span class="st">"exprs"</span></span>
<span id="cb473-748"><a href="#cb473-748" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-749"><a href="#cb473-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-750"><a href="#cb473-750" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDims</span>(spe)</span>
<span id="cb473-751"><a href="#cb473-751" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">reducedDim</span>(spe, <span class="st">"UMAP"</span>))</span>
<span id="cb473-752"><a href="#cb473-752" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-753"><a href="#cb473-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-756"><a href="#cb473-756" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-757"><a href="#cb473-757" aria-hidden="true" tabindex="-1"></a><span class="do">### visualize patient id </span></span>
<span id="cb473-758"><a href="#cb473-758" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-759"><a href="#cb473-759" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-760"><a href="#cb473-760" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-761"><a href="#cb473-761" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-762"><a href="#cb473-762" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP"</span>)</span>
<span id="cb473-763"><a href="#cb473-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-764"><a href="#cb473-764" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-765"><a href="#cb473-765" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"TSNE"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-766"><a href="#cb473-766" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-767"><a href="#cb473-767" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-768"><a href="#cb473-768" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on TSNE"</span>)</span>
<span id="cb473-769"><a href="#cb473-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-770"><a href="#cb473-770" aria-hidden="true" tabindex="-1"></a><span class="do">### visualize region of interest id</span></span>
<span id="cb473-771"><a href="#cb473-771" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-772"><a href="#cb473-772" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"ROI"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-773"><a href="#cb473-773" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-774"><a href="#cb473-774" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>ROI) <span class="sc">+</span></span>
<span id="cb473-775"><a href="#cb473-775" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"ROI ID on UMAP"</span>)</span>
<span id="cb473-776"><a href="#cb473-776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-777"><a href="#cb473-777" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-778"><a href="#cb473-778" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"ROI"</span>, <span class="at">reduction.use =</span> <span class="st">"TSNE"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-779"><a href="#cb473-779" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-780"><a href="#cb473-780" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>ROI) <span class="sc">+</span></span>
<span id="cb473-781"><a href="#cb473-781" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"ROI ID on TSNE"</span>)</span>
<span id="cb473-782"><a href="#cb473-782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-783"><a href="#cb473-783" aria-hidden="true" tabindex="-1"></a><span class="do">### visualize indication</span></span>
<span id="cb473-784"><a href="#cb473-784" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-785"><a href="#cb473-785" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"indication"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-786"><a href="#cb473-786" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-787"><a href="#cb473-787" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication) <span class="sc">+</span></span>
<span id="cb473-788"><a href="#cb473-788" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Indication on UMAP"</span>)</span>
<span id="cb473-789"><a href="#cb473-789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-790"><a href="#cb473-790" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-791"><a href="#cb473-791" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"indication"</span>, <span class="at">reduction.use =</span> <span class="st">"TSNE"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-792"><a href="#cb473-792" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-793"><a href="#cb473-793" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication) <span class="sc">+</span></span>
<span id="cb473-794"><a href="#cb473-794" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Indication on TSNE"</span>)</span>
<span id="cb473-795"><a href="#cb473-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-796"><a href="#cb473-796" aria-hidden="true" tabindex="-1"></a>(p1 <span class="sc">+</span> p2) <span class="sc">/</span> (p3 <span class="sc">+</span> p4) <span class="sc">/</span> (p5 <span class="sc">+</span> p6)</span>
<span id="cb473-797"><a href="#cb473-797" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-798"><a href="#cb473-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-801"><a href="#cb473-801" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-802"><a href="#cb473-802" aria-hidden="true" tabindex="-1"></a><span class="do">### visualize marker expression</span></span>
<span id="cb473-803"><a href="#cb473-803" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-804"><a href="#cb473-804" aria-hidden="true" tabindex="-1"></a>        spe, <span class="at">var =</span> <span class="st">"Ecad"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb473-805"><a href="#cb473-805" aria-hidden="true" tabindex="-1"></a>        <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-806"><a href="#cb473-806" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-807"><a href="#cb473-807" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis</span>(<span class="at">name =</span> <span class="st">"Ecad"</span>) <span class="sc">+</span></span>
<span id="cb473-808"><a href="#cb473-808" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"E-Cadherin expression on UMAP"</span>)</span>
<span id="cb473-809"><a href="#cb473-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-810"><a href="#cb473-810" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-811"><a href="#cb473-811" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"CD45RO"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb473-812"><a href="#cb473-812" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-813"><a href="#cb473-813" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-814"><a href="#cb473-814" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis</span>(<span class="at">name =</span> <span class="st">"CD45RO"</span>) <span class="sc">+</span></span>
<span id="cb473-815"><a href="#cb473-815" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"CD45RO expression on UMAP"</span>)</span>
<span id="cb473-816"><a href="#cb473-816" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-817"><a href="#cb473-817" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-818"><a href="#cb473-818" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"Ecad"</span>, <span class="at">reduction.use =</span> <span class="st">"TSNE"</span>, </span>
<span id="cb473-819"><a href="#cb473-819" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-820"><a href="#cb473-820" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-821"><a href="#cb473-821" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis</span>(<span class="at">name =</span> <span class="st">"Ecad"</span>) <span class="sc">+</span></span>
<span id="cb473-822"><a href="#cb473-822" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Ecad expression on TSNE"</span>)</span>
<span id="cb473-823"><a href="#cb473-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-824"><a href="#cb473-824" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-825"><a href="#cb473-825" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"CD45RO"</span>, <span class="at">reduction.use =</span> <span class="st">"TSNE"</span>, </span>
<span id="cb473-826"><a href="#cb473-826" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-827"><a href="#cb473-827" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-828"><a href="#cb473-828" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis</span>(<span class="at">name =</span> <span class="st">"CD45RO"</span>) <span class="sc">+</span></span>
<span id="cb473-829"><a href="#cb473-829" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"CD45RO expression on TSNE"</span>)</span>
<span id="cb473-830"><a href="#cb473-830" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-831"><a href="#cb473-831" aria-hidden="true" tabindex="-1"></a>(p1 <span class="sc">+</span> p2) <span class="sc">/</span> (p3 <span class="sc">+</span> p4)</span>
<span id="cb473-832"><a href="#cb473-832" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-833"><a href="#cb473-833" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-834"><a href="#cb473-834" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb473-835"><a href="#cb473-835" aria-hidden="true" tabindex="-1"></a>We observe a strong separation of tumor cells (Ecad+ cells) between the patients. Here, each patient was diagnosed with a different tumor type. The separation of tumor cells could be of biological origin since tumor cells tend to display differences in expression between patients and cancer types and/or of technical origin: the panel only contains a single tumor marker (E-Cadherin) and therefore slight technical differences in staining causes visible separation between cells of different patients. Nevertheless, the immune compartment (CD45RO+ cells) mix between patients and we can rule out systematic staining differences between patients.</span>
<span id="cb473-836"><a href="#cb473-836" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb473-837"><a href="#cb473-837" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-838"><a href="#cb473-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-841"><a href="#cb473-841" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-842"><a href="#cb473-842" aria-hidden="true" tabindex="-1"></a><span class="do">### Save objects for further downstream analysis</span></span>
<span id="cb473-843"><a href="#cb473-843" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-844"><a href="#cb473-844" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-845"><a href="#cb473-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-846"><a href="#cb473-846" aria-hidden="true" tabindex="-1"></a><span class="fu">## Batch effect  correction</span></span>
<span id="cb473-847"><a href="#cb473-847" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-848"><a href="#cb473-848" aria-hidden="true" tabindex="-1"></a><span class="fu">### fastMNN correction</span></span>
<span id="cb473-849"><a href="#cb473-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-852"><a href="#cb473-852" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-853"><a href="#cb473-853" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-854"><a href="#cb473-854" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-855"><a href="#cb473-855" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-856"><a href="#cb473-856" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-859"><a href="#cb473-859" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-860"><a href="#cb473-860" aria-hidden="true" tabindex="-1"></a><span class="do">### Perform sample correction</span></span>
<span id="cb473-861"><a href="#cb473-861" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220228</span>)</span>
<span id="cb473-862"><a href="#cb473-862" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> batchelor<span class="sc">::</span><span class="fu">fastMNN</span>(</span>
<span id="cb473-863"><a href="#cb473-863" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-864"><a href="#cb473-864" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch =</span> spe<span class="sc">$</span>patient_id,</span>
<span id="cb473-865"><a href="#cb473-865" aria-hidden="true" tabindex="-1"></a>    <span class="at">auto.merge =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-866"><a href="#cb473-866" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel,</span>
<span id="cb473-867"><a href="#cb473-867" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.type =</span> <span class="st">"exprs"</span></span>
<span id="cb473-868"><a href="#cb473-868" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-869"><a href="#cb473-869" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-870"><a href="#cb473-870" aria-hidden="true" tabindex="-1"></a><span class="do">### Check that order of cells is the same</span></span>
<span id="cb473-871"><a href="#cb473-871" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(<span class="fu">colnames</span>(spe), <span class="fu">colnames</span>(out)))</span>
<span id="cb473-872"><a href="#cb473-872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-873"><a href="#cb473-873" aria-hidden="true" tabindex="-1"></a><span class="do">### Transfer the correction results to the main spe object</span></span>
<span id="cb473-874"><a href="#cb473-874" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(spe, <span class="st">"fastMNN"</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(out, <span class="st">"corrected"</span>)</span>
<span id="cb473-875"><a href="#cb473-875" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-876"><a href="#cb473-876" aria-hidden="true" tabindex="-1"></a><span class="do">### Quality control of correction results</span></span>
<span id="cb473-877"><a href="#cb473-877" aria-hidden="true" tabindex="-1"></a>merge_info <span class="ot">&lt;-</span> <span class="fu">metadata</span>(out)<span class="sc">$</span>merge.info </span>
<span id="cb473-878"><a href="#cb473-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-879"><a href="#cb473-879" aria-hidden="true" tabindex="-1"></a><span class="do">### 1. We observe that Patient4 and Patient2 are most similar with a low batch effect. </span></span>
<span id="cb473-880"><a href="#cb473-880" aria-hidden="true" tabindex="-1"></a><span class="do">### 2. Merging cells of Patient3 into the combined batch of Patient1, Patient2 </span></span>
<span id="cb473-881"><a href="#cb473-881" aria-hidden="true" tabindex="-1"></a><span class="do">### and Patient4 resulted in the highest percentage of lost variance and the </span></span>
<span id="cb473-882"><a href="#cb473-882" aria-hidden="true" tabindex="-1"></a><span class="do">### detection of the largest batch effect. </span></span>
<span id="cb473-883"><a href="#cb473-883" aria-hidden="true" tabindex="-1"></a>merge_info[, <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"right"</span>, <span class="st">"batch.size"</span>)]</span>
<span id="cb473-884"><a href="#cb473-884" aria-hidden="true" tabindex="-1"></a>merge_info<span class="sc">$</span>lost.var</span>
<span id="cb473-885"><a href="#cb473-885" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-886"><a href="#cb473-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-887"><a href="#cb473-887" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-890"><a href="#cb473-890" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-891"><a href="#cb473-891" aria-hidden="true" tabindex="-1"></a><span class="do">### Recompute the UMAP embedding using the corrected low-dimensional </span></span>
<span id="cb473-892"><a href="#cb473-892" aria-hidden="true" tabindex="-1"></a><span class="do">### coordinates for each cell.</span></span>
<span id="cb473-893"><a href="#cb473-893" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220228</span>)</span>
<span id="cb473-894"><a href="#cb473-894" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runUMAP</span>(spe, <span class="at">dimred=</span> <span class="st">"fastMNN"</span>, <span class="at">name =</span> <span class="st">"UMAP_mnnCorrected"</span>)</span>
<span id="cb473-895"><a href="#cb473-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-896"><a href="#cb473-896" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize patient id </span></span>
<span id="cb473-897"><a href="#cb473-897" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-898"><a href="#cb473-898" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-899"><a href="#cb473-899" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-900"><a href="#cb473-900" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-901"><a href="#cb473-901" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP before correction"</span>)</span>
<span id="cb473-902"><a href="#cb473-902" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-903"><a href="#cb473-903" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-904"><a href="#cb473-904" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-905"><a href="#cb473-905" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-906"><a href="#cb473-906" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-907"><a href="#cb473-907" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP after correction"</span>)</span>
<span id="cb473-908"><a href="#cb473-908" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-909"><a href="#cb473-909" aria-hidden="true" tabindex="-1"></a><span class="do">### We observe an imperfect merging of Patient3 into all other samples. </span></span>
<span id="cb473-910"><a href="#cb473-910" aria-hidden="true" tabindex="-1"></a><span class="do">### This was already seen when displaying the merging information above. </span></span>
<span id="cb473-911"><a href="#cb473-911" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1, p2)</span>
<span id="cb473-912"><a href="#cb473-912" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-913"><a href="#cb473-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-916"><a href="#cb473-916" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-917"><a href="#cb473-917" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the expression of selected markers across all cells before and </span></span>
<span id="cb473-918"><a href="#cb473-918" aria-hidden="true" tabindex="-1"></a><span class="do">### after batch correction</span></span>
<span id="cb473-919"><a href="#cb473-919" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb473-920"><a href="#cb473-920" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Ecad"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"FOXP3"</span>, <span class="st">"CD206"</span>, <span class="st">"MPO"</span>, <span class="st">"SMA"</span>, <span class="st">"Ki67"</span></span>
<span id="cb473-921"><a href="#cb473-921" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-922"><a href="#cb473-922" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-923"><a href="#cb473-923" aria-hidden="true" tabindex="-1"></a><span class="do">### Before correction</span></span>
<span id="cb473-924"><a href="#cb473-924" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-925"><a href="#cb473-925" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb473-926"><a href="#cb473-926" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-927"><a href="#cb473-927" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-928"><a href="#cb473-928" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-929"><a href="#cb473-929" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-930"><a href="#cb473-930" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list)</span>
<span id="cb473-931"><a href="#cb473-931" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-932"><a href="#cb473-932" aria-hidden="true" tabindex="-1"></a><span class="do">### After correction</span></span>
<span id="cb473-933"><a href="#cb473-933" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-934"><a href="#cb473-934" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span id="cb473-935"><a href="#cb473-935" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-936"><a href="#cb473-936" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb473-937"><a href="#cb473-937" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-938"><a href="#cb473-938" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-939"><a href="#cb473-939" aria-hidden="true" tabindex="-1"></a><span class="do">### We observe that immune cells across patients are merged after batch </span></span>
<span id="cb473-940"><a href="#cb473-940" aria-hidden="true" tabindex="-1"></a><span class="do">### correction using fastMNN. However, the tumor cells of different patients </span></span>
<span id="cb473-941"><a href="#cb473-941" aria-hidden="true" tabindex="-1"></a><span class="do">### still cluster separately.</span></span>
<span id="cb473-942"><a href="#cb473-942" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list) </span>
<span id="cb473-943"><a href="#cb473-943" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-944"><a href="#cb473-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-945"><a href="#cb473-945" aria-hidden="true" tabindex="-1"></a><span class="fu">### harmony correction</span></span>
<span id="cb473-946"><a href="#cb473-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-949"><a href="#cb473-949" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-950"><a href="#cb473-950" aria-hidden="true" tabindex="-1"></a><span class="do">### harmony returns the corrected low-dimensional coordinates for each cell</span></span>
<span id="cb473-951"><a href="#cb473-951" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(</span>
<span id="cb473-952"><a href="#cb473-952" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-953"><a href="#cb473-953" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset_row =</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel, </span>
<span id="cb473-954"><a href="#cb473-954" aria-hidden="true" tabindex="-1"></a>    <span class="at">exprs_values =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-955"><a href="#cb473-955" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncomponents =</span> <span class="dv">30</span>,</span>
<span id="cb473-956"><a href="#cb473-956" aria-hidden="true" tabindex="-1"></a>    <span class="at">BSPARAM =</span> <span class="fu">ExactParam</span>()</span>
<span id="cb473-957"><a href="#cb473-957" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-958"><a href="#cb473-958" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-959"><a href="#cb473-959" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">230616</span>)</span>
<span id="cb473-960"><a href="#cb473-960" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">RunHarmony</span>(spe, <span class="at">group.by.vars =</span> <span class="st">"patient_id"</span>)</span>
<span id="cb473-961"><a href="#cb473-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-962"><a href="#cb473-962" aria-hidden="true" tabindex="-1"></a><span class="do">### Check that order of cells is the same</span></span>
<span id="cb473-963"><a href="#cb473-963" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(<span class="fu">colnames</span>(spe), <span class="fu">colnames</span>(out)))</span>
<span id="cb473-964"><a href="#cb473-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-965"><a href="#cb473-965" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(spe, <span class="st">"harmony"</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(out, <span class="st">"HARMONY"</span>)</span>
<span id="cb473-966"><a href="#cb473-966" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-967"><a href="#cb473-967" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-970"><a href="#cb473-970" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-971"><a href="#cb473-971" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220228</span>)</span>
<span id="cb473-972"><a href="#cb473-972" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">"harmony"</span>, <span class="at">name =</span> <span class="st">"UMAP_harmony"</span>)</span>
<span id="cb473-973"><a href="#cb473-973" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-974"><a href="#cb473-974" aria-hidden="true" tabindex="-1"></a><span class="do">### visualize patient id</span></span>
<span id="cb473-975"><a href="#cb473-975" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-976"><a href="#cb473-976" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-977"><a href="#cb473-977" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-978"><a href="#cb473-978" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-979"><a href="#cb473-979" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP before correction"</span>)</span>
<span id="cb473-980"><a href="#cb473-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-981"><a href="#cb473-981" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-982"><a href="#cb473-982" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, <span class="at">reduction.use =</span> <span class="st">"UMAP_harmony"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-983"><a href="#cb473-983" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-984"><a href="#cb473-984" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-985"><a href="#cb473-985" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP after correction"</span>)</span>
<span id="cb473-986"><a href="#cb473-986" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-987"><a href="#cb473-987" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2)</span>
<span id="cb473-988"><a href="#cb473-988" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-989"><a href="#cb473-989" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize selected marker expression</span></span>
<span id="cb473-990"><a href="#cb473-990" aria-hidden="true" tabindex="-1"></a><span class="do">### Before correction</span></span>
<span id="cb473-991"><a href="#cb473-991" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-992"><a href="#cb473-992" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>,</span>
<span id="cb473-993"><a href="#cb473-993" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-994"><a href="#cb473-994" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-995"><a href="#cb473-995" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-996"><a href="#cb473-996" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-997"><a href="#cb473-997" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list)</span>
<span id="cb473-998"><a href="#cb473-998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-999"><a href="#cb473-999" aria-hidden="true" tabindex="-1"></a><span class="do">### After correction</span></span>
<span id="cb473-1000"><a href="#cb473-1000" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-1001"><a href="#cb473-1001" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP_harmony"</span>,</span>
<span id="cb473-1002"><a href="#cb473-1002" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1003"><a href="#cb473-1003" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1004"><a href="#cb473-1004" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-1005"><a href="#cb473-1005" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1006"><a href="#cb473-1006" aria-hidden="true" tabindex="-1"></a><span class="do">### We observe a more aggressive merging of cells from different patients </span></span>
<span id="cb473-1007"><a href="#cb473-1007" aria-hidden="true" tabindex="-1"></a><span class="do">### compared to the results after fastMNN correction. Importantly, immune cell </span></span>
<span id="cb473-1008"><a href="#cb473-1008" aria-hidden="true" tabindex="-1"></a><span class="do">### and epithelial markers are expressed in distinct regions of the UMAP.</span></span>
<span id="cb473-1009"><a href="#cb473-1009" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list) </span>
<span id="cb473-1010"><a href="#cb473-1010" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1011"><a href="#cb473-1011" aria-hidden="true" tabindex="-1"></a><span class="fu">### Seurat correction</span></span>
<span id="cb473-1012"><a href="#cb473-1012" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1015"><a href="#cb473-1015" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1016"><a href="#cb473-1016" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">as.Seurat</span>(spe, <span class="at">counts =</span> <span class="st">"counts"</span>, <span class="at">data =</span> <span class="st">"exprs"</span>)</span>
<span id="cb473-1017"><a href="#cb473-1017" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(seurat_obj, <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)))</span>
<span id="cb473-1018"><a href="#cb473-1018" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(seurat_obj, <span class="at">split.by =</span> <span class="st">"patient_id"</span>)</span>
<span id="cb473-1019"><a href="#cb473-1019" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1020"><a href="#cb473-1020" aria-hidden="true" tabindex="-1"></a><span class="do">### Define the features used for integration and perform PCA on cells of each</span></span>
<span id="cb473-1021"><a href="#cb473-1021" aria-hidden="true" tabindex="-1"></a><span class="do">### patient individually</span></span>
<span id="cb473-1022"><a href="#cb473-1022" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel]</span>
<span id="cb473-1023"><a href="#cb473-1023" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1024"><a href="#cb473-1024" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb473-1025"><a href="#cb473-1025" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> seurat_list,</span>
<span id="cb473-1026"><a href="#cb473-1026" aria-hidden="true" tabindex="-1"></a>    <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb473-1027"><a href="#cb473-1027" aria-hidden="true" tabindex="-1"></a>        x <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(x, <span class="at">features =</span> features, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb473-1028"><a href="#cb473-1028" aria-hidden="true" tabindex="-1"></a>        x <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(x, <span class="at">features =</span> features, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">approx =</span> <span class="cn">FALSE</span>)</span>
<span id="cb473-1029"><a href="#cb473-1029" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(x)</span>
<span id="cb473-1030"><a href="#cb473-1030" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb473-1031"><a href="#cb473-1031" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1032"><a href="#cb473-1032" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1033"><a href="#cb473-1033" aria-hidden="true" tabindex="-1"></a>anchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(</span>
<span id="cb473-1034"><a href="#cb473-1034" aria-hidden="true" tabindex="-1"></a>    <span class="at">object.list =</span> seurat_list,</span>
<span id="cb473-1035"><a href="#cb473-1035" aria-hidden="true" tabindex="-1"></a>    <span class="at">anchor.features =</span> features,</span>
<span id="cb473-1036"><a href="#cb473-1036" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction =</span> <span class="st">"rpca"</span>,</span>
<span id="cb473-1037"><a href="#cb473-1037" aria-hidden="true" tabindex="-1"></a>    <span class="at">k.anchor =</span> <span class="dv">20</span></span>
<span id="cb473-1038"><a href="#cb473-1038" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1039"><a href="#cb473-1039" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1040"><a href="#cb473-1040" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> anchors)</span>
<span id="cb473-1041"><a href="#cb473-1041" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1042"><a href="#cb473-1042" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(combined) <span class="ot">&lt;-</span> <span class="st">"integrated"</span></span>
<span id="cb473-1043"><a href="#cb473-1043" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1044"><a href="#cb473-1044" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(combined, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb473-1045"><a href="#cb473-1045" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(combined, <span class="at">npcs =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">approx =</span> <span class="cn">FALSE</span>)</span>
<span id="cb473-1046"><a href="#cb473-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1047"><a href="#cb473-1047" aria-hidden="true" tabindex="-1"></a><span class="do">### Check that order of cells is the same</span></span>
<span id="cb473-1048"><a href="#cb473-1048" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(<span class="fu">colnames</span>(spe), <span class="fu">colnames</span>(combined)))</span>
<span id="cb473-1049"><a href="#cb473-1049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1050"><a href="#cb473-1050" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(spe, <span class="st">"seurat"</span>) <span class="ot">&lt;-</span> <span class="fu">Embeddings</span>(combined, <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb473-1051"><a href="#cb473-1051" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1052"><a href="#cb473-1052" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1055"><a href="#cb473-1055" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1056"><a href="#cb473-1056" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220228</span>)</span>
<span id="cb473-1057"><a href="#cb473-1057" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">"seurat"</span>, <span class="at">name =</span> <span class="st">"UMAP_seurat"</span>) </span>
<span id="cb473-1058"><a href="#cb473-1058" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1059"><a href="#cb473-1059" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize patient id </span></span>
<span id="cb473-1060"><a href="#cb473-1060" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1061"><a href="#cb473-1061" aria-hidden="true" tabindex="-1"></a>        spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, </span>
<span id="cb473-1062"><a href="#cb473-1062" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-1063"><a href="#cb473-1063" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-1064"><a href="#cb473-1064" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-1065"><a href="#cb473-1065" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP before correction"</span>)</span>
<span id="cb473-1066"><a href="#cb473-1066" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1067"><a href="#cb473-1067" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1068"><a href="#cb473-1068" aria-hidden="true" tabindex="-1"></a>        spe, <span class="at">var =</span> <span class="st">"patient_id"</span>, </span>
<span id="cb473-1069"><a href="#cb473-1069" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction.use =</span> <span class="st">"UMAP_seurat"</span>, <span class="at">size =</span> <span class="fl">0.2</span></span>
<span id="cb473-1070"><a href="#cb473-1070" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb473-1071"><a href="#cb473-1071" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-1072"><a href="#cb473-1072" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID on UMAP after correction"</span>)</span>
<span id="cb473-1073"><a href="#cb473-1073" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1074"><a href="#cb473-1074" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2)</span>
<span id="cb473-1075"><a href="#cb473-1075" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1076"><a href="#cb473-1076" aria-hidden="true" tabindex="-1"></a><span class="do">### Before correction</span></span>
<span id="cb473-1077"><a href="#cb473-1077" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-1078"><a href="#cb473-1078" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, </span>
<span id="cb473-1079"><a href="#cb473-1079" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1080"><a href="#cb473-1080" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb473-1081"><a href="#cb473-1081" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-1082"><a href="#cb473-1082" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list)</span>
<span id="cb473-1083"><a href="#cb473-1083" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1084"><a href="#cb473-1084" aria-hidden="true" tabindex="-1"></a><span class="do">### After correction</span></span>
<span id="cb473-1085"><a href="#cb473-1085" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">multi_dittoDimPlot</span>(</span>
<span id="cb473-1086"><a href="#cb473-1086" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> markers, <span class="at">reduction.use =</span> <span class="st">"UMAP_seurat"</span>, </span>
<span id="cb473-1087"><a href="#cb473-1087" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">size =</span> <span class="fl">0.2</span>, <span class="at">list.out =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1088"><a href="#cb473-1088" aria-hidden="true" tabindex="-1"></a>    ) </span>
<span id="cb473-1089"><a href="#cb473-1089" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1090"><a href="#cb473-1090" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(plot_list, <span class="cf">function</span>(x) x <span class="sc">+</span> <span class="fu">scale_color_viridis</span>())</span>
<span id="cb473-1091"><a href="#cb473-1091" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1092"><a href="#cb473-1092" aria-hidden="true" tabindex="-1"></a><span class="do">### Similar to the methods presented above, Seurat integrates immune cells </span></span>
<span id="cb473-1093"><a href="#cb473-1093" aria-hidden="true" tabindex="-1"></a><span class="do">### correctly. When visualizing the patient IDs, slight patient-to-patient </span></span>
<span id="cb473-1094"><a href="#cb473-1094" aria-hidden="true" tabindex="-1"></a><span class="do">### differences within tumor cells can be detected.</span></span>
<span id="cb473-1095"><a href="#cb473-1095" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list) </span>
<span id="cb473-1096"><a href="#cb473-1096" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1097"><a href="#cb473-1097" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1100"><a href="#cb473-1100" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1101"><a href="#cb473-1101" aria-hidden="true" tabindex="-1"></a><span class="do">### Save modified object</span></span>
<span id="cb473-1102"><a href="#cb473-1102" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1103"><a href="#cb473-1103" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1104"><a href="#cb473-1104" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cell phenotyping</span></span>
<span id="cb473-1105"><a href="#cb473-1105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1106"><a href="#cb473-1106" aria-hidden="true" tabindex="-1"></a>A common step during single-cell data analysis is the annotation of cells based on their phenotype. Defining cell phenotypes is often subjective and relies on previous biological knowledge. </span>
<span id="cb473-1107"><a href="#cb473-1107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1108"><a href="#cb473-1108" aria-hidden="true" tabindex="-1"></a>In highly-multiplexed imaging, target proteins or molecules are manually selected based on the biological question at hand. It narrows down the feature space and facilitates the manual annotation of clusters to derive cell phenotypes. </span>
<span id="cb473-1109"><a href="#cb473-1109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1112"><a href="#cb473-1112" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1113"><a href="#cb473-1113" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1114"><a href="#cb473-1114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1115"><a href="#cb473-1115" aria-hidden="true" tabindex="-1"></a><span class="do">###  Sample 2000 cells to visualize cluster membership.</span></span>
<span id="cb473-1116"><a href="#cb473-1116" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220619</span>)</span>
<span id="cb473-1117"><a href="#cb473-1117" aria-hidden="true" tabindex="-1"></a>cur_cells <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(spe)), <span class="dv">2000</span>)</span>
<span id="cb473-1118"><a href="#cb473-1118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1119"><a href="#cb473-1119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1120"><a href="#cb473-1120" aria-hidden="true" tabindex="-1"></a><span class="fu">### Rphenograph</span></span>
<span id="cb473-1123"><a href="#cb473-1123" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1124"><a href="#cb473-1124" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(spe, <span class="st">"exprs"</span>)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel, ])</span>
<span id="cb473-1125"><a href="#cb473-1125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1126"><a href="#cb473-1126" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">230619</span>)</span>
<span id="cb473-1127"><a href="#cb473-1127" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">Rphenograph</span>(mat, <span class="at">k =</span> <span class="dv">45</span>)</span>
<span id="cb473-1128"><a href="#cb473-1128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1129"><a href="#cb473-1129" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">membership</span>(out[[<span class="dv">2</span>]]))</span>
<span id="cb473-1130"><a href="#cb473-1130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1131"><a href="#cb473-1131" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>pg_clusters <span class="ot">&lt;-</span> clusters</span>
<span id="cb473-1132"><a href="#cb473-1132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1133"><a href="#cb473-1133" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1134"><a href="#cb473-1134" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"pg_clusters"</span>,</span>
<span id="cb473-1135"><a href="#cb473-1135" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1136"><a href="#cb473-1136" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1137"><a href="#cb473-1137" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1138"><a href="#cb473-1138" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Phenograph clusters on UMAP"</span>)</span>
<span id="cb473-1139"><a href="#cb473-1139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1140"><a href="#cb473-1140" aria-hidden="true" tabindex="-1"></a><span class="do">### We can observe that some of the clusters only contain cells of a single </span></span>
<span id="cb473-1141"><a href="#cb473-1141" aria-hidden="true" tabindex="-1"></a><span class="do">### patient. This can often be observed in the tumor compartment.</span></span>
<span id="cb473-1142"><a href="#cb473-1142" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1143"><a href="#cb473-1143" aria-hidden="true" tabindex="-1"></a>    spe[, cur_cells],</span>
<span id="cb473-1144"><a href="#cb473-1144" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1145"><a href="#cb473-1145" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1146"><a href="#cb473-1146" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-1147"><a href="#cb473-1147" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"pg_clusters"</span>, <span class="st">"patient_id"</span>),</span>
<span id="cb473-1148"><a href="#cb473-1148" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.colors =</span> <span class="fu">c</span>(</span>
<span id="cb473-1149"><a href="#cb473-1149" aria-hidden="true" tabindex="-1"></a>        <span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>pg_clusters))],</span>
<span id="cb473-1150"><a href="#cb473-1150" aria-hidden="true" tabindex="-1"></a>        <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id</span>
<span id="cb473-1151"><a href="#cb473-1151" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1152"><a href="#cb473-1152" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1153"><a href="#cb473-1153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1154"><a href="#cb473-1154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1157"><a href="#cb473-1157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1158"><a href="#cb473-1158" aria-hidden="true" tabindex="-1"></a><span class="do">### Use the integrated cells in low dimensional embedding for clustering</span></span>
<span id="cb473-1159"><a href="#cb473-1159" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(spe, <span class="st">"fastMNN"</span>)</span>
<span id="cb473-1160"><a href="#cb473-1160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1161"><a href="#cb473-1161" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">230619</span>)</span>
<span id="cb473-1162"><a href="#cb473-1162" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">Rphenograph</span>(mat, <span class="at">k =</span> <span class="dv">45</span>)</span>
<span id="cb473-1163"><a href="#cb473-1163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1164"><a href="#cb473-1164" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">membership</span>(out[[<span class="dv">2</span>]]))</span>
<span id="cb473-1165"><a href="#cb473-1165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1166"><a href="#cb473-1166" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>pg_clusters_corrected <span class="ot">&lt;-</span> clusters</span>
<span id="cb473-1167"><a href="#cb473-1167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1168"><a href="#cb473-1168" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1169"><a href="#cb473-1169" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"pg_clusters_corrected"</span>,</span>
<span id="cb473-1170"><a href="#cb473-1170" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1171"><a href="#cb473-1171" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1172"><a href="#cb473-1172" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1173"><a href="#cb473-1173" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Phenograph clusters on UMAP, integrated cells"</span>)</span>
<span id="cb473-1174"><a href="#cb473-1174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1175"><a href="#cb473-1175" aria-hidden="true" tabindex="-1"></a><span class="do">### Clustering using the integrated embedding leads to clusters that contain </span></span>
<span id="cb473-1176"><a href="#cb473-1176" aria-hidden="true" tabindex="-1"></a><span class="do">### cells of different patients. </span></span>
<span id="cb473-1177"><a href="#cb473-1177" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1178"><a href="#cb473-1178" aria-hidden="true" tabindex="-1"></a>    spe[, cur_cells],</span>
<span id="cb473-1179"><a href="#cb473-1179" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1180"><a href="#cb473-1180" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1181"><a href="#cb473-1181" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-1182"><a href="#cb473-1182" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"pg_clusters_corrected"</span>, <span class="st">"patient_id"</span>),</span>
<span id="cb473-1183"><a href="#cb473-1183" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.colors =</span> <span class="fu">c</span>(<span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>pg_clusters_corrected))],</span>
<span id="cb473-1184"><a href="#cb473-1184" aria-hidden="true" tabindex="-1"></a>        <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-1185"><a href="#cb473-1185" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1186"><a href="#cb473-1186" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1187"><a href="#cb473-1187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1188"><a href="#cb473-1188" aria-hidden="true" tabindex="-1"></a><span class="fu">### Shared nearest neighbour graph</span></span>
<span id="cb473-1189"><a href="#cb473-1189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1192"><a href="#cb473-1192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1193"><a href="#cb473-1193" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(spe, <span class="st">"exprs"</span>)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel, ])</span>
<span id="cb473-1194"><a href="#cb473-1194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1195"><a href="#cb473-1195" aria-hidden="true" tabindex="-1"></a>combinations <span class="ot">&lt;-</span> <span class="fu">clusterSweep</span>(</span>
<span id="cb473-1196"><a href="#cb473-1196" aria-hidden="true" tabindex="-1"></a>    mat,</span>
<span id="cb473-1197"><a href="#cb473-1197" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM =</span> <span class="fu">SNNGraphParam</span>(),</span>
<span id="cb473-1198"><a href="#cb473-1198" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="fu">c</span>(<span class="dv">10</span>L, <span class="dv">20</span>L),</span>
<span id="cb473-1199"><a href="#cb473-1199" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="fu">c</span>(<span class="st">"rank"</span>, <span class="st">"jaccard"</span>),</span>
<span id="cb473-1200"><a href="#cb473-1200" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster.fun =</span> <span class="st">"louvain"</span>,</span>
<span id="cb473-1201"><a href="#cb473-1201" aria-hidden="true" tabindex="-1"></a>    <span class="at">BPPARAM =</span> <span class="fu">MulticoreParam</span>(<span class="at">RNGseed =</span> <span class="dv">220427</span>)</span>
<span id="cb473-1202"><a href="#cb473-1202" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1203"><a href="#cb473-1203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1204"><a href="#cb473-1204" aria-hidden="true" tabindex="-1"></a>sil <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb473-1205"><a href="#cb473-1205" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.list</span>(combinations<span class="sc">$</span>clusters),</span>
<span id="cb473-1206"><a href="#cb473-1206" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">mean</span>(<span class="fu">approxSilhouette</span>(mat, x)<span class="sc">$</span>width),</span>
<span id="cb473-1207"><a href="#cb473-1207" aria-hidden="true" tabindex="-1"></a>    <span class="dv">0</span></span>
<span id="cb473-1208"><a href="#cb473-1208" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1209"><a href="#cb473-1209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1210"><a href="#cb473-1210" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb473-1211"><a href="#cb473-1211" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">method =</span> <span class="fu">names</span>(sil),</span>
<span id="cb473-1212"><a href="#cb473-1212" aria-hidden="true" tabindex="-1"></a>    <span class="at">sil =</span> sil)</span>
<span id="cb473-1213"><a href="#cb473-1213" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-1214"><a href="#cb473-1214" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(method, sil)) <span class="sc">+</span></span>
<span id="cb473-1215"><a href="#cb473-1215" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb473-1216"><a href="#cb473-1216" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb473-1217"><a href="#cb473-1217" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Cluster parameter combination"</span>) <span class="sc">+</span></span>
<span id="cb473-1218"><a href="#cb473-1218" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Average silhouette width"</span>)</span>
<span id="cb473-1219"><a href="#cb473-1219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1220"><a href="#cb473-1220" aria-hidden="true" tabindex="-1"></a>pur <span class="ot">&lt;-</span> <span class="fu">vapply</span>(</span>
<span id="cb473-1221"><a href="#cb473-1221" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.list</span>(combinations<span class="sc">$</span>clusters), </span>
<span id="cb473-1222"><a href="#cb473-1222" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">mean</span>(<span class="fu">neighborPurity</span>(mat, x)<span class="sc">$</span>purity), </span>
<span id="cb473-1223"><a href="#cb473-1223" aria-hidden="true" tabindex="-1"></a>    <span class="dv">0</span></span>
<span id="cb473-1224"><a href="#cb473-1224" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1225"><a href="#cb473-1225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1226"><a href="#cb473-1226" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb473-1227"><a href="#cb473-1227" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">method =</span> <span class="fu">names</span>(pur), <span class="at">pur =</span> pur)</span>
<span id="cb473-1228"><a href="#cb473-1228" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-1229"><a href="#cb473-1229" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(method, pur)) <span class="sc">+</span></span>
<span id="cb473-1230"><a href="#cb473-1230" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb473-1231"><a href="#cb473-1231" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb473-1232"><a href="#cb473-1232" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Cluster parameter combination"</span>) <span class="sc">+</span></span>
<span id="cb473-1233"><a href="#cb473-1233" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Average neighborhood purity"</span>)</span>
<span id="cb473-1234"><a href="#cb473-1234" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1235"><a href="#cb473-1235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1236"><a href="#cb473-1236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1239"><a href="#cb473-1239" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1240"><a href="#cb473-1240" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220620</span>)</span>
<span id="cb473-1241"><a href="#cb473-1241" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(</span>
<span id="cb473-1242"><a href="#cb473-1242" aria-hidden="true" tabindex="-1"></a>    spe[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel,], </span>
<span id="cb473-1243"><a href="#cb473-1243" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.type =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1244"><a href="#cb473-1244" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM =</span> <span class="fu">SNNGraphParam</span>(<span class="at">k=</span><span class="dv">20</span>, </span>
<span id="cb473-1245"><a href="#cb473-1245" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster.fun =</span> <span class="st">"louvain"</span>,</span>
<span id="cb473-1246"><a href="#cb473-1246" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"rank"</span>)</span>
<span id="cb473-1247"><a href="#cb473-1247" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1248"><a href="#cb473-1248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1249"><a href="#cb473-1249" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>nn_clusters <span class="ot">&lt;-</span> clusters</span>
<span id="cb473-1250"><a href="#cb473-1250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1251"><a href="#cb473-1251" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1252"><a href="#cb473-1252" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"nn_clusters"</span>, </span>
<span id="cb473-1253"><a href="#cb473-1253" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1254"><a href="#cb473-1254" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1255"><a href="#cb473-1255" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-1256"><a href="#cb473-1256" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"SNN clusters on UMAP"</span>)</span>
<span id="cb473-1257"><a href="#cb473-1257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1258"><a href="#cb473-1258" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1259"><a href="#cb473-1259" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1260"><a href="#cb473-1260" aria-hidden="true" tabindex="-1"></a>             <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1261"><a href="#cb473-1261" aria-hidden="true" tabindex="-1"></a>             <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1262"><a href="#cb473-1262" aria-hidden="true" tabindex="-1"></a>             <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1263"><a href="#cb473-1263" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"nn_clusters"</span>, <span class="st">"patient_id"</span>),</span>
<span id="cb473-1264"><a href="#cb473-1264" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.colors =</span> <span class="fu">c</span>(<span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>nn_clusters))],</span>
<span id="cb473-1265"><a href="#cb473-1265" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-1266"><a href="#cb473-1266" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1267"><a href="#cb473-1267" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1268"><a href="#cb473-1268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1271"><a href="#cb473-1271" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1272"><a href="#cb473-1272" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220621</span>)</span>
<span id="cb473-1273"><a href="#cb473-1273" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(</span>
<span id="cb473-1274"><a href="#cb473-1274" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-1275"><a href="#cb473-1275" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred =</span> <span class="st">"fastMNN"</span>, </span>
<span id="cb473-1276"><a href="#cb473-1276" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM =</span> <span class="fu">SNNGraphParam</span>(<span class="at">k =</span> <span class="dv">20</span>, </span>
<span id="cb473-1277"><a href="#cb473-1277" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster.fun =</span> <span class="st">"louvain"</span>,</span>
<span id="cb473-1278"><a href="#cb473-1278" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"rank"</span>)</span>
<span id="cb473-1279"><a href="#cb473-1279" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1280"><a href="#cb473-1280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1281"><a href="#cb473-1281" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>nn_clusters_corrected <span class="ot">&lt;-</span> clusters</span>
<span id="cb473-1282"><a href="#cb473-1282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1283"><a href="#cb473-1283" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1284"><a href="#cb473-1284" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"nn_clusters_corrected"</span>, </span>
<span id="cb473-1285"><a href="#cb473-1285" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1286"><a href="#cb473-1286" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1287"><a href="#cb473-1287" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1288"><a href="#cb473-1288" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"SNN clusters on UMAP, integrated cells"</span>)</span>
<span id="cb473-1289"><a href="#cb473-1289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1290"><a href="#cb473-1290" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1291"><a href="#cb473-1291" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1292"><a href="#cb473-1292" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1293"><a href="#cb473-1293" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1294"><a href="#cb473-1294" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1295"><a href="#cb473-1295" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"nn_clusters_corrected"</span>,<span class="st">"patient_id"</span>),</span>
<span id="cb473-1296"><a href="#cb473-1296" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.colors =</span> <span class="fu">c</span>(<span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>nn_clusters_corrected))],</span>
<span id="cb473-1297"><a href="#cb473-1297" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-1298"><a href="#cb473-1298" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1299"><a href="#cb473-1299" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1300"><a href="#cb473-1300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1301"><a href="#cb473-1301" aria-hidden="true" tabindex="-1"></a><span class="fu">### Self organizing maps</span></span>
<span id="cb473-1302"><a href="#cb473-1302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1305"><a href="#cb473-1305" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1306"><a href="#cb473-1306" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FlowSOM and ConsensusClusterPlus clustering</span></span>
<span id="cb473-1307"><a href="#cb473-1307" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220410</span>)</span>
<span id="cb473-1308"><a href="#cb473-1308" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> CATALYST<span class="sc">::</span><span class="fu">cluster</span>(spe, </span>
<span id="cb473-1309"><a href="#cb473-1309" aria-hidden="true" tabindex="-1"></a>               <span class="at">features =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1310"><a href="#cb473-1310" aria-hidden="true" tabindex="-1"></a>               <span class="at">maxK =</span> <span class="dv">30</span>)</span>
<span id="cb473-1311"><a href="#cb473-1311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1312"><a href="#cb473-1312" aria-hidden="true" tabindex="-1"></a><span class="co"># Assess cluster stability</span></span>
<span id="cb473-1313"><a href="#cb473-1313" aria-hidden="true" tabindex="-1"></a><span class="fu">delta_area</span>(spe)</span>
<span id="cb473-1314"><a href="#cb473-1314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1315"><a href="#cb473-1315" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>som_clusters <span class="ot">&lt;-</span> <span class="fu">cluster_ids</span>(spe, <span class="st">"meta13"</span>)</span>
<span id="cb473-1316"><a href="#cb473-1316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1317"><a href="#cb473-1317" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1318"><a href="#cb473-1318" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"som_clusters"</span>, </span>
<span id="cb473-1319"><a href="#cb473-1319" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1320"><a href="#cb473-1320" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1321"><a href="#cb473-1321" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-1322"><a href="#cb473-1322" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"SOM clusters on UMAP"</span>)</span>
<span id="cb473-1323"><a href="#cb473-1323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1324"><a href="#cb473-1324" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1325"><a href="#cb473-1325" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1326"><a href="#cb473-1326" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1327"><a href="#cb473-1327" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1328"><a href="#cb473-1328" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1329"><a href="#cb473-1329" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"som_clusters"</span>, <span class="st">"patient_id"</span>),</span>
<span id="cb473-1330"><a href="#cb473-1330" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.colors =</span> <span class="fu">c</span>(<span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>som_clusters))],</span>
<span id="cb473-1331"><a href="#cb473-1331" aria-hidden="true" tabindex="-1"></a>    <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-1332"><a href="#cb473-1332" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1333"><a href="#cb473-1333" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1334"><a href="#cb473-1334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1337"><a href="#cb473-1337" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1338"><a href="#cb473-1338" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kohonen)</span>
<span id="cb473-1339"><a href="#cb473-1339" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ConsensusClusterPlus)</span>
<span id="cb473-1340"><a href="#cb473-1340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1341"><a href="#cb473-1341" aria-hidden="true" tabindex="-1"></a><span class="do">### Select integrated cells</span></span>
<span id="cb473-1342"><a href="#cb473-1342" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(spe, <span class="st">"fastMNN"</span>)</span>
<span id="cb473-1343"><a href="#cb473-1343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1344"><a href="#cb473-1344" aria-hidden="true" tabindex="-1"></a><span class="do">### Perform SOM clustering</span></span>
<span id="cb473-1345"><a href="#cb473-1345" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220410</span>)</span>
<span id="cb473-1346"><a href="#cb473-1346" aria-hidden="true" tabindex="-1"></a>som.out <span class="ot">&lt;-</span> <span class="fu">clusterRows</span>(mat, <span class="fu">SomParam</span>(<span class="dv">100</span>), <span class="at">full =</span> <span class="cn">TRUE</span>)</span>
<span id="cb473-1347"><a href="#cb473-1347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1348"><a href="#cb473-1348" aria-hidden="true" tabindex="-1"></a><span class="do">### Cluster the 100 SOM codes into larger clusters</span></span>
<span id="cb473-1349"><a href="#cb473-1349" aria-hidden="true" tabindex="-1"></a>ccp <span class="ot">&lt;-</span> <span class="fu">ConsensusClusterPlus</span>(</span>
<span id="cb473-1350"><a href="#cb473-1350" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>(som.out<span class="sc">$</span>objects<span class="sc">$</span>som<span class="sc">$</span>codes[[<span class="dv">1</span>]]),</span>
<span id="cb473-1351"><a href="#cb473-1351" aria-hidden="true" tabindex="-1"></a>    <span class="at">maxK =</span> <span class="dv">30</span>,</span>
<span id="cb473-1352"><a href="#cb473-1352" aria-hidden="true" tabindex="-1"></a>    <span class="at">reps =</span> <span class="dv">100</span>, </span>
<span id="cb473-1353"><a href="#cb473-1353" aria-hidden="true" tabindex="-1"></a>    <span class="at">distance =</span> <span class="st">"euclidean"</span>, </span>
<span id="cb473-1354"><a href="#cb473-1354" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">220410</span>, </span>
<span id="cb473-1355"><a href="#cb473-1355" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot =</span> <span class="cn">NULL</span></span>
<span id="cb473-1356"><a href="#cb473-1356" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1357"><a href="#cb473-1357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1358"><a href="#cb473-1358" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize delta area plot</span></span>
<span id="cb473-1359"><a href="#cb473-1359" aria-hidden="true" tabindex="-1"></a>CATALYST<span class="sc">:::</span><span class="fu">.plot_delta_area</span>(ccp)</span>
<span id="cb473-1360"><a href="#cb473-1360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1361"><a href="#cb473-1361" aria-hidden="true" tabindex="-1"></a><span class="do">### Link ConsensusClusterPlus clusters with SOM codes and save in object</span></span>
<span id="cb473-1362"><a href="#cb473-1362" aria-hidden="true" tabindex="-1"></a>som.cluster <span class="ot">&lt;-</span> ccp[[<span class="dv">16</span>]][[<span class="st">"consensusClass"</span>]][som.out<span class="sc">$</span>clusters]</span>
<span id="cb473-1363"><a href="#cb473-1363" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>som_clusters_corrected <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(som.cluster)</span>
<span id="cb473-1364"><a href="#cb473-1364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1365"><a href="#cb473-1365" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1366"><a href="#cb473-1366" aria-hidden="true" tabindex="-1"></a>    spe, <span class="at">var =</span> <span class="st">"som_clusters_corrected"</span>, </span>
<span id="cb473-1367"><a href="#cb473-1367" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1368"><a href="#cb473-1368" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1369"><a href="#cb473-1369" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb473-1370"><a href="#cb473-1370" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"SOM clusters on UMAP, integrated cells"</span>)</span>
<span id="cb473-1371"><a href="#cb473-1371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1372"><a href="#cb473-1372" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1373"><a href="#cb473-1373" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1374"><a href="#cb473-1374" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>use_channel],</span>
<span id="cb473-1375"><a href="#cb473-1375" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1376"><a href="#cb473-1376" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1377"><a href="#cb473-1377" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"som_clusters_corrected"</span>,<span class="st">"patient_id"</span>),</span>
<span id="cb473-1378"><a href="#cb473-1378" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.colors =</span> <span class="fu">c</span>(<span class="fu">dittoColors</span>(<span class="dv">1</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(spe<span class="sc">$</span>som_clusters_corrected))],</span>
<span id="cb473-1379"><a href="#cb473-1379" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-1380"><a href="#cb473-1380" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1381"><a href="#cb473-1381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1382"><a href="#cb473-1382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1383"><a href="#cb473-1383" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compare between clustering approaches</span></span>
<span id="cb473-1384"><a href="#cb473-1384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1387"><a href="#cb473-1387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1388"><a href="#cb473-1388" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb473-1389"><a href="#cb473-1389" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb473-1390"><a href="#cb473-1390" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb473-1391"><a href="#cb473-1391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1392"><a href="#cb473-1392" aria-hidden="true" tabindex="-1"></a>tab1 <span class="ot">&lt;-</span> <span class="fu">table</span>(</span>
<span id="cb473-1393"><a href="#cb473-1393" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"Rphenograph"</span>, spe<span class="sc">$</span>pg_clusters), </span>
<span id="cb473-1394"><a href="#cb473-1394" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"SNN"</span>, spe<span class="sc">$</span>nn_clusters)</span>
<span id="cb473-1395"><a href="#cb473-1395" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1396"><a href="#cb473-1396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1397"><a href="#cb473-1397" aria-hidden="true" tabindex="-1"></a>tab2 <span class="ot">&lt;-</span> <span class="fu">table</span>(</span>
<span id="cb473-1398"><a href="#cb473-1398" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"Rphenograph"</span>, spe<span class="sc">$</span>pg_clusters), </span>
<span id="cb473-1399"><a href="#cb473-1399" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"SOM"</span>, spe<span class="sc">$</span>som_clusters)</span>
<span id="cb473-1400"><a href="#cb473-1400" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1401"><a href="#cb473-1401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1402"><a href="#cb473-1402" aria-hidden="true" tabindex="-1"></a>tab3 <span class="ot">&lt;-</span> <span class="fu">table</span>(</span>
<span id="cb473-1403"><a href="#cb473-1403" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"SNN"</span>, spe<span class="sc">$</span>nn_clusters), </span>
<span id="cb473-1404"><a href="#cb473-1404" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(<span class="st">"SOM"</span>, spe<span class="sc">$</span>som_clusters)</span>
<span id="cb473-1405"><a href="#cb473-1405" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1406"><a href="#cb473-1406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1407"><a href="#cb473-1407" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab1 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1408"><a href="#cb473-1408" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab2 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1409"><a href="#cb473-1409" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab3 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1410"><a href="#cb473-1410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1411"><a href="#cb473-1411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1412"><a href="#cb473-1412" aria-hidden="true" tabindex="-1"></a><span class="fu">### Further clustering notes</span></span>
<span id="cb473-1415"><a href="#cb473-1415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1416"><a href="#cb473-1416" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb473-1417"><a href="#cb473-1417" aria-hidden="true" tabindex="-1"></a>cluster_celltype <span class="ot">&lt;-</span> <span class="fu">recode</span>(spe<span class="sc">$</span>nn_clusters_corrected,</span>
<span id="cb473-1418"><a href="#cb473-1418" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"Tumor_proliferating"</span>,</span>
<span id="cb473-1419"><a href="#cb473-1419" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Myeloid"</span>,</span>
<span id="cb473-1420"><a href="#cb473-1420" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"Tumor"</span>,</span>
<span id="cb473-1421"><a href="#cb473-1421" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"Tumor"</span>,</span>
<span id="cb473-1422"><a href="#cb473-1422" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"Stroma"</span>,</span>
<span id="cb473-1423"><a href="#cb473-1423" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"Proliferating"</span>,</span>
<span id="cb473-1424"><a href="#cb473-1424" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"Myeloid"</span>,</span>
<span id="cb473-1425"><a href="#cb473-1425" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"Plasma_cell"</span>,</span>
<span id="cb473-1426"><a href="#cb473-1426" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"CD8"</span>,</span>
<span id="cb473-1427"><a href="#cb473-1427" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"CD4"</span>,</span>
<span id="cb473-1428"><a href="#cb473-1428" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"Neutrophil"</span>,</span>
<span id="cb473-1429"><a href="#cb473-1429" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"Bcell"</span>,</span>
<span id="cb473-1430"><a href="#cb473-1430" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"Stroma"</span>)</span>
<span id="cb473-1431"><a href="#cb473-1431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1432"><a href="#cb473-1432" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>cluster_celltype <span class="ot">&lt;-</span> cluster_celltype</span>
<span id="cb473-1433"><a href="#cb473-1433" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1434"><a href="#cb473-1434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1437"><a href="#cb473-1437" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1438"><a href="#cb473-1438" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1439"><a href="#cb473-1439" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1440"><a href="#cb473-1440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1441"><a href="#cb473-1441" aria-hidden="true" tabindex="-1"></a><span class="fu">### Classfication approach</span></span>
<span id="cb473-1442"><a href="#cb473-1442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1445"><a href="#cb473-1445" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1446"><a href="#cb473-1446" aria-hidden="true" tabindex="-1"></a><span class="do">### Manual labeling of cells</span></span>
<span id="cb473-1447"><a href="#cb473-1447" aria-hidden="true" tabindex="-1"></a><span class="co"># if (interactive()) {</span></span>
<span id="cb473-1448"><a href="#cb473-1448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1449"><a href="#cb473-1449" aria-hidden="true" tabindex="-1"></a><span class="co">#     images &lt;- qread(here(dir, "data/images.qs"))</span></span>
<span id="cb473-1450"><a href="#cb473-1450" aria-hidden="true" tabindex="-1"></a><span class="co">#     masks &lt;- qread(here(dir, "data/masks.qs"))</span></span>
<span id="cb473-1451"><a href="#cb473-1451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1452"><a href="#cb473-1452" aria-hidden="true" tabindex="-1"></a><span class="co">#     cytomapperShiny(object = spe, mask = masks, image = images,</span></span>
<span id="cb473-1453"><a href="#cb473-1453" aria-hidden="true" tabindex="-1"></a><span class="co">#                     cell_id = "ObjectNumber", img_id = "sample_id")</span></span>
<span id="cb473-1454"><a href="#cb473-1454" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb473-1455"><a href="#cb473-1455" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1456"><a href="#cb473-1456" aria-hidden="true" tabindex="-1"></a><span class="do">### Define color vectors</span></span>
<span id="cb473-1457"><a href="#cb473-1457" aria-hidden="true" tabindex="-1"></a>celltype <span class="ot">&lt;-</span> <span class="fu">setNames</span>(</span>
<span id="cb473-1458"><a href="#cb473-1458" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"#3F1B03"</span>, <span class="st">"#F4AD31"</span>, <span class="st">"#894F36"</span>, <span class="st">"#1C750C"</span>, <span class="st">"#EF8ECC"</span>,</span>
<span id="cb473-1459"><a href="#cb473-1459" aria-hidden="true" tabindex="-1"></a>        <span class="st">"#6471E2"</span>, <span class="st">"#4DB23B"</span>, <span class="st">"grey"</span>, <span class="st">"#F4800C"</span>, <span class="st">"#BF0A3D"</span>, <span class="st">"#066970"</span></span>
<span id="cb473-1460"><a href="#cb473-1460" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb473-1461"><a href="#cb473-1461" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"Tumor"</span>, <span class="st">"Stroma"</span>, <span class="st">"Myeloid"</span>, <span class="st">"CD8"</span>, <span class="st">"Plasma_cell"</span>,</span>
<span id="cb473-1462"><a href="#cb473-1462" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Treg"</span>, <span class="st">"CD4"</span>, <span class="st">"undefined"</span>, <span class="st">"BnTcell"</span>, <span class="st">"Bcell"</span>, <span class="st">"Neutrophil"</span>)</span>
<span id="cb473-1463"><a href="#cb473-1463" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1464"><a href="#cb473-1464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1465"><a href="#cb473-1465" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype <span class="ot">&lt;-</span> celltype</span>
<span id="cb473-1466"><a href="#cb473-1466" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1467"><a href="#cb473-1467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1470"><a href="#cb473-1470" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1471"><a href="#cb473-1471" aria-hidden="true" tabindex="-1"></a><span class="do">### Read in and consolidate labeled data</span></span>
<span id="cb473-1472"><a href="#cb473-1472" aria-hidden="true" tabindex="-1"></a>label_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(</span>
<span id="cb473-1473"><a href="#cb473-1473" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(dir, <span class="st">"data/gated_cells"</span>), </span>
<span id="cb473-1474"><a href="#cb473-1474" aria-hidden="true" tabindex="-1"></a>    <span class="at">full.names =</span> <span class="cn">TRUE</span>, <span class="at">pattern =</span> <span class="st">".rds$"</span></span>
<span id="cb473-1475"><a href="#cb473-1475" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1476"><a href="#cb473-1476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1477"><a href="#cb473-1477" aria-hidden="true" tabindex="-1"></a><span class="do">### Read in SPE objects</span></span>
<span id="cb473-1478"><a href="#cb473-1478" aria-hidden="true" tabindex="-1"></a>spes <span class="ot">&lt;-</span> <span class="fu">lapply</span>(label_files, readRDS)</span>
<span id="cb473-1479"><a href="#cb473-1479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1480"><a href="#cb473-1480" aria-hidden="true" tabindex="-1"></a><span class="do">### Merge SPE objects</span></span>
<span id="cb473-1481"><a href="#cb473-1481" aria-hidden="true" tabindex="-1"></a>concat_spe <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"cbind"</span>, spes)</span>
<span id="cb473-1482"><a href="#cb473-1482" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1483"><a href="#cb473-1483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1486"><a href="#cb473-1486" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1487"><a href="#cb473-1487" aria-hidden="true" tabindex="-1"></a>filter_labels <span class="ot">&lt;-</span> <span class="cf">function</span>(object, </span>
<span id="cb473-1488"><a href="#cb473-1488" aria-hidden="true" tabindex="-1"></a>                          <span class="at">label =</span> <span class="st">"cytomapper_CellLabel"</span>) {</span>
<span id="cb473-1489"><a href="#cb473-1489" aria-hidden="true" tabindex="-1"></a>    cur_tab <span class="ot">&lt;-</span> <span class="fu">unclass</span>(<span class="fu">table</span>(<span class="fu">colnames</span>(object), object[[label]]))</span>
<span id="cb473-1490"><a href="#cb473-1490" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-1491"><a href="#cb473-1491" aria-hidden="true" tabindex="-1"></a>    cur_labels <span class="ot">&lt;-</span> <span class="fu">colnames</span>(cur_tab)[<span class="fu">apply</span>(cur_tab, <span class="dv">1</span>, which.max)]</span>
<span id="cb473-1492"><a href="#cb473-1492" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(cur_labels) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(cur_tab)</span>
<span id="cb473-1493"><a href="#cb473-1493" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-1494"><a href="#cb473-1494" aria-hidden="true" tabindex="-1"></a>    cur_labels <span class="ot">&lt;-</span> cur_labels[<span class="fu">rowSums</span>(cur_tab) <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb473-1495"><a href="#cb473-1495" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb473-1496"><a href="#cb473-1496" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(cur_labels)</span>
<span id="cb473-1497"><a href="#cb473-1497" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb473-1498"><a href="#cb473-1498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1499"><a href="#cb473-1499" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">filter_labels</span>(concat_spe)</span>
<span id="cb473-1500"><a href="#cb473-1500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1501"><a href="#cb473-1501" aria-hidden="true" tabindex="-1"></a>cur_spe <span class="ot">&lt;-</span> concat_spe[,concat_spe<span class="sc">$</span>cytomapper_CellLabel <span class="sc">!=</span> <span class="st">"Tumor"</span>]</span>
<span id="cb473-1502"><a href="#cb473-1502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1503"><a href="#cb473-1503" aria-hidden="true" tabindex="-1"></a>non_tumor_labels <span class="ot">&lt;-</span> <span class="fu">filter_labels</span>(cur_spe)</span>
<span id="cb473-1504"><a href="#cb473-1504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1505"><a href="#cb473-1505" aria-hidden="true" tabindex="-1"></a>additional_cells <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="fu">names</span>(non_tumor_labels), <span class="fu">names</span>(labels))</span>
<span id="cb473-1506"><a href="#cb473-1506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1507"><a href="#cb473-1507" aria-hidden="true" tabindex="-1"></a>final_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(labels, non_tumor_labels[additional_cells])</span>
<span id="cb473-1508"><a href="#cb473-1508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1509"><a href="#cb473-1509" aria-hidden="true" tabindex="-1"></a><span class="do">### Transfer labels to SPE object</span></span>
<span id="cb473-1510"><a href="#cb473-1510" aria-hidden="true" tabindex="-1"></a>spe_labels <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"unlabeled"</span>, <span class="fu">ncol</span>(spe))</span>
<span id="cb473-1511"><a href="#cb473-1511" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(spe_labels) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(spe)</span>
<span id="cb473-1512"><a href="#cb473-1512" aria-hidden="true" tabindex="-1"></a>spe_labels[<span class="fu">names</span>(final_labels)] <span class="ot">&lt;-</span> final_labels</span>
<span id="cb473-1513"><a href="#cb473-1513" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>cell_labels <span class="ot">&lt;-</span> spe_labels</span>
<span id="cb473-1514"><a href="#cb473-1514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1515"><a href="#cb473-1515" aria-hidden="true" tabindex="-1"></a><span class="do">### Number of cells labeled per patient</span></span>
<span id="cb473-1516"><a href="#cb473-1516" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(spe<span class="sc">$</span>cell_labels, spe<span class="sc">$</span>patient_id)</span>
<span id="cb473-1517"><a href="#cb473-1517" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1518"><a href="#cb473-1518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1521"><a href="#cb473-1521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1522"><a href="#cb473-1522" aria-hidden="true" tabindex="-1"></a><span class="do">### Split between labeled and unlabeled cells</span></span>
<span id="cb473-1523"><a href="#cb473-1523" aria-hidden="true" tabindex="-1"></a>lab_spe <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>cell_labels <span class="sc">!=</span> <span class="st">"unlabeled"</span>]</span>
<span id="cb473-1524"><a href="#cb473-1524" aria-hidden="true" tabindex="-1"></a>unlab_spe <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>cell_labels <span class="sc">==</span> <span class="st">"unlabeled"</span>]</span>
<span id="cb473-1525"><a href="#cb473-1525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1526"><a href="#cb473-1526" aria-hidden="true" tabindex="-1"></a><span class="do">### Randomly split into train and test data</span></span>
<span id="cb473-1527"><a href="#cb473-1527" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">221029</span>)</span>
<span id="cb473-1528"><a href="#cb473-1528" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(<span class="fu">factor</span>(lab_spe<span class="sc">$</span>cell_labels), <span class="at">p =</span> <span class="fl">0.75</span>)</span>
<span id="cb473-1529"><a href="#cb473-1529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1530"><a href="#cb473-1530" aria-hidden="true" tabindex="-1"></a>train_spe <span class="ot">&lt;-</span> lab_spe[,trainIndex<span class="sc">$</span>Resample1]</span>
<span id="cb473-1531"><a href="#cb473-1531" aria-hidden="true" tabindex="-1"></a>test_spe <span class="ot">&lt;-</span> lab_spe[,<span class="sc">-</span>trainIndex<span class="sc">$</span>Resample1]</span>
<span id="cb473-1532"><a href="#cb473-1532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1533"><a href="#cb473-1533" aria-hidden="true" tabindex="-1"></a><span class="do">### Define fit parameters for 5-fold cross validation</span></span>
<span id="cb473-1534"><a href="#cb473-1534" aria-hidden="true" tabindex="-1"></a>fitControl <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">"cv"</span>,<span class="at">number =</span> <span class="dv">5</span>)</span>
<span id="cb473-1535"><a href="#cb473-1535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1536"><a href="#cb473-1536" aria-hidden="true" tabindex="-1"></a><span class="do">### Select the arsinh-transformed counts for training</span></span>
<span id="cb473-1537"><a href="#cb473-1537" aria-hidden="true" tabindex="-1"></a>cur_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(train_spe, <span class="st">"exprs"</span>)[<span class="fu">rowData</span>(train_spe)<span class="sc">$</span>use_channel,])</span>
<span id="cb473-1538"><a href="#cb473-1538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1539"><a href="#cb473-1539" aria-hidden="true" tabindex="-1"></a><span class="do">### Train a random forest classifier</span></span>
<span id="cb473-1540"><a href="#cb473-1540" aria-hidden="true" tabindex="-1"></a>rffit <span class="ot">&lt;-</span> <span class="fu">train</span>(</span>
<span id="cb473-1541"><a href="#cb473-1541" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> cur_mat, </span>
<span id="cb473-1542"><a href="#cb473-1542" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">factor</span>(train_spe<span class="sc">$</span>cell_labels),</span>
<span id="cb473-1543"><a href="#cb473-1543" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"rf"</span>, <span class="at">ntree =</span> <span class="dv">1000</span>,</span>
<span id="cb473-1544"><a href="#cb473-1544" aria-hidden="true" tabindex="-1"></a>    <span class="at">tuneLength =</span> <span class="dv">5</span>,</span>
<span id="cb473-1545"><a href="#cb473-1545" aria-hidden="true" tabindex="-1"></a>    <span class="at">trControl =</span> fitControl</span>
<span id="cb473-1546"><a href="#cb473-1546" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1547"><a href="#cb473-1547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1548"><a href="#cb473-1548" aria-hidden="true" tabindex="-1"></a>rffit</span>
<span id="cb473-1549"><a href="#cb473-1549" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1550"><a href="#cb473-1550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1553"><a href="#cb473-1553" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1554"><a href="#cb473-1554" aria-hidden="true" tabindex="-1"></a><span class="do">### Classifier performance</span></span>
<span id="cb473-1555"><a href="#cb473-1555" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(rffit) <span class="sc">+</span> </span>
<span id="cb473-1556"><a href="#cb473-1556" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="at">data =</span> rffit<span class="sc">$</span>results,</span>
<span id="cb473-1557"><a href="#cb473-1557" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">ymin =</span> Accuracy <span class="sc">-</span> AccuracySD,</span>
<span id="cb473-1558"><a href="#cb473-1558" aria-hidden="true" tabindex="-1"></a>                    <span class="at">ymax =</span> Accuracy <span class="sc">+</span> AccuracySD),</span>
<span id="cb473-1559"><a href="#cb473-1559" aria-hidden="true" tabindex="-1"></a>                <span class="at">width =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb473-1560"><a href="#cb473-1560" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>)</span>
<span id="cb473-1561"><a href="#cb473-1561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1562"><a href="#cb473-1562" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the variable importance of the classifier.</span></span>
<span id="cb473-1563"><a href="#cb473-1563" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">varImp</span>(rffit))</span>
<span id="cb473-1564"><a href="#cb473-1564" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1565"><a href="#cb473-1565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1568"><a href="#cb473-1568" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1569"><a href="#cb473-1569" aria-hidden="true" tabindex="-1"></a><span class="do">### Select the arsinh-transformed counts of the test data</span></span>
<span id="cb473-1570"><a href="#cb473-1570" aria-hidden="true" tabindex="-1"></a>cur_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(test_spe, <span class="st">"exprs"</span>)[<span class="fu">rowData</span>(test_spe)<span class="sc">$</span>use_channel,])</span>
<span id="cb473-1571"><a href="#cb473-1571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1572"><a href="#cb473-1572" aria-hidden="true" tabindex="-1"></a><span class="do">### Predict the cell phenotype labels of the test data</span></span>
<span id="cb473-1573"><a href="#cb473-1573" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">231019</span>)</span>
<span id="cb473-1574"><a href="#cb473-1574" aria-hidden="true" tabindex="-1"></a>cur_pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(rffit, <span class="at">newdata =</span> cur_mat)</span>
<span id="cb473-1575"><a href="#cb473-1575" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1576"><a href="#cb473-1576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1579"><a href="#cb473-1579" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1580"><a href="#cb473-1580" aria-hidden="true" tabindex="-1"></a>cm <span class="ot">&lt;-</span> <span class="fu">confusionMatrix</span>(</span>
<span id="cb473-1581"><a href="#cb473-1581" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> cur_pred, </span>
<span id="cb473-1582"><a href="#cb473-1582" aria-hidden="true" tabindex="-1"></a>    <span class="at">reference =</span> <span class="fu">factor</span>(test_spe<span class="sc">$</span>cell_labels), </span>
<span id="cb473-1583"><a href="#cb473-1583" aria-hidden="true" tabindex="-1"></a>    <span class="at">mode =</span> <span class="st">"everything"</span></span>
<span id="cb473-1584"><a href="#cb473-1584" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1585"><a href="#cb473-1585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1586"><a href="#cb473-1586" aria-hidden="true" tabindex="-1"></a>cm</span>
<span id="cb473-1587"><a href="#cb473-1587" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1588"><a href="#cb473-1588" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1591"><a href="#cb473-1591" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1592"><a href="#cb473-1592" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(cm<span class="sc">$</span>byClass) <span class="sc">|&gt;</span></span>
<span id="cb473-1593"><a href="#cb473-1593" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">class =</span> <span class="fu">sub</span>(<span class="st">"Class: "</span>, <span class="st">""</span>, <span class="fu">rownames</span>(cm<span class="sc">$</span>byClass))) <span class="sc">|&gt;</span></span>
<span id="cb473-1594"><a href="#cb473-1594" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb473-1595"><a href="#cb473-1595" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="dv">1</span> <span class="sc">-</span> Specificity, Sensitivity, </span>
<span id="cb473-1596"><a href="#cb473-1596" aria-hidden="true" tabindex="-1"></a>                 <span class="at">size =</span> Detection.Rate,</span>
<span id="cb473-1597"><a href="#cb473-1597" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> class),</span>
<span id="cb473-1598"><a href="#cb473-1598" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> <span class="dv">21</span>) <span class="sc">+</span> </span>
<span id="cb473-1599"><a href="#cb473-1599" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype) <span class="sc">+</span></span>
<span id="cb473-1600"><a href="#cb473-1600" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> </span>
<span id="cb473-1601"><a href="#cb473-1601" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Sensitivity (TPR)"</span>) <span class="sc">+</span></span>
<span id="cb473-1602"><a href="#cb473-1602" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"1 - Specificity (FPR)"</span>)</span>
<span id="cb473-1603"><a href="#cb473-1603" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1604"><a href="#cb473-1604" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1607"><a href="#cb473-1607" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1608"><a href="#cb473-1608" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">231019</span>)</span>
<span id="cb473-1609"><a href="#cb473-1609" aria-hidden="true" tabindex="-1"></a>cur_pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(rffit, </span>
<span id="cb473-1610"><a href="#cb473-1610" aria-hidden="true" tabindex="-1"></a>                    <span class="at">newdata =</span> cur_mat, </span>
<span id="cb473-1611"><a href="#cb473-1611" aria-hidden="true" tabindex="-1"></a>                    <span class="at">type =</span> <span class="st">"prob"</span>)</span>
<span id="cb473-1612"><a href="#cb473-1612" aria-hidden="true" tabindex="-1"></a>cur_pred<span class="sc">$</span>truth <span class="ot">&lt;-</span> <span class="fu">factor</span>(test_spe<span class="sc">$</span>cell_labels)</span>
<span id="cb473-1613"><a href="#cb473-1613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1614"><a href="#cb473-1614" aria-hidden="true" tabindex="-1"></a>cur_pred <span class="sc">|&gt;</span></span>
<span id="cb473-1615"><a href="#cb473-1615" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> Bcell<span class="sc">:</span>Tumor) <span class="sc">|&gt;</span></span>
<span id="cb473-1616"><a href="#cb473-1616" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb473-1617"><a href="#cb473-1617" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> name, <span class="at">y =</span> value, <span class="at">fill =</span> name), <span class="at">outlier.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb473-1618"><a href="#cb473-1618" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(. <span class="sc">~</span> truth, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb473-1619"><a href="#cb473-1619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)  <span class="sc">+</span></span>
<span id="cb473-1620"><a href="#cb473-1620" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb473-1621"><a href="#cb473-1621" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb473-1622"><a href="#cb473-1622" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)</span>
<span id="cb473-1623"><a href="#cb473-1623" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1624"><a href="#cb473-1624" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1625"><a href="#cb473-1625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1628"><a href="#cb473-1628" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1629"><a href="#cb473-1629" aria-hidden="true" tabindex="-1"></a><span class="do">### Classification of new data</span></span>
<span id="cb473-1630"><a href="#cb473-1630" aria-hidden="true" tabindex="-1"></a><span class="do">### Select the arsinh-transformed counts of the unlabeled data for prediction</span></span>
<span id="cb473-1631"><a href="#cb473-1631" aria-hidden="true" tabindex="-1"></a>cur_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(unlab_spe, <span class="st">"exprs"</span>)[<span class="fu">rowData</span>(unlab_spe)<span class="sc">$</span>use_channel,])</span>
<span id="cb473-1632"><a href="#cb473-1632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1633"><a href="#cb473-1633" aria-hidden="true" tabindex="-1"></a><span class="do">### Predict the cell phenotype labels of the unlabeled data</span></span>
<span id="cb473-1634"><a href="#cb473-1634" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">231014</span>)</span>
<span id="cb473-1635"><a href="#cb473-1635" aria-hidden="true" tabindex="-1"></a>cell_class <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">predict</span>(</span>
<span id="cb473-1636"><a href="#cb473-1636" aria-hidden="true" tabindex="-1"></a>    rffit, </span>
<span id="cb473-1637"><a href="#cb473-1637" aria-hidden="true" tabindex="-1"></a>    <span class="at">newdata =</span> cur_mat, </span>
<span id="cb473-1638"><a href="#cb473-1638" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"raw"</span>)</span>
<span id="cb473-1639"><a href="#cb473-1639" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1640"><a href="#cb473-1640" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cell_class) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(cur_mat)</span>
<span id="cb473-1641"><a href="#cb473-1641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1642"><a href="#cb473-1642" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cell_class)</span>
<span id="cb473-1643"><a href="#cb473-1643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1644"><a href="#cb473-1644" aria-hidden="true" tabindex="-1"></a><span class="do">### Extract prediction probabilities for each cell</span></span>
<span id="cb473-1645"><a href="#cb473-1645" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">231014</span>)</span>
<span id="cb473-1646"><a href="#cb473-1646" aria-hidden="true" tabindex="-1"></a>cell_prob <span class="ot">&lt;-</span> <span class="fu">predict</span>(rffit, <span class="at">newdata =</span> cur_mat, <span class="at">type =</span> <span class="st">"prob"</span>)</span>
<span id="cb473-1647"><a href="#cb473-1647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1648"><a href="#cb473-1648" aria-hidden="true" tabindex="-1"></a><span class="do">### Distribution of maximum probabilities</span></span>
<span id="cb473-1649"><a href="#cb473-1649" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(<span class="at">max_prob =</span> <span class="fu">rowMax</span>(<span class="fu">as.matrix</span>(cell_prob)),</span>
<span id="cb473-1650"><a href="#cb473-1650" aria-hidden="true" tabindex="-1"></a>       <span class="at">type =</span> cell_class) <span class="sc">|&gt;</span></span>
<span id="cb473-1651"><a href="#cb473-1651" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb473-1652"><a href="#cb473-1652" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_density_ridges</span>(<span class="fu">aes</span>(<span class="at">x =</span> max_prob, <span class="at">y =</span> cell_class, <span class="at">fill =</span> cell_class)) <span class="sc">+</span></span>
<span id="cb473-1653"><a href="#cb473-1653" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype) <span class="sc">+</span></span>
<span id="cb473-1654"><a href="#cb473-1654" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb473-1655"><a href="#cb473-1655" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlab</span>(<span class="st">"Maximum probability"</span>) <span class="sc">+</span></span>
<span id="cb473-1656"><a href="#cb473-1656" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ylab</span>(<span class="st">"Cell type"</span>) <span class="sc">+</span> </span>
<span id="cb473-1657"><a href="#cb473-1657" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">1.2</span>))</span>
<span id="cb473-1658"><a href="#cb473-1658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1659"><a href="#cb473-1659" aria-hidden="true" tabindex="-1"></a><span class="do">### Label undefined cells</span></span>
<span id="cb473-1660"><a href="#cb473-1660" aria-hidden="true" tabindex="-1"></a>cell_class[<span class="fu">rowMax</span>(<span class="fu">as.matrix</span>(cell_prob)) <span class="sc">&lt;</span> <span class="fl">0.4</span>] <span class="ot">&lt;-</span> <span class="st">"undefined"</span></span>
<span id="cb473-1661"><a href="#cb473-1661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1662"><a href="#cb473-1662" aria-hidden="true" tabindex="-1"></a><span class="do">### Store labels in SpatialExperiment onject</span></span>
<span id="cb473-1663"><a href="#cb473-1663" aria-hidden="true" tabindex="-1"></a>cell_labels <span class="ot">&lt;-</span> spe<span class="sc">$</span>cell_labels</span>
<span id="cb473-1664"><a href="#cb473-1664" aria-hidden="true" tabindex="-1"></a>cell_labels[<span class="fu">colnames</span>(unlab_spe)] <span class="ot">&lt;-</span> cell_class</span>
<span id="cb473-1665"><a href="#cb473-1665" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>celltype <span class="ot">&lt;-</span> cell_labels </span>
<span id="cb473-1666"><a href="#cb473-1666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1667"><a href="#cb473-1667" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(spe<span class="sc">$</span>celltype, spe<span class="sc">$</span>patient_id)</span>
<span id="cb473-1668"><a href="#cb473-1668" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1669"><a href="#cb473-1669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1672"><a href="#cb473-1672" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1673"><a href="#cb473-1673" aria-hidden="true" tabindex="-1"></a>tab1 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1674"><a href="#cb473-1674" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"Rphenograph"</span>, spe<span class="sc">$</span>pg_clusters))</span>
<span id="cb473-1675"><a href="#cb473-1675" aria-hidden="true" tabindex="-1"></a>tab2 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1676"><a href="#cb473-1676" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"SNN"</span>, spe<span class="sc">$</span>nn_clusters))</span>
<span id="cb473-1677"><a href="#cb473-1677" aria-hidden="true" tabindex="-1"></a>tab3 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1678"><a href="#cb473-1678" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"SOM"</span>, spe<span class="sc">$</span>som_clusters))</span>
<span id="cb473-1679"><a href="#cb473-1679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1680"><a href="#cb473-1680" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab1 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1681"><a href="#cb473-1681" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab2 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1682"><a href="#cb473-1682" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab3 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1683"><a href="#cb473-1683" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1684"><a href="#cb473-1684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1687"><a href="#cb473-1687" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1688"><a href="#cb473-1688" aria-hidden="true" tabindex="-1"></a>tab1 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1689"><a href="#cb473-1689" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"Rphenograph"</span>, spe<span class="sc">$</span>pg_clusters_corrected))</span>
<span id="cb473-1690"><a href="#cb473-1690" aria-hidden="true" tabindex="-1"></a>tab2 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1691"><a href="#cb473-1691" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"SNN"</span>, spe<span class="sc">$</span>nn_clusters_corrected))</span>
<span id="cb473-1692"><a href="#cb473-1692" aria-hidden="true" tabindex="-1"></a>tab3 <span class="ot">&lt;-</span> <span class="fu">table</span>(spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1693"><a href="#cb473-1693" aria-hidden="true" tabindex="-1"></a>              <span class="fu">paste</span>(<span class="st">"SOM"</span>, spe<span class="sc">$</span>som_clusters_corrected))</span>
<span id="cb473-1694"><a href="#cb473-1694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1695"><a href="#cb473-1695" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab1 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1696"><a href="#cb473-1696" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab2 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1697"><a href="#cb473-1697" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">log10</span>(tab3 <span class="sc">+</span> <span class="dv">10</span>), <span class="at">color =</span> <span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb473-1698"><a href="#cb473-1698" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1699"><a href="#cb473-1699" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1702"><a href="#cb473-1702" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1703"><a href="#cb473-1703" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(spe, <span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1704"><a href="#cb473-1704" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1705"><a href="#cb473-1705" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1706"><a href="#cb473-1706" aria-hidden="true" tabindex="-1"></a><span class="fu">## Single cell visualization</span></span>
<span id="cb473-1707"><a href="#cb473-1707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1708"><a href="#cb473-1708" aria-hidden="true" tabindex="-1"></a><span class="fu">### Inital setup</span></span>
<span id="cb473-1711"><a href="#cb473-1711" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1712"><a href="#cb473-1712" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="fu">here</span>(dir, <span class="st">"data/spe.qs"</span>))</span>
<span id="cb473-1713"><a href="#cb473-1713" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDims</span>(spe)</span>
<span id="cb473-1714"><a href="#cb473-1714" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">colData</span>(spe))</span>
<span id="cb473-1715"><a href="#cb473-1715" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1716"><a href="#cb473-1716" aria-hidden="true" tabindex="-1"></a><span class="do">### Define cell phenotype markers</span></span>
<span id="cb473-1717"><a href="#cb473-1717" aria-hidden="true" tabindex="-1"></a>type_markers <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb473-1718"><a href="#cb473-1718" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Ecad"</span>, <span class="st">"CD45RO"</span>, <span class="st">"CD20"</span>, <span class="st">"CD3"</span>, <span class="st">"FOXP3"</span>, <span class="st">"CD206"</span>, <span class="st">"MPO"</span>,</span>
<span id="cb473-1719"><a href="#cb473-1719" aria-hidden="true" tabindex="-1"></a>    <span class="st">"SMA"</span>, <span class="st">"CD8a"</span>, <span class="st">"CD4"</span>, <span class="st">"HLADR"</span>, <span class="st">"CD15"</span>, <span class="st">"CD38"</span>, <span class="st">"PDGFRb"</span></span>
<span id="cb473-1720"><a href="#cb473-1720" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1721"><a href="#cb473-1721" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1722"><a href="#cb473-1722" aria-hidden="true" tabindex="-1"></a><span class="do">### Define cell state markers</span></span>
<span id="cb473-1723"><a href="#cb473-1723" aria-hidden="true" tabindex="-1"></a>state_markers <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb473-1724"><a href="#cb473-1724" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CarbonicAnhydrase"</span>, <span class="st">"Ki67"</span>, <span class="st">"PD1"</span>, <span class="st">"GrzB"</span>, <span class="st">"PDL1"</span>,</span>
<span id="cb473-1725"><a href="#cb473-1725" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ICOS"</span>, <span class="st">"TCF7"</span>, <span class="st">"VISTA"</span></span>
<span id="cb473-1726"><a href="#cb473-1726" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1727"><a href="#cb473-1727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1728"><a href="#cb473-1728" aria-hidden="true" tabindex="-1"></a><span class="do">### Add to spe</span></span>
<span id="cb473-1729"><a href="#cb473-1729" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb473-1730"><a href="#cb473-1730" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(spe) <span class="sc">%in%</span> type_markers, <span class="st">"type"</span>,</span>
<span id="cb473-1731"><a href="#cb473-1731" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ifelse</span>(<span class="fu">rownames</span>(spe) <span class="sc">%in%</span> state_markers, <span class="st">"state"</span>,</span>
<span id="cb473-1732"><a href="#cb473-1732" aria-hidden="true" tabindex="-1"></a>        <span class="st">"other"</span>)</span>
<span id="cb473-1733"><a href="#cb473-1733" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1734"><a href="#cb473-1734" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1735"><a href="#cb473-1735" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cell-type level</span></span>
<span id="cb473-1736"><a href="#cb473-1736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1739"><a href="#cb473-1739" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1740"><a href="#cb473-1740" aria-hidden="true" tabindex="-1"></a><span class="do">### Dimensionality reduction visualization</span></span>
<span id="cb473-1741"><a href="#cb473-1741" aria-hidden="true" tabindex="-1"></a><span class="do">### Interpreting these UMAP, tSNE plots is not trivial, but local neighborhoods </span></span>
<span id="cb473-1742"><a href="#cb473-1742" aria-hidden="true" tabindex="-1"></a><span class="do">### in the plot can suggest similarity in expression for given cells.</span></span>
<span id="cb473-1743"><a href="#cb473-1743" aria-hidden="true" tabindex="-1"></a><span class="do">### tSNE/UMAP aim to preserve the distances between each cell and its neighbors in the high-dimensional space.</span></span>
<span id="cb473-1744"><a href="#cb473-1744" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1745"><a href="#cb473-1745" aria-hidden="true" tabindex="-1"></a><span class="do">### UMAP colored by cell type and expression - dittoDimPlot</span></span>
<span id="cb473-1746"><a href="#cb473-1746" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1747"><a href="#cb473-1747" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-1748"><a href="#cb473-1748" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"celltype"</span>, </span>
<span id="cb473-1749"><a href="#cb473-1749" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span id="cb473-1750"><a href="#cb473-1750" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-1751"><a href="#cb473-1751" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1752"><a href="#cb473-1752" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1753"><a href="#cb473-1753" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype) <span class="sc">+</span></span>
<span id="cb473-1754"><a href="#cb473-1754" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb473-1755"><a href="#cb473-1755" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Cell types on UMAP, integrated cells"</span>)</span>
<span id="cb473-1756"><a href="#cb473-1756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1757"><a href="#cb473-1757" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-1758"><a href="#cb473-1758" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-1759"><a href="#cb473-1759" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"Ecad"</span>, </span>
<span id="cb473-1760"><a href="#cb473-1760" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-1761"><a href="#cb473-1761" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>, </span>
<span id="cb473-1762"><a href="#cb473-1762" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>, </span>
<span id="cb473-1763"><a href="#cb473-1763" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1764"><a href="#cb473-1764" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1765"><a href="#cb473-1765" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1766"><a href="#cb473-1766" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_viridis</span>()</span>
<span id="cb473-1767"><a href="#cb473-1767" aria-hidden="true" tabindex="-1"></a>    <span class="co"># coord_fixed()</span></span>
<span id="cb473-1768"><a href="#cb473-1768" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb473-1769"><a href="#cb473-1769" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span>
<span id="cb473-1770"><a href="#cb473-1770" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1771"><a href="#cb473-1771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1774"><a href="#cb473-1774" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1775"><a href="#cb473-1775" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP colored by expression for all markers - plotReducedDim</span></span>
<span id="cb473-1776"><a href="#cb473-1776" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb473-1777"><a href="#cb473-1777" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>],</span>
<span id="cb473-1778"><a href="#cb473-1778" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) {</span>
<span id="cb473-1779"><a href="#cb473-1779" aria-hidden="true" tabindex="-1"></a>        p <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(</span>
<span id="cb473-1780"><a href="#cb473-1780" aria-hidden="true" tabindex="-1"></a>            spe,</span>
<span id="cb473-1781"><a href="#cb473-1781" aria-hidden="true" tabindex="-1"></a>            <span class="at">dimred =</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span id="cb473-1782"><a href="#cb473-1782" aria-hidden="true" tabindex="-1"></a>            <span class="at">colour_by =</span> x,</span>
<span id="cb473-1783"><a href="#cb473-1783" aria-hidden="true" tabindex="-1"></a>            <span class="at">by_exprs_values =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-1784"><a href="#cb473-1784" aria-hidden="true" tabindex="-1"></a>            <span class="at">point_size =</span> <span class="fl">0.2</span></span>
<span id="cb473-1785"><a href="#cb473-1785" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb473-1786"><a href="#cb473-1786" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(p)</span>
<span id="cb473-1787"><a href="#cb473-1787" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb473-1788"><a href="#cb473-1788" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1789"><a href="#cb473-1789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1790"><a href="#cb473-1790" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plot_list)</span>
<span id="cb473-1791"><a href="#cb473-1791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1792"><a href="#cb473-1792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1795"><a href="#cb473-1795" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1796"><a href="#cb473-1796" aria-hidden="true" tabindex="-1"></a><span class="do">### Heatmap visualization, it is often useful to visualize single-cell </span></span>
<span id="cb473-1797"><a href="#cb473-1797" aria-hidden="true" tabindex="-1"></a><span class="do">### expression per cell type in form of a heatmap</span></span>
<span id="cb473-1798"><a href="#cb473-1798" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">220818</span>)</span>
<span id="cb473-1799"><a href="#cb473-1799" aria-hidden="true" tabindex="-1"></a>cur_cells <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(spe)), <span class="dv">4000</span>)</span>
<span id="cb473-1800"><a href="#cb473-1800" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1801"><a href="#cb473-1801" aria-hidden="true" tabindex="-1"></a><span class="do">### Heatmap visualization</span></span>
<span id="cb473-1802"><a href="#cb473-1802" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1803"><a href="#cb473-1803" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1804"><a href="#cb473-1804" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>],</span>
<span id="cb473-1805"><a href="#cb473-1805" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1806"><a href="#cb473-1806" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb473-1807"><a href="#cb473-1807" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1808"><a href="#cb473-1808" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-1809"><a href="#cb473-1809" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"indication"</span>, <span class="st">"patient_id"</span>),</span>
<span id="cb473-1810"><a href="#cb473-1810" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-1811"><a href="#cb473-1811" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication,</span>
<span id="cb473-1812"><a href="#cb473-1812" aria-hidden="true" tabindex="-1"></a>        <span class="at">patient_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id,</span>
<span id="cb473-1813"><a href="#cb473-1813" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype</span>
<span id="cb473-1814"><a href="#cb473-1814" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1815"><a href="#cb473-1815" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1816"><a href="#cb473-1816" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1817"><a href="#cb473-1817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1820"><a href="#cb473-1820" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1821"><a href="#cb473-1821" aria-hidden="true" tabindex="-1"></a><span class="do">### Visualize the mean marker expression per cell type for all cells</span></span>
<span id="cb473-1822"><a href="#cb473-1822" aria-hidden="true" tabindex="-1"></a><span class="do">### aggregate by cell type</span></span>
<span id="cb473-1823"><a href="#cb473-1823" aria-hidden="true" tabindex="-1"></a>celltype_mean <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">aggregateAcrossCells</span>(</span>
<span id="cb473-1824"><a href="#cb473-1824" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as</span>(spe, <span class="st">"SingleCellExperiment"</span>),  </span>
<span id="cb473-1825"><a href="#cb473-1825" aria-hidden="true" tabindex="-1"></a>    <span class="at">ids =</span> spe<span class="sc">$</span>celltype, </span>
<span id="cb473-1826"><a href="#cb473-1826" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="st">"mean"</span>,</span>
<span id="cb473-1827"><a href="#cb473-1827" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.assay.type =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1828"><a href="#cb473-1828" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>]</span>
<span id="cb473-1829"><a href="#cb473-1829" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1830"><a href="#cb473-1830" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1831"><a href="#cb473-1831" aria-hidden="true" tabindex="-1"></a><span class="do">### No scaling</span></span>
<span id="cb473-1832"><a href="#cb473-1832" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1833"><a href="#cb473-1833" aria-hidden="true" tabindex="-1"></a>    celltype_mean,</span>
<span id="cb473-1834"><a href="#cb473-1834" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1835"><a href="#cb473-1835" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb473-1836"><a href="#cb473-1836" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-1837"><a href="#cb473-1837" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-1838"><a href="#cb473-1838" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"ncells"</span>),</span>
<span id="cb473-1839"><a href="#cb473-1839" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-1840"><a href="#cb473-1840" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype,</span>
<span id="cb473-1841"><a href="#cb473-1841" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncells =</span> <span class="fu">plasma</span>(<span class="dv">100</span>)</span>
<span id="cb473-1842"><a href="#cb473-1842" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1843"><a href="#cb473-1843" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1844"><a href="#cb473-1844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1845"><a href="#cb473-1845" aria-hidden="true" tabindex="-1"></a><span class="do">### Scaled to max</span></span>
<span id="cb473-1846"><a href="#cb473-1846" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1847"><a href="#cb473-1847" aria-hidden="true" tabindex="-1"></a>    celltype_mean,</span>
<span id="cb473-1848"><a href="#cb473-1848" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1849"><a href="#cb473-1849" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb473-1850"><a href="#cb473-1850" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaled.to.max =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-1851"><a href="#cb473-1851" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors.max.scaled =</span> <span class="fu">inferno</span>(<span class="dv">100</span>),</span>
<span id="cb473-1852"><a href="#cb473-1852" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"ncells"</span>),</span>
<span id="cb473-1853"><a href="#cb473-1853" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-1854"><a href="#cb473-1854" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype,</span>
<span id="cb473-1855"><a href="#cb473-1855" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncells =</span> <span class="fu">plasma</span>(<span class="dv">100</span>)</span>
<span id="cb473-1856"><a href="#cb473-1856" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1857"><a href="#cb473-1857" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1858"><a href="#cb473-1858" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1859"><a href="#cb473-1859" aria-hidden="true" tabindex="-1"></a><span class="do">### # Z score scaled</span></span>
<span id="cb473-1860"><a href="#cb473-1860" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-1861"><a href="#cb473-1861" aria-hidden="true" tabindex="-1"></a>    celltype_mean,</span>
<span id="cb473-1862"><a href="#cb473-1862" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1863"><a href="#cb473-1863" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb473-1864"><a href="#cb473-1864" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"ncells"</span>),</span>
<span id="cb473-1865"><a href="#cb473-1865" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-1866"><a href="#cb473-1866" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype,</span>
<span id="cb473-1867"><a href="#cb473-1867" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncells =</span> <span class="fu">plasma</span>(<span class="dv">100</span>)</span>
<span id="cb473-1868"><a href="#cb473-1868" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-1869"><a href="#cb473-1869" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1870"><a href="#cb473-1870" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1871"><a href="#cb473-1871" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1874"><a href="#cb473-1874" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1875"><a href="#cb473-1875" aria-hidden="true" tabindex="-1"></a><span class="do">### Violin plot visualization - plotExpression</span></span>
<span id="cb473-1876"><a href="#cb473-1876" aria-hidden="true" tabindex="-1"></a><span class="fu">plotExpression</span>(</span>
<span id="cb473-1877"><a href="#cb473-1877" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-1878"><a href="#cb473-1878" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>],</span>
<span id="cb473-1879"><a href="#cb473-1879" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"celltype"</span>, </span>
<span id="cb473-1880"><a href="#cb473-1880" aria-hidden="true" tabindex="-1"></a>    <span class="at">exprs_values =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-1881"><a href="#cb473-1881" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by =</span> <span class="st">"celltype"</span></span>
<span id="cb473-1882"><a href="#cb473-1882" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1883"><a href="#cb473-1883" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span>  <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>))<span class="sc">+</span></span>
<span id="cb473-1884"><a href="#cb473-1884" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1885"><a href="#cb473-1885" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1886"><a href="#cb473-1886" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1887"><a href="#cb473-1887" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1890"><a href="#cb473-1890" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1891"><a href="#cb473-1891" aria-hidden="true" tabindex="-1"></a><span class="do">### Scatter plot visualization</span></span>
<span id="cb473-1892"><a href="#cb473-1892" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoScatterPlot</span>(</span>
<span id="cb473-1893"><a href="#cb473-1893" aria-hidden="true" tabindex="-1"></a>    spe,</span>
<span id="cb473-1894"><a href="#cb473-1894" aria-hidden="true" tabindex="-1"></a>    <span class="at">x.var =</span> <span class="st">"CD3"</span>,</span>
<span id="cb473-1895"><a href="#cb473-1895" aria-hidden="true" tabindex="-1"></a>    <span class="at">y.var =</span> <span class="st">"CD20"</span>,</span>
<span id="cb473-1896"><a href="#cb473-1896" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.x =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-1897"><a href="#cb473-1897" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.y =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-1898"><a href="#cb473-1898" aria-hidden="true" tabindex="-1"></a>    <span class="at">color.var =</span> <span class="st">"celltype"</span></span>
<span id="cb473-1899"><a href="#cb473-1899" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1900"><a href="#cb473-1900" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype) <span class="sc">+</span></span>
<span id="cb473-1901"><a href="#cb473-1901" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Scatterplot for CD3/CD20 labelled by celltype"</span>)</span>
<span id="cb473-1902"><a href="#cb473-1902" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1903"><a href="#cb473-1903" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1906"><a href="#cb473-1906" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1907"><a href="#cb473-1907" aria-hidden="true" tabindex="-1"></a><span class="do">### by sample_id - percentage</span></span>
<span id="cb473-1908"><a href="#cb473-1908" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoBarPlot</span>(spe, <span class="at">var =</span> <span class="st">"celltype"</span>, <span class="at">group.by =</span> <span class="st">"sample_id"</span>) <span class="sc">+</span></span>
<span id="cb473-1909"><a href="#cb473-1909" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1910"><a href="#cb473-1910" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1911"><a href="#cb473-1911" aria-hidden="true" tabindex="-1"></a><span class="do">### by patient_id - percentage</span></span>
<span id="cb473-1912"><a href="#cb473-1912" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoBarPlot</span>(spe, <span class="at">var =</span> <span class="st">"celltype"</span>, <span class="at">group.by =</span> <span class="st">"patient_id"</span>) <span class="sc">+</span></span>
<span id="cb473-1913"><a href="#cb473-1913" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1914"><a href="#cb473-1914" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1915"><a href="#cb473-1915" aria-hidden="true" tabindex="-1"></a><span class="do">### by patient_id - count</span></span>
<span id="cb473-1916"><a href="#cb473-1916" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoBarPlot</span>(spe, <span class="at">scale =</span> <span class="st">"count"</span>,<span class="at">var =</span> <span class="st">"celltype"</span>, <span class="at">group.by =</span> <span class="st">"patient_id"</span>) <span class="sc">+</span></span>
<span id="cb473-1917"><a href="#cb473-1917" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1918"><a href="#cb473-1918" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1919"><a href="#cb473-1919" aria-hidden="true" tabindex="-1"></a><span class="do">### We can see that cell type frequencies change between samples/patients and that the highest proportion/counts of plasma cells and stromal cells can be observed for Patient 2 and Patient 4, respectively.</span></span>
<span id="cb473-1920"><a href="#cb473-1920" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1921"><a href="#cb473-1921" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1924"><a href="#cb473-1924" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1925"><a href="#cb473-1925" aria-hidden="true" tabindex="-1"></a><span class="do">### CATALYST-based visualization</span></span>
<span id="cb473-1926"><a href="#cb473-1926" aria-hidden="true" tabindex="-1"></a><span class="do">### Save SPE in CATALYST-compatible object with renamed colData entries and</span></span>
<span id="cb473-1927"><a href="#cb473-1927" aria-hidden="true" tabindex="-1"></a><span class="do">### new metadata information</span></span>
<span id="cb473-1928"><a href="#cb473-1928" aria-hidden="true" tabindex="-1"></a>spe_cat <span class="ot">&lt;-</span> spe</span>
<span id="cb473-1929"><a href="#cb473-1929" aria-hidden="true" tabindex="-1"></a>spe_cat<span class="sc">$</span>sample_id <span class="ot">&lt;-</span> <span class="fu">factor</span>(spe<span class="sc">$</span>sample_id)</span>
<span id="cb473-1930"><a href="#cb473-1930" aria-hidden="true" tabindex="-1"></a>spe_cat<span class="sc">$</span>condition <span class="ot">&lt;-</span> <span class="fu">factor</span>(spe<span class="sc">$</span>indication)</span>
<span id="cb473-1931"><a href="#cb473-1931" aria-hidden="true" tabindex="-1"></a>spe_cat<span class="sc">$</span>cluster_id <span class="ot">&lt;-</span> <span class="fu">factor</span>(spe<span class="sc">$</span>celltype)</span>
<span id="cb473-1932"><a href="#cb473-1932" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1933"><a href="#cb473-1933" aria-hidden="true" tabindex="-1"></a><span class="do">### Add celltype information to metadata</span></span>
<span id="cb473-1934"><a href="#cb473-1934" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(spe_cat)<span class="sc">$</span>cluster_codes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb473-1935"><a href="#cb473-1935" aria-hidden="true" tabindex="-1"></a>    <span class="at">celltype =</span> <span class="fu">factor</span>(spe_cat<span class="sc">$</span>celltype)</span>
<span id="cb473-1936"><a href="#cb473-1936" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1937"><a href="#cb473-1937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1938"><a href="#cb473-1938" aria-hidden="true" tabindex="-1"></a><span class="do">### Pseudobulk-level MDS plot</span></span>
<span id="cb473-1939"><a href="#cb473-1939" aria-hidden="true" tabindex="-1"></a><span class="do">### MDS pseudobulk by cell type to highlight expression similarities between </span></span>
<span id="cb473-1940"><a href="#cb473-1940" aria-hidden="true" tabindex="-1"></a><span class="do">### cell types and subsequently for each celltype-sample-combination.</span></span>
<span id="cb473-1941"><a href="#cb473-1941" aria-hidden="true" tabindex="-1"></a>CATALYST<span class="sc">::</span><span class="fu">pbMDS</span>(</span>
<span id="cb473-1942"><a href="#cb473-1942" aria-hidden="true" tabindex="-1"></a>    spe_cat,</span>
<span id="cb473-1943"><a href="#cb473-1943" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"cluster_id"</span>,</span>
<span id="cb473-1944"><a href="#cb473-1944" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">rownames</span>(spe_cat)[<span class="fu">rowData</span>(spe_cat)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>],</span>
<span id="cb473-1945"><a href="#cb473-1945" aria-hidden="true" tabindex="-1"></a>    <span class="at">label_by =</span> <span class="st">"cluster_id"</span>,</span>
<span id="cb473-1946"><a href="#cb473-1946" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="st">"celltype"</span></span>
<span id="cb473-1947"><a href="#cb473-1947" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1948"><a href="#cb473-1948" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe_cat)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1949"><a href="#cb473-1949" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1950"><a href="#cb473-1950" aria-hidden="true" tabindex="-1"></a><span class="do">### MDS pseudobulk by cell type and sample_id</span></span>
<span id="cb473-1951"><a href="#cb473-1951" aria-hidden="true" tabindex="-1"></a><span class="do">### We can see that the pseudobulk-expression profile of neutrophils seems </span></span>
<span id="cb473-1952"><a href="#cb473-1952" aria-hidden="true" tabindex="-1"></a><span class="do">### markedly distinct from the other cell types, while comparable cell types </span></span>
<span id="cb473-1953"><a href="#cb473-1953" aria-hidden="true" tabindex="-1"></a><span class="do">### such as the T cell subtypes group together. Furthermore, pseudobulk </span></span>
<span id="cb473-1954"><a href="#cb473-1954" aria-hidden="true" tabindex="-1"></a><span class="do">## cell-type profiles from SCCHN appear different from the other indications.</span></span>
<span id="cb473-1955"><a href="#cb473-1955" aria-hidden="true" tabindex="-1"></a>CATALYST<span class="sc">::</span><span class="fu">pbMDS</span>(</span>
<span id="cb473-1956"><a href="#cb473-1956" aria-hidden="true" tabindex="-1"></a>    spe_cat, </span>
<span id="cb473-1957"><a href="#cb473-1957" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"both"</span>, </span>
<span id="cb473-1958"><a href="#cb473-1958" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">rownames</span>(spe_cat)[<span class="fu">rowData</span>(spe_cat)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>], </span>
<span id="cb473-1959"><a href="#cb473-1959" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="st">"celltype"</span>, </span>
<span id="cb473-1960"><a href="#cb473-1960" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_by =</span> <span class="st">"condition"</span>, </span>
<span id="cb473-1961"><a href="#cb473-1961" aria-hidden="true" tabindex="-1"></a>    <span class="at">size_by =</span> <span class="cn">TRUE</span></span>
<span id="cb473-1962"><a href="#cb473-1962" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1963"><a href="#cb473-1963" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe_cat)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype)</span>
<span id="cb473-1964"><a href="#cb473-1964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1965"><a href="#cb473-1965" aria-hidden="true" tabindex="-1"></a><span class="do">### Reduced dimension plot on CLR of proportions</span></span>
<span id="cb473-1966"><a href="#cb473-1966" aria-hidden="true" tabindex="-1"></a><span class="do">### The output plots aim to illustrate the degree of similarity between </span></span>
<span id="cb473-1967"><a href="#cb473-1967" aria-hidden="true" tabindex="-1"></a><span class="do">### cell types based on sample proportions.</span></span>
<span id="cb473-1968"><a href="#cb473-1968" aria-hidden="true" tabindex="-1"></a><span class="do">### CLR on cluster proportions across samples</span></span>
<span id="cb473-1969"><a href="#cb473-1969" aria-hidden="true" tabindex="-1"></a><span class="do">### We can again observe that neutrophils have a divergent profile also in terms of their sample proportions.</span></span>
<span id="cb473-1970"><a href="#cb473-1970" aria-hidden="true" tabindex="-1"></a>CATALYST<span class="sc">::</span><span class="fu">clrDR</span>(</span>
<span id="cb473-1971"><a href="#cb473-1971" aria-hidden="true" tabindex="-1"></a>    spe_cat, </span>
<span id="cb473-1972"><a href="#cb473-1972" aria-hidden="true" tabindex="-1"></a>    <span class="at">dr =</span> <span class="st">"PCA"</span>, </span>
<span id="cb473-1973"><a href="#cb473-1973" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"cluster_id"</span>, </span>
<span id="cb473-1974"><a href="#cb473-1974" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="st">"celltype"</span>, </span>
<span id="cb473-1975"><a href="#cb473-1975" aria-hidden="true" tabindex="-1"></a>    <span class="at">label_by =</span> <span class="st">"cluster_id"</span>, </span>
<span id="cb473-1976"><a href="#cb473-1976" aria-hidden="true" tabindex="-1"></a>    <span class="at">arrow_col =</span> <span class="st">"sample_id"</span>, </span>
<span id="cb473-1977"><a href="#cb473-1977" aria-hidden="true" tabindex="-1"></a>    <span class="at">point_pal =</span> <span class="fu">metadata</span>(spe_cat)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype</span>
<span id="cb473-1978"><a href="#cb473-1978" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-1979"><a href="#cb473-1979" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1980"><a href="#cb473-1980" aria-hidden="true" tabindex="-1"></a><span class="do">### Pseudobulk expression boxplot</span></span>
<span id="cb473-1981"><a href="#cb473-1981" aria-hidden="true" tabindex="-1"></a><span class="do">### Notably, CD15 levels are elevated in SCCHN in comparison to all other </span></span>
<span id="cb473-1982"><a href="#cb473-1982" aria-hidden="true" tabindex="-1"></a><span class="do">### indications for most cell types.</span></span>
<span id="cb473-1983"><a href="#cb473-1983" aria-hidden="true" tabindex="-1"></a>CATALYST<span class="sc">::</span><span class="fu">plotPbExprs</span>(</span>
<span id="cb473-1984"><a href="#cb473-1984" aria-hidden="true" tabindex="-1"></a>    spe_cat, </span>
<span id="cb473-1985"><a href="#cb473-1985" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="st">"celltype"</span>, </span>
<span id="cb473-1986"><a href="#cb473-1986" aria-hidden="true" tabindex="-1"></a>    <span class="at">facet_by =</span> <span class="st">"cluster_id"</span>, </span>
<span id="cb473-1987"><a href="#cb473-1987" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncol =</span> <span class="dv">2</span>, </span>
<span id="cb473-1988"><a href="#cb473-1988" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">rownames</span>(spe_cat)[<span class="fu">rowData</span>(spe_cat)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>]</span>
<span id="cb473-1989"><a href="#cb473-1989" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-1990"><a href="#cb473-1990" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe_cat)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication)</span>
<span id="cb473-1991"><a href="#cb473-1991" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-1992"><a href="#cb473-1992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1993"><a href="#cb473-1993" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sample level</span></span>
<span id="cb473-1994"><a href="#cb473-1994" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-1997"><a href="#cb473-1997" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-1998"><a href="#cb473-1998" aria-hidden="true" tabindex="-1"></a><span class="do">### Dimensionality reduction visualization</span></span>
<span id="cb473-1999"><a href="#cb473-1999" aria-hidden="true" tabindex="-1"></a><span class="do">## UMAP colored by cell type and expression - dittoDimPlot</span></span>
<span id="cb473-2000"><a href="#cb473-2000" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-2001"><a href="#cb473-2001" aria-hidden="true" tabindex="-1"></a>    spe,</span>
<span id="cb473-2002"><a href="#cb473-2002" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb473-2003"><a href="#cb473-2003" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>,</span>
<span id="cb473-2004"><a href="#cb473-2004" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-2005"><a href="#cb473-2005" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-2006"><a href="#cb473-2006" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">FALSE</span></span>
<span id="cb473-2007"><a href="#cb473-2007" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2008"><a href="#cb473-2008" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>sample_id) <span class="sc">+</span></span>
<span id="cb473-2009"><a href="#cb473-2009" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb473-2010"><a href="#cb473-2010" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Sample ID"</span>)</span>
<span id="cb473-2011"><a href="#cb473-2011" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2012"><a href="#cb473-2012" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-2013"><a href="#cb473-2013" aria-hidden="true" tabindex="-1"></a>    spe,</span>
<span id="cb473-2014"><a href="#cb473-2014" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb473-2015"><a href="#cb473-2015" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span id="cb473-2016"><a href="#cb473-2016" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-2017"><a href="#cb473-2017" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-2018"><a href="#cb473-2018" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">FALSE</span></span>
<span id="cb473-2019"><a href="#cb473-2019" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2020"><a href="#cb473-2020" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>sample_id) <span class="sc">+</span></span>
<span id="cb473-2021"><a href="#cb473-2021" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb473-2022"><a href="#cb473-2022" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Sample ID"</span>)</span>
<span id="cb473-2023"><a href="#cb473-2023" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2024"><a href="#cb473-2024" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-2025"><a href="#cb473-2025" aria-hidden="true" tabindex="-1"></a>    spe,</span>
<span id="cb473-2026"><a href="#cb473-2026" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"patient_id"</span>,</span>
<span id="cb473-2027"><a href="#cb473-2027" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP"</span>,</span>
<span id="cb473-2028"><a href="#cb473-2028" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-2029"><a href="#cb473-2029" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">FALSE</span></span>
<span id="cb473-2030"><a href="#cb473-2030" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2031"><a href="#cb473-2031" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-2032"><a href="#cb473-2032" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb473-2033"><a href="#cb473-2033" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID"</span>)</span>
<span id="cb473-2034"><a href="#cb473-2034" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2035"><a href="#cb473-2035" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">dittoDimPlot</span>(</span>
<span id="cb473-2036"><a href="#cb473-2036" aria-hidden="true" tabindex="-1"></a>    spe,</span>
<span id="cb473-2037"><a href="#cb473-2037" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"patient_id"</span>,</span>
<span id="cb473-2038"><a href="#cb473-2038" aria-hidden="true" tabindex="-1"></a>    <span class="at">reduction.use =</span> <span class="st">"UMAP_mnnCorrected"</span>,</span>
<span id="cb473-2039"><a href="#cb473-2039" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.2</span>,</span>
<span id="cb473-2040"><a href="#cb473-2040" aria-hidden="true" tabindex="-1"></a>    <span class="at">do.label =</span> <span class="cn">FALSE</span></span>
<span id="cb473-2041"><a href="#cb473-2041" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2042"><a href="#cb473-2042" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id) <span class="sc">+</span></span>
<span id="cb473-2043"><a href="#cb473-2043" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb473-2044"><a href="#cb473-2044" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Patient ID"</span>)</span>
<span id="cb473-2045"><a href="#cb473-2045" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2046"><a href="#cb473-2046" aria-hidden="true" tabindex="-1"></a><span class="do">### fastMNN approach (right side of the plot) leads to mixing of cells across </span></span>
<span id="cb473-2047"><a href="#cb473-2047" aria-hidden="true" tabindex="-1"></a><span class="do">### samples/patients and thus batch effect correction.</span></span>
<span id="cb473-2048"><a href="#cb473-2048" aria-hidden="true" tabindex="-1"></a>(p1 <span class="sc">+</span> p2) <span class="sc">/</span> (p3 <span class="sc">+</span> p4)</span>
<span id="cb473-2049"><a href="#cb473-2049" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-2050"><a href="#cb473-2050" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2053"><a href="#cb473-2053" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-2054"><a href="#cb473-2054" aria-hidden="true" tabindex="-1"></a><span class="do">### Heatmap visualization to highlight biological differences across samples/patients.</span></span>
<span id="cb473-2055"><a href="#cb473-2055" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-2056"><a href="#cb473-2056" aria-hidden="true" tabindex="-1"></a>    spe[,cur_cells], </span>
<span id="cb473-2057"><a href="#cb473-2057" aria-hidden="true" tabindex="-1"></a>    <span class="at">genes =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>],</span>
<span id="cb473-2058"><a href="#cb473-2058" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-2059"><a href="#cb473-2059" aria-hidden="true" tabindex="-1"></a>    <span class="at">order.by =</span> <span class="fu">c</span>(<span class="st">"patient_id"</span>,<span class="st">"sample_id"</span>),</span>
<span id="cb473-2060"><a href="#cb473-2060" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb473-2061"><a href="#cb473-2061" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-2062"><a href="#cb473-2062" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>), </span>
<span id="cb473-2063"><a href="#cb473-2063" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"indication"</span>, <span class="st">"patient_id"</span>, <span class="st">"sample_id"</span>),</span>
<span id="cb473-2064"><a href="#cb473-2064" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-2065"><a href="#cb473-2065" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype,</span>
<span id="cb473-2066"><a href="#cb473-2066" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication,</span>
<span id="cb473-2067"><a href="#cb473-2067" aria-hidden="true" tabindex="-1"></a>        <span class="at">patient_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id,</span>
<span id="cb473-2068"><a href="#cb473-2068" aria-hidden="true" tabindex="-1"></a>        <span class="at">sample_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>sample_id)</span>
<span id="cb473-2069"><a href="#cb473-2069" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2070"><a href="#cb473-2070" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2071"><a href="#cb473-2071" aria-hidden="true" tabindex="-1"></a><span class="do">### mean expression by patient_id</span></span>
<span id="cb473-2072"><a href="#cb473-2072" aria-hidden="true" tabindex="-1"></a>patient_mean <span class="ot">&lt;-</span> <span class="fu">aggregateAcrossCells</span>(</span>
<span id="cb473-2073"><a href="#cb473-2073" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as</span>(spe, <span class="st">"SingleCellExperiment"</span>),  </span>
<span id="cb473-2074"><a href="#cb473-2074" aria-hidden="true" tabindex="-1"></a>    <span class="at">ids =</span> spe<span class="sc">$</span>patient_id, </span>
<span id="cb473-2075"><a href="#cb473-2075" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="st">"mean"</span>,</span>
<span id="cb473-2076"><a href="#cb473-2076" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.assay.type =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-2077"><a href="#cb473-2077" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>]</span>
<span id="cb473-2078"><a href="#cb473-2078" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2079"><a href="#cb473-2079" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2080"><a href="#cb473-2080" aria-hidden="true" tabindex="-1"></a><span class="do">### No scaling</span></span>
<span id="cb473-2081"><a href="#cb473-2081" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-2082"><a href="#cb473-2082" aria-hidden="true" tabindex="-1"></a>    patient_mean,</span>
<span id="cb473-2083"><a href="#cb473-2083" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-2084"><a href="#cb473-2084" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb473-2085"><a href="#cb473-2085" aria-hidden="true" tabindex="-1"></a>    <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb473-2086"><a href="#cb473-2086" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors =</span> <span class="fu">viridis</span>(<span class="dv">100</span>),</span>
<span id="cb473-2087"><a href="#cb473-2087" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"patient_id"</span>,<span class="st">"indication"</span>,<span class="st">"ncells"</span>),</span>
<span id="cb473-2088"><a href="#cb473-2088" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-2089"><a href="#cb473-2089" aria-hidden="true" tabindex="-1"></a>        <span class="at">patient_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id,</span>
<span id="cb473-2090"><a href="#cb473-2090" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication,</span>
<span id="cb473-2091"><a href="#cb473-2091" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncells =</span> <span class="fu">plasma</span>(<span class="dv">100</span>)</span>
<span id="cb473-2092"><a href="#cb473-2092" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-2093"><a href="#cb473-2093" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2094"><a href="#cb473-2094" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2095"><a href="#cb473-2095" aria-hidden="true" tabindex="-1"></a><span class="do">### Max expression scaling</span></span>
<span id="cb473-2096"><a href="#cb473-2096" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(</span>
<span id="cb473-2097"><a href="#cb473-2097" aria-hidden="true" tabindex="-1"></a>    patient_mean,</span>
<span id="cb473-2098"><a href="#cb473-2098" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">"exprs"</span>, </span>
<span id="cb473-2099"><a href="#cb473-2099" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb473-2100"><a href="#cb473-2100" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaled.to.max =</span>  <span class="cn">TRUE</span>,</span>
<span id="cb473-2101"><a href="#cb473-2101" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap.colors.max.scaled =</span> <span class="fu">inferno</span>(<span class="dv">100</span>),</span>
<span id="cb473-2102"><a href="#cb473-2102" aria-hidden="true" tabindex="-1"></a>    <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"patient_id"</span>,<span class="st">"indication"</span>,<span class="st">"ncells"</span>),</span>
<span id="cb473-2103"><a href="#cb473-2103" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb473-2104"><a href="#cb473-2104" aria-hidden="true" tabindex="-1"></a>        <span class="at">patient_id =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id,</span>
<span id="cb473-2105"><a href="#cb473-2105" aria-hidden="true" tabindex="-1"></a>        <span class="at">indication =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication,</span>
<span id="cb473-2106"><a href="#cb473-2106" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncells =</span> <span class="fu">plasma</span>(<span class="dv">100</span>)</span>
<span id="cb473-2107"><a href="#cb473-2107" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb473-2108"><a href="#cb473-2108" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2109"><a href="#cb473-2109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-2110"><a href="#cb473-2110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2113"><a href="#cb473-2113" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-2114"><a href="#cb473-2114" aria-hidden="true" tabindex="-1"></a><span class="do">### Barplot visualization</span></span>
<span id="cb473-2115"><a href="#cb473-2115" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoBarPlot</span>(</span>
<span id="cb473-2116"><a href="#cb473-2116" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-2117"><a href="#cb473-2117" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"patient_id"</span>, </span>
<span id="cb473-2118"><a href="#cb473-2118" aria-hidden="true" tabindex="-1"></a>    <span class="at">group.by =</span> <span class="st">"celltype"</span></span>
<span id="cb473-2119"><a href="#cb473-2119" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2120"><a href="#cb473-2120" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>patient_id)</span>
<span id="cb473-2121"><a href="#cb473-2121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2122"><a href="#cb473-2122" aria-hidden="true" tabindex="-1"></a><span class="do">### </span></span>
<span id="cb473-2123"><a href="#cb473-2123" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoBarPlot</span>(</span>
<span id="cb473-2124"><a href="#cb473-2124" aria-hidden="true" tabindex="-1"></a>    spe, </span>
<span id="cb473-2125"><a href="#cb473-2125" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="st">"sample_id"</span>, </span>
<span id="cb473-2126"><a href="#cb473-2126" aria-hidden="true" tabindex="-1"></a>    <span class="at">group.by =</span> <span class="st">"celltype"</span></span>
<span id="cb473-2127"><a href="#cb473-2127" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb473-2128"><a href="#cb473-2128" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>sample_id)</span>
<span id="cb473-2129"><a href="#cb473-2129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-2130"><a href="#cb473-2130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2131"><a href="#cb473-2131" aria-hidden="true" tabindex="-1"></a><span class="fu">### ComplexHeatmap</span></span>
<span id="cb473-2132"><a href="#cb473-2132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2135"><a href="#cb473-2135" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-2136"><a href="#cb473-2136" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">22</span>)</span>
<span id="cb473-2137"><a href="#cb473-2137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2138"><a href="#cb473-2138" aria-hidden="true" tabindex="-1"></a><span class="do">### 1. Heatmap bodies </span><span class="al">###</span></span>
<span id="cb473-2139"><a href="#cb473-2139" aria-hidden="true" tabindex="-1"></a><span class="do">### Heatmap body color</span></span>
<span id="cb473-2140"><a href="#cb473-2140" aria-hidden="true" tabindex="-1"></a>col_exprs <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(</span>
<span id="cb473-2141"><a href="#cb473-2141" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb473-2142"><a href="#cb473-2142" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"#440154FF"</span>, <span class="st">"#3B518BFF"</span>, <span class="st">"#20938CFF"</span>, <span class="st">"#6ACD5AFF"</span>, <span class="st">"#FDE725FF"</span>)</span>
<span id="cb473-2143"><a href="#cb473-2143" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2144"><a href="#cb473-2144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2145"><a href="#cb473-2145" aria-hidden="true" tabindex="-1"></a><span class="do">### Create Heatmap objects</span></span>
<span id="cb473-2146"><a href="#cb473-2146" aria-hidden="true" tabindex="-1"></a><span class="do">### By cell type markers</span></span>
<span id="cb473-2147"><a href="#cb473-2147" aria-hidden="true" tabindex="-1"></a>celltype_mean <span class="ot">&lt;-</span> <span class="fu">aggregateAcrossCells</span>(</span>
<span id="cb473-2148"><a href="#cb473-2148" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as</span>(spe, <span class="st">"SingleCellExperiment"</span>),</span>
<span id="cb473-2149"><a href="#cb473-2149" aria-hidden="true" tabindex="-1"></a>    <span class="at">ids =</span> spe<span class="sc">$</span>celltype,</span>
<span id="cb473-2150"><a href="#cb473-2150" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="st">"mean"</span>,</span>
<span id="cb473-2151"><a href="#cb473-2151" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.assay.type =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-2152"><a href="#cb473-2152" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"type"</span>]</span>
<span id="cb473-2153"><a href="#cb473-2153" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2154"><a href="#cb473-2154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2155"><a href="#cb473-2155" aria-hidden="true" tabindex="-1"></a>h_type <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb473-2156"><a href="#cb473-2156" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>(<span class="fu">assay</span>(celltype_mean, <span class="st">"exprs"</span>)),</span>
<span id="cb473-2157"><a href="#cb473-2157" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title =</span> <span class="st">"type_markers"</span>,</span>
<span id="cb473-2158"><a href="#cb473-2158" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> col_exprs,</span>
<span id="cb473-2159"><a href="#cb473-2159" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"mean exprs"</span>,</span>
<span id="cb473-2160"><a href="#cb473-2160" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-2161"><a href="#cb473-2161" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_names =</span> <span class="cn">TRUE</span></span>
<span id="cb473-2162"><a href="#cb473-2162" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2163"><a href="#cb473-2163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2164"><a href="#cb473-2164" aria-hidden="true" tabindex="-1"></a><span class="co"># By cell state markers</span></span>
<span id="cb473-2165"><a href="#cb473-2165" aria-hidden="true" tabindex="-1"></a>cellstate_mean <span class="ot">&lt;-</span> <span class="fu">aggregateAcrossCells</span>(</span>
<span id="cb473-2166"><a href="#cb473-2166" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as</span>(spe, <span class="st">"SingleCellExperiment"</span>),</span>
<span id="cb473-2167"><a href="#cb473-2167" aria-hidden="true" tabindex="-1"></a>    <span class="at">ids =</span> spe<span class="sc">$</span>celltype,</span>
<span id="cb473-2168"><a href="#cb473-2168" aria-hidden="true" tabindex="-1"></a>    <span class="at">statistics =</span> <span class="st">"mean"</span>,</span>
<span id="cb473-2169"><a href="#cb473-2169" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.assay.type =</span> <span class="st">"exprs"</span>,</span>
<span id="cb473-2170"><a href="#cb473-2170" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rownames</span>(spe)[<span class="fu">rowData</span>(spe)<span class="sc">$</span>marker_class <span class="sc">==</span> <span class="st">"state"</span>]</span>
<span id="cb473-2171"><a href="#cb473-2171" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2172"><a href="#cb473-2172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2173"><a href="#cb473-2173" aria-hidden="true" tabindex="-1"></a>h_state <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb473-2174"><a href="#cb473-2174" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>(<span class="fu">assay</span>(cellstate_mean, <span class="st">"exprs"</span>)),</span>
<span id="cb473-2175"><a href="#cb473-2175" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title =</span> <span class="st">"state_markers"</span>,</span>
<span id="cb473-2176"><a href="#cb473-2176" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> col_exprs,</span>
<span id="cb473-2177"><a href="#cb473-2177" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"mean exprs"</span>,</span>
<span id="cb473-2178"><a href="#cb473-2178" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-2179"><a href="#cb473-2179" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_names =</span> <span class="cn">TRUE</span></span>
<span id="cb473-2180"><a href="#cb473-2180" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2181"><a href="#cb473-2181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2182"><a href="#cb473-2182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2183"><a href="#cb473-2183" aria-hidden="true" tabindex="-1"></a><span class="do">### 2. Heatmap annotation </span><span class="al">###</span></span>
<span id="cb473-2184"><a href="#cb473-2184" aria-hidden="true" tabindex="-1"></a><span class="do">### 2.1  Metadata features</span></span>
<span id="cb473-2185"><a href="#cb473-2185" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">&lt;-</span> <span class="fu">colData</span>(celltype_mean) <span class="sc">|&gt;</span> </span>
<span id="cb473-2186"><a href="#cb473-2186" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb473-2187"><a href="#cb473-2187" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(celltype, ncells)</span>
<span id="cb473-2188"><a href="#cb473-2188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2189"><a href="#cb473-2189" aria-hidden="true" tabindex="-1"></a><span class="do">### Proportion of indication per celltype</span></span>
<span id="cb473-2190"><a href="#cb473-2190" aria-hidden="true" tabindex="-1"></a>indication <span class="ot">&lt;-</span> <span class="fu">unclass</span>(<span class="fu">prop.table</span>(<span class="fu">table</span>(spe<span class="sc">$</span>celltype, spe<span class="sc">$</span>indication), <span class="at">margin =</span> <span class="dv">1</span>))</span>
<span id="cb473-2191"><a href="#cb473-2191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2192"><a href="#cb473-2192" aria-hidden="true" tabindex="-1"></a><span class="do">### Number of contributing patients per celltype</span></span>
<span id="cb473-2193"><a href="#cb473-2193" aria-hidden="true" tabindex="-1"></a>cluster_PID <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe) <span class="sc">|&gt;</span></span>
<span id="cb473-2194"><a href="#cb473-2194" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb473-2195"><a href="#cb473-2195" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(celltype, patient_id) <span class="sc">|&gt;</span></span>
<span id="cb473-2196"><a href="#cb473-2196" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(celltype) <span class="sc">|&gt;</span> <span class="fu">table</span>() <span class="sc">|&gt;</span></span>
<span id="cb473-2197"><a href="#cb473-2197" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb473-2198"><a href="#cb473-2198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2199"><a href="#cb473-2199" aria-hidden="true" tabindex="-1"></a>n_PID <span class="ot">&lt;-</span> cluster_PID <span class="sc">|&gt;</span></span>
<span id="cb473-2200"><a href="#cb473-2200" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(Freq <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">|&gt;</span></span>
<span id="cb473-2201"><a href="#cb473-2201" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(celltype) <span class="sc">|&gt;</span></span>
<span id="cb473-2202"><a href="#cb473-2202" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">count</span>(<span class="at">name =</span> <span class="st">"n_PID"</span>) <span class="sc">|&gt;</span></span>
<span id="cb473-2203"><a href="#cb473-2203" aria-hidden="true" tabindex="-1"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"celltype"</span>)</span>
<span id="cb473-2204"><a href="#cb473-2204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2205"><a href="#cb473-2205" aria-hidden="true" tabindex="-1"></a><span class="do">### Create HeatmapAnnotation objects</span></span>
<span id="cb473-2206"><a href="#cb473-2206" aria-hidden="true" tabindex="-1"></a>ha_anno <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb473-2207"><a href="#cb473-2207" aria-hidden="true" tabindex="-1"></a>    <span class="at">celltype =</span> anno<span class="sc">$</span>celltype,</span>
<span id="cb473-2208"><a href="#cb473-2208" aria-hidden="true" tabindex="-1"></a>    <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-2209"><a href="#cb473-2209" aria-hidden="true" tabindex="-1"></a>    <span class="at">gap =</span> <span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>),</span>
<span id="cb473-2210"><a href="#cb473-2210" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">celltype =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>celltype),</span>
<span id="cb473-2211"><a href="#cb473-2211" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span></span>
<span id="cb473-2212"><a href="#cb473-2212" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2213"><a href="#cb473-2213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2214"><a href="#cb473-2214" aria-hidden="true" tabindex="-1"></a>ha_meta <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb473-2215"><a href="#cb473-2215" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_cells =</span> <span class="fu">anno_barplot</span>(anno<span class="sc">$</span>ncells, <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>)),</span>
<span id="cb473-2216"><a href="#cb473-2216" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_PID =</span> <span class="fu">anno_barplot</span>(n_PID, <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>)),</span>
<span id="cb473-2217"><a href="#cb473-2217" aria-hidden="true" tabindex="-1"></a>    <span class="at">indication =</span> <span class="fu">anno_barplot</span>(indication, <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>),</span>
<span id="cb473-2218"><a href="#cb473-2218" aria-hidden="true" tabindex="-1"></a>    <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">fill =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication)),</span>
<span id="cb473-2219"><a href="#cb473-2219" aria-hidden="true" tabindex="-1"></a>    <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-2220"><a href="#cb473-2220" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation_name_rot =</span> <span class="dv">90</span>,</span>
<span id="cb473-2221"><a href="#cb473-2221" aria-hidden="true" tabindex="-1"></a>    <span class="at">gap =</span> <span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>),</span>
<span id="cb473-2222"><a href="#cb473-2222" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span></span>
<span id="cb473-2223"><a href="#cb473-2223" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2224"><a href="#cb473-2224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2225"><a href="#cb473-2225" aria-hidden="true" tabindex="-1"></a><span class="do">### 2.2 Spatial features</span></span>
<span id="cb473-2226"><a href="#cb473-2226" aria-hidden="true" tabindex="-1"></a><span class="do">### Add number of neighbors to spe object (saved in colPair)</span></span>
<span id="cb473-2227"><a href="#cb473-2227" aria-hidden="true" tabindex="-1"></a>spe<span class="sc">$</span>n_neighbors <span class="ot">&lt;-</span> <span class="fu">countLnodeHits</span>(<span class="fu">colPair</span>(spe, <span class="st">"neighborhood"</span>))</span>
<span id="cb473-2228"><a href="#cb473-2228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2229"><a href="#cb473-2229" aria-hidden="true" tabindex="-1"></a><span class="do">### Select spatial features and average over celltypes</span></span>
<span id="cb473-2230"><a href="#cb473-2230" aria-hidden="true" tabindex="-1"></a>spatial <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe) <span class="sc">|&gt;</span></span>
<span id="cb473-2231"><a href="#cb473-2231" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb473-2232"><a href="#cb473-2232" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(area, celltype, n_neighbors)</span>
<span id="cb473-2233"><a href="#cb473-2233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2234"><a href="#cb473-2234" aria-hidden="true" tabindex="-1"></a>spatial <span class="ot">&lt;-</span> spatial <span class="sc">|&gt;</span></span>
<span id="cb473-2235"><a href="#cb473-2235" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>celltype) <span class="sc">|&gt;</span></span>
<span id="cb473-2236"><a href="#cb473-2236" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aggregate</span>(<span class="at">by =</span> <span class="fu">list</span>(<span class="at">celltype =</span> spatial<span class="sc">$</span>celltype), <span class="at">FUN =</span> mean) <span class="sc">|&gt;</span></span>
<span id="cb473-2237"><a href="#cb473-2237" aria-hidden="true" tabindex="-1"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"celltype"</span>)</span>
<span id="cb473-2238"><a href="#cb473-2238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2239"><a href="#cb473-2239" aria-hidden="true" tabindex="-1"></a><span class="do">### Create HeatmapAnnotation object</span></span>
<span id="cb473-2240"><a href="#cb473-2240" aria-hidden="true" tabindex="-1"></a>ha_spatial <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb473-2241"><a href="#cb473-2241" aria-hidden="true" tabindex="-1"></a>    <span class="at">area =</span> spatial<span class="sc">$</span>area,</span>
<span id="cb473-2242"><a href="#cb473-2242" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_neighbors =</span> spatial<span class="sc">$</span>n_neighbors,</span>
<span id="cb473-2243"><a href="#cb473-2243" aria-hidden="true" tabindex="-1"></a>    <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb473-2244"><a href="#cb473-2244" aria-hidden="true" tabindex="-1"></a>    <span class="at">gap =</span> <span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">"mm"</span>),</span>
<span id="cb473-2245"><a href="#cb473-2245" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span></span>
<span id="cb473-2246"><a href="#cb473-2246" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2247"><a href="#cb473-2247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2248"><a href="#cb473-2248" aria-hidden="true" tabindex="-1"></a><span class="do">### 3. Plot rich heatmap </span><span class="al">###</span></span>
<span id="cb473-2249"><a href="#cb473-2249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2250"><a href="#cb473-2250" aria-hidden="true" tabindex="-1"></a><span class="do">### Create HeatmapList object</span></span>
<span id="cb473-2251"><a href="#cb473-2251" aria-hidden="true" tabindex="-1"></a>h_list <span class="ot">&lt;-</span> h_type <span class="sc">+</span></span>
<span id="cb473-2252"><a href="#cb473-2252" aria-hidden="true" tabindex="-1"></a>    h_state <span class="sc">+</span></span>
<span id="cb473-2253"><a href="#cb473-2253" aria-hidden="true" tabindex="-1"></a>    ha_anno <span class="sc">+</span></span>
<span id="cb473-2254"><a href="#cb473-2254" aria-hidden="true" tabindex="-1"></a>    ha_spatial <span class="sc">+</span></span>
<span id="cb473-2255"><a href="#cb473-2255" aria-hidden="true" tabindex="-1"></a>    ha_meta</span>
<span id="cb473-2256"><a href="#cb473-2256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2257"><a href="#cb473-2257" aria-hidden="true" tabindex="-1"></a><span class="do">### Add customized legend for anno_barplot()</span></span>
<span id="cb473-2258"><a href="#cb473-2258" aria-hidden="true" tabindex="-1"></a>lgd <span class="ot">&lt;-</span> <span class="fu">Legend</span>(</span>
<span id="cb473-2259"><a href="#cb473-2259" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"indication"</span>,</span>
<span id="cb473-2260"><a href="#cb473-2260" aria-hidden="true" tabindex="-1"></a>    <span class="at">at =</span> <span class="fu">colnames</span>(indication),</span>
<span id="cb473-2261"><a href="#cb473-2261" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend_gp =</span> <span class="fu">gpar</span>(<span class="at">fill =</span> <span class="fu">metadata</span>(spe)<span class="sc">$</span>color_vectors<span class="sc">$</span>indication)</span>
<span id="cb473-2262"><a href="#cb473-2262" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb473-2263"><a href="#cb473-2263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2264"><a href="#cb473-2264" aria-hidden="true" tabindex="-1"></a><span class="do">### Plot</span></span>
<span id="cb473-2265"><a href="#cb473-2265" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(h_list, <span class="at">annotation_legend_list =</span> <span class="fu">list</span>(lgd))</span>
<span id="cb473-2266"><a href="#cb473-2266" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb473-2267"><a href="#cb473-2267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2268"><a href="#cb473-2268" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reference</span></span>
<span id="cb473-2269"><a href="#cb473-2269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2270"><a href="#cb473-2270" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">Analysis workflow for IMC data</span><span class="co">](https://bodenmillergroup.github.io/IMCDataAnalysis/)</span></span>
<span id="cb473-2271"><a href="#cb473-2271" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">Working with CellProfiler data in R</span><span class="co">](https://complexinterface.com/research/projects/image-analysis/)</span></span>
<span id="cb473-2272"><a href="#cb473-2272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2273"><a href="#cb473-2273" aria-hidden="true" tabindex="-1"></a><span class="fu">## Sessioninfo</span></span>
<span id="cb473-2274"><a href="#cb473-2274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb473-2277"><a href="#cb473-2277" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb473-2278"><a href="#cb473-2278" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb473-2279"><a href="#cb473-2279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2026 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a>, <a href="https://www.python.org/"><i class="fa-brands fa-python" aria-label="python"></i></a> and <a href="https://quarto.org/">Quarto</a></span></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


<script src="../../../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>